{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f9045d0-6fd7-4b09-922a-365a72b6b06a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import itertools\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "10818620-52ba-4b42-b6ae-6044de2bb977",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../snakemake/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "7d9fcd93-f652-44d9-836d-b8d1ce968585",
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_ab = '../'+expand(config['data']['filt_ab'], species='human')[0]\n",
    "lib_meta = '../'+expand(config['data']['meta'], species='human')[0]\n",
    "cerberus_h5 = '../'+expand(config['data']['cerb_annot'], species='human')[0]\n",
    "ca_plus = '../'+expand(config['data']['ca_plus'], species='human')[0]\n",
    "swan_file = '../'+expand(config['data']['sg'], species='human')[0]\n",
    "gtf = '../'+expand(config['data']['cerb_gtf'], species='human')[0]\n",
    "\n",
    "phastcons = '../'+config['ref']['phastcons100']['txt']\n",
    "\n",
    "gene_subset = 'polya'\n",
    "\n",
    "biosamp_name_map = '../'+expand(config['ref']['biosamp_map'])[0]\n",
    "\n",
    "cage_meta = '../'+expand(config['cage']['meta'], species='human')[0]\n",
    "rampage_meta = '../'+expand(config['rampage']['meta'], species='human')[0]\n",
    "\n",
    "min_tpm = 1\n",
    "gene_subset = 'protein_coding'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "37db4253-5b18-4900-bbe5-199bfd136ea0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n"
     ]
    }
   ],
   "source": [
    "gtf_df = pr.read_gtf(gtf, rename_attr=True, duplicate_attr=True).as_df()\n",
    "gtf_df = cerberus.add_stable_gid(gtf_df)\n",
    "gtf_df = pr.PyRanges(gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7b615c31-47c9-43d9-bf24-0a5c69d3b05d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df = gtf_df.df\n",
    "gene_df = gene_df.loc[gene_df.Feature=='gene']\n",
    "gene_df = gene_df[['gene_id', 'gene_name']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb810bd4-0827-4679-8793-eae11c547865",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fee98c7f-9b2a-4857-be4c-62a97b16abc5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c7a856e0-c3cf-4907-8bf3-acd4132c8495",
   "metadata": {},
   "outputs": [],
   "source": [
    "## mean intron length\n",
    "introns = gtf_df.features.introns(by='transcript').df\n",
    "introns['len'] = (introns.Start-introns.End).abs()\n",
    "temp = introns[['gene_id', 'transcript_id', 'len']].groupby(['gene_id', 'transcript_id']).sum().reset_index()\n",
    "temp = temp[['gene_id', 'len']].groupby('gene_id').mean().reset_index().rename({'len':'mean_intron_len'}, axis=1)\n",
    "gene_df = gene_df.merge(temp, how='left', on='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1d01f3e0-e001-4c04-a727-da7590f58600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mean_intron_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>51925.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>CFH</td>\n",
       "      <td>65430.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000001461</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>40274.700000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004487</td>\n",
       "      <td>KDM1A</td>\n",
       "      <td>60957.750000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000007933</td>\n",
       "      <td>FMO3</td>\n",
       "      <td>16727.250000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name  mean_intron_len\n",
       "0  ENSG00000000460  C1orf112     51925.666667\n",
       "1  ENSG00000000971       CFH     65430.333333\n",
       "2  ENSG00000001461    NIPAL3     40274.700000\n",
       "3  ENSG00000004487     KDM1A     60957.750000\n",
       "4  ENSG00000007933      FMO3     16727.250000"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "4366a556-4d82-4bfe-91e5-ac43dc53883a",
   "metadata": {},
   "outputs": [],
   "source": [
    "## mean # introns\n",
    "temp = introns[['gene_id', 'transcript_id', 'len']].groupby(['gene_id', 'transcript_id']).count().reset_index()\n",
    "temp = temp[['gene_id', 'len']].groupby('gene_id').mean().reset_index().rename({'len':'mean_n_intron'}, axis=1)\n",
    "gene_df = gene_df.merge(temp, how='left', on='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "08cf995e-80f3-4dd2-ad5d-a3911102e31a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mean_intron_len</th>\n",
       "      <th>mean_n_intron</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>51925.666667</td>\n",
       "      <td>21.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>CFH</td>\n",
       "      <td>65430.333333</td>\n",
       "      <td>13.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000001461</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>40274.700000</td>\n",
       "      <td>8.800000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004487</td>\n",
       "      <td>KDM1A</td>\n",
       "      <td>60957.750000</td>\n",
       "      <td>17.875000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000007933</td>\n",
       "      <td>FMO3</td>\n",
       "      <td>16727.250000</td>\n",
       "      <td>6.250000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name  mean_intron_len  mean_n_intron\n",
       "0  ENSG00000000460  C1orf112     51925.666667      21.666667\n",
       "1  ENSG00000000971       CFH     65430.333333      13.333333\n",
       "2  ENSG00000001461    NIPAL3     40274.700000       8.800000\n",
       "3  ENSG00000004487     KDM1A     60957.750000      17.875000\n",
       "4  ENSG00000007933      FMO3     16727.250000       6.250000"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "445bc91c-5172-4aa7-8942-68fa39839e49",
   "metadata": {},
   "outputs": [],
   "source": [
    "## exon len % 3 (internal exons only)\n",
    "gtf_df = gtf_df.as_df()\n",
    "exons = gtf_df.loc[gtf_df.Feature=='exon']\n",
    "\n",
    "# remove monoexonic or two-exon transcripts\n",
    "temp = exons[['transcript_id', 'Start']].groupby('transcript_id').count().reset_index().rename({'Start': 'n_exons'}, axis=1)\n",
    "temp = temp.loc[temp.n_exons>2]\n",
    "exons = exons.loc[exons.transcript_id.isin(temp.transcript_id.tolist())]\n",
    "\n",
    "# restrict to internal exons\n",
    "exons = exons.reset_index(drop=True)\n",
    "drop_inds = exons[['transcript_id']].drop_duplicates(keep='first').index.tolist()+\\\n",
    "            exons[['transcript_id']].drop_duplicates(keep='last').index.tolist()\n",
    "keep_inds = list(set(exons.index.tolist())-set(drop_inds))\n",
    "exons = exons.loc[keep_inds]\n",
    "\n",
    "# compute len of exons\n",
    "exons['exon_len'] = (exons.Start-exons.End).abs()\n",
    "\n",
    "# mod 3 info\n",
    "exons['exon_len_mod_3'] = exons.exon_len % 3 \n",
    "exons['frame_preserving'] = exons.exon_len % 3 == 0\n",
    "\n",
    "# % of frame preserving exons / gene\n",
    "keep_cols = ['Chromosome', 'Start', 'End', 'Strand', 'gene_id', 'exon_len_mod_3', 'frame_preserving']\n",
    "exons = exons[keep_cols].drop_duplicates()\n",
    "exons = exons[['gene_id', 'frame_preserving']].groupby('gene_id').agg({'frame_preserving': ['count', 'sum']}).reset_index().rename({'count':'n_exons', 'sum':'n_frame_preserving_exons'}, axis=1)\n",
    "exons.columns = ['gene_id', 'n_exons', 'n_frame_preserving_exons']\n",
    "exons['perc_frame_pres_exons'] = (exons.n_frame_preserving_exons/exons.n_exons)*100\n",
    "gene_df = gene_df.merge(exons, how='left', on='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "ecc77bc9-6d39-40fb-a107-0a672eb89df6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mean_intron_len</th>\n",
       "      <th>mean_n_intron</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>n_frame_preserving_exons</th>\n",
       "      <th>perc_frame_pres_exons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>51925.666667</td>\n",
       "      <td>21.666667</td>\n",
       "      <td>25.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>52.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>CFH</td>\n",
       "      <td>65430.333333</td>\n",
       "      <td>13.333333</td>\n",
       "      <td>23.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>82.608696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000001461</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>40274.700000</td>\n",
       "      <td>8.800000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>18.181818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004487</td>\n",
       "      <td>KDM1A</td>\n",
       "      <td>60957.750000</td>\n",
       "      <td>17.875000</td>\n",
       "      <td>23.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>21.739130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000007933</td>\n",
       "      <td>FMO3</td>\n",
       "      <td>16727.250000</td>\n",
       "      <td>6.250000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>28.571429</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name  mean_intron_len  mean_n_intron  n_exons  \\\n",
       "0  ENSG00000000460  C1orf112     51925.666667      21.666667     25.0   \n",
       "1  ENSG00000000971       CFH     65430.333333      13.333333     23.0   \n",
       "2  ENSG00000001461    NIPAL3     40274.700000       8.800000     11.0   \n",
       "3  ENSG00000004487     KDM1A     60957.750000      17.875000     23.0   \n",
       "4  ENSG00000007933      FMO3     16727.250000       6.250000      7.0   \n",
       "\n",
       "   n_frame_preserving_exons  perc_frame_pres_exons  \n",
       "0                      13.0              52.000000  \n",
       "1                      19.0              82.608696  \n",
       "2                       2.0              18.181818  \n",
       "3                       5.0              21.739130  \n",
       "4                       2.0              28.571429  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "cbcbd0fb-af6e-4f40-b565-21b7995b1196",
   "metadata": {},
   "outputs": [],
   "source": [
    "## conservation\n",
    "\n",
    "min_cons_score = 250\n",
    "\n",
    "# read conserved elements in, threshold for a certain score\n",
    "cons = pd.read_csv(phastcons, sep='\\t', header=None,\n",
    "                 names=['bin', 'Chromosome', 'Start', 'End', 'idk1', 'score'])\n",
    "cons = cons[['Chromosome', 'Start', 'End', 'score']]\n",
    "cons = cons.loc[cons.score >= min_cons_score]\n",
    "cons = pr.PyRanges(cons)\n",
    "\n",
    "# get exons\n",
    "exons = gtf_df[['Chromosome', 'Start', 'End', 'gene_name',\n",
    "                 'gene_id', 'transcript_id', 'Feature']]\n",
    "exons = exons.loc[gtf_df.Feature=='exon']\n",
    "exons['exon_len'] = (exons.Start-exons.End).abs()\n",
    "exons = pr.PyRanges(exons)\n",
    "exons = exons.join(cons, report_overlap=True)\n",
    "\n",
    "# get the % length covered in each thing\n",
    "exons = exons.df\n",
    "keep_cols = ['Chromosome', 'Start', 'End', 'gene_name', 'gene_id', 'transcript_id', 'exon_len', 'Overlap']\n",
    "gb_cols = ['Chromosome', 'Start', 'End', 'gene_name', 'gene_id', 'transcript_id', 'exon_len']\n",
    "temp = exons[keep_cols].groupby(gb_cols, observed=True).sum().reset_index().rename({'Overlap': 'cons_len'}, axis=1)\n",
    "\n",
    "# get total len of transcript and len covered of each transcript \n",
    "keep_cols = ['gene_name', 'gene_id', 'transcript_id', 'exon_len', 'cons_len']\n",
    "gb_cols = ['gene_name', 'gene_id', 'transcript_id']\n",
    "temp = temp[keep_cols].groupby(gb_cols,observed=True).sum().reset_index().rename({'exon_len': 't_len'}, axis=1)\n",
    "\n",
    "# now get mean % exonic conservation per gene\n",
    "temp['perc_cons'] = (temp['cons_len']/temp['t_len'])*100\n",
    "keep_cols = ['gene_id', 'perc_cons']\n",
    "gb_cols = ['gene_id']\n",
    "temp = temp[keep_cols].groupby(gb_cols, observed=True).mean().reset_index().rename({'perc_cons':'mean_perc_cons'}, axis=1)\n",
    "gene_df = gene_df.merge(temp, how='left', on='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "87b5b9fd-03a2-4a1d-b855-8cf7a1db2c51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>mean_intron_len</th>\n",
       "      <th>mean_n_intron</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>n_frame_preserving_exons</th>\n",
       "      <th>perc_frame_pres_exons</th>\n",
       "      <th>mean_perc_cons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>51925.666667</td>\n",
       "      <td>21.666667</td>\n",
       "      <td>25.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>52.000000</td>\n",
       "      <td>61.028481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000971</td>\n",
       "      <td>CFH</td>\n",
       "      <td>65430.333333</td>\n",
       "      <td>13.333333</td>\n",
       "      <td>23.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>82.608696</td>\n",
       "      <td>26.370211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000001461</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>40274.700000</td>\n",
       "      <td>8.800000</td>\n",
       "      <td>11.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>18.181818</td>\n",
       "      <td>38.732814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004487</td>\n",
       "      <td>KDM1A</td>\n",
       "      <td>60957.750000</td>\n",
       "      <td>17.875000</td>\n",
       "      <td>23.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>21.739130</td>\n",
       "      <td>70.077454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000007933</td>\n",
       "      <td>FMO3</td>\n",
       "      <td>16727.250000</td>\n",
       "      <td>6.250000</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>28.571429</td>\n",
       "      <td>34.694776</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id gene_name  mean_intron_len  mean_n_intron  n_exons  \\\n",
       "0  ENSG00000000460  C1orf112     51925.666667      21.666667     25.0   \n",
       "1  ENSG00000000971       CFH     65430.333333      13.333333     23.0   \n",
       "2  ENSG00000001461    NIPAL3     40274.700000       8.800000     11.0   \n",
       "3  ENSG00000004487     KDM1A     60957.750000      17.875000     23.0   \n",
       "4  ENSG00000007933      FMO3     16727.250000       6.250000      7.0   \n",
       "\n",
       "   n_frame_preserving_exons  perc_frame_pres_exons  mean_perc_cons  \n",
       "0                      13.0              52.000000       61.028481  \n",
       "1                      19.0              82.608696       26.370211  \n",
       "2                       2.0              18.181818       38.732814  \n",
       "3                       5.0              21.739130       70.077454  \n",
       "4                       2.0              28.571429       34.694776  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d806c49e-2227-4981-85c1-86b3ca1da17d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mean_perc_cons</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>41.472387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>79.513616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000000419</td>\n",
       "      <td>57.704859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000000457</td>\n",
       "      <td>55.789980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000000460</td>\n",
       "      <td>61.028481</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id  mean_perc_cons\n",
       "0  ENSG00000000003       41.472387\n",
       "1  ENSG00000000005       79.513616\n",
       "2  ENSG00000000419       57.704859\n",
       "3  ENSG00000000457       55.789980\n",
       "4  ENSG00000000460       61.028481"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ce71cb3-11c0-4add-91fa-c6870807743f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c79e28c-6ea1-4f7a-9796-31acce146764",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
