{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4a6a7910-a490-43ff-8801-9900e4bc0530",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ab3a9a29-a63c-448e-bc46-d1e37eb3da45",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../snakemake/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "455c11a8-9c97-4f6d-b4a1-71650f09105a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = '../'+expand(config['data']['ab'], species='human')[0]\n",
    "talon_ab = '../'+expand(config['data']['talon_filt_ab'], species='human')[0]\n",
    "lapa_ab = '../'+expand(config['data']['lapa_filt_ab'], species='human')[0]\n",
    "filt_ab = '../'+expand(config['data']['filt_ab'], species='human')[0]\n",
    "read_annot = '../'+expand(config['data']['read_annot'], species='human')[0]\n",
    "t_metadata = '../'+expand(config['ref']['cerberus']['t_info'], species='human')[0]\n",
    "lib_meta = '../'+expand(config['data']['meta'], species='human')[0]\n",
    "swan_file = '../'+expand(config['data']['sg'], species='human')[0]\n",
    "cerberus_h5 = '../'+expand(config['data']['cerb_annot'], species='human')[0]\n",
    "cerb_t_metadata = '../'+expand(config['data']['t_info'], species='human')[0]\n",
    "major_isos = '../'+expand(config['data']['major_isos'], species='human', obs_col='sample')[0]\n",
    "pp_summary = '../'+expand(config['data']['p_pred']['summary'], species='human')[0]\n",
    "ref_t_metadata = '../'+expand(config['ref']['t_info'], species='human')[0]\n",
    "ref_g_metadata = '../'+expand(config['ref']['g_info'], species='human')[0]\n",
    "\n",
    "spike_filt_ab = '../'+config['data']['spikes']['filt_ab']\n",
    "gene_spike_ab = '../'+config['data']['spikes']['gene_ab']\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "min_tpm = 1\n",
    "gene_subset = 'protein_coding'\n",
    "obs_col = 'sample'\n",
    "go_gene_subset = 'protein_coding'\n",
    "\n",
    "m_lib_meta = '../'+expand(config['data']['meta'], species='mouse')[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40fa0ef4-1393-4fe6-9305-20c93a39587d",
   "metadata": {},
   "source": [
    "## SIRV novel isoform detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d43c9673-48c4-4110-bbf8-012ba1894332",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../data/spike/talon/annot_talon_abundance_filtered.tsv'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "spike_filt_ab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1f74b4b4-d935-4df9-8d62-8b1b85a7ccfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(spike_filt_ab, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "621d6cd4-9266-438a-99f1-724562a2042d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>cortex_wt_f_1_1_labeled_mouse</th>\n",
       "      <th>cortex_wt_f_1_2_labeled_mouse</th>\n",
       "      <th>cortex_wt_m_2_1_labeled_mouse</th>\n",
       "      <th>cortex_wt_m_2_2_labeled_mouse</th>\n",
       "      <th>hippocampus_wt_f_1_1_labeled_mouse</th>\n",
       "      <th>hippocampus_wt_f_1_2_labeled_mouse</th>\n",
       "      <th>hippocampus_5x_m_1_1_labeled_mouse</th>\n",
       "      <th>hippocampus_5x_m_1_2_labeled_mouse</th>\n",
       "      <th>cortex_5x_m_1_1_labeled_mouse</th>\n",
       "      <th>cortex_5x_m_1_2_labeled_mouse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>350</th>\n",
       "      <td>11</td>\n",
       "      <td>3633</td>\n",
       "      <td>SIRV6.1</td>\n",
       "      <td>spikesT000003633</td>\n",
       "      <td>spikesG000000011</td>\n",
       "      <td>spikesT000003633</td>\n",
       "      <td>3</td>\n",
       "      <td>788</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>11</td>\n",
       "      <td>3637</td>\n",
       "      <td>SIRV6.1</td>\n",
       "      <td>spikesT000003637</td>\n",
       "      <td>spikesG000000011</td>\n",
       "      <td>spikesT000003637</td>\n",
       "      <td>2</td>\n",
       "      <td>658</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>11</td>\n",
       "      <td>3677</td>\n",
       "      <td>SIRV6.1</td>\n",
       "      <td>spikesT000003677</td>\n",
       "      <td>spikesG000000011</td>\n",
       "      <td>spikesT000003677</td>\n",
       "      <td>2</td>\n",
       "      <td>1356</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>11</td>\n",
       "      <td>3680</td>\n",
       "      <td>SIRV6.1</td>\n",
       "      <td>spikesT000003680</td>\n",
       "      <td>spikesG000000011</td>\n",
       "      <td>spikesT000003680</td>\n",
       "      <td>1</td>\n",
       "      <td>302</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>196</td>\n",
       "      <td>3735</td>\n",
       "      <td>spikesG000000196</td>\n",
       "      <td>spikesT000003735</td>\n",
       "      <td>spikesG000000196</td>\n",
       "      <td>spikesT000003735</td>\n",
       "      <td>1</td>\n",
       "      <td>729</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>Antisense</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 237 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_ID  transcript_ID     annot_gene_id annot_transcript_id  \\\n",
       "350       11           3633           SIRV6.1    spikesT000003633   \n",
       "351       11           3637           SIRV6.1    spikesT000003637   \n",
       "352       11           3677           SIRV6.1    spikesT000003677   \n",
       "353       11           3680           SIRV6.1    spikesT000003680   \n",
       "354      196           3735  spikesG000000196    spikesT000003735   \n",
       "\n",
       "      annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "350  spikesG000000011      spikesT000003633        3     788        Known   \n",
       "351  spikesG000000011      spikesT000003637        2     658        Known   \n",
       "352  spikesG000000011      spikesT000003677        2    1356        Known   \n",
       "353  spikesG000000011      spikesT000003680        1     302        Known   \n",
       "354  spikesG000000196      spikesT000003735        1     729    Antisense   \n",
       "\n",
       "    transcript_novelty  ... cortex_wt_f_1_1_labeled_mouse  \\\n",
       "350                ISM  ...                             0   \n",
       "351                ISM  ...                             0   \n",
       "352                ISM  ...                             0   \n",
       "353                ISM  ...                             0   \n",
       "354          Antisense  ...                             0   \n",
       "\n",
       "     cortex_wt_f_1_2_labeled_mouse  cortex_wt_m_2_1_labeled_mouse  \\\n",
       "350                              0                              0   \n",
       "351                              0                              0   \n",
       "352                              0                              1   \n",
       "353                              0                              0   \n",
       "354                              0                              0   \n",
       "\n",
       "     cortex_wt_m_2_2_labeled_mouse  hippocampus_wt_f_1_1_labeled_mouse  \\\n",
       "350                              0                                   0   \n",
       "351                              0                                   0   \n",
       "352                              2                                   0   \n",
       "353                              0                                   0   \n",
       "354                              1                                   0   \n",
       "\n",
       "     hippocampus_wt_f_1_2_labeled_mouse  hippocampus_5x_m_1_1_labeled_mouse  \\\n",
       "350                                   0                                   0   \n",
       "351                                   0                                   0   \n",
       "352                                   0                                   0   \n",
       "353                                   0                                   0   \n",
       "354                                   0                                   0   \n",
       "\n",
       "     hippocampus_5x_m_1_2_labeled_mouse  cortex_5x_m_1_1_labeled_mouse  \\\n",
       "350                                   0                              0   \n",
       "351                                   0                              0   \n",
       "352                                   1                              1   \n",
       "353                                   0                              0   \n",
       "354                                   0                              0   \n",
       "\n",
       "     cortex_5x_m_1_2_labeled_mouse  \n",
       "350                              3  \n",
       "351                              1  \n",
       "352                              0  \n",
       "353                              1  \n",
       "354                              0  \n",
       "\n",
       "[5 rows x 237 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b305b471-f5fd-4373-a0a9-0934a62114bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Antisense</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISM</th>\n",
       "      <td>141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    annot_transcript_id\n",
       "transcript_novelty                     \n",
       "Antisense                             9\n",
       "ISM                                 141\n",
       "NIC                                  21\n",
       "NNC                                  14"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.transcript_novelty!='Known']\n",
    "temp = temp[['annot_transcript_id', 'transcript_novelty']].groupby('transcript_novelty').count()\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2d06060-f7fc-438f-93d4-d0c9b5c74f15",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "051d349a-3950-433a-9a3a-9e58c6d0e05a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "648ccdd0-9c4a-4f98-95a4-101e63aa9a1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7c53e8f-f09b-4180-be5b-7601641c0a35",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f2976b1-9dd1-4c4f-a663-c336e7d758b7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d819f5-89ac-4abd-82ca-9d8f023f86c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a0e0561d-e6a8-4585-a68f-4315565220f6",
   "metadata": {},
   "source": [
    "## 230720 ok so the sirv \"genes\" overlap, but how many genes do we call per sirv \"chromosome\"?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "01e0d269-d119-4a6a-ae2e-1bf83566d65f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(lapa_ab, sep='\\t')\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "291acdfb-9b84-4039-9cad-ecc32e1e8233",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Antisense</th>\n",
       "      <td>23316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISM</th>\n",
       "      <td>80937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intergenic</th>\n",
       "      <td>1803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>193913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>79168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>23323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    annot_transcript_id\n",
       "transcript_novelty                     \n",
       "Antisense                         23316\n",
       "ISM                               80937\n",
       "Intergenic                         1803\n",
       "Known                            193913\n",
       "NIC                               79168\n",
       "NNC                               23323"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['transcript_novelty', 'annot_transcript_id']].groupby('transcript_novelty').count()\n",
    "# these look like they've just been filtered for reproducibility and to remove genomic\n",
    "# figured out through breadcrumbs trails that these have indeed been filtered for reproducibility: \n",
    "# https://sandbox.zenodo.org/record/1095599"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7adc073d-8227-43a0-b9de-ed5115947ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 230720 ok so the sirv \"genes\" overlap, but what do they look like on the brwoser?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0b9f25e-206c-413b-a63e-12628f8a269a",
   "metadata": {},
   "source": [
    "## 230720 Is there only one sirv / ercc gene per \"chromosome\"? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a68b4119-c0d2-4067-92c1-c8a3e6daf42b",
   "metadata": {},
   "outputs": [],
   "source": [
    "spike_gtf = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/refs/gencode_v29_sirv4_ercc.gtf'\n",
    "df = pr.read_gtf(spike_gtf, duplicate_attr=True).as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a25b6abc-f4c7-4b37-b586-f920c9f08a2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ERCC-00002', 'ERCC-00003', 'ERCC-00004', 'ERCC-00009',\n",
       "       'ERCC-00012', 'ERCC-00013', 'ERCC-00014', 'ERCC-00016',\n",
       "       'ERCC-00017', 'ERCC-00019', 'ERCC-00022', 'ERCC-00024',\n",
       "       'ERCC-00025', 'ERCC-00028', 'ERCC-00031', 'ERCC-00033',\n",
       "       'ERCC-00034', 'ERCC-00035', 'ERCC-00039', 'ERCC-00040',\n",
       "       'ERCC-00041', 'ERCC-00042', 'ERCC-00043', 'ERCC-00044',\n",
       "       'ERCC-00046', 'ERCC-00048', 'ERCC-00051', 'ERCC-00053',\n",
       "       'ERCC-00054', 'ERCC-00057', 'ERCC-00058', 'ERCC-00059',\n",
       "       'ERCC-00060', 'ERCC-00061', 'ERCC-00062', 'ERCC-00067',\n",
       "       'ERCC-00069', 'ERCC-00071', 'ERCC-00073', 'ERCC-00074',\n",
       "       'ERCC-00075', 'ERCC-00076', 'ERCC-00077', 'ERCC-00078',\n",
       "       'ERCC-00079', 'ERCC-00081', 'ERCC-00083', 'ERCC-00084',\n",
       "       'ERCC-00085', 'ERCC-00086', 'ERCC-00092', 'ERCC-00095',\n",
       "       'ERCC-00096', 'ERCC-00097', 'ERCC-00098', 'ERCC-00099',\n",
       "       'ERCC-00104', 'ERCC-00108', 'ERCC-00109', 'ERCC-00111',\n",
       "       'ERCC-00112', 'ERCC-00113', 'ERCC-00116', 'ERCC-00117',\n",
       "       'ERCC-00120', 'ERCC-00123', 'ERCC-00126', 'ERCC-00130',\n",
       "       'ERCC-00131', 'ERCC-00134', 'ERCC-00136', 'ERCC-00137',\n",
       "       'ERCC-00138', 'ERCC-00142', 'ERCC-00143', 'ERCC-00144',\n",
       "       'ERCC-00145', 'ERCC-00147', 'ERCC-00148', 'ERCC-00150',\n",
       "       'ERCC-00154', 'ERCC-00156', 'ERCC-00157', 'ERCC-00158',\n",
       "       'ERCC-00160', 'ERCC-00162', 'ERCC-00163', 'ERCC-00164',\n",
       "       'ERCC-00165', 'ERCC-00168', 'ERCC-00170', 'ERCC-00171', 'SIRV1',\n",
       "       'SIRV2', 'SIRV3', 'SIRV4', 'SIRV5', 'SIRV6', 'SIRV7', 'SIRV4001',\n",
       "       'SIRV4002', 'SIRV4003', 'SIRV6001', 'SIRV6002', 'SIRV6003',\n",
       "       'SIRV8001', 'SIRV8002', 'SIRV8003', 'SIRV10001', 'SIRV10002',\n",
       "       'SIRV10003', 'SIRV12001', 'SIRV12002', 'SIRV12003'], dtype=object)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.loc[(df.Chromosome.str.contains('ERCC'))|(df.Chromosome.str.contains('SIRV'))]\n",
    "df.Chromosome = df.Chromosome.astype('str')\n",
    "df.Chromosome.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "98153dca-8b14-4f88-af47-d30139684d9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SIRV1', 'SIRV2', 'SIRV3', 'SIRV4', 'SIRV5', 'SIRV6', 'SIRV7',\n",
       "       'SIRV4001', 'SIRV4002', 'SIRV4003', 'SIRV6001', 'SIRV6002',\n",
       "       'SIRV6003', 'SIRV8001', 'SIRV8002', 'SIRV8003', 'SIRV10001',\n",
       "       'SIRV10002', 'SIRV10003', 'SIRV12001', 'SIRV12002', 'SIRV12003'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Chromosome.str.contains('SIRV'),'Chromosome'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5d139d85-b186-4f05-8cdd-306d0b0efb1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ERCC-00002', 'ERCC-00003', 'ERCC-00004', 'ERCC-00009',\n",
       "       'ERCC-00012', 'ERCC-00013', 'ERCC-00014', 'ERCC-00016',\n",
       "       'ERCC-00017', 'ERCC-00019', 'ERCC-00022', 'ERCC-00024',\n",
       "       'ERCC-00025', 'ERCC-00028', 'ERCC-00031', 'ERCC-00033',\n",
       "       'ERCC-00034', 'ERCC-00035', 'ERCC-00039', 'ERCC-00040',\n",
       "       'ERCC-00041', 'ERCC-00042', 'ERCC-00043', 'ERCC-00044',\n",
       "       'ERCC-00046', 'ERCC-00048', 'ERCC-00051', 'ERCC-00053',\n",
       "       'ERCC-00054', 'ERCC-00057', 'ERCC-00058', 'ERCC-00059',\n",
       "       'ERCC-00060', 'ERCC-00061', 'ERCC-00062', 'ERCC-00067',\n",
       "       'ERCC-00069', 'ERCC-00071', 'ERCC-00073', 'ERCC-00074',\n",
       "       'ERCC-00075', 'ERCC-00076', 'ERCC-00077', 'ERCC-00078',\n",
       "       'ERCC-00079', 'ERCC-00081', 'ERCC-00083', 'ERCC-00084',\n",
       "       'ERCC-00085', 'ERCC-00086', 'ERCC-00092', 'ERCC-00095',\n",
       "       'ERCC-00096', 'ERCC-00097', 'ERCC-00098', 'ERCC-00099',\n",
       "       'ERCC-00104', 'ERCC-00108', 'ERCC-00109', 'ERCC-00111',\n",
       "       'ERCC-00112', 'ERCC-00113', 'ERCC-00116', 'ERCC-00117',\n",
       "       'ERCC-00120', 'ERCC-00123', 'ERCC-00126', 'ERCC-00130',\n",
       "       'ERCC-00131', 'ERCC-00134', 'ERCC-00136', 'ERCC-00137',\n",
       "       'ERCC-00138', 'ERCC-00142', 'ERCC-00143', 'ERCC-00144',\n",
       "       'ERCC-00145', 'ERCC-00147', 'ERCC-00148', 'ERCC-00150',\n",
       "       'ERCC-00154', 'ERCC-00156', 'ERCC-00157', 'ERCC-00158',\n",
       "       'ERCC-00160', 'ERCC-00162', 'ERCC-00163', 'ERCC-00164',\n",
       "       'ERCC-00165', 'ERCC-00168', 'ERCC-00170', 'ERCC-00171'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Chromosome.str.contains('ERCC'),'Chromosome'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d9f5ccd5-1531-4cdb-8d73-8020a0d7522d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>exon_assignment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>ERCC</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>1045</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ERCC-00002A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERCC_spike_in</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source Feature  Start   End Score Strand Frame      gene_id  \\\n",
       "0  ERCC-00002   ERCC    exon      0  1045     0      +     .  ERCC-00002A   \n",
       "\n",
       "  gene_type  ... transcript_support_level  tag havana_transcript exon_number  \\\n",
       "0       NaN  ...                      NaN  NaN               NaN         NaN   \n",
       "\n",
       "  exon_id  ont protein_id ccdsid   gene_biotype exon_assignment  \n",
       "0     NaN  NaN        NaN    NaN  ERCC_spike_in             NaN  \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Chromosome == 'ERCC-00002']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8b8d5f93-ff25-44e9-81a5-279368ceb2a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>SIRV3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>SIRV4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>SIRV5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>SIRV6</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome  n_genes\n",
       "92       SIRV1        2\n",
       "99       SIRV2        3\n",
       "100      SIRV3        4\n",
       "101      SIRV4        3\n",
       "105      SIRV5        2\n",
       "106      SIRV6        3"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df[['Chromosome', 'gene_id']].drop_duplicates()\n",
    "temp = temp.groupby('Chromosome').nunique().reset_index().rename({'gene_id':'n_genes'}, axis=1)\n",
    "temp = temp.loc[temp.n_genes>1]\n",
    "# so sirvs 1-6 have more than one \"gene\" per chromosome, but where are they located? are they overlapping\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5329a4eb-97f4-4a63-bc69-50b9e6952684",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chromosome SIRV1\n",
      "Gene SIRV1B min coord: 10582\n",
      "Gene SIRV1B max coord: 11643\n",
      "Gene SIRV1A min coord: 1000\n",
      "Gene SIRV1A max coord: 11643\n",
      "\n",
      "Chromosome SIRV2\n",
      "Gene SIRV2B min coord: 1108\n",
      "Gene SIRV2B max coord: 1631\n",
      "Gene SIRV2C min coord: 4033\n",
      "Gene SIRV2C max coord: 4457\n",
      "Gene SIRV2A min coord: 1000\n",
      "Gene SIRV2A max coord: 5911\n",
      "\n",
      "Chromosome SIRV3\n",
      "Gene SIRV3A min coord: 1944\n",
      "Gene SIRV3A max coord: 8939\n",
      "Gene SIRV3B min coord: 1000\n",
      "Gene SIRV3B max coord: 1982\n",
      "Gene SIRV3C min coord: 8759\n",
      "Gene SIRV3C max coord: 9943\n",
      "Gene SIRV3D min coord: 4601\n",
      "Gene SIRV3D max coord: 4762\n",
      "\n",
      "Chromosome SIRV4\n",
      "Gene SIRV4C min coord: 1000\n",
      "Gene SIRV4C max coord: 3403\n",
      "Gene SIRV4A min coord: 8322\n",
      "Gene SIRV4A max coord: 15122\n",
      "Gene SIRV4B min coord: 3637\n",
      "Gene SIRV4B max coord: 5158\n",
      "\n",
      "Chromosome SIRV5\n",
      "Gene SIRV5A min coord: 1000\n",
      "Gene SIRV5A max coord: 13606\n",
      "Gene SIRV5B min coord: 2177\n",
      "Gene SIRV5B max coord: 2406\n",
      "\n",
      "Chromosome SIRV6\n",
      "Gene SIRV6A min coord: 1000\n",
      "Gene SIRV6A max coord: 11837\n",
      "Gene SIRV6B min coord: 1544\n",
      "Gene SIRV6B max coord: 1820\n",
      "Gene SIRV6C min coord: 2358\n",
      "Gene SIRV6C max coord: 2547\n"
     ]
    }
   ],
   "source": [
    "for c in temp.Chromosome.unique():\n",
    "    temp2 = df.loc[df.Chromosome == c]\n",
    "    print()\n",
    "    print(f'Chromosome {c}')\n",
    "    for g in temp2.gene_id.unique():\n",
    "        max_coord = df.loc[df.gene_id == g][['Start', 'End']].max().max()\n",
    "        min_coord = df.loc[df.gene_id == g][['Start', 'End']].min().min()\n",
    "        print(f'Gene {g} min coord: {min_coord}')\n",
    "        print(f'Gene {g} max coord: {max_coord}')\n",
    "        \n",
    "# they do overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe2b6f54-33db-40e5-ade8-1f28c6c0c399",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cf9512c-bb81-4804-967b-ff6bdf0f8857",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cfd1a800-e38d-4397-8418-b5c5dc752b81",
   "metadata": {},
   "source": [
    "## 230720 are SIRVs / ERCCs in our unfiltered matrix?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6cf31f8c-b0b5-4ab8-89f7-e58a359e09a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(talon_ab, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "37c2133d-672f-4008-bebf-2ab3310554b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>calu3_1_2</th>\n",
       "      <th>brodmann_area_46_1_1</th>\n",
       "      <th>brodmann_area_46_7_1</th>\n",
       "      <th>brodmann_area_46_5_1</th>\n",
       "      <th>brodmann_area_46_6_1</th>\n",
       "      <th>brodmann_area_46_8_1</th>\n",
       "      <th>brodmann_area_46_4_1</th>\n",
       "      <th>brodmann_area_46_2_1</th>\n",
       "      <th>brodmann_area_46_9_1</th>\n",
       "      <th>brodmann_area_46_3_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>2</td>\n",
       "      <td>535</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A-201</td>\n",
       "      <td>3</td>\n",
       "      <td>1187</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A-202</td>\n",
       "      <td>2</td>\n",
       "      <td>590</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000453576.2</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>AL627309.1-204</td>\n",
       "      <td>2</td>\n",
       "      <td>336</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID  transcript_ID      annot_gene_id annot_transcript_id  \\\n",
       "0        3              4  ENSG00000278267.1   ENST00000619216.1   \n",
       "1        4              6  ENSG00000243485.5   ENST00000469289.1   \n",
       "2        6              8  ENSG00000237613.2   ENST00000417324.1   \n",
       "3        6              9  ENSG00000237613.2   ENST00000461467.1   \n",
       "4       10             19  ENSG00000238009.6   ENST00000453576.2   \n",
       "\n",
       "  annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "0       MIR6859-1         MIR6859-1-201        1      68        Known   \n",
       "1     MIR1302-2HG       MIR1302-2HG-201        2     535        Known   \n",
       "2         FAM138A           FAM138A-201        3    1187        Known   \n",
       "3         FAM138A           FAM138A-202        2     590        Known   \n",
       "4      AL627309.1        AL627309.1-204        2     336        Known   \n",
       "\n",
       "  transcript_novelty  ... calu3_1_2  brodmann_area_46_1_1  \\\n",
       "0              Known  ...         0                     0   \n",
       "1              Known  ...         0                     0   \n",
       "2              Known  ...         0                     0   \n",
       "3              Known  ...         0                     0   \n",
       "4              Known  ...         0                     0   \n",
       "\n",
       "   brodmann_area_46_7_1  brodmann_area_46_5_1  brodmann_area_46_6_1  \\\n",
       "0                     0                     0                     0   \n",
       "1                     0                     0                     0   \n",
       "2                     0                     0                     0   \n",
       "3                     0                     0                     0   \n",
       "4                     0                     0                     0   \n",
       "\n",
       "   brodmann_area_46_8_1  brodmann_area_46_4_1  brodmann_area_46_2_1  \\\n",
       "0                     0                     0                     0   \n",
       "1                     0                     0                     0   \n",
       "2                     0                     0                     0   \n",
       "3                     0                     0                     0   \n",
       "4                     0                     0                     0   \n",
       "\n",
       "   brodmann_area_46_9_1  brodmann_area_46_3_1  \n",
       "0                     0                     0  \n",
       "1                     0                     0  \n",
       "2                     0                     0  \n",
       "3                     0                     0  \n",
       "4                     0                     0  \n",
       "\n",
       "[5 rows x 149 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "544649b7-54d4-4c2b-a369-5fbc61096b07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>exon_assignment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>1000</td>\n",
       "      <td>1484</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6337</td>\n",
       "      <td>6473</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6560</td>\n",
       "      <td>6813</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7814</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10282</td>\n",
       "      <td>10366</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10444</td>\n",
       "      <td>10786</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>1006</td>\n",
       "      <td>1484</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV102_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6337</td>\n",
       "      <td>6813</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV102_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7814</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV102_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10282</td>\n",
       "      <td>10366</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV102_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>1000</td>\n",
       "      <td>1484</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6337</td>\n",
       "      <td>6473</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6560</td>\n",
       "      <td>6813</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7814</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10282</td>\n",
       "      <td>10366</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6449</td>\n",
       "      <td>6473</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6560</td>\n",
       "      <td>6813</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7814</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10282</td>\n",
       "      <td>10366</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10593</td>\n",
       "      <td>10640</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>1000</td>\n",
       "      <td>1484</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV106_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7808</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV106_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10553</td>\n",
       "      <td>10786</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV106_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV107_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10882</td>\n",
       "      <td>11242</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV107_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>11403</td>\n",
       "      <td>11643</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV107_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome           Source Feature  Start    End Score Strand Frame  \\\n",
       "98       SIRV1  LexogenSIRVData    exon   1000   1484     0      -     .   \n",
       "99       SIRV1  LexogenSIRVData    exon   6337   6473     0      -     .   \n",
       "100      SIRV1  LexogenSIRVData    exon   6560   6813     0      -     .   \n",
       "101      SIRV1  LexogenSIRVData    exon   7552   7814     0      -     .   \n",
       "102      SIRV1  LexogenSIRVData    exon  10282  10366     0      -     .   \n",
       "103      SIRV1  LexogenSIRVData    exon  10444  10786     0      -     .   \n",
       "104      SIRV1  LexogenSIRVData    exon   1006   1484     0      -     .   \n",
       "105      SIRV1  LexogenSIRVData    exon   6337   6813     0      -     .   \n",
       "106      SIRV1  LexogenSIRVData    exon   7552   7814     0      -     .   \n",
       "107      SIRV1  LexogenSIRVData    exon  10282  10366     0      -     .   \n",
       "108      SIRV1  LexogenSIRVData    exon   1000   1484     0      -     .   \n",
       "109      SIRV1  LexogenSIRVData    exon   6337   6473     0      -     .   \n",
       "110      SIRV1  LexogenSIRVData    exon   6560   6813     0      -     .   \n",
       "111      SIRV1  LexogenSIRVData    exon   7552   7814     0      -     .   \n",
       "112      SIRV1  LexogenSIRVData    exon  10282  10366     0      -     .   \n",
       "113      SIRV1  LexogenSIRVData    exon  10647  10791     0      -     .   \n",
       "114      SIRV1  LexogenSIRVData    exon   6449   6473     0      -     .   \n",
       "115      SIRV1  LexogenSIRVData    exon   6560   6813     0      -     .   \n",
       "116      SIRV1  LexogenSIRVData    exon   7552   7814     0      -     .   \n",
       "117      SIRV1  LexogenSIRVData    exon  10282  10366     0      -     .   \n",
       "118      SIRV1  LexogenSIRVData    exon  10593  10640     0      -     .   \n",
       "119      SIRV1  LexogenSIRVData    exon   1000   1484     0      -     .   \n",
       "120      SIRV1  LexogenSIRVData    exon   7552   7808     0      -     .   \n",
       "121      SIRV1  LexogenSIRVData    exon  10553  10786     0      -     .   \n",
       "122      SIRV1  LexogenSIRVData    exon  10647  10791     0      -     .   \n",
       "123      SIRV1  LexogenSIRVData    exon  10882  11242     0      -     .   \n",
       "124      SIRV1  LexogenSIRVData    exon  11403  11643     0      -     .   \n",
       "\n",
       "    gene_id gene_type  ... transcript_support_level  tag havana_transcript  \\\n",
       "98   SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "99   SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "100  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "101  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "102  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "103  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "104  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "105  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "106  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "107  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "108  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "109  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "110  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "111  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "112  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "113  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "114  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "115  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "116  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "117  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "118  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "119  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "120  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "121  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "122  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "123  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "124  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "\n",
       "    exon_number exon_id  ont protein_id ccdsid gene_biotype exon_assignment  \n",
       "98          NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_0  \n",
       "99          NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_1  \n",
       "100         NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_2  \n",
       "101         NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_3  \n",
       "102         NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_4  \n",
       "103         NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_5  \n",
       "104         NaN     NaN  NaN        NaN    NaN          NaN       SIRV102_0  \n",
       "105         NaN     NaN  NaN        NaN    NaN          NaN       SIRV102_1  \n",
       "106         NaN     NaN  NaN        NaN    NaN          NaN       SIRV102_2  \n",
       "107         NaN     NaN  NaN        NaN    NaN          NaN       SIRV102_3  \n",
       "108         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_0  \n",
       "109         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_1  \n",
       "110         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_2  \n",
       "111         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_3  \n",
       "112         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_4  \n",
       "113         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_5  \n",
       "114         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_0  \n",
       "115         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_1  \n",
       "116         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_2  \n",
       "117         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_3  \n",
       "118         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_4  \n",
       "119         NaN     NaN  NaN        NaN    NaN          NaN       SIRV106_0  \n",
       "120         NaN     NaN  NaN        NaN    NaN          NaN       SIRV106_1  \n",
       "121         NaN     NaN  NaN        NaN    NaN          NaN       SIRV106_2  \n",
       "122         NaN     NaN  NaN        NaN    NaN          NaN       SIRV107_0  \n",
       "123         NaN     NaN  NaN        NaN    NaN          NaN       SIRV107_1  \n",
       "124         NaN     NaN  NaN        NaN    NaN          NaN       SIRV107_2  \n",
       "\n",
       "[27 rows x 26 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.annot_gene_id.str.contains('SIRV')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e156d11-a858-4764-b0f5-4f0f11ad045e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
