{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4a6a7910-a490-43ff-8801-9900e4bc0530",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab3a9a29-a63c-448e-bc46-d1e37eb3da45",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../snakemake/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "455c11a8-9c97-4f6d-b4a1-71650f09105a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = '../'+expand(config['data']['ab'], species='human')[0]\n",
    "talon_ab = '../'+expand(config['data']['talon_filt_ab'], species='human')[0]\n",
    "lapa_ab = '../'+expand(config['data']['lapa_filt_ab'], species='human')[0]\n",
    "filt_ab = '../'+expand(config['data']['filt_ab'], species='human')[0]\n",
    "read_annot = '../'+expand(config['data']['read_annot'], species='human')[0]\n",
    "t_metadata = '../'+expand(config['ref']['cerberus']['t_info'], species='human')[0]\n",
    "lib_meta = '../'+expand(config['data']['meta'], species='human')[0]\n",
    "swan_file = '../'+expand(config['data']['sg'], species='human')[0]\n",
    "cerberus_h5 = '../'+expand(config['data']['cerb_annot'], species='human')[0]\n",
    "cerb_t_metadata = '../'+expand(config['data']['t_info'], species='human')[0]\n",
    "major_isos = '../'+expand(config['data']['major_isos'], species='human', obs_col='sample')[0]\n",
    "pp_summary = '../'+expand(config['data']['p_pred']['summary'], species='human')[0]\n",
    "ref_t_metadata = '../'+expand(config['ref']['t_info'], species='human')[0]\n",
    "ref_g_metadata = '../'+expand(config['ref']['g_info'], species='human')[0]\n",
    "\n",
    "spike_filt_ab = '../'+config['data']['spikes']['ca_ab']\n",
    "spike_ca = '../'+config['data']['spikes']['ca_annot']\n",
    "gene_spike_ab = '../'+config['data']['spikes']['gene_ab']\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "min_tpm = 1\n",
    "gene_subset = 'protein_coding'\n",
    "obs_col = 'sample'\n",
    "go_gene_subset = 'protein_coding'\n",
    "\n",
    "m_lib_meta = '../'+expand(config['data']['meta'], species='mouse')[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40fa0ef4-1393-4fe6-9305-20c93a39587d",
   "metadata": {},
   "source": [
    "## SIRV novel isoform detection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1f74b4b4-d935-4df9-8d62-8b1b85a7ccfc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating ic TPM values\n",
      "Enforcing minimum TPM\n",
      "Total # ics detected: 214\n",
      "# ics >= 1 tpm: 213\n",
      "Number of ics reported: 213\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/scripts/utils.py:2118: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[tpm_col] = (df[d]*1000000)/df[total_col]\n",
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/scripts/utils.py:2117: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[total_col] = df[d].sum()\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(spike_filt_ab, sep='\\t')\n",
    "df, ic_ids = get_tpm_table(df,\n",
    "                   how='ic',\n",
    "                   species='spikes',\n",
    "                   min_tpm=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "29e99310-ed49-4c67-b736-94aeec04a73f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add novelty info\n",
    "ca = cerberus.read(spike_ca)\n",
    "ic = ca.ic.copy(deep=True)\n",
    "ic = ic.loc[ic.Name.isin(ic_ids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a058c672-0080-4f44-9864-5550b4ca6b7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only the sirvs\n",
    "ic = ic.loc[ic.gene_id.str.contains('SIRV')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f46c9d4a-3abb-4e8a-a706-1ac4fd32634a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ISM</th>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Name\n",
       "novelty      \n",
       "ISM         5\n",
       "Known      79\n",
       "NIC         3\n",
       "NNC        38"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ic[['Name', 'novelty']].groupby('novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "39171e0a-ec5f-4d2f-aa84-85aae4f3cedf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Coordinates</th>\n",
       "      <th>Name</th>\n",
       "      <th>source</th>\n",
       "      <th>novelty</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>178</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>-</td>\n",
       "      <td>5788-4800-4687-4479-4338-4094-3963-3825-3665-3...</td>\n",
       "      <td>SIRV2_6</td>\n",
       "      <td>lapa</td>\n",
       "      <td>NNC</td>\n",
       "      <td>SIRV2</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>SIRV3</td>\n",
       "      <td>+</td>\n",
       "      <td>2005-4568-4779-6057-7986-8124-8207-8755</td>\n",
       "      <td>SIRV3_12</td>\n",
       "      <td>lapa</td>\n",
       "      <td>NNC</td>\n",
       "      <td>SIRV3</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>SIRV4</td>\n",
       "      <td>+</td>\n",
       "      <td>1343-1678-1885-2389</td>\n",
       "      <td>SIRV4_8</td>\n",
       "      <td>lapa</td>\n",
       "      <td>NNC</td>\n",
       "      <td>SIRV4</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>SIRV4</td>\n",
       "      <td>+</td>\n",
       "      <td>1346-1678-1885-2385</td>\n",
       "      <td>SIRV4_9</td>\n",
       "      <td>lapa</td>\n",
       "      <td>NNC</td>\n",
       "      <td>SIRV4</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>SIRV5</td>\n",
       "      <td>+</td>\n",
       "      <td>1087-3298-3404-3483-3643-5380-5450-5543-5626-6...</td>\n",
       "      <td>SIRV5_13</td>\n",
       "      <td>lapa</td>\n",
       "      <td>NNC</td>\n",
       "      <td>SIRV5</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome Strand                                        Coordinates  \\\n",
       "178      SIRV2      -  5788-4800-4687-4479-4338-4094-3963-3825-3665-3...   \n",
       "179      SIRV3      +            2005-4568-4779-6057-7986-8124-8207-8755   \n",
       "180      SIRV4      +                                1343-1678-1885-2389   \n",
       "181      SIRV4      +                                1346-1678-1885-2385   \n",
       "183      SIRV5      +  1087-3298-3404-3483-3643-5380-5450-5543-5626-6...   \n",
       "\n",
       "         Name source novelty gene_id  ic  \n",
       "178   SIRV2_6   lapa     NNC   SIRV2   6  \n",
       "179  SIRV3_12   lapa     NNC   SIRV3  12  \n",
       "180   SIRV4_8   lapa     NNC   SIRV4   8  \n",
       "181   SIRV4_9   lapa     NNC   SIRV4   9  \n",
       "183  SIRV5_13   lapa     NNC   SIRV5  13  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ic.loc[ic.novelty=='NNC'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fed5479c-0348-4476-b5e2-cbe449911ec1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b6b6b10-9764-416d-9e10-cbf666fea038",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45a13c30-1fd1-4bc0-9c80-58f45de8a107",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "b305b471-f5fd-4373-a0a9-0934a62114bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ISM_rescue</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    annot_transcript_id\n",
       "transcript_novelty                     \n",
       "ISM_rescue                           11\n",
       "Known                               255\n",
       "NIC                                  21\n",
       "NNC                                  14"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.transcript_novelty!='Known']\n",
    "temp = df.copy(deep=True)\n",
    "temp = temp[['annot_transcript_id', 'transcript_novelty']].groupby('transcript_novelty').count()\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "e2d06060-f7fc-438f-93d4-d0c9b5c74f15",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "264\n",
      "226\n",
      "Enforcing minimum TPM\n",
      "Total # isoformss detected: 467\n",
      "# isoformss >= 1 tpm: 446\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/miniconda3/envs/base_clone/lib/python3.7/site-packages/ipykernel_launcher.py:28: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "/Users/fairliereese/miniconda3/envs/base_clone/lib/python3.7/site-packages/ipykernel_launcher.py:27: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n"
     ]
    }
   ],
   "source": [
    "meta_df = pd.DataFrame()\n",
    "meta_df['dataset'] = get_datasets('human')\n",
    "meta_df['species'] = 'human'\n",
    "temp = pd.DataFrame()\n",
    "temp['dataset'] = get_datasets('mouse')\n",
    "temp['species'] = 'mouse'\n",
    "meta_df = pd.concat([meta_df, temp], axis=0)\n",
    "meta_df['dataset'] = meta_df['dataset']+'_labeled_'+meta_df['species']\n",
    "meta_df.head()\n",
    "# # ?dataset_cols = [d+'_labeled_human' for d in get_datasets('human')]\n",
    "# # dataset_cols += [d+'_labeled_mouse' for d in get_datasets('mouse')]\n",
    "dataset_cols = meta_df.dataset.tolist()\n",
    "print(len(dataset_cols))\n",
    "dataset_cols = list(set(dataset_cols)&set(df.columns.tolist()))\n",
    "print(len(dataset_cols))\n",
    "\n",
    "id_col = 'annot_transcript_id'\n",
    "\n",
    "df = df[dataset_cols+[id_col]]\n",
    "df.set_index(id_col, inplace=True)\n",
    "\n",
    "# compute TPM\n",
    "tpm_cols = []\n",
    "for d in dataset_cols:\n",
    "    tpm_col = '{}_tpm'.format(d)\n",
    "    total_col = '{}_total'.format(d)\n",
    "    df[total_col] = df[d].sum()\n",
    "    df[tpm_col] = (df[d]*1000000)/df[total_col]\n",
    "    tpm_cols.append(tpm_col)\n",
    "df = df[tpm_cols]\n",
    "\n",
    "# enforce tpm threshold\n",
    "if min_tpm:\n",
    "    print('Enforcing minimum TPM')\n",
    "    print('Total # {}s detected: {}'.format('isoforms', len(df.index)))\n",
    "    df = df.loc[(df >= min_tpm).any(axis=1)]\n",
    "    print('# {}s >= {} tpm: {}'.format('isoforms', min_tpm, len(df.index)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "8424559f-f8e4-4fe8-8db3-6b7d8225e8ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "866e9ac6-15aa-4f2f-a23c-e1ebdd2e2811",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = pd.read_csv(spike_filt_ab, sep='\\t')\n",
    "ab = ab[['annot_transcript_id', 'transcript_novelty', 'annot_gene_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "bcf73e32-8b5d-4169-b88e-99578a84bcd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.merge(ab, how='left', left_index=True, right_on='annot_transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "567c1263-8f32-4506-8970-27a640e4dfab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "462    gSpikein_ERCC-00165\n",
       "463    gSpikein_ERCC-00168\n",
       "464    gSpikein_ERCC-00170\n",
       "465    gSpikein_ERCC-00170\n",
       "466    gSpikein_ERCC-00171\n",
       "Name: annot_gene_id, dtype: object"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.annot_gene_id.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "d7c53e8f-f09b-4180-be5b-7601641c0a35",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ISM_rescue</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    annot_transcript_id\n",
       "transcript_novelty                     \n",
       "ISM_rescue                           12\n",
       "Known                               242\n",
       "NIC                                  31\n",
       "NNC                                  20"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# temp = df.loc[df.transcript_novelty!='Known']\n",
    "temp = df.copy(deep=True)\n",
    "temp = temp.loc[temp.annot_gene_id.str.contains('SIRV')]\n",
    "temp\n",
    "temp = temp[['annot_transcript_id', 'transcript_novelty']].groupby('transcript_novelty').count()\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "8a322926-c5ad-409d-ba81-20b812a7f9be",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['ic_id'] = df.annot_transcript_id.str.split('#', n=1, expand=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "0cbf8d17-e6e1-4678-b47e-5e68ea9a549e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ISM_rescue</th>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ic_id\n",
       "transcript_novelty       \n",
       "ISM_rescue             12\n",
       "Known                  82\n",
       "NIC                    21\n",
       "NNC                    14"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# temp = df.loc[df.transcript_novelty!='Known']\n",
    "temp = df.copy(deep=True)\n",
    "temp = temp.loc[temp.annot_gene_id.str.contains('SIRV')]\n",
    "temp\n",
    "temp = temp[['ic_id', 'transcript_novelty']].groupby('transcript_novelty').nunique()\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "9f2976b1-9dd1-4c4f-a663-c336e7d758b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>right_cardiac_atrium_3_1_labeled_human_tpm</th>\n",
       "      <th>gastroc_4d_f_2_labeled_mouse_tpm</th>\n",
       "      <th>mesenteric_fat_pad_1_1_labeled_human_tpm</th>\n",
       "      <th>psoas_muscle_2_1_labeled_human_tpm</th>\n",
       "      <th>adrenal_4d_m_1_labeled_mouse_tpm</th>\n",
       "      <th>adrenal_25d_f_2_labeled_mouse_tpm</th>\n",
       "      <th>hl60_m1_12hr_1_2_labeled_human_tpm</th>\n",
       "      <th>hl60_m2_24hr_1_2_labeled_human_tpm</th>\n",
       "      <th>hippocampus_5x_m_1_2_labeled_mouse_tpm</th>\n",
       "      <th>k562_3_2_labeled_human_tpm</th>\n",
       "      <th>...</th>\n",
       "      <th>brodmann_area_46_2_1_labeled_human_tpm</th>\n",
       "      <th>adrenal_4d_f_2_labeled_mouse_tpm</th>\n",
       "      <th>hippocampus_5x_f_1_1_labeled_mouse_tpm</th>\n",
       "      <th>brodmann_area_46_3_1_labeled_human_tpm</th>\n",
       "      <th>ovary_1_1_labeled_human_tpm</th>\n",
       "      <th>heart_right_ventricle_3_1_labeled_human_tpm</th>\n",
       "      <th>h9_panc_progen_1_1_labeled_human_tpm</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>annot_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>70921.985816</td>\n",
       "      <td>224322.781196</td>\n",
       "      <td>106743.271848</td>\n",
       "      <td>211453.744493</td>\n",
       "      <td>186820.169701</td>\n",
       "      <td>215034.373943</td>\n",
       "      <td>163531.669866</td>\n",
       "      <td>154126.213592</td>\n",
       "      <td>147679.324895</td>\n",
       "      <td>161666.666667</td>\n",
       "      <td>...</td>\n",
       "      <td>186400.318080</td>\n",
       "      <td>225219.102128</td>\n",
       "      <td>29268.292683</td>\n",
       "      <td>176291.598604</td>\n",
       "      <td>96418.732782</td>\n",
       "      <td>203551.912568</td>\n",
       "      <td>47945.205479</td>\n",
       "      <td>tSpikein_ERCC-00002</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>317.198503</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6456.602885</td>\n",
       "      <td>149.515942</td>\n",
       "      <td>1070.663812</td>\n",
       "      <td>383.877159</td>\n",
       "      <td>809.061489</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>7222.222222</td>\n",
       "      <td>...</td>\n",
       "      <td>65.538530</td>\n",
       "      <td>1681.273475</td>\n",
       "      <td>2439.024390</td>\n",
       "      <td>85.412596</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8587.041374</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>tSpikein_ERCC-00002#0</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>328</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.884420</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>169.052181</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>404.530744</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1111.111111</td>\n",
       "      <td>...</td>\n",
       "      <td>34.953883</td>\n",
       "      <td>107.315328</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>32.029724</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>97.580016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>tSpikein_ERCC-00003</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>329</th>\n",
       "      <td>4255.319149</td>\n",
       "      <td>7485.884667</td>\n",
       "      <td>5140.610826</td>\n",
       "      <td>5380.502404</td>\n",
       "      <td>6503.943483</td>\n",
       "      <td>6874.788685</td>\n",
       "      <td>2111.324376</td>\n",
       "      <td>2022.653722</td>\n",
       "      <td>7434.197308</td>\n",
       "      <td>6666.666667</td>\n",
       "      <td>...</td>\n",
       "      <td>8288.439440</td>\n",
       "      <td>5151.135754</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>9352.679286</td>\n",
       "      <td>2066.115702</td>\n",
       "      <td>4488.680718</td>\n",
       "      <td>4109.589041</td>\n",
       "      <td>tSpikein_ERCC-00003#0</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>330</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>33.628140</td>\n",
       "      <td>37.378986</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.369235</td>\n",
       "      <td>35.771776</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>10.676575</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>tSpikein_ERCC-00003#1</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>190.319102</td>\n",
       "      <td>302.388872</td>\n",
       "      <td>437.165820</td>\n",
       "      <td>224.273913</td>\n",
       "      <td>394.455088</td>\n",
       "      <td>191.938580</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>602.772755</td>\n",
       "      <td>555.555556</td>\n",
       "      <td>...</td>\n",
       "      <td>384.492710</td>\n",
       "      <td>357.717761</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>320.297236</td>\n",
       "      <td>688.705234</td>\n",
       "      <td>487.900078</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>tSpikein_ERCC-00165</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.369235</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>21.353149</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>tSpikein_ERCC-00168</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>13.107706</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>tSpikein_ERCC-00170</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>63.439701</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.884420</td>\n",
       "      <td>74.757971</td>\n",
       "      <td>112.701454</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>200.924252</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>139.815531</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>128.118894</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>97.580016</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>tSpikein_ERCC-00170#0</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>5673.758865</td>\n",
       "      <td>11165.387299</td>\n",
       "      <td>907.166616</td>\n",
       "      <td>15233.547432</td>\n",
       "      <td>17829.776100</td>\n",
       "      <td>24230.812577</td>\n",
       "      <td>8061.420345</td>\n",
       "      <td>9708.737864</td>\n",
       "      <td>2411.091019</td>\n",
       "      <td>17222.222222</td>\n",
       "      <td>...</td>\n",
       "      <td>16865.248413</td>\n",
       "      <td>14058.307995</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>18150.176697</td>\n",
       "      <td>2066.115702</td>\n",
       "      <td>17759.562842</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>tSpikein_ERCC-00171</td>\n",
       "      <td>Known</td>\n",
       "      <td>gSpikein_ERCC-00171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>141 rows Ã— 229 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     right_cardiac_atrium_3_1_labeled_human_tpm  \\\n",
       "326                                70921.985816   \n",
       "327                                    0.000000   \n",
       "328                                    0.000000   \n",
       "329                                 4255.319149   \n",
       "330                                    0.000000   \n",
       "..                                          ...   \n",
       "462                                    0.000000   \n",
       "463                                    0.000000   \n",
       "464                                    0.000000   \n",
       "465                                    0.000000   \n",
       "466                                 5673.758865   \n",
       "\n",
       "     gastroc_4d_f_2_labeled_mouse_tpm  \\\n",
       "326                     224322.781196   \n",
       "327                        317.198503   \n",
       "328                          0.000000   \n",
       "329                       7485.884667   \n",
       "330                          0.000000   \n",
       "..                                ...   \n",
       "462                        190.319102   \n",
       "463                          0.000000   \n",
       "464                          0.000000   \n",
       "465                         63.439701   \n",
       "466                      11165.387299   \n",
       "\n",
       "     mesenteric_fat_pad_1_1_labeled_human_tpm  \\\n",
       "326                             106743.271848   \n",
       "327                                  0.000000   \n",
       "328                                  0.000000   \n",
       "329                               5140.610826   \n",
       "330                                  0.000000   \n",
       "..                                        ...   \n",
       "462                                302.388872   \n",
       "463                                  0.000000   \n",
       "464                                  0.000000   \n",
       "465                                  0.000000   \n",
       "466                                907.166616   \n",
       "\n",
       "     psoas_muscle_2_1_labeled_human_tpm  adrenal_4d_m_1_labeled_mouse_tpm  \\\n",
       "326                       211453.744493                     186820.169701   \n",
       "327                         6456.602885                        149.515942   \n",
       "328                          100.884420                          0.000000   \n",
       "329                         5380.502404                       6503.943483   \n",
       "330                           33.628140                         37.378986   \n",
       "..                                  ...                               ...   \n",
       "462                          437.165820                        224.273913   \n",
       "463                            0.000000                          0.000000   \n",
       "464                            0.000000                          0.000000   \n",
       "465                          100.884420                         74.757971   \n",
       "466                        15233.547432                      17829.776100   \n",
       "\n",
       "     adrenal_25d_f_2_labeled_mouse_tpm  hl60_m1_12hr_1_2_labeled_human_tpm  \\\n",
       "326                      215034.373943                       163531.669866   \n",
       "327                        1070.663812                          383.877159   \n",
       "328                         169.052181                            0.000000   \n",
       "329                        6874.788685                         2111.324376   \n",
       "330                           0.000000                            0.000000   \n",
       "..                                 ...                                 ...   \n",
       "462                         394.455088                          191.938580   \n",
       "463                           0.000000                            0.000000   \n",
       "464                           0.000000                            0.000000   \n",
       "465                         112.701454                            0.000000   \n",
       "466                       24230.812577                         8061.420345   \n",
       "\n",
       "     hl60_m2_24hr_1_2_labeled_human_tpm  \\\n",
       "326                       154126.213592   \n",
       "327                          809.061489   \n",
       "328                          404.530744   \n",
       "329                         2022.653722   \n",
       "330                            0.000000   \n",
       "..                                  ...   \n",
       "462                            0.000000   \n",
       "463                            0.000000   \n",
       "464                            0.000000   \n",
       "465                            0.000000   \n",
       "466                         9708.737864   \n",
       "\n",
       "     hippocampus_5x_m_1_2_labeled_mouse_tpm  k562_3_2_labeled_human_tpm  ...  \\\n",
       "326                           147679.324895               161666.666667  ...   \n",
       "327                                0.000000                 7222.222222  ...   \n",
       "328                                0.000000                 1111.111111  ...   \n",
       "329                             7434.197308                 6666.666667  ...   \n",
       "330                                0.000000                    0.000000  ...   \n",
       "..                                      ...                         ...  ...   \n",
       "462                              602.772755                  555.555556  ...   \n",
       "463                                0.000000                    0.000000  ...   \n",
       "464                                0.000000                    0.000000  ...   \n",
       "465                              200.924252                    0.000000  ...   \n",
       "466                             2411.091019                17222.222222  ...   \n",
       "\n",
       "     brodmann_area_46_2_1_labeled_human_tpm  adrenal_4d_f_2_labeled_mouse_tpm  \\\n",
       "326                           186400.318080                     225219.102128   \n",
       "327                               65.538530                       1681.273475   \n",
       "328                               34.953883                        107.315328   \n",
       "329                             8288.439440                       5151.135754   \n",
       "330                                4.369235                         35.771776   \n",
       "..                                      ...                               ...   \n",
       "462                              384.492710                        357.717761   \n",
       "463                                4.369235                          0.000000   \n",
       "464                               13.107706                          0.000000   \n",
       "465                              139.815531                          0.000000   \n",
       "466                            16865.248413                      14058.307995   \n",
       "\n",
       "     hippocampus_5x_f_1_1_labeled_mouse_tpm  \\\n",
       "326                            29268.292683   \n",
       "327                             2439.024390   \n",
       "328                                0.000000   \n",
       "329                                0.000000   \n",
       "330                                0.000000   \n",
       "..                                      ...   \n",
       "462                                0.000000   \n",
       "463                                0.000000   \n",
       "464                                0.000000   \n",
       "465                                0.000000   \n",
       "466                                0.000000   \n",
       "\n",
       "     brodmann_area_46_3_1_labeled_human_tpm  ovary_1_1_labeled_human_tpm  \\\n",
       "326                           176291.598604                 96418.732782   \n",
       "327                               85.412596                     0.000000   \n",
       "328                               32.029724                     0.000000   \n",
       "329                             9352.679286                  2066.115702   \n",
       "330                               10.676575                     0.000000   \n",
       "..                                      ...                          ...   \n",
       "462                              320.297236                   688.705234   \n",
       "463                               21.353149                     0.000000   \n",
       "464                                0.000000                     0.000000   \n",
       "465                              128.118894                     0.000000   \n",
       "466                            18150.176697                  2066.115702   \n",
       "\n",
       "     heart_right_ventricle_3_1_labeled_human_tpm  \\\n",
       "326                                203551.912568   \n",
       "327                                  8587.041374   \n",
       "328                                    97.580016   \n",
       "329                                  4488.680718   \n",
       "330                                     0.000000   \n",
       "..                                           ...   \n",
       "462                                   487.900078   \n",
       "463                                     0.000000   \n",
       "464                                     0.000000   \n",
       "465                                    97.580016   \n",
       "466                                 17759.562842   \n",
       "\n",
       "     h9_panc_progen_1_1_labeled_human_tpm    annot_transcript_id  \\\n",
       "326                          47945.205479    tSpikein_ERCC-00002   \n",
       "327                              0.000000  tSpikein_ERCC-00002#0   \n",
       "328                              0.000000    tSpikein_ERCC-00003   \n",
       "329                           4109.589041  tSpikein_ERCC-00003#0   \n",
       "330                              0.000000  tSpikein_ERCC-00003#1   \n",
       "..                                    ...                    ...   \n",
       "462                              0.000000    tSpikein_ERCC-00165   \n",
       "463                              0.000000    tSpikein_ERCC-00168   \n",
       "464                              0.000000    tSpikein_ERCC-00170   \n",
       "465                              0.000000  tSpikein_ERCC-00170#0   \n",
       "466                              0.000000    tSpikein_ERCC-00171   \n",
       "\n",
       "     transcript_novelty        annot_gene_id  \n",
       "326               Known  gSpikein_ERCC-00002  \n",
       "327               Known  gSpikein_ERCC-00002  \n",
       "328               Known  gSpikein_ERCC-00003  \n",
       "329               Known  gSpikein_ERCC-00003  \n",
       "330               Known  gSpikein_ERCC-00003  \n",
       "..                  ...                  ...  \n",
       "462               Known  gSpikein_ERCC-00165  \n",
       "463               Known  gSpikein_ERCC-00168  \n",
       "464               Known  gSpikein_ERCC-00170  \n",
       "465               Known  gSpikein_ERCC-00170  \n",
       "466               Known  gSpikein_ERCC-00171  \n",
       "\n",
       "[141 rows x 229 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.annot_transcript_id.str.contains('ERCC')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23d819f5-89ac-4abd-82ca-9d8f023f86c9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a0e0561d-e6a8-4585-a68f-4315565220f6",
   "metadata": {},
   "source": [
    "## 230720 ok so the sirv \"genes\" overlap, but how many genes do we call per sirv \"chromosome\"?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "01e0d269-d119-4a6a-ae2e-1bf83566d65f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(lapa_ab, sep='\\t')\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "291acdfb-9b84-4039-9cad-ecc32e1e8233",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Antisense</th>\n",
       "      <td>23316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ISM</th>\n",
       "      <td>80937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intergenic</th>\n",
       "      <td>1803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>193913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NIC</th>\n",
       "      <td>79168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NNC</th>\n",
       "      <td>23323</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    annot_transcript_id\n",
       "transcript_novelty                     \n",
       "Antisense                         23316\n",
       "ISM                               80937\n",
       "Intergenic                         1803\n",
       "Known                            193913\n",
       "NIC                               79168\n",
       "NNC                               23323"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['transcript_novelty', 'annot_transcript_id']].groupby('transcript_novelty').count()\n",
    "# these look like they've just been filtered for reproducibility and to remove genomic\n",
    "# figured out through breadcrumbs trails that these have indeed been filtered for reproducibility: \n",
    "# https://sandbox.zenodo.org/record/1095599"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7adc073d-8227-43a0-b9de-ed5115947ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 230720 ok so the sirv \"genes\" overlap, but what do they look like on the brwoser?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0b9f25e-206c-413b-a63e-12628f8a269a",
   "metadata": {},
   "source": [
    "## 230720 Is there only one sirv / ercc gene per \"chromosome\"? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "a68b4119-c0d2-4067-92c1-c8a3e6daf42b",
   "metadata": {},
   "outputs": [],
   "source": [
    "spike_gtf = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/rnawg/refs/gencode_v29_sirv4_ercc.gtf'\n",
    "df = pr.read_gtf(spike_gtf, duplicate_attr=True).as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a25b6abc-f4c7-4b37-b586-f920c9f08a2a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ERCC-00002', 'ERCC-00003', 'ERCC-00004', 'ERCC-00009',\n",
       "       'ERCC-00012', 'ERCC-00013', 'ERCC-00014', 'ERCC-00016',\n",
       "       'ERCC-00017', 'ERCC-00019', 'ERCC-00022', 'ERCC-00024',\n",
       "       'ERCC-00025', 'ERCC-00028', 'ERCC-00031', 'ERCC-00033',\n",
       "       'ERCC-00034', 'ERCC-00035', 'ERCC-00039', 'ERCC-00040',\n",
       "       'ERCC-00041', 'ERCC-00042', 'ERCC-00043', 'ERCC-00044',\n",
       "       'ERCC-00046', 'ERCC-00048', 'ERCC-00051', 'ERCC-00053',\n",
       "       'ERCC-00054', 'ERCC-00057', 'ERCC-00058', 'ERCC-00059',\n",
       "       'ERCC-00060', 'ERCC-00061', 'ERCC-00062', 'ERCC-00067',\n",
       "       'ERCC-00069', 'ERCC-00071', 'ERCC-00073', 'ERCC-00074',\n",
       "       'ERCC-00075', 'ERCC-00076', 'ERCC-00077', 'ERCC-00078',\n",
       "       'ERCC-00079', 'ERCC-00081', 'ERCC-00083', 'ERCC-00084',\n",
       "       'ERCC-00085', 'ERCC-00086', 'ERCC-00092', 'ERCC-00095',\n",
       "       'ERCC-00096', 'ERCC-00097', 'ERCC-00098', 'ERCC-00099',\n",
       "       'ERCC-00104', 'ERCC-00108', 'ERCC-00109', 'ERCC-00111',\n",
       "       'ERCC-00112', 'ERCC-00113', 'ERCC-00116', 'ERCC-00117',\n",
       "       'ERCC-00120', 'ERCC-00123', 'ERCC-00126', 'ERCC-00130',\n",
       "       'ERCC-00131', 'ERCC-00134', 'ERCC-00136', 'ERCC-00137',\n",
       "       'ERCC-00138', 'ERCC-00142', 'ERCC-00143', 'ERCC-00144',\n",
       "       'ERCC-00145', 'ERCC-00147', 'ERCC-00148', 'ERCC-00150',\n",
       "       'ERCC-00154', 'ERCC-00156', 'ERCC-00157', 'ERCC-00158',\n",
       "       'ERCC-00160', 'ERCC-00162', 'ERCC-00163', 'ERCC-00164',\n",
       "       'ERCC-00165', 'ERCC-00168', 'ERCC-00170', 'ERCC-00171', 'SIRV1',\n",
       "       'SIRV2', 'SIRV3', 'SIRV4', 'SIRV5', 'SIRV6', 'SIRV7', 'SIRV4001',\n",
       "       'SIRV4002', 'SIRV4003', 'SIRV6001', 'SIRV6002', 'SIRV6003',\n",
       "       'SIRV8001', 'SIRV8002', 'SIRV8003', 'SIRV10001', 'SIRV10002',\n",
       "       'SIRV10003', 'SIRV12001', 'SIRV12002', 'SIRV12003'], dtype=object)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.loc[(df.Chromosome.str.contains('ERCC'))|(df.Chromosome.str.contains('SIRV'))]\n",
    "df.Chromosome = df.Chromosome.astype('str')\n",
    "df.Chromosome.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "98153dca-8b14-4f88-af47-d30139684d9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['SIRV1', 'SIRV2', 'SIRV3', 'SIRV4', 'SIRV5', 'SIRV6', 'SIRV7',\n",
       "       'SIRV4001', 'SIRV4002', 'SIRV4003', 'SIRV6001', 'SIRV6002',\n",
       "       'SIRV6003', 'SIRV8001', 'SIRV8002', 'SIRV8003', 'SIRV10001',\n",
       "       'SIRV10002', 'SIRV10003', 'SIRV12001', 'SIRV12002', 'SIRV12003'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Chromosome.str.contains('SIRV'),'Chromosome'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "5d139d85-b186-4f05-8cdd-306d0b0efb1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ERCC-00002', 'ERCC-00003', 'ERCC-00004', 'ERCC-00009',\n",
       "       'ERCC-00012', 'ERCC-00013', 'ERCC-00014', 'ERCC-00016',\n",
       "       'ERCC-00017', 'ERCC-00019', 'ERCC-00022', 'ERCC-00024',\n",
       "       'ERCC-00025', 'ERCC-00028', 'ERCC-00031', 'ERCC-00033',\n",
       "       'ERCC-00034', 'ERCC-00035', 'ERCC-00039', 'ERCC-00040',\n",
       "       'ERCC-00041', 'ERCC-00042', 'ERCC-00043', 'ERCC-00044',\n",
       "       'ERCC-00046', 'ERCC-00048', 'ERCC-00051', 'ERCC-00053',\n",
       "       'ERCC-00054', 'ERCC-00057', 'ERCC-00058', 'ERCC-00059',\n",
       "       'ERCC-00060', 'ERCC-00061', 'ERCC-00062', 'ERCC-00067',\n",
       "       'ERCC-00069', 'ERCC-00071', 'ERCC-00073', 'ERCC-00074',\n",
       "       'ERCC-00075', 'ERCC-00076', 'ERCC-00077', 'ERCC-00078',\n",
       "       'ERCC-00079', 'ERCC-00081', 'ERCC-00083', 'ERCC-00084',\n",
       "       'ERCC-00085', 'ERCC-00086', 'ERCC-00092', 'ERCC-00095',\n",
       "       'ERCC-00096', 'ERCC-00097', 'ERCC-00098', 'ERCC-00099',\n",
       "       'ERCC-00104', 'ERCC-00108', 'ERCC-00109', 'ERCC-00111',\n",
       "       'ERCC-00112', 'ERCC-00113', 'ERCC-00116', 'ERCC-00117',\n",
       "       'ERCC-00120', 'ERCC-00123', 'ERCC-00126', 'ERCC-00130',\n",
       "       'ERCC-00131', 'ERCC-00134', 'ERCC-00136', 'ERCC-00137',\n",
       "       'ERCC-00138', 'ERCC-00142', 'ERCC-00143', 'ERCC-00144',\n",
       "       'ERCC-00145', 'ERCC-00147', 'ERCC-00148', 'ERCC-00150',\n",
       "       'ERCC-00154', 'ERCC-00156', 'ERCC-00157', 'ERCC-00158',\n",
       "       'ERCC-00160', 'ERCC-00162', 'ERCC-00163', 'ERCC-00164',\n",
       "       'ERCC-00165', 'ERCC-00168', 'ERCC-00170', 'ERCC-00171'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Chromosome.str.contains('ERCC'),'Chromosome'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d9f5ccd5-1531-4cdb-8d73-8020a0d7522d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>exon_assignment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>ERCC</td>\n",
       "      <td>exon</td>\n",
       "      <td>0</td>\n",
       "      <td>1045</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ERCC-00002A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERCC_spike_in</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Source Feature  Start   End Score Strand Frame      gene_id  \\\n",
       "0  ERCC-00002   ERCC    exon      0  1045     0      +     .  ERCC-00002A   \n",
       "\n",
       "  gene_type  ... transcript_support_level  tag havana_transcript exon_number  \\\n",
       "0       NaN  ...                      NaN  NaN               NaN         NaN   \n",
       "\n",
       "  exon_id  ont protein_id ccdsid   gene_biotype exon_assignment  \n",
       "0     NaN  NaN        NaN    NaN  ERCC_spike_in             NaN  \n",
       "\n",
       "[1 rows x 26 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.Chromosome == 'ERCC-00002']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "8b8d5f93-ff25-44e9-81a5-279368ceb2a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SIRV2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>SIRV3</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>SIRV4</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>SIRV5</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>SIRV6</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome  n_genes\n",
       "92       SIRV1        2\n",
       "99       SIRV2        3\n",
       "100      SIRV3        4\n",
       "101      SIRV4        3\n",
       "105      SIRV5        2\n",
       "106      SIRV6        3"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df[['Chromosome', 'gene_id']].drop_duplicates()\n",
    "temp = temp.groupby('Chromosome').nunique().reset_index().rename({'gene_id':'n_genes'}, axis=1)\n",
    "temp = temp.loc[temp.n_genes>1]\n",
    "# so sirvs 1-6 have more than one \"gene\" per chromosome, but where are they located? are they overlapping\n",
    "temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5329a4eb-97f4-4a63-bc69-50b9e6952684",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Chromosome SIRV1\n",
      "Gene SIRV1B min coord: 10582\n",
      "Gene SIRV1B max coord: 11643\n",
      "Gene SIRV1A min coord: 1000\n",
      "Gene SIRV1A max coord: 11643\n",
      "\n",
      "Chromosome SIRV2\n",
      "Gene SIRV2B min coord: 1108\n",
      "Gene SIRV2B max coord: 1631\n",
      "Gene SIRV2C min coord: 4033\n",
      "Gene SIRV2C max coord: 4457\n",
      "Gene SIRV2A min coord: 1000\n",
      "Gene SIRV2A max coord: 5911\n",
      "\n",
      "Chromosome SIRV3\n",
      "Gene SIRV3A min coord: 1944\n",
      "Gene SIRV3A max coord: 8939\n",
      "Gene SIRV3B min coord: 1000\n",
      "Gene SIRV3B max coord: 1982\n",
      "Gene SIRV3C min coord: 8759\n",
      "Gene SIRV3C max coord: 9943\n",
      "Gene SIRV3D min coord: 4601\n",
      "Gene SIRV3D max coord: 4762\n",
      "\n",
      "Chromosome SIRV4\n",
      "Gene SIRV4C min coord: 1000\n",
      "Gene SIRV4C max coord: 3403\n",
      "Gene SIRV4A min coord: 8322\n",
      "Gene SIRV4A max coord: 15122\n",
      "Gene SIRV4B min coord: 3637\n",
      "Gene SIRV4B max coord: 5158\n",
      "\n",
      "Chromosome SIRV5\n",
      "Gene SIRV5A min coord: 1000\n",
      "Gene SIRV5A max coord: 13606\n",
      "Gene SIRV5B min coord: 2177\n",
      "Gene SIRV5B max coord: 2406\n",
      "\n",
      "Chromosome SIRV6\n",
      "Gene SIRV6A min coord: 1000\n",
      "Gene SIRV6A max coord: 11837\n",
      "Gene SIRV6B min coord: 1544\n",
      "Gene SIRV6B max coord: 1820\n",
      "Gene SIRV6C min coord: 2358\n",
      "Gene SIRV6C max coord: 2547\n"
     ]
    }
   ],
   "source": [
    "for c in temp.Chromosome.unique():\n",
    "    temp2 = df.loc[df.Chromosome == c]\n",
    "    print()\n",
    "    print(f'Chromosome {c}')\n",
    "    for g in temp2.gene_id.unique():\n",
    "        max_coord = df.loc[df.gene_id == g][['Start', 'End']].max().max()\n",
    "        min_coord = df.loc[df.gene_id == g][['Start', 'End']].min().min()\n",
    "        print(f'Gene {g} min coord: {min_coord}')\n",
    "        print(f'Gene {g} max coord: {max_coord}')\n",
    "        \n",
    "# they do overlap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe2b6f54-33db-40e5-ade8-1f28c6c0c399",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cf9512c-bb81-4804-967b-ff6bdf0f8857",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cfd1a800-e38d-4397-8418-b5c5dc752b81",
   "metadata": {},
   "source": [
    "## 230720 are SIRVs / ERCCs in our unfiltered matrix?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6cf31f8c-b0b5-4ab8-89f7-e58a359e09a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(talon_ab, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "37c2133d-672f-4008-bebf-2ab3310554b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>calu3_1_2</th>\n",
       "      <th>brodmann_area_46_1_1</th>\n",
       "      <th>brodmann_area_46_7_1</th>\n",
       "      <th>brodmann_area_46_5_1</th>\n",
       "      <th>brodmann_area_46_6_1</th>\n",
       "      <th>brodmann_area_46_8_1</th>\n",
       "      <th>brodmann_area_46_4_1</th>\n",
       "      <th>brodmann_area_46_2_1</th>\n",
       "      <th>brodmann_area_46_9_1</th>\n",
       "      <th>brodmann_area_46_3_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>2</td>\n",
       "      <td>535</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A-201</td>\n",
       "      <td>3</td>\n",
       "      <td>1187</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A-202</td>\n",
       "      <td>2</td>\n",
       "      <td>590</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000453576.2</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>AL627309.1-204</td>\n",
       "      <td>2</td>\n",
       "      <td>336</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID  transcript_ID      annot_gene_id annot_transcript_id  \\\n",
       "0        3              4  ENSG00000278267.1   ENST00000619216.1   \n",
       "1        4              6  ENSG00000243485.5   ENST00000469289.1   \n",
       "2        6              8  ENSG00000237613.2   ENST00000417324.1   \n",
       "3        6              9  ENSG00000237613.2   ENST00000461467.1   \n",
       "4       10             19  ENSG00000238009.6   ENST00000453576.2   \n",
       "\n",
       "  annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "0       MIR6859-1         MIR6859-1-201        1      68        Known   \n",
       "1     MIR1302-2HG       MIR1302-2HG-201        2     535        Known   \n",
       "2         FAM138A           FAM138A-201        3    1187        Known   \n",
       "3         FAM138A           FAM138A-202        2     590        Known   \n",
       "4      AL627309.1        AL627309.1-204        2     336        Known   \n",
       "\n",
       "  transcript_novelty  ... calu3_1_2  brodmann_area_46_1_1  \\\n",
       "0              Known  ...         0                     0   \n",
       "1              Known  ...         0                     0   \n",
       "2              Known  ...         0                     0   \n",
       "3              Known  ...         0                     0   \n",
       "4              Known  ...         0                     0   \n",
       "\n",
       "   brodmann_area_46_7_1  brodmann_area_46_5_1  brodmann_area_46_6_1  \\\n",
       "0                     0                     0                     0   \n",
       "1                     0                     0                     0   \n",
       "2                     0                     0                     0   \n",
       "3                     0                     0                     0   \n",
       "4                     0                     0                     0   \n",
       "\n",
       "   brodmann_area_46_8_1  brodmann_area_46_4_1  brodmann_area_46_2_1  \\\n",
       "0                     0                     0                     0   \n",
       "1                     0                     0                     0   \n",
       "2                     0                     0                     0   \n",
       "3                     0                     0                     0   \n",
       "4                     0                     0                     0   \n",
       "\n",
       "   brodmann_area_46_9_1  brodmann_area_46_3_1  \n",
       "0                     0                     0  \n",
       "1                     0                     0  \n",
       "2                     0                     0  \n",
       "3                     0                     0  \n",
       "4                     0                     0  \n",
       "\n",
       "[5 rows x 149 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "544649b7-54d4-4c2b-a369-5fbc61096b07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>gene_biotype</th>\n",
       "      <th>exon_assignment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>1000</td>\n",
       "      <td>1484</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6337</td>\n",
       "      <td>6473</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6560</td>\n",
       "      <td>6813</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7814</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10282</td>\n",
       "      <td>10366</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10444</td>\n",
       "      <td>10786</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV101_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>1006</td>\n",
       "      <td>1484</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV102_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6337</td>\n",
       "      <td>6813</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV102_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7814</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV102_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10282</td>\n",
       "      <td>10366</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV102_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>1000</td>\n",
       "      <td>1484</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6337</td>\n",
       "      <td>6473</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6560</td>\n",
       "      <td>6813</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7814</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10282</td>\n",
       "      <td>10366</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>113</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV103_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6449</td>\n",
       "      <td>6473</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>6560</td>\n",
       "      <td>6813</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7814</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10282</td>\n",
       "      <td>10366</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10593</td>\n",
       "      <td>10640</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV105_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>1000</td>\n",
       "      <td>1484</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV106_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>120</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>7552</td>\n",
       "      <td>7808</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV106_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10553</td>\n",
       "      <td>10786</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV106_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10647</td>\n",
       "      <td>10791</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV107_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>123</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>10882</td>\n",
       "      <td>11242</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV107_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>SIRV1</td>\n",
       "      <td>LexogenSIRVData</td>\n",
       "      <td>exon</td>\n",
       "      <td>11403</td>\n",
       "      <td>11643</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>SIRV1A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIRV107_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chromosome           Source Feature  Start    End Score Strand Frame  \\\n",
       "98       SIRV1  LexogenSIRVData    exon   1000   1484     0      -     .   \n",
       "99       SIRV1  LexogenSIRVData    exon   6337   6473     0      -     .   \n",
       "100      SIRV1  LexogenSIRVData    exon   6560   6813     0      -     .   \n",
       "101      SIRV1  LexogenSIRVData    exon   7552   7814     0      -     .   \n",
       "102      SIRV1  LexogenSIRVData    exon  10282  10366     0      -     .   \n",
       "103      SIRV1  LexogenSIRVData    exon  10444  10786     0      -     .   \n",
       "104      SIRV1  LexogenSIRVData    exon   1006   1484     0      -     .   \n",
       "105      SIRV1  LexogenSIRVData    exon   6337   6813     0      -     .   \n",
       "106      SIRV1  LexogenSIRVData    exon   7552   7814     0      -     .   \n",
       "107      SIRV1  LexogenSIRVData    exon  10282  10366     0      -     .   \n",
       "108      SIRV1  LexogenSIRVData    exon   1000   1484     0      -     .   \n",
       "109      SIRV1  LexogenSIRVData    exon   6337   6473     0      -     .   \n",
       "110      SIRV1  LexogenSIRVData    exon   6560   6813     0      -     .   \n",
       "111      SIRV1  LexogenSIRVData    exon   7552   7814     0      -     .   \n",
       "112      SIRV1  LexogenSIRVData    exon  10282  10366     0      -     .   \n",
       "113      SIRV1  LexogenSIRVData    exon  10647  10791     0      -     .   \n",
       "114      SIRV1  LexogenSIRVData    exon   6449   6473     0      -     .   \n",
       "115      SIRV1  LexogenSIRVData    exon   6560   6813     0      -     .   \n",
       "116      SIRV1  LexogenSIRVData    exon   7552   7814     0      -     .   \n",
       "117      SIRV1  LexogenSIRVData    exon  10282  10366     0      -     .   \n",
       "118      SIRV1  LexogenSIRVData    exon  10593  10640     0      -     .   \n",
       "119      SIRV1  LexogenSIRVData    exon   1000   1484     0      -     .   \n",
       "120      SIRV1  LexogenSIRVData    exon   7552   7808     0      -     .   \n",
       "121      SIRV1  LexogenSIRVData    exon  10553  10786     0      -     .   \n",
       "122      SIRV1  LexogenSIRVData    exon  10647  10791     0      -     .   \n",
       "123      SIRV1  LexogenSIRVData    exon  10882  11242     0      -     .   \n",
       "124      SIRV1  LexogenSIRVData    exon  11403  11643     0      -     .   \n",
       "\n",
       "    gene_id gene_type  ... transcript_support_level  tag havana_transcript  \\\n",
       "98   SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "99   SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "100  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "101  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "102  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "103  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "104  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "105  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "106  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "107  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "108  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "109  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "110  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "111  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "112  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "113  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "114  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "115  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "116  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "117  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "118  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "119  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "120  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "121  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "122  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "123  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "124  SIRV1A       NaN  ...                      NaN  NaN               NaN   \n",
       "\n",
       "    exon_number exon_id  ont protein_id ccdsid gene_biotype exon_assignment  \n",
       "98          NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_0  \n",
       "99          NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_1  \n",
       "100         NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_2  \n",
       "101         NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_3  \n",
       "102         NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_4  \n",
       "103         NaN     NaN  NaN        NaN    NaN          NaN       SIRV101_5  \n",
       "104         NaN     NaN  NaN        NaN    NaN          NaN       SIRV102_0  \n",
       "105         NaN     NaN  NaN        NaN    NaN          NaN       SIRV102_1  \n",
       "106         NaN     NaN  NaN        NaN    NaN          NaN       SIRV102_2  \n",
       "107         NaN     NaN  NaN        NaN    NaN          NaN       SIRV102_3  \n",
       "108         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_0  \n",
       "109         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_1  \n",
       "110         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_2  \n",
       "111         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_3  \n",
       "112         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_4  \n",
       "113         NaN     NaN  NaN        NaN    NaN          NaN       SIRV103_5  \n",
       "114         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_0  \n",
       "115         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_1  \n",
       "116         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_2  \n",
       "117         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_3  \n",
       "118         NaN     NaN  NaN        NaN    NaN          NaN       SIRV105_4  \n",
       "119         NaN     NaN  NaN        NaN    NaN          NaN       SIRV106_0  \n",
       "120         NaN     NaN  NaN        NaN    NaN          NaN       SIRV106_1  \n",
       "121         NaN     NaN  NaN        NaN    NaN          NaN       SIRV106_2  \n",
       "122         NaN     NaN  NaN        NaN    NaN          NaN       SIRV107_0  \n",
       "123         NaN     NaN  NaN        NaN    NaN          NaN       SIRV107_1  \n",
       "124         NaN     NaN  NaN        NaN    NaN          NaN       SIRV107_2  \n",
       "\n",
       "[27 rows x 26 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.annot_gene_id.str.contains('SIRV')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e156d11-a858-4764-b0f5-4f0f11ad045e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
