{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9888f570-a5e4-4e19-8471-e5ced81ee843",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5ce077b0-3d12-40b5-a7c3-23d115f58698",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "54dd6c46-9c9e-4b0a-abf2-bd842259d0c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_gtf = '../'+expand(config['ref']['talon']['gtf'], species='human')[0]\n",
    "m_ab_file = '../'+expand(config['lr']['talon']['ab'], species='human')[0]\n",
    "m_filt_ab_file = '../'+expand(config['lr']['talon']['filt_ab'], species='human')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "06a12190-0dec-499e-af94-453067d749ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Antisense</th>\n",
       "      <td>9795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fusion</th>\n",
       "      <td>577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intergenic</th>\n",
       "      <td>742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>180287</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              annot_transcript_id\n",
       "gene_novelty                     \n",
       "Antisense                    9795\n",
       "Fusion                        577\n",
       "Intergenic                    742\n",
       "Known                      180287"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.read_csv(m_filt_ab_file, sep='\\t')\n",
    "df2[['annot_transcript_id', 'gene_novelty']].groupby('gene_novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b05aa8b8-5f0f-4501-a67f-f5cd1545893f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>h9_chondro_1_3</th>\n",
       "      <th>h9_panc_progen_1_2</th>\n",
       "      <th>hepg2_2_1</th>\n",
       "      <th>hl60_m1_12hr_1_2</th>\n",
       "      <th>hl60_m2_72hr_1_2</th>\n",
       "      <th>k562_3_2</th>\n",
       "      <th>lower_lobe_of_right_lung_1_1</th>\n",
       "      <th>ovary_2_1</th>\n",
       "      <th>pgp1_endo_1_1</th>\n",
       "      <th>right_cardiac_atrium_3_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>99127</th>\n",
       "      <td>62815</td>\n",
       "      <td>234794</td>\n",
       "      <td>TALONG000062815</td>\n",
       "      <td>TALONT000234794</td>\n",
       "      <td>TALONG000062815</td>\n",
       "      <td>TALONT000234794</td>\n",
       "      <td>5</td>\n",
       "      <td>1540</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99485</th>\n",
       "      <td>63684</td>\n",
       "      <td>238972</td>\n",
       "      <td>TALONG000063684</td>\n",
       "      <td>TALONT000238972</td>\n",
       "      <td>TALONG000063684</td>\n",
       "      <td>TALONT000238972</td>\n",
       "      <td>7</td>\n",
       "      <td>3704</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99498</th>\n",
       "      <td>63684</td>\n",
       "      <td>239129</td>\n",
       "      <td>TALONG000063684</td>\n",
       "      <td>TALONT000239129</td>\n",
       "      <td>TALONG000063684</td>\n",
       "      <td>TALONT000239129</td>\n",
       "      <td>5</td>\n",
       "      <td>3678</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101261</th>\n",
       "      <td>66573</td>\n",
       "      <td>259108</td>\n",
       "      <td>TALONG000066573</td>\n",
       "      <td>TALONT000259108</td>\n",
       "      <td>TALONG000066573</td>\n",
       "      <td>TALONT000259108</td>\n",
       "      <td>13</td>\n",
       "      <td>3250</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101549</th>\n",
       "      <td>67522</td>\n",
       "      <td>262589</td>\n",
       "      <td>TALONG000067522</td>\n",
       "      <td>TALONT000262589</td>\n",
       "      <td>TALONG000067522</td>\n",
       "      <td>TALONT000262589</td>\n",
       "      <td>2</td>\n",
       "      <td>3645</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_ID  transcript_ID    annot_gene_id annot_transcript_id  \\\n",
       "99127     62815         234794  TALONG000062815     TALONT000234794   \n",
       "99485     63684         238972  TALONG000063684     TALONT000238972   \n",
       "99498     63684         239129  TALONG000063684     TALONT000239129   \n",
       "101261    66573         259108  TALONG000066573     TALONT000259108   \n",
       "101549    67522         262589  TALONG000067522     TALONT000262589   \n",
       "\n",
       "        annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "99127   TALONG000062815       TALONT000234794        5    1540       Fusion   \n",
       "99485   TALONG000063684       TALONT000238972        7    3704       Fusion   \n",
       "99498   TALONG000063684       TALONT000239129        5    3678       Fusion   \n",
       "101261  TALONG000066573       TALONT000259108       13    3250       Fusion   \n",
       "101549  TALONG000067522       TALONT000262589        2    3645       Fusion   \n",
       "\n",
       "       transcript_novelty  ... h9_chondro_1_3  h9_panc_progen_1_2  hepg2_2_1  \\\n",
       "99127                 ISM  ...              1                   5          5   \n",
       "99485              Fusion  ...              0                   0          0   \n",
       "99498                 ISM  ...             14                   0          0   \n",
       "101261                ISM  ...              0                   0          0   \n",
       "101549                ISM  ...              0                   0          0   \n",
       "\n",
       "        hl60_m1_12hr_1_2  hl60_m2_72hr_1_2  k562_3_2  \\\n",
       "99127                  1                 1        16   \n",
       "99485                  0                 0         1   \n",
       "99498                  6                 2         3   \n",
       "101261                 0                 0         0   \n",
       "101549                 0                 0         0   \n",
       "\n",
       "        lower_lobe_of_right_lung_1_1  ovary_2_1  pgp1_endo_1_1  \\\n",
       "99127                              4          7              4   \n",
       "99485                              3          8              0   \n",
       "99498                              0          0              1   \n",
       "101261                             1          0              0   \n",
       "101549                             0          0              0   \n",
       "\n",
       "        right_cardiac_atrium_3_1  \n",
       "99127                          6  \n",
       "99485                          0  \n",
       "99498                          0  \n",
       "101261                         0  \n",
       "101549                         0  \n",
       "\n",
       "[5 rows x 149 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.loc[df2.gene_novelty=='Fusion'].head(15).tail(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "4c7857c8-bd4f-43c9-86db-da3a292fc1cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>h9_chondro_1_3</th>\n",
       "      <th>h9_panc_progen_1_2</th>\n",
       "      <th>hepg2_2_1</th>\n",
       "      <th>hl60_m1_12hr_1_2</th>\n",
       "      <th>hl60_m2_72hr_1_2</th>\n",
       "      <th>k562_3_2</th>\n",
       "      <th>lower_lobe_of_right_lung_1_1</th>\n",
       "      <th>ovary_2_1</th>\n",
       "      <th>pgp1_endo_1_1</th>\n",
       "      <th>right_cardiac_atrium_3_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24936</th>\n",
       "      <td>14914</td>\n",
       "      <td>54193</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>ENST00000497410.5</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>BRD9-214</td>\n",
       "      <td>2</td>\n",
       "      <td>1916</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24937</th>\n",
       "      <td>14914</td>\n",
       "      <td>54194</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>ENST00000483234.5</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>BRD9-205</td>\n",
       "      <td>6</td>\n",
       "      <td>2211</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24938</th>\n",
       "      <td>14914</td>\n",
       "      <td>54197</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>ENST00000495265.5</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>BRD9-212</td>\n",
       "      <td>16</td>\n",
       "      <td>2843</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24939</th>\n",
       "      <td>14914</td>\n",
       "      <td>54199</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>ENST00000483173.5</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>BRD9-204</td>\n",
       "      <td>16</td>\n",
       "      <td>2120</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24940</th>\n",
       "      <td>14914</td>\n",
       "      <td>54200</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>ENST00000467963.5</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>BRD9-202</td>\n",
       "      <td>16</td>\n",
       "      <td>2156</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "      <td>23</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24941</th>\n",
       "      <td>14914</td>\n",
       "      <td>54201</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>ENST00000519112.5</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>BRD9-217</td>\n",
       "      <td>7</td>\n",
       "      <td>861</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97352</th>\n",
       "      <td>14914</td>\n",
       "      <td>214706</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>TALONT000214706</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>TALONT000214706</td>\n",
       "      <td>16</td>\n",
       "      <td>2724</td>\n",
       "      <td>Known</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97354</th>\n",
       "      <td>59968</td>\n",
       "      <td>214728</td>\n",
       "      <td>TALONG000059968</td>\n",
       "      <td>TALONT000214728</td>\n",
       "      <td>TALONG000059968</td>\n",
       "      <td>TALONT000214728</td>\n",
       "      <td>16</td>\n",
       "      <td>2661</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97362</th>\n",
       "      <td>14914</td>\n",
       "      <td>214877</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>TALONT000214877</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>TALONT000214877</td>\n",
       "      <td>8</td>\n",
       "      <td>2918</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97367</th>\n",
       "      <td>14914</td>\n",
       "      <td>214926</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>TALONT000214926</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>TALONT000214926</td>\n",
       "      <td>12</td>\n",
       "      <td>2055</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97373</th>\n",
       "      <td>14914</td>\n",
       "      <td>214959</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>TALONT000214959</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>TALONT000214959</td>\n",
       "      <td>15</td>\n",
       "      <td>2412</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97391</th>\n",
       "      <td>14914</td>\n",
       "      <td>215158</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>TALONT000215158</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>TALONT000215158</td>\n",
       "      <td>14</td>\n",
       "      <td>2517</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97452</th>\n",
       "      <td>14914</td>\n",
       "      <td>215761</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>TALONT000215761</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>TALONT000215761</td>\n",
       "      <td>2</td>\n",
       "      <td>410</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>14</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182058</th>\n",
       "      <td>14914</td>\n",
       "      <td>1179185</td>\n",
       "      <td>ENSG00000028310.17</td>\n",
       "      <td>TALONT001179185</td>\n",
       "      <td>BRD9</td>\n",
       "      <td>TALONT001179185</td>\n",
       "      <td>13</td>\n",
       "      <td>2516</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_ID  transcript_ID       annot_gene_id annot_transcript_id  \\\n",
       "24936     14914          54193  ENSG00000028310.17   ENST00000497410.5   \n",
       "24937     14914          54194  ENSG00000028310.17   ENST00000483234.5   \n",
       "24938     14914          54197  ENSG00000028310.17   ENST00000495265.5   \n",
       "24939     14914          54199  ENSG00000028310.17   ENST00000483173.5   \n",
       "24940     14914          54200  ENSG00000028310.17   ENST00000467963.5   \n",
       "24941     14914          54201  ENSG00000028310.17   ENST00000519112.5   \n",
       "97352     14914         214706  ENSG00000028310.17     TALONT000214706   \n",
       "97354     59968         214728     TALONG000059968     TALONT000214728   \n",
       "97362     14914         214877  ENSG00000028310.17     TALONT000214877   \n",
       "97367     14914         214926  ENSG00000028310.17     TALONT000214926   \n",
       "97373     14914         214959  ENSG00000028310.17     TALONT000214959   \n",
       "97391     14914         215158  ENSG00000028310.17     TALONT000215158   \n",
       "97452     14914         215761  ENSG00000028310.17     TALONT000215761   \n",
       "182058    14914        1179185  ENSG00000028310.17     TALONT001179185   \n",
       "\n",
       "        annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "24936              BRD9              BRD9-214        2    1916        Known   \n",
       "24937              BRD9              BRD9-205        6    2211        Known   \n",
       "24938              BRD9              BRD9-212       16    2843        Known   \n",
       "24939              BRD9              BRD9-204       16    2120        Known   \n",
       "24940              BRD9              BRD9-202       16    2156        Known   \n",
       "24941              BRD9              BRD9-217        7     861        Known   \n",
       "97352              BRD9       TALONT000214706       16    2724        Known   \n",
       "97354   TALONG000059968       TALONT000214728       16    2661       Fusion   \n",
       "97362              BRD9       TALONT000214877        8    2918        Known   \n",
       "97367              BRD9       TALONT000214926       12    2055        Known   \n",
       "97373              BRD9       TALONT000214959       15    2412        Known   \n",
       "97391              BRD9       TALONT000215158       14    2517        Known   \n",
       "97452              BRD9       TALONT000215761        2     410        Known   \n",
       "182058             BRD9       TALONT001179185       13    2516        Known   \n",
       "\n",
       "       transcript_novelty  ... h9_chondro_1_3  h9_panc_progen_1_2  hepg2_2_1  \\\n",
       "24936               Known  ...              0                   0          0   \n",
       "24937               Known  ...              0                   1          0   \n",
       "24938               Known  ...              0                   1          0   \n",
       "24939               Known  ...              0                   0          0   \n",
       "24940               Known  ...             11                   0          7   \n",
       "24941               Known  ...              0                   0          0   \n",
       "97352                 NIC  ...              4                   0          1   \n",
       "97354                 ISM  ...              3                   0          1   \n",
       "97362                 ISM  ...              0                   2          0   \n",
       "97367                 ISM  ...              0                   2          0   \n",
       "97373                 ISM  ...              3                   1          0   \n",
       "97391                 ISM  ...              0                   5          1   \n",
       "97452                 ISM  ...              0                   1          2   \n",
       "182058                ISM  ...              0                   3          0   \n",
       "\n",
       "        hl60_m1_12hr_1_2  hl60_m2_72hr_1_2  k562_3_2  \\\n",
       "24936                  1                 0         0   \n",
       "24937                  1                 1         1   \n",
       "24938                  0                 0         0   \n",
       "24939                  0                 0         0   \n",
       "24940                  2                17        23   \n",
       "24941                  0                 0         0   \n",
       "97352                  0                 2        10   \n",
       "97354                  0                 2         8   \n",
       "97362                  3                 2         3   \n",
       "97367                  3                 0         0   \n",
       "97373                  0                 1         3   \n",
       "97391                  3                 4         3   \n",
       "97452                 22                 0        19   \n",
       "182058                 1                 0         0   \n",
       "\n",
       "        lower_lobe_of_right_lung_1_1  ovary_2_1  pgp1_endo_1_1  \\\n",
       "24936                              1          0              1   \n",
       "24937                              0          0              0   \n",
       "24938                              0          2              0   \n",
       "24939                              0          0              0   \n",
       "24940                             11         13              8   \n",
       "24941                              0          0              0   \n",
       "97352                              6          8              2   \n",
       "97354                              2          3              2   \n",
       "97362                              2          1              0   \n",
       "97367                              0          0              1   \n",
       "97373                              0          0              0   \n",
       "97391                              0          0              0   \n",
       "97452                              9          0             14   \n",
       "182058                             0          0              0   \n",
       "\n",
       "        right_cardiac_atrium_3_1  \n",
       "24936                          0  \n",
       "24937                          0  \n",
       "24938                          0  \n",
       "24939                          0  \n",
       "24940                         13  \n",
       "24941                          0  \n",
       "97352                          2  \n",
       "97354                          3  \n",
       "97362                          2  \n",
       "97367                          0  \n",
       "97373                          2  \n",
       "97391                          0  \n",
       "97452                          0  \n",
       "182058                         0  \n",
       "\n",
       "[14 rows x 149 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.loc[(df2.annot_transcript_id=='TALONT000214728')|(df2.annot_gene_name=='BRD9')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00c6a644-19a4-47e9-87f2-39c4229eb764",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53df5839-3d39-4ac7-9e97-7f1d925793b2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33e93aff-7050-45d7-b11c-9b70b43f404f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1d613e5-c47f-4aa4-93d5-612052deb68a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "9f93f66e-f23f-4d16-a2d0-2804f9c0e9dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = pr.read_gtf(m_gtf, duplicate_attr=True).df\n",
    "gtf_df['tag'] = gtf_df.tag.astype(str)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "6d7a8088-70ad-4859-bac6-1edabdd9cac9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>772</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11868</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11868</td>\n",
       "      <td>12227</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12612</td>\n",
       "      <td>12721</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>776</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13220</td>\n",
       "      <td>14409</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>...</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>DDX11L1-202</td>\n",
       "      <td>1</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>3</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251291</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248858917</td>\n",
       "      <td>248859033</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-228</td>\n",
       "      <td>4</td>\n",
       "      <td>nan</td>\n",
       "      <td>OTTHUMT00000382606.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00002188413.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251292</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>248857974</td>\n",
       "      <td>248858385</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000171163.15</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>ZNF692-228</td>\n",
       "      <td>4</td>\n",
       "      <td>nan</td>\n",
       "      <td>OTTHUMT00000382606.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSE00002170424.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251293</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>248912689</td>\n",
       "      <td>248912795</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000200495.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251294</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>248912689</td>\n",
       "      <td>248912795</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000200495.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RNU6-1205P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251295</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>248912689</td>\n",
       "      <td>248912795</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSG00000200495.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>RNU6-1205P-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSE00001438388.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>250524 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome   Source     Feature      Start        End Score Strand  \\\n",
       "772          chr1   HAVANA        gene      11868      14409     .      +   \n",
       "773          chr1   HAVANA  transcript      11868      14409     .      +   \n",
       "774          chr1   HAVANA        exon      11868      12227     .      +   \n",
       "775          chr1   HAVANA        exon      12612      12721     .      +   \n",
       "776          chr1   HAVANA        exon      13220      14409     .      +   \n",
       "...           ...      ...         ...        ...        ...   ...    ...   \n",
       "251291       chr1   HAVANA        exon  248858917  248859033     .      -   \n",
       "251292       chr1   HAVANA        exon  248857974  248858385     .      -   \n",
       "251293       chr1  ENSEMBL        gene  248912689  248912795     .      -   \n",
       "251294       chr1  ENSEMBL  transcript  248912689  248912795     .      -   \n",
       "251295       chr1  ENSEMBL        exon  248912689  248912795     .      -   \n",
       "\n",
       "       Frame             gene_id                           gene_type  ...  \\\n",
       "772        .   ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "773        .   ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "774        .   ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "775        .   ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "776        .   ENSG00000223972.5  transcribed_unprocessed_pseudogene  ...   \n",
       "...      ...                 ...                                 ...  ...   \n",
       "251291     .  ENSG00000171163.15                      protein_coding  ...   \n",
       "251292     .  ENSG00000171163.15                      protein_coding  ...   \n",
       "251293     .   ENSG00000200495.1                               snRNA  ...   \n",
       "251294     .   ENSG00000200495.1                               snRNA  ...   \n",
       "251295     .   ENSG00000200495.1                               snRNA  ...   \n",
       "\n",
       "             transcript_type transcript_name transcript_support_level    tag  \\\n",
       "772                      NaN             NaN                      NaN    nan   \n",
       "773     processed_transcript     DDX11L1-202                        1  basic   \n",
       "774     processed_transcript     DDX11L1-202                        1  basic   \n",
       "775     processed_transcript     DDX11L1-202                        1  basic   \n",
       "776     processed_transcript     DDX11L1-202                        1  basic   \n",
       "...                      ...             ...                      ...    ...   \n",
       "251291       retained_intron      ZNF692-228                        4    nan   \n",
       "251292       retained_intron      ZNF692-228                        4    nan   \n",
       "251293                   NaN             NaN                      NaN    nan   \n",
       "251294                 snRNA  RNU6-1205P-201                       NA  basic   \n",
       "251295                 snRNA  RNU6-1205P-201                       NA  basic   \n",
       "\n",
       "           havana_transcript exon_number            exon_id  ont protein_id  \\\n",
       "772                      NaN         NaN                NaN  NaN        NaN   \n",
       "773     OTTHUMT00000362751.1         NaN                NaN  NaN        NaN   \n",
       "774     OTTHUMT00000362751.1           1  ENSE00002234944.1  NaN        NaN   \n",
       "775     OTTHUMT00000362751.1           2  ENSE00003582793.1  NaN        NaN   \n",
       "776     OTTHUMT00000362751.1           3  ENSE00002312635.1  NaN        NaN   \n",
       "...                      ...         ...                ...  ...        ...   \n",
       "251291  OTTHUMT00000382606.1           1  ENSE00002188413.1  NaN        NaN   \n",
       "251292  OTTHUMT00000382606.1           2  ENSE00002170424.1  NaN        NaN   \n",
       "251293                   NaN         NaN                NaN  NaN        NaN   \n",
       "251294                   NaN         NaN                NaN  NaN        NaN   \n",
       "251295                   NaN           1  ENSE00001438388.1  NaN        NaN   \n",
       "\n",
       "       ccdsid  \n",
       "772       NaN  \n",
       "773       NaN  \n",
       "774       NaN  \n",
       "775       NaN  \n",
       "776       NaN  \n",
       "...       ...  \n",
       "251291    NaN  \n",
       "251292    NaN  \n",
       "251293    NaN  \n",
       "251294    NaN  \n",
       "251295    NaN  \n",
       "\n",
       "[250524 rows x 24 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.loc[gtf_df.Chromosome=='chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "70397a92-d2b3-4435-ab01-671f37fb23e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Source, Feature, Start, End, Score, Strand, Frame, gene_id, gene_type, gene_name, level, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, protein_id, ccdsid]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# are all \"overlapping_locus\" tags just genes?\n",
    "gtf_df.loc[(gtf_df.tag.str.contains('overlapping_locus'))&(gtf_df.Feature!='gene')]\n",
    "# yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "baf3a207-bb28-4dcd-ba26-fdf3307b28d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>10451</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           tag\n",
       "Feature       \n",
       "gene     10451"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what entries have the \"overlapping_locus\" tag?\n",
    "temp = gtf_df.loc[gtf_df.tag.str.contains('overlapping_locus')].copy(deep=True)\n",
    "temp[['tag', 'Feature']].groupby('Feature').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "36963d20-638c-4c28-8979-36fc9f91312d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Source, Feature, Start, End, Score, Strand, Frame, gene_id, gene_type, gene_name, level, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, ont, protein_id, ccdsid]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there are no entries with the \"readthrough_gene\" tag\n",
    "gtf_df.loc[gtf_df.tag.str.contains('readthrough_gene')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2aca2628-1f5d-4439-8ed1-d183a0c49a53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>6706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UTR</th>\n",
       "      <td>5761</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>15759</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>1879</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tag\n",
       "Feature           \n",
       "CDS           6706\n",
       "UTR           5761\n",
       "exon         15759\n",
       "start_codon    804\n",
       "stop_codon     974\n",
       "transcript    1879"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what entries have the \"readthrough_transcript\" tag?\n",
    "temp = gtf_df.loc[gtf_df.tag.str.contains('readthrough_transcript')].copy(deep=True)\n",
    "temp[['tag', 'Feature']].groupby('Feature').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "06ebcb7b-5fcc-4f64-8a25-82158811b86f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>6490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UTR</th>\n",
       "      <td>5640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>15397</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>775</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>1833</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               tag\n",
       "Feature           \n",
       "CDS           6490\n",
       "UTR           5640\n",
       "exon         15397\n",
       "start_codon    775\n",
       "stop_codon     945\n",
       "transcript    1833"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get a list of genes w/ the overlapping_locus tags\n",
    "o_genes = gtf_df.loc[gtf_df.tag.str.contains('overlapping_locus'), 'gene_id'].unique().tolist()\n",
    "\n",
    "# how many transcripts from genes w/ overlapping locus tag have the 'read_trhough_transcript' tag?\n",
    "temp = gtf_df.loc[gtf_df.gene_id.isin(o_genes)].copy(deep=True)\n",
    "temp = temp.loc[temp.tag.str.contains('readthrough_transcript')].copy(deep=True)\n",
    "temp[['tag', 'Feature']].groupby('Feature').count()\n",
    "\n",
    "# so there are a few transcrpts with the readthrough_transcript tag that \n",
    "# don't belong to the overlapping_locus genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "2bfec637-f43c-4c09-b29a-91acba59f6fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# more throughough check for concordance of overlapping locus and \n",
    "temp = gtf_df.loc[gtf_df.Feature=='transcript'][['gene_id', 'gene_name', 'tag', 'transcript_id']].copy(deep=True)\n",
    "temp['readthrough_transcript'] = temp.tag.str.contains('readthrough_transcript')\n",
    "temp.drop('tag', inplace=True, axis=1)\n",
    "temp.drop_duplicates(inplace=True)\n",
    "\n",
    "temp2 = gtf_df.loc[gtf_df.Feature=='gene'][['gene_id', 'tag']].copy(deep=True)\n",
    "temp2['overlapping_locus'] = temp2.tag.str.contains('overlapping_locus')\n",
    "temp2.drop('tag', axis=1, inplace=True)\n",
    "temp2.drop_duplicates(inplace=True)\n",
    "\n",
    "temp = temp.merge(temp2, how='outer', on='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "6d9b7465-02e7-44fd-abf6-d7051b92a32c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>readthrough_transcript</th>\n",
       "      <th>overlapping_locus</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>155960</td>\n",
       "      <td>155780</td>\n",
       "      <td>155960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>49102</td>\n",
       "      <td>49102</td>\n",
       "      <td>49102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>1833</td>\n",
       "      <td>1833</td>\n",
       "      <td>1833</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   readthrough_transcript  overlapping_locus  gene_id  gene_name  \\\n",
       "0                   False              False   155960     155780   \n",
       "1                   False               True    49102      49102   \n",
       "2                    True              False       46         46   \n",
       "3                    True               True     1833       1833   \n",
       "\n",
       "   transcript_id  \n",
       "0         155960  \n",
       "1          49102  \n",
       "2             46  \n",
       "3           1833  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.groupby(['readthrough_transcript', 'overlapping_locus']).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c24011b4-888f-4c5e-b5a4-7d3b42802a7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>readthrough_transcript</th>\n",
       "      <th>overlapping_locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1447</th>\n",
       "      <td>ENSG00000285721.1</td>\n",
       "      <td>AL031281.2</td>\n",
       "      <td>ENST00000648594.1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5516</th>\n",
       "      <td>ENSG00000280778.1</td>\n",
       "      <td>AC243547.3</td>\n",
       "      <td>ENST00000630636.1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5517</th>\n",
       "      <td>ENSG00000280778.1</td>\n",
       "      <td>AC243547.3</td>\n",
       "      <td>ENST00000625258.1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12653</th>\n",
       "      <td>ENSG00000285839.1</td>\n",
       "      <td>AL445685.3</td>\n",
       "      <td>ENST00000648686.1</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16180</th>\n",
       "      <td>ENSG00000158887.15</td>\n",
       "      <td>MPZ</td>\n",
       "      <td>ENST00000533357.4</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id   gene_name      transcript_id  \\\n",
       "1447    ENSG00000285721.1  AL031281.2  ENST00000648594.1   \n",
       "5516    ENSG00000280778.1  AC243547.3  ENST00000630636.1   \n",
       "5517    ENSG00000280778.1  AC243547.3  ENST00000625258.1   \n",
       "12653   ENSG00000285839.1  AL445685.3  ENST00000648686.1   \n",
       "16180  ENSG00000158887.15         MPZ  ENST00000533357.4   \n",
       "\n",
       "       readthrough_transcript  overlapping_locus  \n",
       "1447                     True              False  \n",
       "5516                     True              False  \n",
       "5517                     True              False  \n",
       "12653                    True              False  \n",
       "16180                    True              False  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[(temp.readthrough_transcript==True)&(temp.overlapping_locus==False)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4c3b7d34-e4cc-41e7-9d2e-112656bd9bdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>readthrough_transcript</th>\n",
       "      <th>overlapping_locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>204</th>\n",
       "      <td>ENSG00000240409.1</td>\n",
       "      <td>MTATP8P1</td>\n",
       "      <td>ENST00000467115.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>205</th>\n",
       "      <td>ENSG00000248527.1</td>\n",
       "      <td>MTATP6P1</td>\n",
       "      <td>ENST00000514057.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>230</th>\n",
       "      <td>ENSG00000285268.1</td>\n",
       "      <td>AL669831.7</td>\n",
       "      <td>ENST00000644482.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>248</th>\n",
       "      <td>ENSG00000272438.1</td>\n",
       "      <td>AL645608.6</td>\n",
       "      <td>ENST00000607769.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249</th>\n",
       "      <td>ENSG00000230699.2</td>\n",
       "      <td>AL645608.2</td>\n",
       "      <td>ENST00000448179.1</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id   gene_name      transcript_id  readthrough_transcript  \\\n",
       "204  ENSG00000240409.1    MTATP8P1  ENST00000467115.1                   False   \n",
       "205  ENSG00000248527.1    MTATP6P1  ENST00000514057.1                   False   \n",
       "230  ENSG00000285268.1  AL669831.7  ENST00000644482.1                   False   \n",
       "248  ENSG00000272438.1  AL645608.6  ENST00000607769.1                   False   \n",
       "249  ENSG00000230699.2  AL645608.2  ENST00000448179.1                   False   \n",
       "\n",
       "     overlapping_locus  \n",
       "204               True  \n",
       "205               True  \n",
       "230               True  \n",
       "248               True  \n",
       "249               True  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[(temp.readthrough_transcript==False)&(temp.overlapping_locus==True)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "5b7ee917-a180-45b5-abfd-674223450f54",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>readthrough_transcript</th>\n",
       "      <th>overlapping_locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16179</th>\n",
       "      <td>ENSG00000158887.15</td>\n",
       "      <td>MPZ</td>\n",
       "      <td>ENST00000463290.5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16180</th>\n",
       "      <td>ENSG00000158887.15</td>\n",
       "      <td>MPZ</td>\n",
       "      <td>ENST00000533357.4</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16181</th>\n",
       "      <td>ENSG00000158887.15</td>\n",
       "      <td>MPZ</td>\n",
       "      <td>ENST00000491222.5</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16182</th>\n",
       "      <td>ENSG00000158887.15</td>\n",
       "      <td>MPZ</td>\n",
       "      <td>ENST00000526189.2</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16183</th>\n",
       "      <td>ENSG00000158887.15</td>\n",
       "      <td>MPZ</td>\n",
       "      <td>ENST00000476410.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16184</th>\n",
       "      <td>ENSG00000158887.15</td>\n",
       "      <td>MPZ</td>\n",
       "      <td>ENST00000488271.1</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gene_id gene_name      transcript_id  \\\n",
       "16179  ENSG00000158887.15       MPZ  ENST00000463290.5   \n",
       "16180  ENSG00000158887.15       MPZ  ENST00000533357.4   \n",
       "16181  ENSG00000158887.15       MPZ  ENST00000491222.5   \n",
       "16182  ENSG00000158887.15       MPZ  ENST00000526189.2   \n",
       "16183  ENSG00000158887.15       MPZ  ENST00000476410.1   \n",
       "16184  ENSG00000158887.15       MPZ  ENST00000488271.1   \n",
       "\n",
       "       readthrough_transcript  overlapping_locus  \n",
       "16179                   False              False  \n",
       "16180                    True              False  \n",
       "16181                   False              False  \n",
       "16182                   False              False  \n",
       "16183                   False              False  \n",
       "16184                   False              False  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.gene_name=='MPZ']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a60deee3-c751-4634-a823-7fb8603fe71f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>readthrough_transcript</th>\n",
       "      <th>overlapping_locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>844</th>\n",
       "      <td>ENSG00000251503.8</td>\n",
       "      <td>CENPS-CORT</td>\n",
       "      <td>ENST00000400900.6</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845</th>\n",
       "      <td>ENSG00000251503.8</td>\n",
       "      <td>CENPS-CORT</td>\n",
       "      <td>ENST00000470413.6</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>846</th>\n",
       "      <td>ENSG00000251503.8</td>\n",
       "      <td>CENPS-CORT</td>\n",
       "      <td>ENST00000602787.6</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847</th>\n",
       "      <td>ENSG00000251503.8</td>\n",
       "      <td>CENPS-CORT</td>\n",
       "      <td>ENST00000602296.6</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>848</th>\n",
       "      <td>ENSG00000251503.8</td>\n",
       "      <td>CENPS-CORT</td>\n",
       "      <td>ENST00000465026.1</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               gene_id   gene_name      transcript_id  readthrough_transcript  \\\n",
       "844  ENSG00000251503.8  CENPS-CORT  ENST00000400900.6                    True   \n",
       "845  ENSG00000251503.8  CENPS-CORT  ENST00000470413.6                    True   \n",
       "846  ENSG00000251503.8  CENPS-CORT  ENST00000602787.6                    True   \n",
       "847  ENSG00000251503.8  CENPS-CORT  ENST00000602296.6                    True   \n",
       "848  ENSG00000251503.8  CENPS-CORT  ENST00000465026.1                    True   \n",
       "\n",
       "     overlapping_locus  \n",
       "844               True  \n",
       "845               True  \n",
       "846               True  \n",
       "847               True  \n",
       "848               True  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[temp.gene_name=='CENPS-CORT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38b438e0-8852-41bf-b087-0a9a08e9f376",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "a1462471-5908-45e9-853e-3f62208208aa",
   "metadata": {},
   "source": [
    "## looking at abundance mats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f3c95af2-fdbb-4864-bd03-d75a0274d548",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(m_ab_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "25c2326b-2d9d-4d15-976e-2dad8c3f2630",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Known', 'Antisense', 'Intergenic', 'Fusion'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.gene_novelty.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "48f82fe1-cb9c-4eef-bfa4-6159d8b6989e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>annot_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Antisense</td>\n",
       "      <td>222873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fusion</td>\n",
       "      <td>43465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Intergenic</td>\n",
       "      <td>64034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Known</td>\n",
       "      <td>2552024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_novelty  annot_gene_id\n",
       "0    Antisense         222873\n",
       "1       Fusion          43465\n",
       "2   Intergenic          64034\n",
       "3        Known        2552024"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['annot_gene_id', 'gene_novelty']].groupby('gene_novelty').count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c878439f-c2df-465c-af0c-ee65a344197b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>adrenal_gland_1_2</th>\n",
       "      <th>cortex_5x_f_1_2</th>\n",
       "      <th>cortex_14d_f_2</th>\n",
       "      <th>cortex_2mo_m_2</th>\n",
       "      <th>gastroc_14d_f_2</th>\n",
       "      <th>gastroc_25d_m_2</th>\n",
       "      <th>gastroc_4d_f_2</th>\n",
       "      <th>heart_18-20mo_m_1</th>\n",
       "      <th>hippocampus_wt_m_2_1</th>\n",
       "      <th>hippocampus_18-20mo_f_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64257</th>\n",
       "      <td>55774</td>\n",
       "      <td>142467</td>\n",
       "      <td>TALONG000055774</td>\n",
       "      <td>TALONT000142467</td>\n",
       "      <td>TALONG000055774</td>\n",
       "      <td>TALONT000142467</td>\n",
       "      <td>13</td>\n",
       "      <td>1934</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64383</th>\n",
       "      <td>55808</td>\n",
       "      <td>142593</td>\n",
       "      <td>TALONG000055808</td>\n",
       "      <td>TALONT000142593</td>\n",
       "      <td>TALONG000055808</td>\n",
       "      <td>TALONT000142593</td>\n",
       "      <td>9</td>\n",
       "      <td>3331</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64754</th>\n",
       "      <td>55869</td>\n",
       "      <td>142964</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142964</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142964</td>\n",
       "      <td>8</td>\n",
       "      <td>2352</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64757</th>\n",
       "      <td>55869</td>\n",
       "      <td>142967</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142967</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142967</td>\n",
       "      <td>9</td>\n",
       "      <td>2518</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64759</th>\n",
       "      <td>55869</td>\n",
       "      <td>142969</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142969</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142969</td>\n",
       "      <td>4</td>\n",
       "      <td>1035</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882153</th>\n",
       "      <td>127991</td>\n",
       "      <td>2960363</td>\n",
       "      <td>TALONG000127991</td>\n",
       "      <td>TALONT002960363</td>\n",
       "      <td>TALONG000127991</td>\n",
       "      <td>TALONT002960363</td>\n",
       "      <td>1</td>\n",
       "      <td>1207</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882155</th>\n",
       "      <td>127991</td>\n",
       "      <td>2960365</td>\n",
       "      <td>TALONG000127991</td>\n",
       "      <td>TALONT002960365</td>\n",
       "      <td>TALONG000127991</td>\n",
       "      <td>TALONT002960365</td>\n",
       "      <td>6</td>\n",
       "      <td>1507</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NNC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882291</th>\n",
       "      <td>180891</td>\n",
       "      <td>2960501</td>\n",
       "      <td>TALONG000180891</td>\n",
       "      <td>TALONT002960501</td>\n",
       "      <td>TALONG000180891</td>\n",
       "      <td>TALONT002960501</td>\n",
       "      <td>18</td>\n",
       "      <td>3855</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882323</th>\n",
       "      <td>180891</td>\n",
       "      <td>2960533</td>\n",
       "      <td>TALONG000180891</td>\n",
       "      <td>TALONT002960533</td>\n",
       "      <td>TALONG000180891</td>\n",
       "      <td>TALONT002960533</td>\n",
       "      <td>1</td>\n",
       "      <td>1894</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882371</th>\n",
       "      <td>261990</td>\n",
       "      <td>2960581</td>\n",
       "      <td>TALONG000261990</td>\n",
       "      <td>TALONT002960581</td>\n",
       "      <td>TALONG000261990</td>\n",
       "      <td>TALONT002960581</td>\n",
       "      <td>9</td>\n",
       "      <td>1948</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43465 rows × 138 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_ID  transcript_ID    annot_gene_id annot_transcript_id  \\\n",
       "64257      55774         142467  TALONG000055774     TALONT000142467   \n",
       "64383      55808         142593  TALONG000055808     TALONT000142593   \n",
       "64754      55869         142964  TALONG000055869     TALONT000142964   \n",
       "64757      55869         142967  TALONG000055869     TALONT000142967   \n",
       "64759      55869         142969  TALONG000055869     TALONT000142969   \n",
       "...          ...            ...              ...                 ...   \n",
       "2882153   127991        2960363  TALONG000127991     TALONT002960363   \n",
       "2882155   127991        2960365  TALONG000127991     TALONT002960365   \n",
       "2882291   180891        2960501  TALONG000180891     TALONT002960501   \n",
       "2882323   180891        2960533  TALONG000180891     TALONT002960533   \n",
       "2882371   261990        2960581  TALONG000261990     TALONT002960581   \n",
       "\n",
       "         annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "64257    TALONG000055774       TALONT000142467       13    1934       Fusion   \n",
       "64383    TALONG000055808       TALONT000142593        9    3331       Fusion   \n",
       "64754    TALONG000055869       TALONT000142964        8    2352       Fusion   \n",
       "64757    TALONG000055869       TALONT000142967        9    2518       Fusion   \n",
       "64759    TALONG000055869       TALONT000142969        4    1035       Fusion   \n",
       "...                  ...                   ...      ...     ...          ...   \n",
       "2882153  TALONG000127991       TALONT002960363        1    1207       Fusion   \n",
       "2882155  TALONG000127991       TALONT002960365        6    1507       Fusion   \n",
       "2882291  TALONG000180891       TALONT002960501       18    3855       Fusion   \n",
       "2882323  TALONG000180891       TALONT002960533        1    1894       Fusion   \n",
       "2882371  TALONG000261990       TALONT002960581        9    1948       Fusion   \n",
       "\n",
       "        transcript_novelty  ... adrenal_gland_1_2  cortex_5x_f_1_2  \\\n",
       "64257               Fusion  ...                 0                0   \n",
       "64383               Fusion  ...                 0                1   \n",
       "64754               Fusion  ...                 0                2   \n",
       "64757                  NIC  ...                 0                3   \n",
       "64759                  NIC  ...                 0                0   \n",
       "...                    ...  ...               ...              ...   \n",
       "2882153            Genomic  ...                 1                0   \n",
       "2882155                NNC  ...                 0                0   \n",
       "2882291                NIC  ...                 0                0   \n",
       "2882323            Genomic  ...                 1                0   \n",
       "2882371             Fusion  ...                 0                0   \n",
       "\n",
       "         cortex_14d_f_2  cortex_2mo_m_2  gastroc_14d_f_2  gastroc_25d_m_2  \\\n",
       "64257                 0               0                0                0   \n",
       "64383                 1               0                0                0   \n",
       "64754                 0               0                0                0   \n",
       "64757                 0               0                0                0   \n",
       "64759                 0               0                0                0   \n",
       "...                 ...             ...              ...              ...   \n",
       "2882153               0               0                1                0   \n",
       "2882155               0               0                0                0   \n",
       "2882291               0               1                0                0   \n",
       "2882323               0               0                0                0   \n",
       "2882371               0               0                0                0   \n",
       "\n",
       "         gastroc_4d_f_2  heart_18-20mo_m_1  hippocampus_wt_m_2_1  \\\n",
       "64257                 0                  0                     0   \n",
       "64383                 0                  0                     2   \n",
       "64754                 0                  0                     1   \n",
       "64757                 0                  0                     0   \n",
       "64759                 0                  0                     0   \n",
       "...                 ...                ...                   ...   \n",
       "2882153               0                  0                     0   \n",
       "2882155               0                  0                     1   \n",
       "2882291               0                  0                     0   \n",
       "2882323               0                  0                     0   \n",
       "2882371               1                  0                     0   \n",
       "\n",
       "         hippocampus_18-20mo_f_2  \n",
       "64257                          0  \n",
       "64383                          1  \n",
       "64754                          0  \n",
       "64757                          0  \n",
       "64759                          0  \n",
       "...                          ...  \n",
       "2882153                        0  \n",
       "2882155                        0  \n",
       "2882291                        0  \n",
       "2882323                        0  \n",
       "2882371                        0  \n",
       "\n",
       "[43465 rows x 138 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.gene_novelty=='Fusion']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3af2222f-17ad-408e-9fc2-0a718aebe0b3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
