{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9888f570-a5e4-4e19-8471-e5ced81ee843",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5ce077b0-3d12-40b5-a7c3-23d115f58698",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "54dd6c46-9c9e-4b0a-abf2-bd842259d0c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "m_gtf = '../'+expand(config['ref']['talon']['gtf'], species='human')[0]\n",
    "m_ab_file = '../'+expand(config['lr']['talon']['ab'], species='mouse')[0]\n",
    "m_filt_ab_file = '../'+expand(config['lr']['talon']['filt_ab'], species='mouse')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "06a12190-0dec-499e-af94-453067d749ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Antisense</th>\n",
       "      <td>1886</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fusion</th>\n",
       "      <td>342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Intergenic</th>\n",
       "      <td>885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>118994</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              annot_transcript_id\n",
       "gene_novelty                     \n",
       "Antisense                    1886\n",
       "Fusion                        342\n",
       "Intergenic                    885\n",
       "Known                      118994"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.read_csv(m_filt_ab_file, sep='\\t')\n",
    "df2[['annot_transcript_id', 'gene_novelty']].groupby('gene_novelty').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9f93f66e-f23f-4d16-a2d0-2804f9c0e9dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = pr.read_gtf(m_gtf, duplicate_attr=True).df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6d7a8088-70ad-4859-bac6-1edabdd9cac9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>772</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>3073252</td>\n",
       "      <td>3074322</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000102693.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>TEC</td>\n",
       "      <td>4933401J01Rik-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000127109.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001343744.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>776</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>3102015</td>\n",
       "      <td>3102125</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000064842.1</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>snRNA</td>\n",
       "      <td>Gm26206-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116270</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>195240909</td>\n",
       "      <td>195241007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099208.1</td>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>...</td>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>Gm27940-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116271</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>195240909</td>\n",
       "      <td>195241007</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000099208.1</td>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>...</td>\n",
       "      <td>misc_RNA</td>\n",
       "      <td>Gm27940-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001316215.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116272</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>195259298</td>\n",
       "      <td>195259848</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104297.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116273</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>195259298</td>\n",
       "      <td>195259848</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104297.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>TEC</td>\n",
       "      <td>Gm38046-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000128998.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116274</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>195259298</td>\n",
       "      <td>195259848</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000104297.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>...</td>\n",
       "      <td>TEC</td>\n",
       "      <td>Gm38046-201</td>\n",
       "      <td>NA</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTMUST00000128998.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001337914.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>115503 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome   Source     Feature      Start        End Score Strand  \\\n",
       "772          chr1   HAVANA        gene    3073252    3074322     .      +   \n",
       "773          chr1   HAVANA  transcript    3073252    3074322     .      +   \n",
       "774          chr1   HAVANA        exon    3073252    3074322     .      +   \n",
       "775          chr1  ENSEMBL        gene    3102015    3102125     .      +   \n",
       "776          chr1  ENSEMBL  transcript    3102015    3102125     .      +   \n",
       "...           ...      ...         ...        ...        ...   ...    ...   \n",
       "116270       chr1  ENSEMBL  transcript  195240909  195241007     .      -   \n",
       "116271       chr1  ENSEMBL        exon  195240909  195241007     .      -   \n",
       "116272       chr1   HAVANA        gene  195259298  195259848     .      -   \n",
       "116273       chr1   HAVANA  transcript  195259298  195259848     .      -   \n",
       "116274       chr1   HAVANA        exon  195259298  195259848     .      -   \n",
       "\n",
       "       Frame               gene_id gene_type  ... transcript_type  \\\n",
       "772        .  ENSMUSG00000102693.1       TEC  ...             NaN   \n",
       "773        .  ENSMUSG00000102693.1       TEC  ...             TEC   \n",
       "774        .  ENSMUSG00000102693.1       TEC  ...             TEC   \n",
       "775        .  ENSMUSG00000064842.1     snRNA  ...             NaN   \n",
       "776        .  ENSMUSG00000064842.1     snRNA  ...           snRNA   \n",
       "...      ...                   ...       ...  ...             ...   \n",
       "116270     .  ENSMUSG00000099208.1  misc_RNA  ...        misc_RNA   \n",
       "116271     .  ENSMUSG00000099208.1  misc_RNA  ...        misc_RNA   \n",
       "116272     .  ENSMUSG00000104297.1       TEC  ...             NaN   \n",
       "116273     .  ENSMUSG00000104297.1       TEC  ...             TEC   \n",
       "116274     .  ENSMUSG00000104297.1       TEC  ...             TEC   \n",
       "\n",
       "          transcript_name transcript_support_level    tag  \\\n",
       "772                   NaN                      NaN    NaN   \n",
       "773     4933401J01Rik-201                       NA  basic   \n",
       "774     4933401J01Rik-201                       NA  basic   \n",
       "775                   NaN                      NaN    NaN   \n",
       "776           Gm26206-201                       NA  basic   \n",
       "...                   ...                      ...    ...   \n",
       "116270        Gm27940-201                       NA  basic   \n",
       "116271        Gm27940-201                       NA  basic   \n",
       "116272                NaN                      NaN    NaN   \n",
       "116273        Gm38046-201                       NA  basic   \n",
       "116274        Gm38046-201                       NA  basic   \n",
       "\n",
       "           havana_transcript exon_number               exon_id protein_id  \\\n",
       "772                      NaN         NaN                   NaN        NaN   \n",
       "773     OTTMUST00000127109.1         NaN                   NaN        NaN   \n",
       "774     OTTMUST00000127109.1           1  ENSMUSE00001343744.1        NaN   \n",
       "775                      NaN         NaN                   NaN        NaN   \n",
       "776                      NaN         NaN                   NaN        NaN   \n",
       "...                      ...         ...                   ...        ...   \n",
       "116270                   NaN         NaN                   NaN        NaN   \n",
       "116271                   NaN           1  ENSMUSE00001316215.1        NaN   \n",
       "116272                   NaN         NaN                   NaN        NaN   \n",
       "116273  OTTMUST00000128998.1         NaN                   NaN        NaN   \n",
       "116274  OTTMUST00000128998.1           1  ENSMUSE00001337914.1        NaN   \n",
       "\n",
       "       ccdsid  ont  \n",
       "772       NaN  NaN  \n",
       "773       NaN  NaN  \n",
       "774       NaN  NaN  \n",
       "775       NaN  NaN  \n",
       "776       NaN  NaN  \n",
       "...       ...  ...  \n",
       "116270    NaN  NaN  \n",
       "116271    NaN  NaN  \n",
       "116272    NaN  NaN  \n",
       "116273    NaN  NaN  \n",
       "116274    NaN  NaN  \n",
       "\n",
       "[115503 rows x 24 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf_df.loc[gtf_df.Chromosome=='chr1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "70397a92-d2b3-4435-ab01-671f37fb23e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Source, Feature, Start, End, Score, Strand, Frame, gene_id, gene_type, gene_name, level, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, protein_id, ccdsid, ont]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# are all \"overlapping_locus\" tags just genes?\n",
    "gtf_df.loc[(gtf_df.tag.str.contains('overlapping_locus'))&(gtf_df.Feature!='gene')]\n",
    "# yes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "baf3a207-bb28-4dcd-ba26-fdf3307b28d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>4859</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          tag\n",
       "Feature      \n",
       "gene     4859"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what entries have the \"overlapping_locus\" tag?\n",
    "temp = gtf_df.loc[gtf_df.tag.str.contains('overlapping_locus')].copy(deep=True)\n",
    "temp[['tag', 'Feature']].groupby('Feature').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "36963d20-638c-4c28-8979-36fc9f91312d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Source, Feature, Start, End, Score, Strand, Frame, gene_id, gene_type, gene_name, level, havana_gene, transcript_id, transcript_type, transcript_name, transcript_support_level, tag, havana_transcript, exon_number, exon_id, protein_id, ccdsid, ont]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 24 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# there are no entries with the \"readthrough_gene\" tag\n",
    "gtf_df['tag'] = gtf_df.tag.astype(str)\n",
    "gtf_df.loc[gtf_df.tag.str.contains('readthrough_gene')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "2aca2628-1f5d-4439-8ed1-d183a0c49a53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>2265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UTR</th>\n",
       "      <td>1425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>3948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>299</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>508</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              tag\n",
       "Feature          \n",
       "CDS          2265\n",
       "UTR          1425\n",
       "exon         3948\n",
       "start_codon   285\n",
       "stop_codon    299\n",
       "transcript    508"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what entries have the \"readthrough_transcript\" tag?\n",
    "temp = gtf_df.loc[gtf_df.tag.str.contains('readthrough_transcript')].copy(deep=True)\n",
    "temp[['tag', 'Feature']].groupby('Feature').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "06ebcb7b-5fcc-4f64-8a25-82158811b86f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Feature</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>2110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UTR</th>\n",
       "      <td>1368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>3731</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              tag\n",
       "Feature          \n",
       "CDS          2110\n",
       "UTR          1368\n",
       "exon         3731\n",
       "start_codon   269\n",
       "stop_codon    283\n",
       "transcript    477"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get a list of genes w/ the overlapping_locus tags\n",
    "o_genes = gtf_df.loc[gtf_df.tag.str.contains('overlapping_locus'), 'gene_id'].unique().tolist()\n",
    "\n",
    "# how many transcripts from genes w/ overlapping locus tag have the 'read_trhough_transcript' tag?\n",
    "temp = gtf_df.loc[gtf_df.gene_id.isin(o_genes)].copy(deep=True)\n",
    "temp = temp.loc[temp.tag.str.contains('readthrough_transcript')].copy(deep=True)\n",
    "temp[['tag', 'Feature']].groupby('Feature').count()\n",
    "\n",
    "# so there are a few transcrpts with the readthrough_transcript tag that \n",
    "# don't belong to the overlapping_locus genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2bfec637-f43c-4c09-b29a-91acba59f6fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# more throughough check for concordance of overlapping locus and \n",
    "temp = gtf_df.loc[gtf_df.Feature=='transcript'][['gene_id', 'gene_name', 'tag', 'transcript_id']].copy(deep=True)\n",
    "temp['readthrough_transcript'] = temp.tag.str.contains('readthrough_transcript')\n",
    "temp.drop('tag', inplace=True, axis=1)\n",
    "temp.drop_duplicates(inplace=True)\n",
    "\n",
    "temp2 = gtf_df.loc[gtf_df.Feature=='gene'][['gene_id', 'tag']].copy(deep=True)\n",
    "temp2['overlapping_locus'] = temp2.tag.str.contains('overlapping_locus')\n",
    "temp2.drop('tag', axis=1, inplace=True)\n",
    "temp2.drop_duplicates(inplace=True)\n",
    "\n",
    "temp = temp.merge(temp2, how='outer', on='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6d9b7465-02e7-44fd-abf6-d7051b92a32c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>readthrough_transcript</th>\n",
       "      <th>overlapping_locus</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>125121</td>\n",
       "      <td>124941</td>\n",
       "      <td>125121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>16413</td>\n",
       "      <td>16413</td>\n",
       "      <td>16413</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   readthrough_transcript  overlapping_locus  gene_id  gene_name  \\\n",
       "0                   False              False   125121     124941   \n",
       "1                   False               True    16413      16413   \n",
       "2                    True              False       31         31   \n",
       "3                    True               True      477        477   \n",
       "\n",
       "   transcript_id  \n",
       "0         125121  \n",
       "1          16413  \n",
       "2             31  \n",
       "3            477  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.groupby(['readthrough_transcript', 'overlapping_locus']).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c24011b4-888f-4c5e-b5a4-7d3b42802a7b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>readthrough_transcript</th>\n",
       "      <th>overlapping_locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>20181</th>\n",
       "      <td>ENSMUSG00000106407.1</td>\n",
       "      <td>Gm43439</td>\n",
       "      <td>ENSMUST00000138710.2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20702</th>\n",
       "      <td>ENSMUSG00000051777.6</td>\n",
       "      <td>Iqcj</td>\n",
       "      <td>ENSMUST00000063263.4</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20703</th>\n",
       "      <td>ENSMUSG00000102422.1</td>\n",
       "      <td>Iqschfp</td>\n",
       "      <td>ENSMUST00000182006.3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25730</th>\n",
       "      <td>ENSMUSG00000105103.1</td>\n",
       "      <td>Gm43191</td>\n",
       "      <td>ENSMUST00000140672.2</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35006</th>\n",
       "      <td>ENSMUSG00000105617.4</td>\n",
       "      <td>Gm43809</td>\n",
       "      <td>ENSMUST00000200694.3</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_id gene_name         transcript_id  \\\n",
       "20181  ENSMUSG00000106407.1   Gm43439  ENSMUST00000138710.2   \n",
       "20702  ENSMUSG00000051777.6      Iqcj  ENSMUST00000063263.4   \n",
       "20703  ENSMUSG00000102422.1   Iqschfp  ENSMUST00000182006.3   \n",
       "25730  ENSMUSG00000105103.1   Gm43191  ENSMUST00000140672.2   \n",
       "35006  ENSMUSG00000105617.4   Gm43809  ENSMUST00000200694.3   \n",
       "\n",
       "       readthrough_transcript  overlapping_locus  \n",
       "20181                    True              False  \n",
       "20702                    True              False  \n",
       "20703                    True              False  \n",
       "25730                    True              False  \n",
       "35006                    True              False  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[(temp.readthrough_transcript==True)&(temp.overlapping_locus==False)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "4c3b7d34-e4cc-41e7-9d2e-112656bd9bdf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>readthrough_transcript</th>\n",
       "      <th>overlapping_locus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>ENSMUST00000134384.7</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>ENSMUST00000027036.10</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>ENSMUST00000150971.7</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>ENSMUST00000119612.8</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>ENSMUSG00000025903.14</td>\n",
       "      <td>Lypla1</td>\n",
       "      <td>ENSMUST00000137887.7</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_id gene_name          transcript_id  \\\n",
       "193  ENSMUSG00000025903.14    Lypla1   ENSMUST00000134384.7   \n",
       "194  ENSMUSG00000025903.14    Lypla1  ENSMUST00000027036.10   \n",
       "195  ENSMUSG00000025903.14    Lypla1   ENSMUST00000150971.7   \n",
       "196  ENSMUSG00000025903.14    Lypla1   ENSMUST00000119612.8   \n",
       "197  ENSMUSG00000025903.14    Lypla1   ENSMUST00000137887.7   \n",
       "\n",
       "     readthrough_transcript  overlapping_locus  \n",
       "193                   False               True  \n",
       "194                   False               True  \n",
       "195                   False               True  \n",
       "196                   False               True  \n",
       "197                   False               True  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.loc[(temp.readthrough_transcript==False)&(temp.overlapping_locus==True)].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a60deee3-c751-4634-a823-7fb8603fe71f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38b438e0-8852-41bf-b087-0a9a08e9f376",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f83c30e8-d9af-43e2-9501-84bd616ecd5a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f3c95af2-fdbb-4864-bd03-d75a0274d548",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(m_ab_file, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "25c2326b-2d9d-4d15-976e-2dad8c3f2630",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Known', 'Antisense', 'Intergenic', 'Fusion'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.gene_novelty.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "48f82fe1-cb9c-4eef-bfa4-6159d8b6989e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>annot_gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Antisense</td>\n",
       "      <td>222873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fusion</td>\n",
       "      <td>43465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Intergenic</td>\n",
       "      <td>64034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Known</td>\n",
       "      <td>2552024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_novelty  annot_gene_id\n",
       "0    Antisense         222873\n",
       "1       Fusion          43465\n",
       "2   Intergenic          64034\n",
       "3        Known        2552024"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['annot_gene_id', 'gene_novelty']].groupby('gene_novelty').count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c878439f-c2df-465c-af0c-ee65a344197b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>adrenal_gland_1_2</th>\n",
       "      <th>cortex_5x_f_1_2</th>\n",
       "      <th>cortex_14d_f_2</th>\n",
       "      <th>cortex_2mo_m_2</th>\n",
       "      <th>gastroc_14d_f_2</th>\n",
       "      <th>gastroc_25d_m_2</th>\n",
       "      <th>gastroc_4d_f_2</th>\n",
       "      <th>heart_18-20mo_m_1</th>\n",
       "      <th>hippocampus_wt_m_2_1</th>\n",
       "      <th>hippocampus_18-20mo_f_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64257</th>\n",
       "      <td>55774</td>\n",
       "      <td>142467</td>\n",
       "      <td>TALONG000055774</td>\n",
       "      <td>TALONT000142467</td>\n",
       "      <td>TALONG000055774</td>\n",
       "      <td>TALONT000142467</td>\n",
       "      <td>13</td>\n",
       "      <td>1934</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64383</th>\n",
       "      <td>55808</td>\n",
       "      <td>142593</td>\n",
       "      <td>TALONG000055808</td>\n",
       "      <td>TALONT000142593</td>\n",
       "      <td>TALONG000055808</td>\n",
       "      <td>TALONT000142593</td>\n",
       "      <td>9</td>\n",
       "      <td>3331</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64754</th>\n",
       "      <td>55869</td>\n",
       "      <td>142964</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142964</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142964</td>\n",
       "      <td>8</td>\n",
       "      <td>2352</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64757</th>\n",
       "      <td>55869</td>\n",
       "      <td>142967</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142967</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142967</td>\n",
       "      <td>9</td>\n",
       "      <td>2518</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64759</th>\n",
       "      <td>55869</td>\n",
       "      <td>142969</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142969</td>\n",
       "      <td>TALONG000055869</td>\n",
       "      <td>TALONT000142969</td>\n",
       "      <td>4</td>\n",
       "      <td>1035</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882153</th>\n",
       "      <td>127991</td>\n",
       "      <td>2960363</td>\n",
       "      <td>TALONG000127991</td>\n",
       "      <td>TALONT002960363</td>\n",
       "      <td>TALONG000127991</td>\n",
       "      <td>TALONT002960363</td>\n",
       "      <td>1</td>\n",
       "      <td>1207</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882155</th>\n",
       "      <td>127991</td>\n",
       "      <td>2960365</td>\n",
       "      <td>TALONG000127991</td>\n",
       "      <td>TALONT002960365</td>\n",
       "      <td>TALONG000127991</td>\n",
       "      <td>TALONT002960365</td>\n",
       "      <td>6</td>\n",
       "      <td>1507</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NNC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882291</th>\n",
       "      <td>180891</td>\n",
       "      <td>2960501</td>\n",
       "      <td>TALONG000180891</td>\n",
       "      <td>TALONT002960501</td>\n",
       "      <td>TALONG000180891</td>\n",
       "      <td>TALONT002960501</td>\n",
       "      <td>18</td>\n",
       "      <td>3855</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882323</th>\n",
       "      <td>180891</td>\n",
       "      <td>2960533</td>\n",
       "      <td>TALONG000180891</td>\n",
       "      <td>TALONT002960533</td>\n",
       "      <td>TALONG000180891</td>\n",
       "      <td>TALONT002960533</td>\n",
       "      <td>1</td>\n",
       "      <td>1894</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Genomic</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882371</th>\n",
       "      <td>261990</td>\n",
       "      <td>2960581</td>\n",
       "      <td>TALONG000261990</td>\n",
       "      <td>TALONT002960581</td>\n",
       "      <td>TALONG000261990</td>\n",
       "      <td>TALONT002960581</td>\n",
       "      <td>9</td>\n",
       "      <td>1948</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>43465 rows × 138 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_ID  transcript_ID    annot_gene_id annot_transcript_id  \\\n",
       "64257      55774         142467  TALONG000055774     TALONT000142467   \n",
       "64383      55808         142593  TALONG000055808     TALONT000142593   \n",
       "64754      55869         142964  TALONG000055869     TALONT000142964   \n",
       "64757      55869         142967  TALONG000055869     TALONT000142967   \n",
       "64759      55869         142969  TALONG000055869     TALONT000142969   \n",
       "...          ...            ...              ...                 ...   \n",
       "2882153   127991        2960363  TALONG000127991     TALONT002960363   \n",
       "2882155   127991        2960365  TALONG000127991     TALONT002960365   \n",
       "2882291   180891        2960501  TALONG000180891     TALONT002960501   \n",
       "2882323   180891        2960533  TALONG000180891     TALONT002960533   \n",
       "2882371   261990        2960581  TALONG000261990     TALONT002960581   \n",
       "\n",
       "         annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "64257    TALONG000055774       TALONT000142467       13    1934       Fusion   \n",
       "64383    TALONG000055808       TALONT000142593        9    3331       Fusion   \n",
       "64754    TALONG000055869       TALONT000142964        8    2352       Fusion   \n",
       "64757    TALONG000055869       TALONT000142967        9    2518       Fusion   \n",
       "64759    TALONG000055869       TALONT000142969        4    1035       Fusion   \n",
       "...                  ...                   ...      ...     ...          ...   \n",
       "2882153  TALONG000127991       TALONT002960363        1    1207       Fusion   \n",
       "2882155  TALONG000127991       TALONT002960365        6    1507       Fusion   \n",
       "2882291  TALONG000180891       TALONT002960501       18    3855       Fusion   \n",
       "2882323  TALONG000180891       TALONT002960533        1    1894       Fusion   \n",
       "2882371  TALONG000261990       TALONT002960581        9    1948       Fusion   \n",
       "\n",
       "        transcript_novelty  ... adrenal_gland_1_2  cortex_5x_f_1_2  \\\n",
       "64257               Fusion  ...                 0                0   \n",
       "64383               Fusion  ...                 0                1   \n",
       "64754               Fusion  ...                 0                2   \n",
       "64757                  NIC  ...                 0                3   \n",
       "64759                  NIC  ...                 0                0   \n",
       "...                    ...  ...               ...              ...   \n",
       "2882153            Genomic  ...                 1                0   \n",
       "2882155                NNC  ...                 0                0   \n",
       "2882291                NIC  ...                 0                0   \n",
       "2882323            Genomic  ...                 1                0   \n",
       "2882371             Fusion  ...                 0                0   \n",
       "\n",
       "         cortex_14d_f_2  cortex_2mo_m_2  gastroc_14d_f_2  gastroc_25d_m_2  \\\n",
       "64257                 0               0                0                0   \n",
       "64383                 1               0                0                0   \n",
       "64754                 0               0                0                0   \n",
       "64757                 0               0                0                0   \n",
       "64759                 0               0                0                0   \n",
       "...                 ...             ...              ...              ...   \n",
       "2882153               0               0                1                0   \n",
       "2882155               0               0                0                0   \n",
       "2882291               0               1                0                0   \n",
       "2882323               0               0                0                0   \n",
       "2882371               0               0                0                0   \n",
       "\n",
       "         gastroc_4d_f_2  heart_18-20mo_m_1  hippocampus_wt_m_2_1  \\\n",
       "64257                 0                  0                     0   \n",
       "64383                 0                  0                     2   \n",
       "64754                 0                  0                     1   \n",
       "64757                 0                  0                     0   \n",
       "64759                 0                  0                     0   \n",
       "...                 ...                ...                   ...   \n",
       "2882153               0                  0                     0   \n",
       "2882155               0                  0                     1   \n",
       "2882291               0                  0                     0   \n",
       "2882323               0                  0                     0   \n",
       "2882371               1                  0                     0   \n",
       "\n",
       "         hippocampus_18-20mo_f_2  \n",
       "64257                          0  \n",
       "64383                          1  \n",
       "64754                          0  \n",
       "64757                          0  \n",
       "64759                          0  \n",
       "...                          ...  \n",
       "2882153                        0  \n",
       "2882155                        0  \n",
       "2882291                        0  \n",
       "2882323                        0  \n",
       "2882371                        0  \n",
       "\n",
       "[43465 rows x 138 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.gene_novelty=='Fusion']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3af2222f-17ad-408e-9fc2-0a718aebe0b3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
