{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d27a75b0-fa47-48b4-bf0a-8f582d04cb64",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from proc_revisions.utils import *\n",
    "from proc_revisions.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e718314b-45cd-42bf-a99a-39b665f29404",
   "metadata": {},
   "outputs": [],
   "source": [
    "od = '../../proc_revisions/'\n",
    "config_file = f'{od}/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1a0e6588-e6ce-4e57-99ae-24d3052113f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = od+expand(config['lr']['talon']['ab'], species='human')[0]\n",
    "talon_filt_ab = od+expand(config['lr']['talon']['filt_ab'], species='human')[0]\n",
    "filt_ab = od+expand(config['lr']['cerberus']['filt_ab'], species='human')[0]\n",
    "read_annot = od+expand(config['lr']['talon']['full_annot'], species='human')[0]\n",
    "t_metadata = od+expand(config['ref']['cerberus']['new_gtf_t_info'], species='human')[0]\n",
    "lib_meta = od+expand(config['lr']['meta'], species='human')[0]\n",
    "swan_file = od+expand(config['lr']['swan']['sg'], species='human')[0]\n",
    "read_lens = od+expand(config['lr']['read_len_meta'], species='human')[0]\n",
    "gtf = od+expand(config['lr']['cerberus']['gtf'], species='human')[0]\n",
    "\n",
    "sr_ab = od+expand(config['sr']['ab'], species='human')[0]\n",
    "\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "min_tpm = 1\n",
    "gene_subset = 'polya'\n",
    "obs_col = 'sample'\n",
    "go_gene_subset = 'protein_coding'\n",
    "\n",
    "m_lib_meta = od+expand(config['lr']['meta'], species='mouse')[0]\n",
    "m_ab = od+expand(config['lr']['talon']['ab'], species='mouse')[0]\n",
    "m_talon_filt_ab = od+expand(config['lr']['talon']['filt_ab'], species='mouse')[0]\n",
    "m_read_annot = od+expand(config['lr']['talon']['full_annot'], species='mouse')[0]\n",
    "m_t_metadata = od+expand(config['ref']['cerberus']['new_gtf_t_info'], species='mouse')[0]\n",
    "m_swan_file = od+expand(config['lr']['swan']['sg'], species='mouse')[0]\n",
    "\n",
    "mouse_ver = 'vM25_cerberus'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a5baf79b-7396-4710-8278-1fcbe264efd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(filt_ab, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ad609055-07ce-4d59-b430-f0c095f37d70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>brodmann_area_46_5_1</th>\n",
       "      <th>h1_1_2</th>\n",
       "      <th>hffc6_1_1</th>\n",
       "      <th>hl60_m1_72hr_1_1</th>\n",
       "      <th>h1_de_1_1</th>\n",
       "      <th>hl60_m0_1_2</th>\n",
       "      <th>aorta_2_1</th>\n",
       "      <th>h9_chondro_1_2</th>\n",
       "      <th>hepg2_1_1</th>\n",
       "      <th>mucosa_of_descending_colon_2_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>52125</td>\n",
       "      <td>187550,187550</td>\n",
       "      <td>ENSG00000121410.11</td>\n",
       "      <td>ENSG00000121410[1,1,2]</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG[1,1,2]</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1722.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>121</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>52125</td>\n",
       "      <td>187551,187551</td>\n",
       "      <td>ENSG00000121410.11</td>\n",
       "      <td>ENSG00000121410[1,4,2]</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG[1,4,2]</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2301.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>52125</td>\n",
       "      <td>1049645,1049645</td>\n",
       "      <td>ENSG00000121410.11</td>\n",
       "      <td>ENSG00000121410[1,5,2]</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG[1,5,2]</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1791.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>52125</td>\n",
       "      <td>187549</td>\n",
       "      <td>ENSG00000121410.11</td>\n",
       "      <td>ENSG00000121410[2,2,1]</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG[2,2,1]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2134.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>52125</td>\n",
       "      <td>187549</td>\n",
       "      <td>ENSG00000121410.11</td>\n",
       "      <td>ENSG00000121410[2,2,2]</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1BG[2,2,2]</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2134.0</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>10</td>\n",
       "      <td>1</td>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>35</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID    transcript_ID       annot_gene_id     annot_transcript_id  \\\n",
       "0    52125    187550,187550  ENSG00000121410.11  ENSG00000121410[1,1,2]   \n",
       "1    52125    187551,187551  ENSG00000121410.11  ENSG00000121410[1,4,2]   \n",
       "2    52125  1049645,1049645  ENSG00000121410.11  ENSG00000121410[1,5,2]   \n",
       "3    52125           187549  ENSG00000121410.11  ENSG00000121410[2,2,1]   \n",
       "4    52125           187549  ENSG00000121410.11  ENSG00000121410[2,2,2]   \n",
       "\n",
       "  annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "0            A1BG           A1BG[1,1,2]      8.0  1722.0        Known   \n",
       "1            A1BG           A1BG[1,4,2]      7.0  2301.0        Known   \n",
       "2            A1BG           A1BG[1,5,2]      9.0  1791.0        Known   \n",
       "3            A1BG           A1BG[2,2,1]      2.0  2134.0        Known   \n",
       "4            A1BG           A1BG[2,2,2]      2.0  2134.0        Known   \n",
       "\n",
       "  transcript_novelty  ... brodmann_area_46_5_1  h1_1_2  hffc6_1_1  \\\n",
       "0              Known  ...                    0       0          0   \n",
       "1              Known  ...                    0       0          0   \n",
       "2                NIC  ...                    0       0          0   \n",
       "3              Known  ...                    1       0          0   \n",
       "4              Known  ...                   10       1         12   \n",
       "\n",
       "   hl60_m1_72hr_1_1  h1_de_1_1  hl60_m0_1_2  aorta_2_1  h9_chondro_1_2  \\\n",
       "0                 0          0            0          0               0   \n",
       "1                 0          0            0          0               0   \n",
       "2                 0          0            0          0               0   \n",
       "3                 0          0            1          0               0   \n",
       "4                 0          3            1          0              35   \n",
       "\n",
       "   hepg2_1_1  mucosa_of_descending_colon_2_1  \n",
       "0        121                               0  \n",
       "1          0                               0  \n",
       "2          6                               0  \n",
       "3          1                               0  \n",
       "4          0                               1  \n",
       "\n",
       "[5 rows x 149 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fdc94e23-613c-42a2-ae18-73ac55c37dbb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>brodmann_area_46_5_1</th>\n",
       "      <th>h1_1_2</th>\n",
       "      <th>hffc6_1_1</th>\n",
       "      <th>hl60_m1_72hr_1_1</th>\n",
       "      <th>h1_de_1_1</th>\n",
       "      <th>hl60_m0_1_2</th>\n",
       "      <th>aorta_2_1</th>\n",
       "      <th>h9_chondro_1_2</th>\n",
       "      <th>hepg2_1_1</th>\n",
       "      <th>mucosa_of_descending_colon_2_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>200296</th>\n",
       "      <td>59968</td>\n",
       "      <td>214728,214728</td>\n",
       "      <td>TALONG000059968</td>\n",
       "      <td>TALONG000059968[1,1,1]</td>\n",
       "      <td>TALONG000059968</td>\n",
       "      <td>TALONG000059968[1,1,1]</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2661.0</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200299</th>\n",
       "      <td>61086</td>\n",
       "      <td>223753</td>\n",
       "      <td>TALONG000061086</td>\n",
       "      <td>TALONG000061086[1,1,1]</td>\n",
       "      <td>TALONG000061086</td>\n",
       "      <td>TALONG000061086[1,1,1]</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1367.0</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200300</th>\n",
       "      <td>61086</td>\n",
       "      <td>224292</td>\n",
       "      <td>TALONG000061086</td>\n",
       "      <td>TALONG000061086[1,2,1]</td>\n",
       "      <td>TALONG000061086</td>\n",
       "      <td>TALONG000061086[1,2,1]</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1363.0</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200301</th>\n",
       "      <td>61194</td>\n",
       "      <td>224539</td>\n",
       "      <td>TALONG000061194</td>\n",
       "      <td>TALONG000061194[1,1,1]</td>\n",
       "      <td>TALONG000061194</td>\n",
       "      <td>TALONG000061194[1,1,1]</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3650.0</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>200302</th>\n",
       "      <td>61194</td>\n",
       "      <td>225029</td>\n",
       "      <td>TALONG000061194</td>\n",
       "      <td>TALONG000061194[1,2,2]</td>\n",
       "      <td>TALONG000061194</td>\n",
       "      <td>TALONG000061194[1,2,2]</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1273.0</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 149 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_ID  transcript_ID    annot_gene_id     annot_transcript_id  \\\n",
       "200296    59968  214728,214728  TALONG000059968  TALONG000059968[1,1,1]   \n",
       "200299    61086         223753  TALONG000061086  TALONG000061086[1,1,1]   \n",
       "200300    61086         224292  TALONG000061086  TALONG000061086[1,2,1]   \n",
       "200301    61194         224539  TALONG000061194  TALONG000061194[1,1,1]   \n",
       "200302    61194         225029  TALONG000061194  TALONG000061194[1,2,2]   \n",
       "\n",
       "        annot_gene_name   annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "200296  TALONG000059968  TALONG000059968[1,1,1]     16.0  2661.0       Fusion   \n",
       "200299  TALONG000061086  TALONG000061086[1,1,1]      6.0  1367.0       Fusion   \n",
       "200300  TALONG000061086  TALONG000061086[1,2,1]      6.0  1363.0       Fusion   \n",
       "200301  TALONG000061194  TALONG000061194[1,1,1]      5.0  3650.0       Fusion   \n",
       "200302  TALONG000061194  TALONG000061194[1,2,2]      4.0  1273.0       Fusion   \n",
       "\n",
       "       transcript_novelty  ... brodmann_area_46_5_1  h1_1_2  hffc6_1_1  \\\n",
       "200296                ISM  ...                    5       5          2   \n",
       "200299                ISM  ...                    0       0          0   \n",
       "200300                ISM  ...                    0       1          0   \n",
       "200301                ISM  ...                    1       0          0   \n",
       "200302                ISM  ...                    0       0          0   \n",
       "\n",
       "        hl60_m1_72hr_1_1  h1_de_1_1  hl60_m0_1_2  aorta_2_1  h9_chondro_1_2  \\\n",
       "200296                 0         13            2          0               1   \n",
       "200299                 0          0            0          0               0   \n",
       "200300                 0          0            0          0               0   \n",
       "200301                 0          0            0          0               0   \n",
       "200302                 0          0            0          0               0   \n",
       "\n",
       "        hepg2_1_1  mucosa_of_descending_colon_2_1  \n",
       "200296          5                               3  \n",
       "200299          0                               0  \n",
       "200300          0                               0  \n",
       "200301          0                               0  \n",
       "200302          0                               0  \n",
       "\n",
       "[5 rows x 149 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get only the readthrough genes\n",
    "df = df.loc[df.gene_novelty=='Fusion']\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "99065f3f-5f67-4967-ab4d-9bf529ab235d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_fusion_sample_t_coords(ab, gtf, min_tpm, sample, species, ofile):\n",
    "    \"\"\"\n",
    "    Get genomic start / stop coords for transcripts expressed in\n",
    "    a given sample\n",
    "    \"\"\"\n",
    "\n",
    "    df = pd.read_csv(ab, sep='\\t')\n",
    "    tids = df.loc[df.gene_novelty=='Fusion', 'annot_transcript_id'].tolist()\n",
    "    print(len(tids))\n",
    "    df = get_det_table(df,\n",
    "                       groupby='sample',\n",
    "                       how='iso',\n",
    "                       min_tpm=min_tpm,\n",
    "                       species=species)\n",
    "    df = df.transpose()\n",
    "    tids2 = df.loc[df[sample]==True].index.tolist()\n",
    "    tids = list(set(tids)&set(tids2))\n",
    "\n",
    "    gtf_df = pr.read_gtf(gtf).df\n",
    "    gtf_df = gtf_df.loc[gtf_df.transcript_id.isin(tids)]\n",
    "    \n",
    "    gtf_df = gtf_df.loc[gtf_df.Feature=='transcript']\n",
    "    gtf_df = gtf_df[['Chromosome', 'Start', 'End', 'Strand', 'gene_id', 'transcript_id',]]\n",
    "    \n",
    "    gtgtf_dff.to_csv(ofile, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "33dfc2c8-0c07-4d73-8008-2e12d7d413fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "548\n",
      "Calculating iso TPM values\n",
      "Enforcing minimum TPM\n",
      "Total # isos detected: 237022\n",
      "# isos >= 1 tpm: 213032\n",
      "Number of isos reported: 213032\n",
      "Found 51 total samples\n"
     ]
    }
   ],
   "source": [
    "# get the transcripts that are fusion that are expressed in each sample\n",
    "df = get_fusion_sample_t_coords(filt_ab,\n",
    "               gtf,\n",
    "               1,\n",
    "               'gm12878',\n",
    "               'human',\n",
    "              ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a9b99f54-cab3-4b6e-8012-3c626b9ec6e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>108449</th>\n",
       "      <td>chr1</td>\n",
       "      <td>103525927</td>\n",
       "      <td>103555289</td>\n",
       "      <td>+</td>\n",
       "      <td>TALONG000169854</td>\n",
       "      <td>TALONG000169854[2,2,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108465</th>\n",
       "      <td>chr1</td>\n",
       "      <td>103525648</td>\n",
       "      <td>103555289</td>\n",
       "      <td>+</td>\n",
       "      <td>TALONG000169854</td>\n",
       "      <td>TALONG000169854[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108511</th>\n",
       "      <td>chr1</td>\n",
       "      <td>168576554</td>\n",
       "      <td>168581452</td>\n",
       "      <td>+</td>\n",
       "      <td>TALONG000189600</td>\n",
       "      <td>TALONG000189600[1,2,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108515</th>\n",
       "      <td>chr1</td>\n",
       "      <td>168576554</td>\n",
       "      <td>168581452</td>\n",
       "      <td>+</td>\n",
       "      <td>TALONG000189600</td>\n",
       "      <td>TALONG000189600[1,1,1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108533</th>\n",
       "      <td>chr1</td>\n",
       "      <td>192609265</td>\n",
       "      <td>192660310</td>\n",
       "      <td>+</td>\n",
       "      <td>TALONG000190575</td>\n",
       "      <td>TALONG000190575[1,1,1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      Start        End Strand          gene_id  \\\n",
       "108449       chr1  103525927  103555289      +  TALONG000169854   \n",
       "108465       chr1  103525648  103555289      +  TALONG000169854   \n",
       "108511       chr1  168576554  168581452      +  TALONG000189600   \n",
       "108515       chr1  168576554  168581452      +  TALONG000189600   \n",
       "108533       chr1  192609265  192660310      +  TALONG000190575   \n",
       "\n",
       "                 transcript_id  \n",
       "108449  TALONG000169854[2,2,1]  \n",
       "108465  TALONG000169854[1,1,1]  \n",
       "108511  TALONG000189600[1,2,1]  \n",
       "108515  TALONG000189600[1,1,1]  \n",
       "108533  TALONG000190575[1,1,1]  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de326d7b-d425-4901-99c5-419a604f0629",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8962754-7c16-4d79-b880-0e66d369db19",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8bc5cb9-81a7-4200-ad0f-78a44c300946",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
