{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "460e29c5-1880-4cca-a671-c0ea66056bb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from proc_revisions.utils import *\n",
    "from proc_revisions.plotting import *\n",
    "# from proc_revisions.mane_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3602f5f3-4332-487d-be6f-1dd13c183b81",
   "metadata": {},
   "outputs": [],
   "source": [
    "od = '../../proc_revisions/'\n",
    "config_file = f'{od}/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e1ed9148-6ff0-4385-a61c-73441c406ef2",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = 'human'\n",
    "ab = od+expand(config['lr']['talon']['fusion_fix']['ab'], species='human')[0]\n",
    "filt_ab = od+expand(config['lr']['cerberus']['filt_ab'], species=species)[0]\n",
    "gtf = od+expand(config['lr']['cerberus']['gtf'], species=species)[0]\n",
    "lib_meta = od+expand(config['lr']['meta'], species=species)[0]\n",
    "cerberus_h5 = od+expand(config['lr']['cerberus']['ca_triplets'], species=species, obs_col='sample')[0]\n",
    "swan_file = od+expand(config['lr']['swan']['sg'], species=species)[0]\n",
    "ref_gtf = od+expand(config['ref']['new_gtf'], species=species)[0]\n",
    "pp_summary = od+expand(config['lr']['protein_pred']['summary'], species='human')[0]\n",
    "gtf_file = od+expand(config['lr']['cerberus']['gtf'], species='human')[0]\n",
    "\n",
    "gtex_gtf_file = od+expand(config['gtex']['gtf'], species='human')[0]\n",
    "\n",
    "gene_subset = 'polya'\n",
    "min_tpm = 1\n",
    "\n",
    "species = 'mouse'\n",
    "m_filt_ab = od+expand(config['lr']['cerberus']['filt_ab'], species=species)[0]\n",
    "m_gtf = od+expand(config['lr']['cerberus']['gtf'], species=species)[0]\n",
    "m_lib_meta = od+expand(config['lr']['meta'], species=species)[0]\n",
    "m_cerberus_h5 = od+expand(config['lr']['cerberus']['ca_triplets'], species=species, obs_col='sample')[0]\n",
    "m_swan_file = od+expand(config['lr']['swan']['sg'], species=species)[0]\n",
    "m_ref_gtf = od+expand(config['ref']['new_gtf'], species=species)[0]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "756df5df-349d-4152-8053-b7eb69842356",
   "metadata": {},
   "source": [
    "## Human -- how many v47 transcripts do we detect by biotype, novelty?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9dd7ad60-4e3b-49f4-b9bb-4d419bd54b18",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ic_id(df):\n",
    "    coord_col = 'ic'\n",
    "    if 'ic' not in df.columns:\n",
    "        coord_col = 'Coordinates'\n",
    "    df['ic_id'] = df.Chromosome.astype(str)+'_'+\\\n",
    "                  df.Strand.astype(str)+'_'+\\\n",
    "                  df[coord_col].astype(str)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "913f0109-1ece-46bb-9fe1-26976af57c38",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gtf = '/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/proc_revisions/ref/human/gencode.vM36.annotation.gtf'\n",
    "ref_gtf_df = pr.read_gtf(ref_gtf)\n",
    "ref_ics = cerberus.get_ic(ref_gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9bf1d41b-eca4-40e4-91a1-e991c58a7d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gtf_df = ref_gtf_df.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "273a0b24-ebdd-45e5-8b92-ab62a9f93adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add biotype \n",
    "ref_gtf_df[['gene_id', 'gene_type']].groupby('gene_type').nunique()\n",
    "ref_ics = ref_ics.merge(ref_gtf_df[['gene_id', 'gene_type']].drop_duplicates(),\n",
    "                        how='left', \n",
    "                        on='gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55cbb5ed-3df3-4056-b6b3-7949443dc519",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "573c19c2-b2f5-425f-b24e-7f024ae337e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../proc_revisions/data/mouse/lr/cerberus/cerberus.gtf\n"
     ]
    }
   ],
   "source": [
    "print(m_gtf)\n",
    "gtf_df = pr.read_gtf(m_gtf)\n",
    "ics = cerberus.get_ic(gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "bd1a3ab7-a14d-4acc-b2fd-710bff1adc2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = gtf_df.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d370bbd6-fd23-4e5e-a87e-96b6e6cbf2ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get ids for uniq chr+strand+ic pairs\n",
    "ref_ics = get_ic_id(ref_ics)\n",
    "ics = get_ic_id(ics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "709cc386-7ce8-4631-a2d1-155d1b402697",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,1,1]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,1,2]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,1,3]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,1,8]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,10,1]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,10,2]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,10,8]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,11,1]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,11,2]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,11,7]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,11,8]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,12,1]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,12,2]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,12,3]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,12,7]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,12,8]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,2,1]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,2,2]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,2,3]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,2,8]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,7,1]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,7,2]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,7,3]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,7,8]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,9,1]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,9,2]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001143[1,9,8]</td>\n",
       "      <td>ENSMUSG00000001143.13</td>\n",
       "      <td>36445049-36443609-36443490-36439726-36439608-3...</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Strand               transcript_id                gene_id  \\\n",
       "1        chr1      -   ENSMUSG00000001143[1,1,1]  ENSMUSG00000001143.13   \n",
       "2        chr1      -   ENSMUSG00000001143[1,1,2]  ENSMUSG00000001143.13   \n",
       "3        chr1      -   ENSMUSG00000001143[1,1,3]  ENSMUSG00000001143.13   \n",
       "4        chr1      -   ENSMUSG00000001143[1,1,8]  ENSMUSG00000001143.13   \n",
       "5        chr1      -  ENSMUSG00000001143[1,10,1]  ENSMUSG00000001143.13   \n",
       "6        chr1      -  ENSMUSG00000001143[1,10,2]  ENSMUSG00000001143.13   \n",
       "7        chr1      -  ENSMUSG00000001143[1,10,8]  ENSMUSG00000001143.13   \n",
       "8        chr1      -  ENSMUSG00000001143[1,11,1]  ENSMUSG00000001143.13   \n",
       "9        chr1      -  ENSMUSG00000001143[1,11,2]  ENSMUSG00000001143.13   \n",
       "10       chr1      -  ENSMUSG00000001143[1,11,7]  ENSMUSG00000001143.13   \n",
       "11       chr1      -  ENSMUSG00000001143[1,11,8]  ENSMUSG00000001143.13   \n",
       "12       chr1      -  ENSMUSG00000001143[1,12,1]  ENSMUSG00000001143.13   \n",
       "13       chr1      -  ENSMUSG00000001143[1,12,2]  ENSMUSG00000001143.13   \n",
       "14       chr1      -  ENSMUSG00000001143[1,12,3]  ENSMUSG00000001143.13   \n",
       "15       chr1      -  ENSMUSG00000001143[1,12,7]  ENSMUSG00000001143.13   \n",
       "16       chr1      -  ENSMUSG00000001143[1,12,8]  ENSMUSG00000001143.13   \n",
       "19       chr1      -   ENSMUSG00000001143[1,2,1]  ENSMUSG00000001143.13   \n",
       "20       chr1      -   ENSMUSG00000001143[1,2,2]  ENSMUSG00000001143.13   \n",
       "21       chr1      -   ENSMUSG00000001143[1,2,3]  ENSMUSG00000001143.13   \n",
       "22       chr1      -   ENSMUSG00000001143[1,2,8]  ENSMUSG00000001143.13   \n",
       "24       chr1      -   ENSMUSG00000001143[1,7,1]  ENSMUSG00000001143.13   \n",
       "25       chr1      -   ENSMUSG00000001143[1,7,2]  ENSMUSG00000001143.13   \n",
       "26       chr1      -   ENSMUSG00000001143[1,7,3]  ENSMUSG00000001143.13   \n",
       "27       chr1      -   ENSMUSG00000001143[1,7,8]  ENSMUSG00000001143.13   \n",
       "28       chr1      -   ENSMUSG00000001143[1,9,1]  ENSMUSG00000001143.13   \n",
       "29       chr1      -   ENSMUSG00000001143[1,9,2]  ENSMUSG00000001143.13   \n",
       "30       chr1      -   ENSMUSG00000001143[1,9,8]  ENSMUSG00000001143.13   \n",
       "\n",
       "                                                   ic  \\\n",
       "1   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "2   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "3   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "4   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "5   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "6   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "7   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "8   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "9   36445049-36443609-36443490-36439726-36439608-3...   \n",
       "10  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "11  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "12  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "13  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "14  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "15  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "16  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "19  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "20  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "21  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "22  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "24  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "25  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "26  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "27  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "28  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "29  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "30  36445049-36443609-36443490-36439726-36439608-3...   \n",
       "\n",
       "                                                ic_id  \n",
       "1   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "2   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "3   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "4   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "5   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "6   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "7   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "8   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "9   chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "10  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "11  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "12  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "13  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "14  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "15  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "16  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "19  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "20  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "21  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "22  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "24  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "25  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "26  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "27  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "28  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "29  chr1_-_36445049-36443609-36443490-36439726-364...  \n",
       "30  chr1_-_36445049-36443609-36443490-36439726-364...  "
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ics = cerberus.get_ic(pr.PyRanges(gtf_df))\n",
    "ics = get_ic_id(ics)\n",
    "ics.loc[ics.ic_id.str.startswith('chr1_-_36445049-36443609-36443490-36439726-364')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "73a8310b-05c7-4552-bc2e-710ba2580343",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000001171.13</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>36484130-36482690-36482571-36478807-36478689-3...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36484130-36482690-36482571-36478807-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000115011.8</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>36484130-36482690-36482571-36478807-36478689-3...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36484130-36482690-36482571-36478807-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>886</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000123583.2</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>36484130-36483223-36483117-36482690-36482600-3...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36484130-36483223-36483117-36482690-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>908</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000125304.8</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>36484130-36482690-36482571-36478807-36478689-3...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36484130-36482690-36482571-36478807-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1120</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000134594.8</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>36478689-36477913-36477830-36467542-36467380-3...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36478689-36477913-36477830-36467542-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1477</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000152088.2</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>36467380-36467297-36467182-36464038</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36467380-36467297-36467182-36464038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3610</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000192969.6</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>36478689-36477913-36477830-36467542-36467380-3...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36478689-36477913-36477830-36467542-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3697</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000193502.2</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>-</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome Strand          transcript_id                gene_id  \\\n",
       "2          chr1      -  ENSMUST00000001171.13  ENSMUSG00000001143.14   \n",
       "777        chr1      -   ENSMUST00000115011.8  ENSMUSG00000001143.14   \n",
       "886        chr1      -   ENSMUST00000123583.2  ENSMUSG00000001143.14   \n",
       "908        chr1      -   ENSMUST00000125304.8  ENSMUSG00000001143.14   \n",
       "1120       chr1      -   ENSMUST00000134594.8  ENSMUSG00000001143.14   \n",
       "1477       chr1      -   ENSMUST00000152088.2  ENSMUSG00000001143.14   \n",
       "3610       chr1      -   ENSMUST00000192969.6  ENSMUSG00000001143.14   \n",
       "3697       chr1      -   ENSMUST00000193502.2  ENSMUSG00000001143.14   \n",
       "\n",
       "                                                     ic       gene_type  \\\n",
       "2     36484130-36482690-36482571-36478807-36478689-3...  protein_coding   \n",
       "777   36484130-36482690-36482571-36478807-36478689-3...  protein_coding   \n",
       "886   36484130-36483223-36483117-36482690-36482600-3...  protein_coding   \n",
       "908   36484130-36482690-36482571-36478807-36478689-3...  protein_coding   \n",
       "1120  36478689-36477913-36477830-36467542-36467380-3...  protein_coding   \n",
       "1477                36467380-36467297-36467182-36464038  protein_coding   \n",
       "3610  36478689-36477913-36477830-36467542-36467380-3...  protein_coding   \n",
       "3697                                                  -  protein_coding   \n",
       "\n",
       "                                                  ic_id  \n",
       "2     chr1_-_36484130-36482690-36482571-36478807-364...  \n",
       "777   chr1_-_36484130-36482690-36482571-36478807-364...  \n",
       "886   chr1_-_36484130-36483223-36483117-36482690-364...  \n",
       "908   chr1_-_36484130-36482690-36482571-36478807-364...  \n",
       "1120  chr1_-_36478689-36477913-36477830-36467542-364...  \n",
       "1477         chr1_-_36467380-36467297-36467182-36464038  \n",
       "3610  chr1_-_36478689-36477913-36477830-36467542-364...  \n",
       "3697                                           chr1_-_-  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_ics.loc[ref_ics.gene_id=='ENSMUSG00000001143.14']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "a5286570-18a9-4403-96ae-ebb24b378738",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>exon_number</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>91282</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>36458951</td>\n",
       "      <td>36484352</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91283</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>36458951</td>\n",
       "      <td>36478749</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lman2l-207</td>\n",
       "      <td>3</td>\n",
       "      <td>cds_start_NF</td>\n",
       "      <td>OTTMUST00000127607.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSMUSP00000141806.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91284</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>36478689</td>\n",
       "      <td>36478749</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lman2l-207</td>\n",
       "      <td>3</td>\n",
       "      <td>cds_start_NF</td>\n",
       "      <td>OTTMUST00000127607.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001344612.2</td>\n",
       "      <td>ENSMUSP00000141806.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91285</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>36478689</td>\n",
       "      <td>36478749</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lman2l-207</td>\n",
       "      <td>3</td>\n",
       "      <td>cds_start_NF</td>\n",
       "      <td>OTTMUST00000127607.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001344612.2</td>\n",
       "      <td>ENSMUSP00000141806.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91286</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>36477830</td>\n",
       "      <td>36477913</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>Lman2l-207</td>\n",
       "      <td>3</td>\n",
       "      <td>cds_start_NF</td>\n",
       "      <td>OTTMUST00000127607.1</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSE00001245826.2</td>\n",
       "      <td>ENSMUSP00000141806.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91391</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>36478689</td>\n",
       "      <td>36478807</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>Lman2l-203</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000051152.1</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSE00001241012.2</td>\n",
       "      <td>ENSMUSP00000137344.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91392</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>36477830</td>\n",
       "      <td>36477913</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>Lman2l-203</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000051152.1</td>\n",
       "      <td>5</td>\n",
       "      <td>ENSMUSE00001212118.2</td>\n",
       "      <td>ENSMUSP00000137344.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91393</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>36467472</td>\n",
       "      <td>36467542</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>nonsense_mediated_decay</td>\n",
       "      <td>Lman2l-203</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000051152.1</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSMUSE00000807125.2</td>\n",
       "      <td>ENSMUSP00000137344.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91394</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>36482825</td>\n",
       "      <td>36484338</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>Lman2l-208</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000127134.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91395</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>36482825</td>\n",
       "      <td>36484338</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>Lman2l-208</td>\n",
       "      <td>NA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OTTMUST00000127134.1</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSE00001337714.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>114 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome  Source     Feature     Start       End Score Strand Frame  \\\n",
       "91282       chr1  HAVANA        gene  36458951  36484352     .      -     .   \n",
       "91283       chr1  HAVANA  transcript  36458951  36478749     .      -     .   \n",
       "91284       chr1  HAVANA        exon  36478689  36478749     .      -     .   \n",
       "91285       chr1  HAVANA         CDS  36478689  36478749     .      -     2   \n",
       "91286       chr1  HAVANA        exon  36477830  36477913     .      -     .   \n",
       "...          ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "91391       chr1  HAVANA         UTR  36478689  36478807     .      -     .   \n",
       "91392       chr1  HAVANA         UTR  36477830  36477913     .      -     .   \n",
       "91393       chr1  HAVANA         UTR  36467472  36467542     .      -     .   \n",
       "91394       chr1  HAVANA  transcript  36482825  36484338     .      -     .   \n",
       "91395       chr1  HAVANA        exon  36482825  36484338     .      -     .   \n",
       "\n",
       "                     gene_id       gene_type  ...          transcript_type  \\\n",
       "91282  ENSMUSG00000001143.14  protein_coding  ...                      NaN   \n",
       "91283  ENSMUSG00000001143.14  protein_coding  ...           protein_coding   \n",
       "91284  ENSMUSG00000001143.14  protein_coding  ...           protein_coding   \n",
       "91285  ENSMUSG00000001143.14  protein_coding  ...           protein_coding   \n",
       "91286  ENSMUSG00000001143.14  protein_coding  ...           protein_coding   \n",
       "...                      ...             ...  ...                      ...   \n",
       "91391  ENSMUSG00000001143.14  protein_coding  ...  nonsense_mediated_decay   \n",
       "91392  ENSMUSG00000001143.14  protein_coding  ...  nonsense_mediated_decay   \n",
       "91393  ENSMUSG00000001143.14  protein_coding  ...  nonsense_mediated_decay   \n",
       "91394  ENSMUSG00000001143.14  protein_coding  ...          retained_intron   \n",
       "91395  ENSMUSG00000001143.14  protein_coding  ...          retained_intron   \n",
       "\n",
       "      transcript_name transcript_support_level           tag  \\\n",
       "91282             NaN                      NaN           NaN   \n",
       "91283      Lman2l-207                        3  cds_start_NF   \n",
       "91284      Lman2l-207                        3  cds_start_NF   \n",
       "91285      Lman2l-207                        3  cds_start_NF   \n",
       "91286      Lman2l-207                        3  cds_start_NF   \n",
       "...               ...                      ...           ...   \n",
       "91391      Lman2l-203                        5           NaN   \n",
       "91392      Lman2l-203                        5           NaN   \n",
       "91393      Lman2l-203                        5           NaN   \n",
       "91394      Lman2l-208                       NA           NaN   \n",
       "91395      Lman2l-208                       NA           NaN   \n",
       "\n",
       "          havana_transcript exon_number               exon_id  \\\n",
       "91282                   NaN         NaN                   NaN   \n",
       "91283  OTTMUST00000127607.1         NaN                   NaN   \n",
       "91284  OTTMUST00000127607.1           1  ENSMUSE00001344612.2   \n",
       "91285  OTTMUST00000127607.1           1  ENSMUSE00001344612.2   \n",
       "91286  OTTMUST00000127607.1           2  ENSMUSE00001245826.2   \n",
       "...                     ...         ...                   ...   \n",
       "91391  OTTMUST00000051152.1           4  ENSMUSE00001241012.2   \n",
       "91392  OTTMUST00000051152.1           5  ENSMUSE00001212118.2   \n",
       "91393  OTTMUST00000051152.1           6  ENSMUSE00000807125.2   \n",
       "91394  OTTMUST00000127134.1         NaN                   NaN   \n",
       "91395  OTTMUST00000127134.1           1  ENSMUSE00001337714.2   \n",
       "\n",
       "                 protein_id ccdsid  ont  \n",
       "91282                   NaN    NaN  NaN  \n",
       "91283  ENSMUSP00000141806.2    NaN  NaN  \n",
       "91284  ENSMUSP00000141806.2    NaN  NaN  \n",
       "91285  ENSMUSP00000141806.2    NaN  NaN  \n",
       "91286  ENSMUSP00000141806.2    NaN  NaN  \n",
       "...                     ...    ...  ...  \n",
       "91391  ENSMUSP00000137344.2    NaN  NaN  \n",
       "91392  ENSMUSP00000137344.2    NaN  NaN  \n",
       "91393  ENSMUSP00000137344.2    NaN  NaN  \n",
       "91394                   NaN    NaN  NaN  \n",
       "91395                   NaN    NaN  NaN  \n",
       "\n",
       "[114 rows x 25 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_gtf_df.loc[ref_gtf_df.gene_id.str.startswith('ENSMUSG00000001143')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dc58bff-dbaf-4a88-b8a1-99f985414ea2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c4995c8-ef59-4f0c-bcfb-acfef17e2883",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf5cc140-59ad-4b4f-9dc0-3658581bb261",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6d77046-e086-42ca-b331-e96f2ddbb22d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "afcf9c87-efb2-418c-99f8-46e1e1aa5b44",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/proc_revisions/utils.py:2110: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[tpm_col] = (df[d]*1000000)/df[total_col]\n",
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/proc_revisions/utils.py:2109: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[total_col] = df[d].sum()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enforcing minimum TPM\n",
      "Total # isos detected: 153398\n",
      "# isos >= 1 tpm: 142504\n",
      "Number of isos reported: 142504\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>gastroc_36d_f_1</th>\n",
       "      <th>hippocampus_2mo_m_2</th>\n",
       "      <th>gastroc_14d_f_1</th>\n",
       "      <th>cortex_wt_m_1_2</th>\n",
       "      <th>heart_2mo_f_2</th>\n",
       "      <th>adrenal_25d_m_2</th>\n",
       "      <th>heart_14d_f_1</th>\n",
       "      <th>adrenal_10d_m_2</th>\n",
       "      <th>...</th>\n",
       "      <th>hippocampus_18-20mo_f_1</th>\n",
       "      <th>gastroc_18-20mo_f_1</th>\n",
       "      <th>gastroc_25d_m_1</th>\n",
       "      <th>cortex_wt_f_1_2</th>\n",
       "      <th>adrenal_36d_m_1</th>\n",
       "      <th>forelimb_e11_1_1</th>\n",
       "      <th>adrenal_gland_1_2</th>\n",
       "      <th>hippocampus_18-20mo_m_2</th>\n",
       "      <th>gid</th>\n",
       "      <th>gene_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000051951[2,2,3]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.586226</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.307316</td>\n",
       "      <td>ENSMUSG00000051951</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000089699[1,1,1]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSMUSG00000089699</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000103161[1,1,1]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSMUSG00000103161</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000025902[1,2,1]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.543755</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000025902[1,2,3]</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>ENSMUSG00000025902</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 131 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index        annot_transcript_id  gastroc_36d_f_1  hippocampus_2mo_m_2  \\\n",
       "0      0  ENSMUSG00000051951[2,2,3]              0.0             1.586226   \n",
       "1      1  ENSMUSG00000089699[1,1,1]              0.0             0.000000   \n",
       "2      2  ENSMUSG00000103161[1,1,1]              0.0             0.000000   \n",
       "3      3  ENSMUSG00000025902[1,2,1]              0.0             0.000000   \n",
       "4      4  ENSMUSG00000025902[1,2,3]              0.0             0.000000   \n",
       "\n",
       "   gastroc_14d_f_1  cortex_wt_m_1_2  heart_2mo_f_2  adrenal_25d_m_2  \\\n",
       "0              0.0              0.0            0.0              0.0   \n",
       "1              0.0              0.0            0.0              0.0   \n",
       "2              0.0              0.0            0.0              0.0   \n",
       "3              0.0              0.0            0.0              0.0   \n",
       "4              0.0              0.0            0.0              0.0   \n",
       "\n",
       "   heart_14d_f_1  adrenal_10d_m_2  ...  hippocampus_18-20mo_f_1  \\\n",
       "0            0.0              0.0  ...                      0.0   \n",
       "1            0.0              0.0  ...                      0.0   \n",
       "2            0.0              0.0  ...                      0.0   \n",
       "3            0.0              0.0  ...                      0.0   \n",
       "4            0.0              0.0  ...                      0.0   \n",
       "\n",
       "   gastroc_18-20mo_f_1  gastroc_25d_m_1  cortex_wt_f_1_2  adrenal_36d_m_1  \\\n",
       "0                  0.0              0.0         0.000000              0.0   \n",
       "1                  0.0              0.0         0.000000              0.0   \n",
       "2                  0.0              0.0         0.000000              0.0   \n",
       "3                  0.0              0.0         0.543755              0.0   \n",
       "4                  0.0              0.0         0.000000              0.0   \n",
       "\n",
       "   forelimb_e11_1_1  adrenal_gland_1_2  hippocampus_18-20mo_m_2  \\\n",
       "0               0.0                0.0                 1.307316   \n",
       "1               0.0                0.0                 0.000000   \n",
       "2               0.0                0.0                 0.000000   \n",
       "3               0.0                0.0                 0.000000   \n",
       "4               0.0                0.0                 0.000000   \n",
       "\n",
       "                  gid  gene_novelty  \n",
       "0  ENSMUSG00000051951         Known  \n",
       "1  ENSMUSG00000089699         Known  \n",
       "2  ENSMUSG00000103161         Known  \n",
       "3  ENSMUSG00000025902         Known  \n",
       "4  ENSMUSG00000025902         Known  \n",
       "\n",
       "[5 rows x 131 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get expressed novel genes\n",
    "df = pd.read_csv(m_filt_ab, sep='\\t')\n",
    "df, _ = get_tpm_table(df, \n",
    "                      how='iso',\n",
    "                      min_tpm=1,\n",
    "                      gene_subset=None,\n",
    "                      species='mouse')\n",
    "\n",
    "df.reset_index(inplace=True)\n",
    "df['gid'] = df.annot_transcript_id.str.split('[', expand=True)[0]\n",
    "df.head()\n",
    "df.reset_index(inplace=True)\n",
    "df.rename({'gid_stable':'gid'}, axis=1, inplace=True)\n",
    "ab_df = pd.read_csv(m_filt_ab, sep='\\t')\n",
    "ab_df['gid'] = ab_df.annot_transcript_id.str.split('[', expand=True)[0]\n",
    "ab_df = ab_df[['gid', 'gene_novelty']].drop_duplicates()\n",
    "ab_df.loc[ab_df.gid.duplicated(keep=False)].sort_values(by='gid')\n",
    "assert len(ab_df.loc[ab_df.gid.duplicated(keep=False)].index) == 0\n",
    "df = df.merge(ab_df, how='left', on='gid')\n",
    "df.loc[df.gene_novelty=='Fusion', 'gene_novelty'] = 'Readthrough'\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8a7b18d1-58b9-4955-8294-54152855cb74",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Known', 'Readthrough', 'Intergenic'], dtype=object)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.gene_novelty.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "46292b9a-9765-4d1c-a8cb-a5d1dbe70e01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_t</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intergenic</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>142366</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Readthrough</th>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 n_t\n",
       "gene_novelty        \n",
       "Intergenic        29\n",
       "Known         142366\n",
       "Readthrough      109"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['gene_novelty', 'annot_transcript_id']].groupby('gene_novelty').nunique().rename({'annot_transcript_id':'n_t'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "0c5734fa-f1ce-4f11-8d14-97390f43394a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_g</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intergenic</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Known</th>\n",
       "      <td>24913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Readthrough</th>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                n_g\n",
       "gene_novelty       \n",
       "Intergenic       24\n",
       "Known         24913\n",
       "Readthrough      55"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[['gene_novelty', 'gid']].groupby('gene_novelty').nunique().rename({'gid':'n_g'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "be71a979-5600-4e73-a1aa-5a666071bafa",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.rename({'annot_transcript_id':'transcript_id'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3701943e-8510-47f2-9658-1977531dfa20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add gene novelties to each ic\n",
    "ics = ics.merge(df[['gene_novelty', 'transcript_id']],\n",
    "                how='inner', \n",
    "                on='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d8c4e0ba-a313-4633-9a88-537118603882",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "142504\n"
     ]
    }
   ],
   "source": [
    "print(len(ics.loc[ics.gene_novelty.notnull()].transcript_id.unique()))\n",
    "ics = ics.loc[ics.gene_novelty.notnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "802573b8-4cba-4e69-8196-82a1df68a12e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# limit to non-monoexonic\n",
    "ics.loc[ics.ic_id.str.endswith('-')].head()\n",
    "ics = ics.loc[~ics.ic_id.str.endswith('-')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b8e28aa4-3016-41f0-8758-428d36d481c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# limit to non dupe\n",
    "temp = ics[['gene_novelty', 'ic_id']].drop_duplicates()\n",
    "temp = temp.loc[temp.ic_id.duplicated(keep=False)].sort_values(by='ic_id')\n",
    "assert len(temp.index) == 0\n",
    "\n",
    "ics = ics[['gene_novelty', 'ic_id']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5bb38bb0-9e7d-40ae-a034-b4802ac9aa2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Known</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Known</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Known</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Known</td>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Known</td>\n",
       "      <td>chr1_-_36445049-36443609-36443519-36439726-364...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_novelty                                              ic_id\n",
       "0         Known  chr1_-_36445049-36443609-36443490-36439726-364...\n",
       "4         Known  chr1_-_36445049-36443609-36443490-36439726-364...\n",
       "7         Known  chr1_-_36445049-36443609-36443490-36439726-364...\n",
       "11        Known  chr1_-_36445049-36443609-36443490-36439726-364...\n",
       "15        Known  chr1_-_36445049-36443609-36443519-36439726-364..."
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ics.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "fa87120b-dd8f-4535-a190-f7668816ddc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000266.9</td>\n",
       "      <td>ENSMUSG00000026535.10</td>\n",
       "      <td>173810239-173805003-173804830-173802712-173802...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_173810239-173805003-173804830-173802712...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000000834.4</td>\n",
       "      <td>ENSMUSG00000000817.11</td>\n",
       "      <td>161615512-161614719-161614673-161610573-161610...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_161615512-161614719-161614673-161610573...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000001171.13</td>\n",
       "      <td>ENSMUSG00000001143.14</td>\n",
       "      <td>36484130-36482690-36482571-36478807-36478689-3...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36484130-36482690-36482571-36478807-364...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000001172.12</td>\n",
       "      <td>ENSMUSG00000079610.10</td>\n",
       "      <td>36586166-36581946-36581842-36581143-36580939-3...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_36586166-36581946-36581842-36581143-365...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUST00000001339.6</td>\n",
       "      <td>ENSMUSG00000001305.6</td>\n",
       "      <td>186481360-186472060-186471800-186469882-186469...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>chr1_-_186481360-186472060-186471800-186469882...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand          transcript_id                gene_id  \\\n",
       "0       chr1      -   ENSMUST00000000266.9  ENSMUSG00000026535.10   \n",
       "1       chr1      -   ENSMUST00000000834.4  ENSMUSG00000000817.11   \n",
       "2       chr1      -  ENSMUST00000001171.13  ENSMUSG00000001143.14   \n",
       "3       chr1      -  ENSMUST00000001172.12  ENSMUSG00000079610.10   \n",
       "4       chr1      -   ENSMUST00000001339.6   ENSMUSG00000001305.6   \n",
       "\n",
       "                                                  ic       gene_type  \\\n",
       "0  173810239-173805003-173804830-173802712-173802...  protein_coding   \n",
       "1  161615512-161614719-161614673-161610573-161610...  protein_coding   \n",
       "2  36484130-36482690-36482571-36478807-36478689-3...  protein_coding   \n",
       "3  36586166-36581946-36581842-36581143-36580939-3...  protein_coding   \n",
       "4  186481360-186472060-186471800-186469882-186469...  protein_coding   \n",
       "\n",
       "                                               ic_id  \n",
       "0  chr1_-_173810239-173805003-173804830-173802712...  \n",
       "1  chr1_-_161615512-161614719-161614673-161610573...  \n",
       "2  chr1_-_36484130-36482690-36482571-36478807-364...  \n",
       "3  chr1_-_36586166-36581946-36581842-36581143-365...  \n",
       "4  chr1_-_186481360-186472060-186471800-186469882...  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_ics.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "efcadb43-10f8-4565-b7b0-d45dd19c82f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge = ics[['ic_id', 'gene_novelty']].merge(\n",
    "            ref_ics[['ic_id', 'gene_type']].drop_duplicates(),\n",
    "            how='left',\n",
    "            on='ic_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "96ac8521-f5a5-4cc6-b4b9-eb2959e7720f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic_id</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>in_v36</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               ic_id gene_novelty gene_type  \\\n",
       "0  chr1_-_36445049-36443609-36443490-36439726-364...        Known       NaN   \n",
       "\n",
       "   in_v36  \n",
       "0   False  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "t = merge.ic_id.values[0]\n",
    "merge.loc[merge.ic_id==t]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59a8c942-8862-4bb1-adf1-bc970b200084",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c752efee-a06a-44f0-8219-c8d5b21ae2ba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ae9b906-97e4-466d-ad13-71b2810b9454",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "084536be-32d6-4f7b-9f32-7797026a2ae5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic_id</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [ic_id, gene_novelty, gene_type]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# where are dupes coming from?\n",
    "merge.loc[merge.ic_id.duplicated(keep=False)].sort_values(by='ic_id')\n",
    "merge.loc[merge.ic_id.duplicated(keep=False)][['gene_type', 'ic_id']].groupby('gene_type').count().rename({'ic_id':'counts'}, axis=1)\n",
    "\n",
    "# so just try limiting it to lncs and protein coding\n",
    "merge = merge.loc[(merge.gene_type.isin(['protein_coding', 'lncRNA']))|\\\n",
    "                  (merge.gene_type.isnull())]\n",
    "merge.loc[merge.ic_id.duplicated(keep=False)][['gene_type', 'ic_id']].groupby('gene_type').count().rename({'ic_id':'counts'}, axis=1)\n",
    "\n",
    "merge.loc[merge.ic_id.duplicated(keep=False)].sort_values(by='ic_id')\n",
    "\n",
    "# who cares cause they're not from novel genes anyway right?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b239018f-a2c9-4cdb-9e3b-9dfaa1787791",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge['in_v36'] = merge.gene_type.notnull()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5e3c5b78-dfb5-48e8-8969-37a300e013f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>n_ic</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>in_v36</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intergenic</th>\n",
       "      <th>False</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">Known</th>\n",
       "      <th>False</th>\n",
       "      <td>78546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>1027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Readthrough</th>\n",
       "      <th>False</th>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      n_ic\n",
       "gene_novelty in_v36       \n",
       "Intergenic   False      29\n",
       "Known        False   78546\n",
       "             True     1027\n",
       "Readthrough  False      84"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nov = merge.loc[merge.gene_novelty!='Known'].copy(deep=True)\n",
    "merge[['ic_id', 'gene_novelty', 'in_v36']].groupby(['gene_novelty', 'in_v36']).nunique().rename({'ic_id':'n_ic'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bdf00521-c4a3-45bd-8ed4-747adf3a6ac3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>n_ic</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>in_v36</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intergenic</th>\n",
       "      <th>NaN</th>\n",
       "      <th>False</th>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Known</th>\n",
       "      <th>lncRNA</th>\n",
       "      <th>True</th>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>protein_coding</th>\n",
       "      <th>True</th>\n",
       "      <td>962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <th>False</th>\n",
       "      <td>78546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Readthrough</th>\n",
       "      <th>NaN</th>\n",
       "      <th>False</th>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     n_ic\n",
       "gene_novelty gene_type      in_v36       \n",
       "Intergenic   NaN            False      29\n",
       "Known        lncRNA         True       65\n",
       "             protein_coding True      962\n",
       "             NaN            False   78546\n",
       "Readthrough  NaN            False      84"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nov = merge.loc[merge.gene_novelty!='Known'].copy(deep=True)\n",
    "merge[['ic_id', 'gene_type', 'gene_novelty', 'in_v36']].groupby(['gene_novelty', 'gene_type', 'in_v36'], dropna=False).nunique().rename({'ic_id':'n_ic'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38bc2f5a-c368-4bdb-98ae-b34d35747dd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge.loc[merge."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebc7a59b-fee5-4c22-9776-2aeb9d881fc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## somethings wrong here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4e48e65-13e5-4586-a08f-99df88630c6c",
   "metadata": {},
   "source": [
    "## What about by novelty category and GENCODE gene biotype?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "1f0ecb38-4b39-456d-a6a6-f45790928ecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get novelty types of each IC from cerberus\n",
    "ca = cerberus.read(m_cerberus_h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a4ad1c53-e907-4ed3-a26a-7512b856d2e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = ca.ic.copy(deep=True)\n",
    "temp.drop('ic', axis=1, inplace=True)\n",
    "temp.rename({'Coordinates':'ic'}, axis=1, inplace=True)\n",
    "temp = get_ic_id(temp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "0a7150ed-e0e1-405a-822a-859fc4147162",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>ic</th>\n",
       "      <th>Name</th>\n",
       "      <th>source</th>\n",
       "      <th>novelty</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000001138_3</td>\n",
       "      <td>vM25,vM21,lapa</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSMUSG00000001138</td>\n",
       "      <td>chr1_+_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000007805_1</td>\n",
       "      <td>vM25,vM21,lapa</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSMUSG00000007805</td>\n",
       "      <td>chr1_+_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000015314_4</td>\n",
       "      <td>vM25,vM21,lapa</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSMUSG00000015314</td>\n",
       "      <td>chr1_+_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000016494_3</td>\n",
       "      <td>vM25,vM21,lapa</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSMUSG00000016494</td>\n",
       "      <td>chr1_+_-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000018196_6</td>\n",
       "      <td>vM25,vM21,lapa</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSMUSG00000018196</td>\n",
       "      <td>chr1_+_-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Strand ic                  Name          source novelty  \\\n",
       "0       chr1      +  -  ENSMUSG00000001138_3  vM25,vM21,lapa   Known   \n",
       "1       chr1      +  -  ENSMUSG00000007805_1  vM25,vM21,lapa   Known   \n",
       "2       chr1      +  -  ENSMUSG00000015314_4  vM25,vM21,lapa   Known   \n",
       "3       chr1      +  -  ENSMUSG00000016494_3  vM25,vM21,lapa   Known   \n",
       "4       chr1      +  -  ENSMUSG00000018196_6  vM25,vM21,lapa   Known   \n",
       "\n",
       "              gene_id     ic_id  \n",
       "0  ENSMUSG00000001138  chr1_+_-  \n",
       "1  ENSMUSG00000007805  chr1_+_-  \n",
       "2  ENSMUSG00000015314  chr1_+_-  \n",
       "3  ENSMUSG00000016494  chr1_+_-  \n",
       "4  ENSMUSG00000018196  chr1_+_-  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "370fc50a-7520-412c-9f4b-b259577b7f90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic_id</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>in_v36</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_-_36445049-36443609-36443490-36439726-364...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_-_36445049-36443609-36443519-36439726-364...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               ic_id gene_novelty gene_type  \\\n",
       "0  chr1_-_36445049-36443609-36443490-36439726-364...        Known       NaN   \n",
       "1  chr1_-_36445049-36443609-36443490-36439726-364...        Known       NaN   \n",
       "2  chr1_-_36445049-36443609-36443490-36439726-364...        Known       NaN   \n",
       "3  chr1_-_36445049-36443609-36443490-36439726-364...        Known       NaN   \n",
       "4  chr1_-_36445049-36443609-36443519-36439726-364...        Known       NaN   \n",
       "\n",
       "   in_v36  \n",
       "0   False  \n",
       "1   False  \n",
       "2   False  \n",
       "3   False  \n",
       "4   False  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7530e67-8753-4abf-9dcc-ac201abb03f9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8728f2a-cc49-46be-831a-703b72e15ed3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbb1b8a0-b353-48ba-a8af-72072a902414",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fb4b702-5507-440f-b87d-147b309e727d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "4ec405d4-7b35-47a3-b572-07665494b77e",
   "metadata": {},
   "outputs": [],
   "source": [
    "merge = merge.merge(temp[['ic_id', 'novelty']], \n",
    "                    how='left',\n",
    "                    on='ic_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "b1f38724-85f5-44d2-8dbe-710bae115416",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge.drop('novelty_x', axis=1, inplace=True)\n",
    "# merge.rename({'novelty_y': 'novelty'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "2cfb571f-e94b-4b9a-8dfa-a18a91597ce3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True, False])"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge.in_v47.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "438025e8-37a6-4206-a3ab-83225d064ab3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic_id</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>in_v47</th>\n",
       "      <th>novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_-_169893787-169888890-169888675-169878819...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1_-_27635064-27625151-27625088-27623926-276...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1_-_27635064-27625151-27625088-27623929-276...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1_-_27635064-27625151-27625088-27623929-276...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NIC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1_-_27635064-27625151-27625088-27623929-276...</td>\n",
       "      <td>Known</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               ic_id gene_novelty gene_type  \\\n",
       "2  chr1_-_169893787-169888890-169888675-169878819...        Known       NaN   \n",
       "6  chr1_-_27635064-27625151-27625088-27623926-276...        Known       NaN   \n",
       "7  chr1_-_27635064-27625151-27625088-27623929-276...        Known       NaN   \n",
       "8  chr1_-_27635064-27625151-27625088-27623929-276...        Known       NaN   \n",
       "9  chr1_-_27635064-27625151-27625088-27623929-276...        Known       NaN   \n",
       "\n",
       "   in_v47 novelty  \n",
       "2   False     NIC  \n",
       "6   False     NNC  \n",
       "7   False     NIC  \n",
       "8   False     NIC  \n",
       "9   False     NNC  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge.loc[merge.in_v47==False].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "3e2143be-4e51-41c6-b684-d6c47c5b9275",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic_id</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>in_v47</th>\n",
       "      <th>novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16383</th>\n",
       "      <td>chr10_+_14838459-14839904-14839985-14840074-14...</td>\n",
       "      <td>Known</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16382</th>\n",
       "      <td>chr10_+_14838459-14839904-14839985-14840074-14...</td>\n",
       "      <td>Known</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14040</th>\n",
       "      <td>chr10_-_5617930-5617400-5617300-5617187</td>\n",
       "      <td>Known</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14041</th>\n",
       "      <td>chr10_-_5617930-5617400-5617300-5617187</td>\n",
       "      <td>Known</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>True</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14180</th>\n",
       "      <td>chr10_-_73252469-73252023-73251762-73250991-73...</td>\n",
       "      <td>Readthrough</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NNC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   ic_id gene_novelty  \\\n",
       "16383  chr10_+_14838459-14839904-14839985-14840074-14...        Known   \n",
       "16382  chr10_+_14838459-14839904-14839985-14840074-14...        Known   \n",
       "14040            chr10_-_5617930-5617400-5617300-5617187        Known   \n",
       "14041            chr10_-_5617930-5617400-5617300-5617187        Known   \n",
       "14180  chr10_-_73252469-73252023-73251762-73250991-73...  Readthrough   \n",
       "\n",
       "            gene_type  in_v47 novelty  \n",
       "16383  protein_coding    True     NNC  \n",
       "16382  protein_coding    True   Known  \n",
       "14040          lncRNA    True   Known  \n",
       "14041          lncRNA    True   Known  \n",
       "14180             NaN   False     NNC  "
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CHECK FOR DUPES\n",
    "merge.loc[merge.ic_id.duplicated(keep=False)].sort_values(by='ic_id').head()\n",
    "# merge.drop_duplicates(inplace=True)\n",
    "# merge.loc[merge.ic_id.duplicated(keep=False)].sort_values(by='ic_id').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "1d9993ca-d591-463a-abe1-3e3b5465f176",
   "metadata": {},
   "outputs": [],
   "source": [
    "# deduplicate using 1 and 0 for known and novel. ANy time \n",
    "# an intron chain is called known, regardless of gene its assigned to, call \n",
    "# it known\n",
    "merge['gb_known_col'] = 0\n",
    "merge.loc[merge.novelty == 'Known', 'gb_known_col'] = 1\n",
    "merge = merge[['ic_id', 'gene_type',\n",
    "               'in_v47', 'gb_known_col']].groupby(['ic_id',\n",
    "                                                   'gene_type',\n",
    "                                                   'in_v47'],\n",
    "                                                  dropna=False).max().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "2d388e5e-e252-41fe-8320-a0b38cd114f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ic_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>in_v47</th>\n",
       "      <th>gb_known_col</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>112049</th>\n",
       "      <td>chr8_-_144314925-144314820</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112050</th>\n",
       "      <td>chr8_-_144314925-144314820</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             ic_id       gene_type  in_v47  gb_known_col\n",
       "112049  chr8_-_144314925-144314820          lncRNA    True             1\n",
       "112050  chr8_-_144314925-144314820  protein_coding    True             1"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge.loc[merge.ic_id.duplicated(keep=False)].sort_values(by='ic_id').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1be87bd3-688e-40f0-9b64-aadc0f5e665c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add known or novel this way\n",
    "merge['novelty'] = 'Known'\n",
    "merge.loc[merge.gb_known_col==0, 'novelty'] = 'Novel'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "06041f22-d028-427e-887e-d1c57ec7c662",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ic_id</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_type</th>\n",
       "      <th>novelty</th>\n",
       "      <th>in_v47</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">lncRNA</th>\n",
       "      <th>Known</th>\n",
       "      <th>True</th>\n",
       "      <td>7051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <th>True</th>\n",
       "      <td>695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">protein_coding</th>\n",
       "      <th>Known</th>\n",
       "      <th>True</th>\n",
       "      <td>66755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <th>True</th>\n",
       "      <td>689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">NaN</th>\n",
       "      <th>Known</th>\n",
       "      <th>False</th>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Novel</th>\n",
       "      <th>False</th>\n",
       "      <td>44563</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               ic_id\n",
       "gene_type      novelty in_v47       \n",
       "lncRNA         Known   True     7051\n",
       "               Novel   True      695\n",
       "protein_coding Known   True    66755\n",
       "               Novel   True      689\n",
       "NaN            Known   False    2000\n",
       "               Novel   False   44563"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge[['ic_id', 'gene_type', 'novelty', 'in_v47']].groupby(['gene_type', 'novelty', 'in_v47'], dropna=False).nunique()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
