{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d29778ca-e6c1-4018-8a0b-ae5e3f72fbe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from proc_revisions.utils import *\n",
    "from proc_revisions.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4c199a76-0756-40bc-ba3d-d3a4831907d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "od = '../../proc_revisions/'\n",
    "config_file = f'{od}/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e454d32d-2d79-4450-8e8b-f0bfc0a94a52",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = od+expand(config['lr']['talon']['fusion_fix']['ab'], species='human')[0]\n",
    "unfilt_ab = od+expand(config['lr']['cerberus']['ab'], species='human')[0] \n",
    "filt_ab = od+expand(config['lr']['cerberus']['filt_ab'], species='human')[0]\n",
    "read_annot = od+expand(config['lr']['talon']['full_annot'], species='human')[0]\n",
    "t_metadata = od+expand(config['ref']['cerberus']['new_gtf_t_info'], species='human')[0]\n",
    "lib_meta = od+expand(config['lr']['meta'], species='human')[0]\n",
    "swan_file = od+expand(config['lr']['swan']['sg'], species='human')[0]\n",
    "cerberus_h5 = od+expand(config['lr']['cerberus']['ca_triplets'], species='human', obs_col='sample')[0]\n",
    "cerb_t_metadata = od+expand(config['lr']['cerberus']['gtf_t_info'], species='human')[0]\n",
    "major_isos = od+expand(config['lr']['analysis']['major_isos'], species='human', obs_col='sample')[0]\n",
    "pi_tpm_table = od+expand(config['lr']['mane']['pi_tpm']['triplet'], species='human', obs_col='sample')[0]\n",
    "\n",
    "ref_t_metadata = od+expand(config['ref']['new_gtf_t_info'], species='human')[0]\n",
    "ref_g_metadata = od+expand(config['ref']['new_gtf_g_info'], species='human')[0]\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "min_tpm = 1\n",
    "gene_subset = 'polya'\n",
    "obs_col = 'sample'\n",
    "go_gene_subset = 'protein_coding'\n",
    "predom_iso_subset = 'protein_coding'\n",
    "\n",
    "m_lib_meta = od+expand(config['lr']['meta'], species='mouse')[0]\n",
    "m_cerberus_h5 = od+expand(config['lr']['cerberus']['ca_triplets'], species='mouse', obs_col='sample')[0]\n",
    "m_swan_file = od+expand(config['lr']['swan']['sg'], species='mouse')[0]\n",
    "m_filt_ab = od+expand(config['lr']['cerberus']['filt_ab'], species='mouse')[0]\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "247f9a85-f528-4267-9329-482f2633cffd",
   "metadata": {},
   "source": [
    "One of the reviewers is worried about how our observations that many genes only have one transcript \n",
    "called might be affected by sequencing depth. To double check this, we'll take the library that the reviewer called out, and ask how often transcripts that only had one transcript in this library also have only one transcript in replicates for the same sample * condition, as well as from the same sample type (brain)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bdf02183-9cee-4b53-8e6a-ce201dae6e9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from ../../proc_revisions/data/mouse/lr/swan/swan_graph.p\n"
     ]
    }
   ],
   "source": [
    "ca = cerberus.read(m_cerberus_h5)\n",
    "sg = swan.read(m_swan_file)\n",
    "filt_ab_df = pd.read_csv(m_filt_ab, sep='\\t')\n",
    "obs_col = 'dataset'\n",
    "min_tpm = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "dd7bb0c4-4259-421d-ae26-e7af70f09d3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# library-level observed triplets\n",
    "df = ca.get_expressed_triplets(sg,\n",
    "                               obs_col=obs_col,\n",
    "                               min_tpm=min_tpm,\n",
    "                               source='dataset_det')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9ff46afb-88d4-49c0-887a-41166b1cf3ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>dataset</th>\n",
       "      <th>n_iso</th>\n",
       "      <th>n_tss</th>\n",
       "      <th>n_ic</th>\n",
       "      <th>n_tes</th>\n",
       "      <th>splicing_ratio</th>\n",
       "      <th>tss_ratio</th>\n",
       "      <th>tes_ratio</th>\n",
       "      <th>spl_ratio</th>\n",
       "      <th>sector</th>\n",
       "      <th>gene_tpm</th>\n",
       "      <th>gname</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [gid, dataset, n_iso, n_tss, n_ic, n_tes, splicing_ratio, tss_ratio, tes_ratio, spl_ratio, sector, gene_tpm, gname, source]\n",
       "Index: []"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get just the stuff with cortex_14d_f_1\n",
    "subset_df = df.loc[df.dataset=='cortex_14d_f_1'].copy(deep=True)\n",
    "subset_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ebddbd89-c427-4407-ba49-718b7bc9cac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENCODE_experiment_id</th>\n",
       "      <th>dataset</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_display</th>\n",
       "      <th>general_tissue_cell_type</th>\n",
       "      <th>fig1_tissue_label</th>\n",
       "      <th>health_status</th>\n",
       "      <th>tissue_or_cell_line</th>\n",
       "      <th>sample_color_hex_code</th>\n",
       "      <th>matching_mouse_samples</th>\n",
       "      <th>reads_post_talon</th>\n",
       "      <th>ENCODE_alignments_id</th>\n",
       "      <th>ENCODE_reads_id</th>\n",
       "      <th>ENCODE_unfiltered_alignments_id</th>\n",
       "      <th>document_urls</th>\n",
       "      <th>document_labels</th>\n",
       "      <th>platform</th>\n",
       "      <th>RIN</th>\n",
       "      <th>spikeins</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCSR989ZYL</td>\n",
       "      <td>a673_1_1</td>\n",
       "      <td>a673</td>\n",
       "      <td>A673</td>\n",
       "      <td>a673</td>\n",
       "      <td>bone</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>#de3700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1330194</td>\n",
       "      <td>ENCFF045ZQI</td>\n",
       "      <td>ENCFF168MIB</td>\n",
       "      <td>ENCFF440LXJ</td>\n",
       "      <td>https://www.encodeproject.org/documents/6d583a...</td>\n",
       "      <td>'ENCODE Long Read RNA-Seq Analysis Protocol fo...</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "      <td>9.8</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCSR989ZYL</td>\n",
       "      <td>a673_1_2</td>\n",
       "      <td>a673</td>\n",
       "      <td>A673</td>\n",
       "      <td>a673</td>\n",
       "      <td>bone</td>\n",
       "      <td>NaN</td>\n",
       "      <td>cell_line</td>\n",
       "      <td>#de3700</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1707207</td>\n",
       "      <td>ENCFF320GCF</td>\n",
       "      <td>ENCFF861BKY</td>\n",
       "      <td>ENCFF240FZT</td>\n",
       "      <td>https://www.encodeproject.org/documents/6d583a...</td>\n",
       "      <td>'ENCODE Long Read RNA-Seq Analysis Protocol fo...</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "      <td>9.8</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCSR081NRO</td>\n",
       "      <td>adrenal_gland_1_1</td>\n",
       "      <td>adrenal_gland</td>\n",
       "      <td>Adrenal gland</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tissue</td>\n",
       "      <td>#e69b00</td>\n",
       "      <td>adrenal_gland,adrenal_10d,adrenal_14d,adrenal_...</td>\n",
       "      <td>765655</td>\n",
       "      <td>ENCFF147OYL</td>\n",
       "      <td>ENCFF211SQY</td>\n",
       "      <td>ENCFF967OHL</td>\n",
       "      <td>https://www.encodeproject.org/documents/54012f...</td>\n",
       "      <td>'non-size selected cDNA libraries for use in P...</td>\n",
       "      <td>Pacific Biosciences Sequel</td>\n",
       "      <td>9.8</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCSR563RLX</td>\n",
       "      <td>adrenal_gland_2_1</td>\n",
       "      <td>adrenal_gland</td>\n",
       "      <td>Adrenal gland</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tissue</td>\n",
       "      <td>#e69b00</td>\n",
       "      <td>adrenal_gland,adrenal_10d,adrenal_14d,adrenal_...</td>\n",
       "      <td>1579294</td>\n",
       "      <td>ENCFF791WUV</td>\n",
       "      <td>ENCFF417ALN</td>\n",
       "      <td>ENCFF900XHI</td>\n",
       "      <td>https://www.encodeproject.org/documents/3baa46...</td>\n",
       "      <td>'PacBio libraries v3 (October, 2020) Protocol ...</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCSR995WKW</td>\n",
       "      <td>adrenal_gland_3_1</td>\n",
       "      <td>adrenal_gland</td>\n",
       "      <td>Adrenal gland</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tissue</td>\n",
       "      <td>#e69b00</td>\n",
       "      <td>adrenal_gland,adrenal_10d,adrenal_14d,adrenal_...</td>\n",
       "      <td>577077</td>\n",
       "      <td>ENCFF243PFI</td>\n",
       "      <td>ENCFF912HPY</td>\n",
       "      <td>ENCFF020MWV</td>\n",
       "      <td>https://www.encodeproject.org/documents/54012f...</td>\n",
       "      <td>'non-size selected cDNA libraries for use in P...</td>\n",
       "      <td>Pacific Biosciences Sequel</td>\n",
       "      <td>9.4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ENCODE_experiment_id            dataset         sample sample_display  \\\n",
       "0          ENCSR989ZYL           a673_1_1           a673           A673   \n",
       "1          ENCSR989ZYL           a673_1_2           a673           A673   \n",
       "2          ENCSR081NRO  adrenal_gland_1_1  adrenal_gland  Adrenal gland   \n",
       "3          ENCSR563RLX  adrenal_gland_2_1  adrenal_gland  Adrenal gland   \n",
       "4          ENCSR995WKW  adrenal_gland_3_1  adrenal_gland  Adrenal gland   \n",
       "\n",
       "  general_tissue_cell_type fig1_tissue_label health_status  \\\n",
       "0                     a673              bone           NaN   \n",
       "1                     a673              bone           NaN   \n",
       "2            adrenal gland     adrenal gland           NaN   \n",
       "3            adrenal gland     adrenal gland           NaN   \n",
       "4            adrenal gland     adrenal gland           NaN   \n",
       "\n",
       "  tissue_or_cell_line sample_color_hex_code  \\\n",
       "0           cell_line               #de3700   \n",
       "1           cell_line               #de3700   \n",
       "2              tissue               #e69b00   \n",
       "3              tissue               #e69b00   \n",
       "4              tissue               #e69b00   \n",
       "\n",
       "                              matching_mouse_samples  reads_post_talon  \\\n",
       "0                                                NaN           1330194   \n",
       "1                                                NaN           1707207   \n",
       "2  adrenal_gland,adrenal_10d,adrenal_14d,adrenal_...            765655   \n",
       "3  adrenal_gland,adrenal_10d,adrenal_14d,adrenal_...           1579294   \n",
       "4  adrenal_gland,adrenal_10d,adrenal_14d,adrenal_...            577077   \n",
       "\n",
       "  ENCODE_alignments_id ENCODE_reads_id ENCODE_unfiltered_alignments_id  \\\n",
       "0          ENCFF045ZQI     ENCFF168MIB                     ENCFF440LXJ   \n",
       "1          ENCFF320GCF     ENCFF861BKY                     ENCFF240FZT   \n",
       "2          ENCFF147OYL     ENCFF211SQY                     ENCFF967OHL   \n",
       "3          ENCFF791WUV     ENCFF417ALN                     ENCFF900XHI   \n",
       "4          ENCFF243PFI     ENCFF912HPY                     ENCFF020MWV   \n",
       "\n",
       "                                       document_urls  \\\n",
       "0  https://www.encodeproject.org/documents/6d583a...   \n",
       "1  https://www.encodeproject.org/documents/6d583a...   \n",
       "2  https://www.encodeproject.org/documents/54012f...   \n",
       "3  https://www.encodeproject.org/documents/3baa46...   \n",
       "4  https://www.encodeproject.org/documents/54012f...   \n",
       "\n",
       "                                     document_labels  \\\n",
       "0  'ENCODE Long Read RNA-Seq Analysis Protocol fo...   \n",
       "1  'ENCODE Long Read RNA-Seq Analysis Protocol fo...   \n",
       "2  'non-size selected cDNA libraries for use in P...   \n",
       "3  'PacBio libraries v3 (October, 2020) Protocol ...   \n",
       "4  'non-size selected cDNA libraries for use in P...   \n",
       "\n",
       "                        platform  RIN  spikeins  \n",
       "0  Pacific Biosciences Sequel II  9.8      True  \n",
       "1  Pacific Biosciences Sequel II  9.8      True  \n",
       "2     Pacific Biosciences Sequel  9.8     False  \n",
       "3  Pacific Biosciences Sequel II  NaN      True  \n",
       "4     Pacific Biosciences Sequel  9.4     False  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = pd.read_csv(lib_meta, sep='\\t')\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34e39334-e216-497b-941a-37a459f9d18e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f207360-3471-48da-beb3-655f9832b4f8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abbe6acc-526a-420f-9c82-aba8c03ce850",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a4431fd-25c1-4342-8e58-e05aac52d8e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339573e1-9316-4d2c-8754-8b3be45b4ffd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37be4adc-4101-421e-be47-e4d4fdb0b5f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
