{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "42b8a611-62ac-4adb-b7a6-91257cead273",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "import sklearn\n",
    "from snakemake.io import expand\n",
    "from sklearn.model_selection import KFold\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.model_selection import cross_val_predict\n",
    "from sklearn.tree import DecisionTreeClassifier, plot_tree\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from proc_revisions.utils import *\n",
    "from proc_revisions.plotting import *\n",
    "from proc_revisions.ss_sj_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c84e7da4-e2b8-4b02-9aad-738e8b6bc9e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "od = '../../proc_revisions/'\n",
    "config_file = f'{od}/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "830861af-3f2a-4ac0-b686-c6bb5af8908d",
   "metadata": {},
   "outputs": [],
   "source": [
    "talon_gtf = od+expand(config['lr']['talon']['gtf'], species='human')[0]\n",
    "talon_ab = od+expand(config['lr']['talon']['ab'], species='human')[0]\n",
    "talon_filt_ab = od+expand(config['lr']['talon']['filt_ab'], species='human')[0]\n",
    "ref_gtf = od+expand(config['ref']['gtf'], species='human')[0]\n",
    "ref_ics = od+expand(config['ref']['cerberus']['ics'], species='human')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "30685241-4e6b-4fbd-b5e0-03696afaa5ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_talon_known_fusion_transcripts(talon_ab,\n",
    "                                       talon_filt_ab,\n",
    "                                       talon_gtf,\n",
    "                                       ref_ics,\n",
    "                                       ref_gtf,\n",
    "                                       wc,\n",
    "                                       ofile_gtf,\n",
    "                                       ofile_ab): \n",
    "    \"\"\"\n",
    "    Fix gene assignments to novel transcripts that have splice sites\n",
    "    that intersect with known readthrough transcripts (v29 or vM21). \n",
    "    Update gene counts accordingly.\n",
    "    * Limit to novel transcripts from known genes\n",
    "    * Limit to novel transcripts that share splice sites with only 2 genes\n",
    "        (one readthrough, one non readthrough)\n",
    "    * Compute what % of splice sites are supported in the novel transcript\n",
    "        by the best match transcript from each of the two annotated genes\n",
    "    * Compute what % of splice sites in each of the two annotated genes\n",
    "        are supported by the novel transcript\n",
    "    * Compute how far away the nearest TSS / TES is for each of the \n",
    "        two annotated genes\n",
    "    * Used a decision tree classifier + manual labels I gave from inspecting \n",
    "        loci to determine whether the gene assignment should be made \n",
    "        to the readthrough gene or the non-readthrough gene\n",
    "    \"\"\"\n",
    "    if wc['species'] == 'human':\n",
    "        ver = 'v40_cerberus'\n",
    "    elif wc['species'] == 'mouse':\n",
    "        ver = 'vM25_cerberus'\n",
    "\n",
    "    gtf_df, _, _ = get_gtf_info(ver=ver, how='gene', add_stable_gid=True)\n",
    "    gtf_df = gtf_df[['gid_stable', 'gname']]\n",
    "    gtf_df.head()\n",
    "\n",
    "    talon_df = pd.read_csv(talon_filt_ab, sep='\\t')\n",
    "    talon_df['gid'] = cerberus.get_stable_gid(talon_df, 'annot_gene_id')\n",
    "\n",
    "    tids = talon_df.loc[(talon_df.transcript_novelty!='Known')&(talon_df.gene_novelty=='Known'), 'annot_transcript_id'].tolist()\n",
    "\n",
    "    df = pr.read_gtf(talon_gtf, rename_attr=True, duplicate_attr=True)\n",
    "    df = cerberus.get_ic(df)\n",
    "    df.rename({'transcript_id':'Name', 'ic': 'Coordinates'}, axis=1, inplace=True)\n",
    "    df['source'] = 'lapa'\n",
    "    \n",
    "    # get gids from annotated fusion genes\n",
    "    rt_df, _, _ = get_gtf_info(ver=ver, how='iso')\n",
    "    rt_df = rt_df.loc[rt_df.readthrough_transcript==True]\n",
    "    rt_df['gid_stable'] = cerberus.get_stable_gid(rt_df, 'gid')\n",
    "    fusion_gids = rt_df.gid_stable.tolist()\n",
    "    \n",
    "    # get reference transcript id to ic name map :'(\n",
    "    ref_df = pr.read_gtf(talon_gtf, rename_attr=True, duplicate_attr=True)\n",
    "    ref_df = cerberus.get_ic(ref_df)\n",
    "    ref_df.head()\n",
    "    \n",
    "    known_df = pd.read_csv(ref_ics, sep='\\t')\n",
    "    known_df['gene_id'] = known_df.Name.str.split('_', expand=True)[0]\n",
    "    known_df['source'] = 'lapa'\n",
    "    \n",
    "    ref_df.rename({'ic':'Coordinates'}, axis=1, inplace=True)\n",
    "    print(len(ref_df.index))\n",
    "    ref_df= ref_df.merge(known_df[['Coordinates', 'gene_id', 'Chromosome', 'Name']], \n",
    "                 how='left', \n",
    "                 on=['Coordinates', 'gene_id', 'Chromosome'])\n",
    "    print(len(ref_df.index))\n",
    "    \n",
    "    known_df = pd.read_csv(ref_ics, sep='\\t')\n",
    "    known_df['gene_id'] = known_df.Name.str.split('_', expand=True)[0]\n",
    "    known_df['source'] = 'lapa'\n",
    "\n",
    "    _, f_ss_ic_df = get_ss_from_ic(df, ['lapa'])\n",
    "    f_ss_ic_df.drop(['source', 'novelty'], axis=1, inplace=True)\n",
    "\n",
    "    _, k_ss_ic_df = get_ss_from_ic(known_df, ['lapa'])\n",
    "    k_ss_ic_df.drop(['source', 'novelty'], axis=1, inplace=True)\n",
    "    \n",
    "    f_ss_ic_df['ss_id'] = f_ss_ic_df.Name+f_ss_ic_df.Chromosome+f_ss_ic_df.Strand+f_ss_ic_df.Start.astype(str)\n",
    "    k_ss_ic_df['ss_id'] = k_ss_ic_df.Name+k_ss_ic_df.Chromosome+k_ss_ic_df.Strand+k_ss_ic_df.Start.astype(str)\n",
    "    \n",
    "    f_ss_ic_df = f_ss_ic_df.merge(k_ss_ic_df, how='left', on=['Chromosome', 'Strand', 'Start', 'ss_type'],\n",
    "                          suffixes=('', '_known'))\n",
    "    f_ss_ic_df['known'] = False\n",
    "    f_ss_ic_df.loc[f_ss_ic_df.Name_known.notnull(), 'known'] = True\n",
    "    \n",
    "    # 1. get all transcript + gene combinations\n",
    "    df = f_ss_ic_df[['gene_id', 'Name', 'gene_id_known', 'known']].drop_duplicates().copy(deep=True)\n",
    "    \n",
    "    # 2. compute number of splice sites / transcript\n",
    "    temp = f_ss_ic_df[['Name', 'ss_id']].groupby('Name').nunique()\n",
    "    temp = temp.reset_index().rename({'ss_id': 'n_ss'}, axis=1)\n",
    "    df = df.merge(temp, how='left', on='Name')\n",
    "    \n",
    "    # 3. add in the novelty of the talon gene and transcript\n",
    "    df = df.merge(talon_df[['transcript_novelty', 'gene_novelty', 'annot_transcript_id']], \n",
    "                  how='left', \n",
    "                  left_on='Name',\n",
    "                  right_on='annot_transcript_id')\n",
    "    df.drop('annot_transcript_id', axis=1, inplace=True)\n",
    "    \n",
    "    # 4. subset for novel transcripts that belong to annotated genes\n",
    "    df = df.loc[(df.transcript_novelty!='Known')&\\\n",
    "                (df.gene_novelty=='Known')]\n",
    "    \n",
    "    # 5. get all transcripts that have >= 1 splice site shared w/ a readthrough gene\n",
    "    tids = df.loc[df.gene_id_known.isin(fusion_gids), 'Name'].unique().tolist()\n",
    "    df = df.loc[df.Name.isin(tids)]\n",
    "    \n",
    "    # 6. get how many unique genes share splice sites with this transcript\n",
    "    temp = df.loc[df.known==True].copy(deep=True)\n",
    "    temp = temp[['Name', 'gene_id_known']].groupby('Name').nunique().reset_index()\n",
    "    temp.rename({'gene_id_known': 'n_genes'}, axis=1, inplace=True)\n",
    "    df = df.merge(temp, how='left', on='Name')\n",
    "    \n",
    "    # 7. get how many unique genes share splice sites with this transcript\n",
    "    temp = df.loc[df.known==True].copy(deep=True)\n",
    "    temp = temp.loc[~temp.gene_id_known.isin(fusion_gids)]\n",
    "    temp = temp[['Name', 'gene_id_known']].groupby('Name').nunique().reset_index()\n",
    "    temp.rename({'gene_id_known': 'n_genes_no_rt'}, axis=1, inplace=True)\n",
    "    df = df.merge(temp, how='left', on='Name')\n",
    "    \n",
    "    # 8. limit to transcripts w/ n_genes_no_rt < n_genes\n",
    "    # and n_genes_no_rt == 1 (only one choice)\n",
    "    # where the gene that was merged with is the non-rt one\n",
    "    df = df.loc[(df.n_genes_no_rt==1)&(df.n_genes>df.n_genes_no_rt)]\n",
    "    df = df.loc[~df.gene_id_known.isin(fusion_gids)]\n",
    "    df.head()\n",
    "    \n",
    "    # 9. get the number of splice sites supported by the transcript w/ the most\n",
    "    # shared splice sites for each transcript + annotated gene combination, \n",
    "    # then merge in based on transcript from obs. transcripts\n",
    "    temp = f_ss_ic_df.copy(deep=True)\n",
    "    temp = temp[['Name', 'ss_id', 'gene_id_known', 'Name_known']].drop_duplicates()\n",
    "    temp = temp.groupby(['Name', 'gene_id_known', 'Name_known']).nunique().reset_index()\n",
    "    temp.rename({'ss_id': 'n_supp_ss',\n",
    "                 'Name_known': 'Name_supp'}, axis=1, inplace=True)\n",
    "    temp = temp.sort_values(by='n_supp_ss', ascending=False)\n",
    "    temp = temp.drop_duplicates(subset=['Name', 'gene_id_known'], keep='first')\n",
    "    df = df.merge(temp, how='left', on=['Name', 'gene_id_known'])\n",
    "    \n",
    "    # 9.5. get the number of splice sites supported by the transcript w/ the most\n",
    "    # shared splice sites for each transcript + annotated gene combination, \n",
    "    # then merge in based on transcript from obs. transcripts\n",
    "    temp = f_ss_ic_df.copy(deep=True)\n",
    "    temp = temp[['Name', 'ss_id', 'gene_id_known', 'Name_known']].drop_duplicates()\n",
    "    temp = temp.loc[temp.gene_id_known.isin(fusion_gids)]\n",
    "    temp = temp.groupby(['Name', 'gene_id_known', 'Name_known']).nunique().reset_index()\n",
    "    temp.rename({'ss_id': 'n_supp_ss_rt',\n",
    "                 'Name_known': 'Name_supp_rt',\n",
    "                 'gene_id_known': 'gene_id_known_rt'}, axis=1, inplace=True)\n",
    "    temp = temp.sort_values(by='n_supp_ss_rt', ascending=False)\n",
    "    temp = temp.drop_duplicates(subset=['Name'], keep='first')\n",
    "    df = df.merge(temp, how='left', on=['Name'])\n",
    "    \n",
    "    # 10. remove entries where talon gene id is already\n",
    "    # the other choice\n",
    "    df['gid'] = cerberus.get_stable_gid(df, 'gene_id')\n",
    "    df = df.loc[df.gid!='gene_id_known']\n",
    "    \n",
    "    # 11. get the total # of splice sites per annotated transcript\n",
    "    temp = k_ss_ic_df[['Name', 'ss_id']].groupby('Name').nunique().reset_index()\n",
    "    temp = temp.rename({'ss_id': 'n_supp_total_ss',\n",
    "                                      'Name': 'Name_supp'}, axis=1)\n",
    "    df = df.merge(temp, how='left', on='Name_supp')\n",
    "    temp = temp.rename({'n_supp_total_ss': 'n_supp_total_ss_rt',\n",
    "                        'Name_supp': 'Name_supp_rt'}, axis=1)\n",
    "    df = df.merge(temp, how='left', on='Name_supp_rt')\n",
    "    \n",
    "    # 12. limit to obs. transcripts that were annotated to readthrough genes\n",
    "    df = df.loc[df.gid.isin(fusion_gids)]\n",
    "    print(len(df.index))\n",
    "    \n",
    "    df['perc_supp_ss'] = (df.n_supp_ss/df.n_ss)*100\n",
    "    df['perc_supp_ss_rt'] = (df.n_supp_ss_rt/df.n_ss)*100\n",
    "    df['perc_supp_annot_ss'] = (df.n_supp_ss/df.n_supp_total_ss)*100\n",
    "    df['perc_supp_annot_ss_rt'] = (df.n_supp_ss_rt/df.n_supp_total_ss_rt)*100\n",
    "    \n",
    "    temp = df.loc[df.gid.isin(fusion_gids)]\n",
    "    print(len(temp.Name.unique()))\n",
    "    \n",
    "    df['perc_supp_diff'] = df['perc_supp_ss']-df['perc_supp_ss_rt']\n",
    "    df['perc_supp_annot_diff'] = df['perc_supp_annot_ss']-df['perc_supp_annot_ss_rt']\n",
    "    df = df.loc[df.perc_supp_annot_ss.notnull()]\n",
    "    \n",
    "    # 14. compute distances from transcript start and end to closest start and end from any transcript in\n",
    "    talon_df = pr.read_gtf(talon_gtf, rename_attr=True, duplicate_attr=True)\n",
    "    talon_df = talon_df.df\n",
    "    talon_df = talon_df.loc[talon_df.Feature=='transcript']\n",
    "    \n",
    "    known_df = pr.read_gtf(ref_gtf, rename_attr=True, duplicate_attr=True)\n",
    "    known_df = known_df.df\n",
    "    known_df = known_df.loc[known_df.Feature=='transcript']\n",
    "    \n",
    "    known_df['gene_id'] = cerberus.get_stable_gid(known_df, 'gene_id')\n",
    "    # add coords to novel thing (df)\n",
    "    temp = talon_df[['transcript_id', 'Start', 'End', 'Chromosome', 'Strand']]\n",
    "    temp.rename({'transcript_id':'Name'}, axis=1, inplace=True)\n",
    "    df = df.merge(temp, how='left', on='Name')\n",
    "    \n",
    "    df = df.drop_duplicates()\n",
    "    temp3 = pd.DataFrame()\n",
    "    temp3['Name'] = df.Name.tolist()\n",
    "    for c in ['gene_id_known', 'gene_id_known_rt']:\n",
    "        for feat in ['End', 'Start']:\n",
    "            temp2 = pd.DataFrame() \n",
    "            for g in df[c].unique().tolist():\n",
    "                if feat == 'Start':\n",
    "                    other_feat = 'End'\n",
    "                elif feat == 'End':\n",
    "                    other_feat = 'Start'\n",
    "                temp_nov = df.loc[df[c] == g].copy(deep=True)\n",
    "                temp_nov = temp_nov[['Name', 'Chromosome', feat, 'Strand']]\n",
    "                temp_nov[other_feat] = temp_nov[feat]+1 \n",
    "\n",
    "                temp_known = known_df.loc[known_df.gene_id == g].copy(deep=True)\n",
    "                temp_known = temp_known[['Chromosome', feat, 'Strand']]\n",
    "                temp_known[other_feat] = temp_known[feat]+1 \n",
    "\n",
    "\n",
    "                temp_nov = pr.PyRanges(temp_nov)\n",
    "                temp_known = pr.PyRanges(temp_known)\n",
    "\n",
    "                # pyranges join and get closest; concat in w/ temp2s\n",
    "                temp_nov = temp_nov.k_nearest(temp_known, k=1, \n",
    "                                   overlap=True,\n",
    "                                   how=None, suffix='_known').df\n",
    "                try:\n",
    "                    temp_nov = temp_nov[['Name', 'Distance']].drop_duplicates()\n",
    "                except:\n",
    "                    import pdb; pdb.set_trace()\n",
    "                d_col = f'{c}_{feat}_dist'\n",
    "                temp_nov.rename({'Distance': d_col}, axis=1, inplace=True)        \n",
    "                # if len(temp_nov.loc[temp_nov[d_col].isnull()].index) > 0:\n",
    "                #     import pdb; pdb.set_trace()\n",
    "                temp2 = pd.concat([temp2, temp_nov], axis=0)\n",
    "            temp3 = temp3.merge(temp2, on='Name', how='left') \n",
    "            \n",
    "    df = df.merge(temp3, how='left', on='Name')  \n",
    "    \n",
    "    # add abs. values\n",
    "    cols = ['gene_id_known_End_dist',\n",
    "            'gene_id_known_Start_dist',\n",
    "            'gene_id_known_rt_End_dist',\n",
    "            'gene_id_known_rt_Start_dist']\n",
    "    for c in cols:\n",
    "        new_col = f'{c}_abs'\n",
    "        df[new_col] = df[c].abs()\n",
    "    \n",
    "    # implement DecisionTree classifier in pandas format w/ thresholds\n",
    "    df['fix'] = False\n",
    "\n",
    "    # start of closest rt transcript must be > 510 bp away\n",
    "    inds = df.loc[df.gene_id_known_rt_Start_dist_abs > 510].index\n",
    "    df.loc[inds, 'fix'] = True\n",
    "\n",
    "    # end of the closest rt transcript must be > 45 kbp away\n",
    "    inds = df.loc[df.gene_id_known_rt_End_dist_abs > 45000].index\n",
    "    df.loc[inds, 'fix'] = True\n",
    "\n",
    "    # % of sss annotated in best matching non-rt transcripts\n",
    "    # must be >94\n",
    "    # % of sss annotated in actual transcript mush be > 84%\n",
    "    inds = df.loc[(df.perc_supp_annot_ss > 94)&(df.perc_supp_ss > 84)].index\n",
    "    df.loc[inds, 'fix'] = True\n",
    "\n",
    "    # df[['fix_needed', 'fix', 'Name']].groupby(['fix_needed', 'fix']).count().reset_index()\n",
    "    \n",
    "    gtf_df = pr.read_gtf(talon_gtf, duplicate_attr=True, rename_attr=True)\n",
    "    gtf_df = gtf_df.df\n",
    "    gtf_df['gid_stable'] = cerberus.get_stable_gid(gtf_df, 'gene_id')\n",
    "\n",
    "    ref_gtf_df = pr.read_gtf(ref_gtf, duplicate_attr=True, rename_attr=True)\n",
    "    ref_gtf_df = ref_gtf_df.df\n",
    "    ref_gtf_df['gid_stable'] = cerberus.get_stable_gid(ref_gtf_df, 'gene_id')\n",
    "    \n",
    "    # update the genes that need fixing to the genes that were non-rt that they \n",
    "    # intersected with\n",
    "    gene_cols = ['Source', 'gene_id', 'gene_name', 'gene_status', 'gene_type', 'talon_gene', 'havana_gene', 'level',\n",
    "                 'antisense_gene', 'gene_antisense_to_IDs', 'intergenic_novel', 'fusion_novel']\n",
    "    talon_gene_cols = ['gene_status', 'talon_gene', 'antisense_gene', 'gene_antisense_to_IDs', 'intergenic_novel', 'fusion_novel']\n",
    "    temp = df.loc[df.fix == True]\n",
    "    temp3 = temp.copy(deep=True)\n",
    "    for ind, entry in temp.iterrows():\n",
    "        t = entry.Name\n",
    "        inds = gtf_df.loc[gtf_df.transcript_id==t].index\n",
    "        g = entry.gene_id_known\n",
    "        dummy_gene_entry = gtf_df.loc[(gtf_df.gid_stable==g)&(gtf_df.Feature=='gene')]\n",
    "        if len(dummy_gene_entry.index) == 0:\n",
    "            # have to pull from reference gtf instead and add corresponding gene entry\n",
    "            dummy_gene_entry = ref_gtf_df.loc[(ref_gtf_df.gid_stable==g)&(ref_gtf_df.Feature=='gene')]\n",
    "            dummy_gene_entry[talon_gene_cols] = np.nan\n",
    "            gtf_df = pd.concat([gtf_df, dummy_gene_entry], axis=0)\n",
    "        try:\n",
    "            assert len(dummy_gene_entry.index) == 1\n",
    "        except:\n",
    "            import pdb; pdb.set_trace()\n",
    "        for c in gene_cols:\n",
    "            gtf_df.loc[inds, c] = dummy_gene_entry[c].values[0]\n",
    "        \n",
    "    # drop stable gid, sort, update ends, and dump\n",
    "    gtf_df.drop('gid_stable', axis=1, inplace=True)\n",
    "    gtf_df = cerberus.sort_gtf(gtf_df)\n",
    "    # mainly ripped out of cerberus code\n",
    "    gtf_temp = gtf_df.copy(deep=True)\n",
    "    for mode in ['tss', 'tes']:\n",
    "        fwd, rev = cerberus.get_stranded_gtf_dfs(gtf_temp)\n",
    "        df = pd.DataFrame()\n",
    "        for strand, temp in zip(['+', '-'], [fwd, rev]):\n",
    "\n",
    "            # fix gene boundaries\n",
    "            temp = cerberus.update_gene_ends(temp, mode, strand)\n",
    "            df = pd.concat([df, temp], ignore_index=True)\n",
    "\n",
    "        gtf_temp = df.copy(deep=True)\n",
    "    pr.PyRanges(gtf_temp).to_gtf(ofile_gtf)\n",
    "    \n",
    "    # also update the unfiltered abundance table\n",
    "    import pdb; pdb.set_trace()\n",
    "    temp = temp3[['Name', 'gene_id_known']]\n",
    "    assert len(temp3.loc[(temp3.perc_supp_ss==100)&(temp3.perc_supp_annot_ss==100)].index) == 0\n",
    "    temp.rename({'Name': 'annot_transcript_id',\n",
    "                 'gene_id_known': 'gid_stable'},\n",
    "                axis=1, inplace=True)\n",
    "    \n",
    "    df = pd.read_csv(talon_ab, sep='\\t')\n",
    "    print(len(df.index))\n",
    "    df['gid_stable'] = cerberus.get_stable_gid(df, 'annot_gene_id')\n",
    "    df[df.columns[:11]].head()\n",
    "    gene_cols = ['gene_ID', 'annot_gene_id', 'annot_gene_name', 'gene_novelty', 'gid_stable']\n",
    "    gene_info = df[gene_cols].drop_duplicates()\n",
    "    gene_info.head()\n",
    "    temp = temp.merge(gene_info, how='left', on='gid_stable')\n",
    "    temp.drop('gid_stable', axis=1, inplace=True)    \n",
    "    df = df.merge(temp, how='left', on='annot_transcript_id', suffixes=('', '_fix'))\n",
    "    fix_inds = df.loc[df.gene_ID_fix.notnull()].index\n",
    "    gene_cols = [g for g in gene_cols if g!='gid_stable']\n",
    "    for g in gene_cols:\n",
    "        fix_col = f'{g}_fix'\n",
    "        df.loc[fix_inds, g] = df.loc[fix_inds, fix_col]\n",
    "        df.drop(fix_col, axis=1, inplace=True)\n",
    "    df.to_csv(ofile_ab, sep='\\t', index=False)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "033555c9-6c24-458c-b32f-96135c115ea4",
   "metadata": {},
   "source": [
    "## 231218 try with my function-ized version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "92626c85-62ec-4394-a229-850a4e2c13ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n",
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n",
      "191401\n",
      "191401\n",
      "253\n",
      "218\n",
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n",
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n",
      "> \u001b[0;32m<ipython-input-103-23fa204640f2>\u001b[0m(301)\u001b[0;36mfix_talon_known_fusion_transcripts\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    299 \u001b[0;31m    \u001b[0;31m# also update the unfiltered abundance table\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    300 \u001b[0;31m    \u001b[0;32mimport\u001b[0m \u001b[0mpdb\u001b[0m\u001b[0;34m;\u001b[0m \u001b[0mpdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_trace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 301 \u001b[0;31m    \u001b[0mtemp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp3\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Name'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gene_id_known'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    302 \u001b[0;31m    \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mperc_supp_ss\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m&\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mperc_supp_annot_ss\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    303 \u001b[0;31m    temp.rename({'Name': 'annot_transcript_id',\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  n\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "> \u001b[0;32m<ipython-input-103-23fa204640f2>\u001b[0m(302)\u001b[0;36mfix_talon_known_fusion_transcripts\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m    300 \u001b[0;31m    \u001b[0;32mimport\u001b[0m \u001b[0mpdb\u001b[0m\u001b[0;34m;\u001b[0m \u001b[0mpdb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_trace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    301 \u001b[0;31m    \u001b[0mtemp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp3\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Name'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'gene_id_known'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m--> 302 \u001b[0;31m    \u001b[0;32massert\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mperc_supp_ss\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m&\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mperc_supp_annot_ss\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0m\u001b[0;32m    303 \u001b[0;31m    temp.rename({'Name': 'annot_transcript_id',\n",
      "\u001b[0m\u001b[0;32m    304 \u001b[0;31m                 'gene_id_known': 'gid_stable'},\n",
      "\u001b[0m\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  temp.head()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "               Name    gene_id_known\n",
      "0   TALONT001122495  ENSG00000143543\n",
      "11  TALONT000930061  ENSG00000116783\n",
      "12  TALONT000930071  ENSG00000116783\n",
      "13  TALONT000930083  ENSG00000116783\n",
      "14  TALONT000983329  ENSG00000156875\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  temp3.loc[(temp3.perc_supp_ss==100)&(temp3.perc_supp_annot_ss==100)]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [gene_id, Name, gene_id_known, known, n_ss, transcript_novelty, gene_novelty, n_genes, n_genes_no_rt, Name_supp, n_supp_ss, gene_id_known_rt, Name_supp_rt, n_supp_ss_rt, gid, n_supp_total_ss, n_supp_total_ss_rt, perc_supp_ss, perc_supp_ss_rt, perc_supp_annot_ss, perc_supp_annot_ss_rt, perc_supp_diff, perc_supp_annot_diff, Start, End, Chromosome, Strand, gene_id_known_End_dist, gene_id_known_Start_dist, gene_id_known_rt_End_dist, gene_id_known_rt_Start_dist, gene_id_known_End_dist_abs, gene_id_known_Start_dist_abs, gene_id_known_rt_End_dist_abs, gene_id_known_rt_Start_dist_abs, fix]\n",
      "Index: []\n",
      "\n",
      "[0 rows x 36 columns]\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "ipdb>  temp['perc_supp_ss'] = 0\n",
      "ipdb>  temp['perc_supp_annot_ss'] = 0\n",
      "ipdb>  c\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5395598\n"
     ]
    }
   ],
   "source": [
    "ofile_gtf = 'test.gtf'\n",
    "ofile_ab = 'test_unfiltered_abundance.tsv'\n",
    "fix_talon_known_fusion_transcripts(talon_ab,\n",
    "                                       talon_filt_ab,\n",
    "                                       talon_gtf,\n",
    "                                       ref_ics,\n",
    "                                       ref_gtf,\n",
    "                                       {'species': 'human'},\n",
    "                                       ofile_gtf,\n",
    "                                       ofile_ab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3947ba69-ce94-4a45-9692-b63b263ba3da",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "326ec4d5-07ce-4c63-b545-4b146a200fb3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "159e89a0-6d6d-403f-9cce-d71dba86ab66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n"
     ]
    }
   ],
   "source": [
    "gtf_df, _, _ = get_gtf_info(ver='v40_cerberus', how='gene', add_stable_gid=True)\n",
    "gtf_df = gtf_df[['gid_stable', 'gname']]\n",
    "gtf_df.head()\n",
    "\n",
    "talon_df = pd.read_csv(talon_filt_ab, sep='\\t')\n",
    "talon_df['gid'] = cerberus.get_stable_gid(talon_df, 'annot_gene_id')\n",
    "\n",
    "tids = talon_df.loc[(talon_df.transcript_novelty!='Known')&(talon_df.gene_novelty=='Known'), 'annot_transcript_id'].tolist()\n",
    "\n",
    "df = pr.read_gtf(talon_gtf, rename_attr=True, duplicate_attr=True)\n",
    "df = cerberus.get_ic(df)\n",
    "df.rename({'transcript_id':'Name', 'ic': 'Coordinates'}, axis=1, inplace=True)\n",
    "df['source'] = 'lapa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8abc929d-32a2-46b8-98c5-49fa525c800e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get gids from annotated fusion genes\n",
    "rt_df, _, _ = get_gtf_info(ver='v40_cerberus', how='iso')\n",
    "rt_df = rt_df.loc[rt_df.readthrough_transcript==True]\n",
    "rt_df['gid_stable'] = cerberus.get_stable_gid(rt_df, 'gid')\n",
    "fusion_gids = rt_df.gid_stable.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "83b7898b-3673-4bb4-abcc-da8c4b414708",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>ic</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>-</td>\n",
       "      <td>TALONT001166745</td>\n",
       "      <td>TALONG000193759</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERCC-00002</td>\n",
       "      <td>+</td>\n",
       "      <td>tSpikein_ERCC-00002</td>\n",
       "      <td>gSpikein_ERCC-00002</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERCC-00003</td>\n",
       "      <td>+</td>\n",
       "      <td>tSpikein_ERCC-00003</td>\n",
       "      <td>gSpikein_ERCC-00003</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERCC-00004</td>\n",
       "      <td>+</td>\n",
       "      <td>tSpikein_ERCC-00004</td>\n",
       "      <td>gSpikein_ERCC-00004</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERCC-00009</td>\n",
       "      <td>-</td>\n",
       "      <td>TALONT000206952</td>\n",
       "      <td>TALONG000058910</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Chromosome Strand        transcript_id              gene_id ic\n",
       "0  ERCC-00002      -      TALONT001166745      TALONG000193759  -\n",
       "1  ERCC-00002      +  tSpikein_ERCC-00002  gSpikein_ERCC-00002  -\n",
       "2  ERCC-00003      +  tSpikein_ERCC-00003  gSpikein_ERCC-00003  -\n",
       "3  ERCC-00004      +  tSpikein_ERCC-00004  gSpikein_ERCC-00004  -\n",
       "4  ERCC-00009      -      TALONT000206952      TALONG000058910  -"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get reference transcript id to ic name map :'(\n",
    "ref_df = pr.read_gtf(talon_gtf, rename_attr=True, duplicate_attr=True)\n",
    "ref_df = cerberus.get_ic(ref_df)\n",
    "ref_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4d8315c2-4af3-4dbf-bf5d-3b45652e6725",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_df = pd.read_csv(ref_ics, sep='\\t')\n",
    "known_df['gene_id'] = known_df.Name.str.split('_', expand=True)[0]\n",
    "known_df['source'] = 'lapa'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ea7ee00b-c0b2-4ec6-b22f-a53290007739",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "191401\n",
      "191401\n"
     ]
    }
   ],
   "source": [
    "ref_df.rename({'ic':'Coordinates'}, axis=1, inplace=True)\n",
    "print(len(ref_df.index))\n",
    "ref_df= ref_df.merge(known_df[['Coordinates', 'gene_id', 'Chromosome', 'Name']], \n",
    "             how='left', \n",
    "             on=['Coordinates', 'gene_id', 'Chromosome'])\n",
    "print(len(ref_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c15b53e4-4a82-4368-9d96-5ca726383b55",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_df = pd.read_csv(ref_ics, sep='\\t')\n",
    "known_df['gene_id'] = known_df.Name.str.split('_', expand=True)[0]\n",
    "known_df['source'] = 'lapa'\n",
    "\n",
    "_, f_ss_ic_df = get_ss_from_ic(df, ['lapa'])\n",
    "f_ss_ic_df.drop(['source', 'novelty'], axis=1, inplace=True)\n",
    "\n",
    "_, k_ss_ic_df = get_ss_from_ic(known_df, ['lapa'])\n",
    "k_ss_ic_df.drop(['source', 'novelty'], axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "0c8ef5d3-030d-481d-9e86-e1222282fc7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_ss_ic_df['ss_id'] = f_ss_ic_df.Name+f_ss_ic_df.Chromosome+f_ss_ic_df.Strand+f_ss_ic_df.Start.astype(str)\n",
    "k_ss_ic_df['ss_id'] = k_ss_ic_df.Name+k_ss_ic_df.Chromosome+k_ss_ic_df.Strand+k_ss_ic_df.Start.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "43473465-90f8-4ab0-b245-8c0084f6c032",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_ss_ic_df = f_ss_ic_df.merge(k_ss_ic_df, how='left', on=['Chromosome', 'Strand', 'Start', 'ss_type'],\n",
    "                          suffixes=('', '_known'))\n",
    "f_ss_ic_df['known'] = False\n",
    "f_ss_ic_df.loc[f_ss_ic_df.Name_known.notnull(), 'known'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b6c8424c-e431-45b4-90f0-f9f0cea5e739",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. get all transcript + gene combinations\n",
    "df = f_ss_ic_df[['gene_id', 'Name', 'gene_id_known', 'known']].drop_duplicates().copy(deep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "26ad76a6-fb68-4bfb-9443-4db9642c5ccc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. compute number of splice sites / transcript\n",
    "temp = f_ss_ic_df[['Name', 'ss_id']].groupby('Name').nunique()\n",
    "temp = temp.reset_index().rename({'ss_id': 'n_ss'}, axis=1)\n",
    "df = df.merge(temp, how='left', on='Name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "25d5ecf6-9d98-4587-b251-52358a6e8409",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3. add in the novelty of the talon gene and transcript\n",
    "df = df.merge(talon_df[['transcript_novelty', 'gene_novelty', 'annot_transcript_id']], \n",
    "              how='left', \n",
    "              left_on='Name',\n",
    "              right_on='annot_transcript_id')\n",
    "df.drop('annot_transcript_id', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ccbefa60-80e0-4c82-8a05-b32cacb71689",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 4. subset for novel transcripts that belong to annotated genes\n",
    "df = df.loc[(df.transcript_novelty!='Known')&\\\n",
    "            (df.gene_novelty=='Known')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c947926c-da09-44b3-86af-c29e6d89c180",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5. get all transcripts that have >= 1 splice site shared w/ a readthrough gene\n",
    "tids = df.loc[df.gene_id_known.isin(fusion_gids), 'Name'].unique().tolist()\n",
    "df = df.loc[df.Name.isin(tids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e14d5929-1c1e-42ab-8cab-e363f89345c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6. get how many unique genes share splice sites with this transcript\n",
    "temp = df.loc[df.known==True].copy(deep=True)\n",
    "temp = temp[['Name', 'gene_id_known']].groupby('Name').nunique().reset_index()\n",
    "temp.rename({'gene_id_known': 'n_genes'}, axis=1, inplace=True)\n",
    "df = df.merge(temp, how='left', on='Name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "1a152801-1d8c-4171-b725-64e9b5c35704",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7. get how many unique genes share splice sites with this transcript\n",
    "temp = df.loc[df.known==True].copy(deep=True)\n",
    "temp = temp.loc[~temp.gene_id_known.isin(fusion_gids)]\n",
    "temp = temp[['Name', 'gene_id_known']].groupby('Name').nunique().reset_index()\n",
    "temp.rename({'gene_id_known': 'n_genes_no_rt'}, axis=1, inplace=True)\n",
    "df = df.merge(temp, how='left', on='Name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "ef6ced26-6647-4c8c-b730-75d4ca249021",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene_id_known</th>\n",
       "      <th>known</th>\n",
       "      <th>n_ss</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_no_rt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000008128.22</td>\n",
       "      <td>TALONT000291637</td>\n",
       "      <td>ENSG00000008128</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000008128.22</td>\n",
       "      <td>TALONT000291641</td>\n",
       "      <td>ENSG00000008128</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000008128.22</td>\n",
       "      <td>TALONT000291742</td>\n",
       "      <td>ENSG00000008128</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000008128.22</td>\n",
       "      <td>TALONT000292018</td>\n",
       "      <td>ENSG00000008128</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000215790.7</td>\n",
       "      <td>TALONT000292468</td>\n",
       "      <td>ENSG00000215790</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "      <td>NNC</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id             Name    gene_id_known  known  n_ss  \\\n",
       "0  ENSG00000008128.22  TALONT000291637  ENSG00000008128   True    12   \n",
       "2  ENSG00000008128.22  TALONT000291641  ENSG00000008128   True    12   \n",
       "4  ENSG00000008128.22  TALONT000291742  ENSG00000008128   True     2   \n",
       "6  ENSG00000008128.22  TALONT000292018  ENSG00000008128   True     6   \n",
       "8   ENSG00000215790.7  TALONT000292468  ENSG00000215790   True    12   \n",
       "\n",
       "  transcript_novelty gene_novelty  n_genes  n_genes_no_rt  \n",
       "0                ISM        Known        2            1.0  \n",
       "2                ISM        Known        2            1.0  \n",
       "4                ISM        Known        2            1.0  \n",
       "6                ISM        Known        2            1.0  \n",
       "8                NNC        Known        2            1.0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 8. limit to transcripts w/ n_genes_no_rt < n_genes\n",
    "# and n_genes_no_rt == 1 (only one choice)\n",
    "# where the gene that was merged with is the non-rt one\n",
    "df = df.loc[(df.n_genes_no_rt==1)&(df.n_genes>df.n_genes_no_rt)]\n",
    "df = df.loc[~df.gene_id_known.isin(fusion_gids)]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "2b7e4e71-2757-4f37-8a04-2e2ad9d1ee20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9. get the number of splice sites supported by the transcript w/ the most\n",
    "# shared splice sites for each transcript + annotated gene combination, \n",
    "# then merge in based on transcript from obs. transcripts\n",
    "temp = f_ss_ic_df.copy(deep=True)\n",
    "temp = temp[['Name', 'ss_id', 'gene_id_known', 'Name_known']].drop_duplicates()\n",
    "temp = temp.groupby(['Name', 'gene_id_known', 'Name_known']).nunique().reset_index()\n",
    "temp.rename({'ss_id': 'n_supp_ss',\n",
    "             'Name_known': 'Name_supp'}, axis=1, inplace=True)\n",
    "temp = temp.sort_values(by='n_supp_ss', ascending=False)\n",
    "temp = temp.drop_duplicates(subset=['Name', 'gene_id_known'], keep='first')\n",
    "df = df.merge(temp, how='left', on=['Name', 'gene_id_known'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1aabce7b-47c0-48ed-a1f5-db04b43a2538",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9.5. get the number of splice sites supported by the transcript w/ the most\n",
    "# shared splice sites for each transcript + annotated gene combination, \n",
    "# then merge in based on transcript from obs. transcripts\n",
    "temp = f_ss_ic_df.copy(deep=True)\n",
    "temp = temp[['Name', 'ss_id', 'gene_id_known', 'Name_known']].drop_duplicates()\n",
    "temp = temp.loc[temp.gene_id_known.isin(fusion_gids)]\n",
    "temp = temp.groupby(['Name', 'gene_id_known', 'Name_known']).nunique().reset_index()\n",
    "temp.rename({'ss_id': 'n_supp_ss_rt',\n",
    "             'Name_known': 'Name_supp_rt',\n",
    "             'gene_id_known': 'gene_id_known_rt'}, axis=1, inplace=True)\n",
    "temp = temp.sort_values(by='n_supp_ss_rt', ascending=False)\n",
    "temp = temp.drop_duplicates(subset=['Name'], keep='first')\n",
    "df = df.merge(temp, how='left', on=['Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "97fe69ee-52fe-4244-a6f5-9d774a0e91cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 10. remove entries where talon gene id is already\n",
    "# the other choice\n",
    "df['gid'] = cerberus.get_stable_gid(df, 'gene_id')\n",
    "df = df.loc[df.gid!='gene_id_known']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e0b4c7c7-514d-4215-a7c2-254218edcacd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 11. get the total # of splice sites per annotated transcript\n",
    "temp = k_ss_ic_df[['Name', 'ss_id']].groupby('Name').nunique().reset_index()\n",
    "temp = temp.rename({'ss_id': 'n_supp_total_ss',\n",
    "                                  'Name': 'Name_supp'}, axis=1)\n",
    "df = df.merge(temp, how='left', on='Name_supp')\n",
    "temp = temp.rename({'n_supp_total_ss': 'n_supp_total_ss_rt',\n",
    "                    'Name_supp': 'Name_supp_rt'}, axis=1)\n",
    "df = df.merge(temp, how='left', on='Name_supp_rt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fa818138-1bc9-469d-bd1f-d7405e3e1415",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "253"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 12. limit to obs. transcripts that were annotated to readthrough genes\n",
    "df = df.loc[df.gid.isin(fusion_gids)]\n",
    "len(df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "4b85a556-9aaa-4a3d-954d-c86a3898e51a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['perc_supp_ss'] = (df.n_supp_ss/df.n_ss)*100\n",
    "df['perc_supp_ss_rt'] = (df.n_supp_ss_rt/df.n_ss)*100\n",
    "df['perc_supp_annot_ss'] = (df.n_supp_ss/df.n_supp_total_ss)*100\n",
    "df['perc_supp_annot_ss_rt'] = (df.n_supp_ss_rt/df.n_supp_total_ss_rt)*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d9c50daf-81cd-485a-ac4c-68375c62cdc7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "218"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = df.loc[df.gid.isin(fusion_gids)]\n",
    "len(temp.Name.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "2cdf5d7e-406a-4d1e-9049-317e775716ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['perc_supp_diff'] = df['perc_supp_ss']-df['perc_supp_ss_rt']\n",
    "df['perc_supp_annot_diff'] = df['perc_supp_annot_ss']-df['perc_supp_annot_ss_rt']\n",
    "df = df.loc[df.perc_supp_annot_ss.notnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6e56d5ef-63b0-49c1-ad20-7b0d29cf9f93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n"
     ]
    }
   ],
   "source": [
    "# 14. compute distances from transcript start and end to closest start and end from any transcript in\n",
    "talon_df = pr.read_gtf(talon_gtf, rename_attr=True, duplicate_attr=True)\n",
    "talon_df = talon_df.df\n",
    "talon_df = talon_df.loc[talon_df.Feature=='transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "13f73d0c-11db-494f-be07-a0a752b34d54",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_df = pr.read_gtf(ref_gtf, rename_attr=True, duplicate_attr=True)\n",
    "known_df = known_df.df\n",
    "known_df = known_df.loc[known_df.Feature=='transcript']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "42c087e7-b375-443b-831b-ebdff010cf0f",
   "metadata": {},
   "outputs": [],
   "source": [
    "known_df['gene_id'] = cerberus.get_stable_gid(known_df, 'gene_id')\n",
    "# add coords to novel thing (df)\n",
    "temp = talon_df[['transcript_id', 'Start', 'End', 'Chromosome', 'Strand']]\n",
    "temp.rename({'transcript_id':'Name'}, axis=1, inplace=True)\n",
    "df = df.merge(temp, how='left', on='Name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1aee21c7-88fd-4443-b018-1b2cc3b06370",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.drop_duplicates()\n",
    "temp3 = pd.DataFrame()\n",
    "temp3['Name'] = df.Name.tolist()\n",
    "for c in ['gene_id_known', 'gene_id_known_rt']:\n",
    "    for feat in ['End', 'Start']:\n",
    "        temp2 = pd.DataFrame() \n",
    "        for g in df[c].unique().tolist():\n",
    "            if feat == 'Start':\n",
    "                other_feat = 'End'\n",
    "            elif feat == 'End':\n",
    "                other_feat = 'Start'\n",
    "            temp_nov = df.loc[df[c] == g].copy(deep=True)\n",
    "            temp_nov = temp_nov[['Name', 'Chromosome', feat, 'Strand']]\n",
    "            temp_nov[other_feat] = temp_nov[feat]+1 \n",
    "\n",
    "            temp_known = known_df.loc[known_df.gene_id == g].copy(deep=True)\n",
    "            temp_known = temp_known[['Chromosome', feat, 'Strand']]\n",
    "            temp_known[other_feat] = temp_known[feat]+1 \n",
    "\n",
    "\n",
    "            temp_nov = pr.PyRanges(temp_nov)\n",
    "            temp_known = pr.PyRanges(temp_known)\n",
    "\n",
    "            # pyranges join and get closest; concat in w/ temp2s\n",
    "            temp_nov = temp_nov.k_nearest(temp_known, k=1, \n",
    "                               overlap=True,\n",
    "                               how=None, suffix='_known').df\n",
    "            try:\n",
    "                temp_nov = temp_nov[['Name', 'Distance']].drop_duplicates()\n",
    "            except:\n",
    "                import pdb; pdb.set_trace()\n",
    "            d_col = f'{c}_{feat}_dist'\n",
    "            temp_nov.rename({'Distance': d_col}, axis=1, inplace=True)        \n",
    "            # if len(temp_nov.loc[temp_nov[d_col].isnull()].index) > 0:\n",
    "            #     import pdb; pdb.set_trace()\n",
    "            temp2 = pd.concat([temp2, temp_nov], axis=0)\n",
    "        temp3 = temp3.merge(temp2, on='Name', how='left')    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "06373e10-b964-4b70-a0f9-259835dfa9c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.merge(temp3, how='left', on='Name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "3b5502fb-abac-4975-a3ee-2e00ef4de760",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add abs. values\n",
    "cols = ['gene_id_known_End_dist',\n",
    "        'gene_id_known_Start_dist',\n",
    "        'gene_id_known_rt_End_dist',\n",
    "        'gene_id_known_rt_Start_dist']\n",
    "for c in cols:\n",
    "    new_col = f'{c}_abs'\n",
    "    df[new_col] = df[c].abs()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "12364c4f-19c8-43bf-ae05-69320ed624a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge with my labels\n",
    "temp = pd.read_csv('labelled_readthrough_fix_candidates.tsv', sep='\\t')\n",
    "df = df.merge(temp, how='left', on='Name')\n",
    "df.to_csv('labelled_readthrough_fix_candidates_with_features.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "2d69d9d2-e931-4239-a255-56168e0f301e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('labelled_readthrough_fix_candidates_with_features.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "00ce6320-4efc-4ec1-aafc-7551406100c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fix_needed</th>\n",
       "      <th>fix</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>could go either way</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>lost cause</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>lost cause</td>\n",
       "      <td>True</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>no</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>no</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>yes</td>\n",
       "      <td>True</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            fix_needed    fix  Name\n",
       "0  could go either way   True     1\n",
       "1           lost cause  False     5\n",
       "2           lost cause   True    12\n",
       "3                   no  False     9\n",
       "4                   no   True     7\n",
       "5                  yes   True    41"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# implement DecisionTree classifier in pandas format w/ thresholds\n",
    "df = pd.read_csv('labelled_readthrough_fix_candidates_with_features.tsv', sep='\\t')\n",
    "df['fix'] = False\n",
    "\n",
    "# start of closest rt transcript must be > 510 bp away\n",
    "inds = df.loc[df.gene_id_known_rt_Start_dist_abs > 510].index\n",
    "df.loc[inds, 'fix'] = True\n",
    "\n",
    "# end of the closest rt transcript must be > 45 kbp away\n",
    "inds = df.loc[df.gene_id_known_rt_End_dist_abs > 45000].index\n",
    "df.loc[inds, 'fix'] = True\n",
    "\n",
    "# % of sss annotated in best matching non-rt transcripts\n",
    "# must be >94\n",
    "# % of sss annotated in actual transcript mush be > 84%\n",
    "inds = df.loc[(df.perc_supp_annot_ss > 94)&(df.perc_supp_annot_ss > 84)].index\n",
    "df.loc[inds, 'fix'] = True\n",
    "\n",
    "df[['fix_needed', 'fix', 'Name']].groupby(['fix_needed', 'fix']).count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "74474ab4-254e-4d9d-b509-3fd4790d92b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Name</th>\n",
       "      <th>gene_id_known</th>\n",
       "      <th>known</th>\n",
       "      <th>n_ss</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>n_genes_no_rt</th>\n",
       "      <th>Name_supp</th>\n",
       "      <th>...</th>\n",
       "      <th>gene_id_known_End_dist</th>\n",
       "      <th>gene_id_known_Start_dist</th>\n",
       "      <th>gene_id_known_rt_End_dist</th>\n",
       "      <th>gene_id_known_rt_Start_dist</th>\n",
       "      <th>gene_id_known_End_dist_abs</th>\n",
       "      <th>gene_id_known_Start_dist_abs</th>\n",
       "      <th>gene_id_known_rt_End_dist_abs</th>\n",
       "      <th>gene_id_known_rt_Start_dist_abs</th>\n",
       "      <th>fix_needed</th>\n",
       "      <th>fix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>ENSG00000260899.1</td>\n",
       "      <td>TALONT000637494</td>\n",
       "      <td>ENSG00000156873</td>\n",
       "      <td>True</td>\n",
       "      <td>18</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000156873_1</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>112</td>\n",
       "      <td>-5986</td>\n",
       "      <td>-7491</td>\n",
       "      <td>2</td>\n",
       "      <td>112</td>\n",
       "      <td>5986</td>\n",
       "      <td>7491</td>\n",
       "      <td>yes</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>ENSG00000260899.1</td>\n",
       "      <td>TALONT000637495</td>\n",
       "      <td>ENSG00000156873</td>\n",
       "      <td>True</td>\n",
       "      <td>14</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000156873_1</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>64</td>\n",
       "      <td>-5986</td>\n",
       "      <td>-7539</td>\n",
       "      <td>2</td>\n",
       "      <td>64</td>\n",
       "      <td>5986</td>\n",
       "      <td>7539</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>ENSG00000260899.1</td>\n",
       "      <td>TALONT000637497</td>\n",
       "      <td>ENSG00000156873</td>\n",
       "      <td>True</td>\n",
       "      <td>16</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000156873_1</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>10</td>\n",
       "      <td>-5986</td>\n",
       "      <td>-7593</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>5986</td>\n",
       "      <td>7593</td>\n",
       "      <td>yes</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>ENSG00000260899.1</td>\n",
       "      <td>TALONT000637508</td>\n",
       "      <td>ENSG00000156873</td>\n",
       "      <td>True</td>\n",
       "      <td>16</td>\n",
       "      <td>NIC</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000156873_5</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>10</td>\n",
       "      <td>-5986</td>\n",
       "      <td>-7593</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>5986</td>\n",
       "      <td>7593</td>\n",
       "      <td>yes</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>ENSG00000260899.1</td>\n",
       "      <td>TALONT000637511</td>\n",
       "      <td>ENSG00000156873</td>\n",
       "      <td>True</td>\n",
       "      <td>16</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000156873_3</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>10</td>\n",
       "      <td>-5986</td>\n",
       "      <td>-7593</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>5986</td>\n",
       "      <td>7593</td>\n",
       "      <td>yes</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>ENSG00000260899.1</td>\n",
       "      <td>TALONT000637622</td>\n",
       "      <td>ENSG00000156873</td>\n",
       "      <td>True</td>\n",
       "      <td>10</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000156873_2</td>\n",
       "      <td>...</td>\n",
       "      <td>-502</td>\n",
       "      <td>0</td>\n",
       "      <td>-2674</td>\n",
       "      <td>-7735</td>\n",
       "      <td>502</td>\n",
       "      <td>0</td>\n",
       "      <td>2674</td>\n",
       "      <td>7735</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>ENSG00000260899.1</td>\n",
       "      <td>TALONT000637792</td>\n",
       "      <td>ENSG00000156873</td>\n",
       "      <td>True</td>\n",
       "      <td>16</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000156873_3</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-5986</td>\n",
       "      <td>-7735</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5986</td>\n",
       "      <td>7735</td>\n",
       "      <td>yes</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>ENSG00000260899.1</td>\n",
       "      <td>TALONT000638355</td>\n",
       "      <td>ENSG00000156873</td>\n",
       "      <td>True</td>\n",
       "      <td>16</td>\n",
       "      <td>ISM</td>\n",
       "      <td>Known</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ENSG00000156873_3</td>\n",
       "      <td>...</td>\n",
       "      <td>-2</td>\n",
       "      <td>0</td>\n",
       "      <td>-5986</td>\n",
       "      <td>-7884</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>5986</td>\n",
       "      <td>7884</td>\n",
       "      <td>yes</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id             Name    gene_id_known  known  n_ss  \\\n",
       "73  ENSG00000260899.1  TALONT000637494  ENSG00000156873   True    18   \n",
       "74  ENSG00000260899.1  TALONT000637495  ENSG00000156873   True    14   \n",
       "75  ENSG00000260899.1  TALONT000637497  ENSG00000156873   True    16   \n",
       "76  ENSG00000260899.1  TALONT000637508  ENSG00000156873   True    16   \n",
       "77  ENSG00000260899.1  TALONT000637511  ENSG00000156873   True    16   \n",
       "78  ENSG00000260899.1  TALONT000637622  ENSG00000156873   True    10   \n",
       "79  ENSG00000260899.1  TALONT000637792  ENSG00000156873   True    16   \n",
       "80  ENSG00000260899.1  TALONT000638355  ENSG00000156873   True    16   \n",
       "\n",
       "   transcript_novelty gene_novelty  n_genes  n_genes_no_rt          Name_supp  \\\n",
       "73                ISM        Known        2            1.0  ENSG00000156873_1   \n",
       "74                ISM        Known        2            1.0  ENSG00000156873_1   \n",
       "75                ISM        Known        2            1.0  ENSG00000156873_1   \n",
       "76                NIC        Known        2            1.0  ENSG00000156873_5   \n",
       "77                ISM        Known        2            1.0  ENSG00000156873_3   \n",
       "78                ISM        Known        2            1.0  ENSG00000156873_2   \n",
       "79                ISM        Known        2            1.0  ENSG00000156873_3   \n",
       "80                ISM        Known        2            1.0  ENSG00000156873_3   \n",
       "\n",
       "    ...  gene_id_known_End_dist gene_id_known_Start_dist  \\\n",
       "73  ...                      -2                      112   \n",
       "74  ...                      -2                       64   \n",
       "75  ...                      -2                       10   \n",
       "76  ...                      -2                       10   \n",
       "77  ...                      -2                       10   \n",
       "78  ...                    -502                        0   \n",
       "79  ...                      -2                        0   \n",
       "80  ...                      -2                        0   \n",
       "\n",
       "   gene_id_known_rt_End_dist  gene_id_known_rt_Start_dist  \\\n",
       "73                     -5986                        -7491   \n",
       "74                     -5986                        -7539   \n",
       "75                     -5986                        -7593   \n",
       "76                     -5986                        -7593   \n",
       "77                     -5986                        -7593   \n",
       "78                     -2674                        -7735   \n",
       "79                     -5986                        -7735   \n",
       "80                     -5986                        -7884   \n",
       "\n",
       "   gene_id_known_End_dist_abs  gene_id_known_Start_dist_abs  \\\n",
       "73                          2                           112   \n",
       "74                          2                            64   \n",
       "75                          2                            10   \n",
       "76                          2                            10   \n",
       "77                          2                            10   \n",
       "78                        502                             0   \n",
       "79                          2                             0   \n",
       "80                          2                             0   \n",
       "\n",
       "    gene_id_known_rt_End_dist_abs  gene_id_known_rt_Start_dist_abs  \\\n",
       "73                           5986                             7491   \n",
       "74                           5986                             7539   \n",
       "75                           5986                             7593   \n",
       "76                           5986                             7593   \n",
       "77                           5986                             7593   \n",
       "78                           2674                             7735   \n",
       "79                           5986                             7735   \n",
       "80                           5986                             7884   \n",
       "\n",
       "    fix_needed   fix  \n",
       "73         yes  True  \n",
       "74         NaN  True  \n",
       "75         yes  True  \n",
       "76         yes  True  \n",
       "77         yes  True  \n",
       "78         NaN  True  \n",
       "79         yes  True  \n",
       "80         yes  True  \n",
       "\n",
       "[8 rows x 37 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tim_gene = 'ENSG00000156873'\n",
    "df.loc[df.gene_id_known==tim_gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "e6a96057-c3d8-430c-b77b-e61006f9c656",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = df.loc[df.fix == True]\n",
    "temp.to_csv('labelled_readthrough_genes_need_fixing.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "34f2cbc9-fe50-41f7-8e87-cedbf2e4b2a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = pd.read_csv('labelled_readthrough_genes_need_fixing.tsv', sep='\\t')\n",
    "assert len(temp.loc[(temp.perc_supp_ss==100)&(temp.perc_supp_annot_ss==100)].index) == 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "d46e0fc1-9c1e-4c0b-8c30-d1fce2b445a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n"
     ]
    }
   ],
   "source": [
    "gtf_df = pr.read_gtf(talon_gtf, duplicate_attr=True, rename_attr=True)\n",
    "gtf_df = gtf_df.df\n",
    "gtf_df['gid_stable'] = cerberus.get_stable_gid(gtf_df, 'gene_id')\n",
    "\n",
    "ref_gtf_df = pr.read_gtf(ref_gtf, duplicate_attr=True, rename_attr=True)\n",
    "ref_gtf_df = ref_gtf_df.df\n",
    "ref_gtf_df['gid_stable'] = cerberus.get_stable_gid(ref_gtf_df, 'gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "7b704996-2df7-4f03-b9a2-bab25433a52f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_id', 'Name', 'gene_id_known', 'known', 'n_ss',\n",
       "       'transcript_novelty', 'gene_novelty', 'n_genes', 'n_genes_no_rt',\n",
       "       'Name_supp', 'n_supp_ss', 'gene_id_known_rt', 'Name_supp_rt',\n",
       "       'n_supp_ss_rt', 'gid', 'n_supp_total_ss', 'n_supp_total_ss_rt',\n",
       "       'perc_supp_ss', 'perc_supp_ss_rt', 'perc_supp_annot_ss',\n",
       "       'perc_supp_annot_ss_rt', 'perc_supp_diff', 'perc_supp_annot_diff',\n",
       "       'Start', 'End', 'Chromosome', 'Strand', 'gene_id_known_End_dist',\n",
       "       'gene_id_known_Start_dist', 'gene_id_known_rt_End_dist',\n",
       "       'gene_id_known_rt_Start_dist', 'gene_id_known_End_dist_abs',\n",
       "       'gene_id_known_Start_dist_abs', 'gene_id_known_rt_End_dist_abs',\n",
       "       'gene_id_known_rt_Start_dist_abs', 'fix_needed', 'fix'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "38511fa0-3cbe-4275-96fb-98c8ce6e4c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# update the genes that need fixing to the genes that were non-rt that they \n",
    "# intersected with\n",
    "gene_cols = ['Source', 'gene_id', 'gene_name', 'gene_status', 'gene_type', 'talon_gene', 'havana_gene', 'level',\n",
    "             'antisense_gene', 'gene_antisense_to_IDs', 'intergenic_novel', 'fusion_novel']\n",
    "talon_gene_cols = ['gene_status', 'talon_gene', 'antisense_gene', 'gene_antisense_to_IDs', 'intergenic_novel', 'fusion_novel']\n",
    "for ind, entry in temp.iterrows():\n",
    "    t = entry.Name\n",
    "    inds = gtf_df.loc[gtf_df.transcript_id==t].index\n",
    "    g = entry.gene_id_known\n",
    "    dummy_gene_entry = gtf_df.loc[(gtf_df.gid_stable==g)&(gtf_df.Feature=='gene')]\n",
    "    if len(dummy_gene_entry.index) == 0:\n",
    "        # have to pull from reference gtf instead and add corresponding gene entry\n",
    "        dummy_gene_entry = ref_gtf_df.loc[(ref_gtf_df.gid_stable==g)&(ref_gtf_df.Feature=='gene')]\n",
    "        dummy_gene_entry[talon_gene_cols] = np.nan\n",
    "        gtf_df = pd.concat([gtf_df, dummy_gene_entry], axis=0)\n",
    "    try:\n",
    "        assert len(dummy_gene_entry.index) == 1\n",
    "    except:\n",
    "        import pdb; pdb.set_trace()\n",
    "    for c in gene_cols:\n",
    "        gtf_df.loc[inds, c] = dummy_gene_entry[c].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "f1c63fb4-f017-4ea7-8a05-7d16ffee77ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop stable gid, sort, update ends, and dump\n",
    "gtf_df.drop('gid_stable', axis=1, inplace=True)\n",
    "gtf_df = cerberus.sort_gtf(gtf_df)\n",
    "# mainly ripped out of cerberus code\n",
    "gtf_temp = gtf_df.copy(deep=True)\n",
    "for mode in ['tss', 'tes']:\n",
    "    fwd, rev = cerberus.get_stranded_gtf_dfs(gtf_temp)\n",
    "    df = pd.DataFrame()\n",
    "    for strand, temp in zip(['+', '-'], [fwd, rev]):\n",
    "\n",
    "        # fix gene boundaries\n",
    "        temp = cerberus.update_gene_ends(temp, mode, strand)\n",
    "        df = pd.concat([df, temp], ignore_index=True)\n",
    "\n",
    "    gtf_temp = df.copy(deep=True)\n",
    "pr.PyRanges(gtf_temp).to_gtf(ofile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "999a5898-c8f4-48ba-8c5d-2bf184b1ac18",
   "metadata": {},
   "outputs": [],
   "source": [
    "ofile = 'annot_readthrough_corrected.gtf'\n",
    "pr.PyRanges(gtf_temp).to_gtf(ofile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3883fdc5-9db9-4662-b097-2110b84aea6c",
   "metadata": {},
   "source": [
    "## Also update the gene abundance matrix with the correct new gene assignments for these transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "382ec3ae-3eb2-47c3-8c70-5d8113c8113b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>gid_stable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TALONT001122495</td>\n",
       "      <td>ENSG00000143543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TALONT000930061</td>\n",
       "      <td>ENSG00000116783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TALONT000930071</td>\n",
       "      <td>ENSG00000116783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TALONT000930083</td>\n",
       "      <td>ENSG00000116783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TALONT000983329</td>\n",
       "      <td>ENSG00000156875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  annot_transcript_id       gid_stable\n",
       "0     TALONT001122495  ENSG00000143543\n",
       "1     TALONT000930061  ENSG00000116783\n",
       "2     TALONT000930071  ENSG00000116783\n",
       "3     TALONT000930083  ENSG00000116783\n",
       "4     TALONT000983329  ENSG00000156875"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = pd.read_csv('labelled_readthrough_genes_need_fixing.tsv', sep='\\t')\n",
    "assert len(temp.loc[(temp.perc_supp_ss==100)&(temp.perc_supp_annot_ss==100)].index) == 0\n",
    "temp = temp[['Name', 'gene_id_known']]\n",
    "temp.rename({'Name': 'annot_transcript_id',\n",
    "             'gene_id_known': 'gid_stable'},\n",
    "            axis=1, inplace=True)\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "35fc3629-3d86-4152-81ba-961f14ad2db3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5395598\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(talon_ab, sep='\\t')\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "84df135b-e261-4d7d-8967-7a48bf75fc28",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>ISM_subtype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>2</td>\n",
       "      <td>535</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A-201</td>\n",
       "      <td>3</td>\n",
       "      <td>1187</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A-202</td>\n",
       "      <td>2</td>\n",
       "      <td>590</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000453576.2</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>AL627309.1-204</td>\n",
       "      <td>2</td>\n",
       "      <td>336</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID  transcript_ID      annot_gene_id annot_transcript_id  \\\n",
       "0        3              4  ENSG00000278267.1   ENST00000619216.1   \n",
       "1        4              6  ENSG00000243485.5   ENST00000469289.1   \n",
       "2        6              8  ENSG00000237613.2   ENST00000417324.1   \n",
       "3        6              9  ENSG00000237613.2   ENST00000461467.1   \n",
       "4       10             19  ENSG00000238009.6   ENST00000453576.2   \n",
       "\n",
       "  annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "0       MIR6859-1         MIR6859-1-201        1      68        Known   \n",
       "1     MIR1302-2HG       MIR1302-2HG-201        2     535        Known   \n",
       "2         FAM138A           FAM138A-201        3    1187        Known   \n",
       "3         FAM138A           FAM138A-202        2     590        Known   \n",
       "4      AL627309.1        AL627309.1-204        2     336        Known   \n",
       "\n",
       "  transcript_novelty ISM_subtype  \n",
       "0              Known        None  \n",
       "1              Known        None  \n",
       "2              Known        None  \n",
       "3              Known        None  \n",
       "4              Known        None  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['gid_stable'] = cerberus.get_stable_gid(df, 'annot_gene_id')\n",
    "df[df.columns[:11]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "0fb854a9-22be-4e7a-bf36-739e5102ff09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>gid_stable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>12</td>\n",
       "      <td>ENSG00000233750.3</td>\n",
       "      <td>CICP27</td>\n",
       "      <td>Known</td>\n",
       "      <td>ENSG00000233750</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID      annot_gene_id annot_gene_name gene_novelty       gid_stable\n",
       "0        3  ENSG00000278267.1       MIR6859-1        Known  ENSG00000278267\n",
       "1        4  ENSG00000243485.5     MIR1302-2HG        Known  ENSG00000243485\n",
       "2        6  ENSG00000237613.2         FAM138A        Known  ENSG00000237613\n",
       "4       10  ENSG00000238009.6      AL627309.1        Known  ENSG00000238009\n",
       "5       12  ENSG00000233750.3          CICP27        Known  ENSG00000233750"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_cols = ['gene_ID', 'annot_gene_id', 'annot_gene_name', 'gene_novelty', 'gid_stable']\n",
    "gene_info = df[gene_cols].drop_duplicates()\n",
    "gene_info.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "0fabd760-e8fd-4794-9f8b-cae4a5ba68d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>gene_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TALONT001122495</td>\n",
       "      <td>ENSG00000143543</td>\n",
       "      <td>3272</td>\n",
       "      <td>ENSG00000143543.14</td>\n",
       "      <td>JTB</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TALONT000930061</td>\n",
       "      <td>ENSG00000116783</td>\n",
       "      <td>1978</td>\n",
       "      <td>ENSG00000116783.14</td>\n",
       "      <td>TNNI3K</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TALONT000930071</td>\n",
       "      <td>ENSG00000116783</td>\n",
       "      <td>1978</td>\n",
       "      <td>ENSG00000116783.14</td>\n",
       "      <td>TNNI3K</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TALONT000930083</td>\n",
       "      <td>ENSG00000116783</td>\n",
       "      <td>1978</td>\n",
       "      <td>ENSG00000116783.14</td>\n",
       "      <td>TNNI3K</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TALONT000983329</td>\n",
       "      <td>ENSG00000156875</td>\n",
       "      <td>2365</td>\n",
       "      <td>ENSG00000156875.13</td>\n",
       "      <td>MFSD14A</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  annot_transcript_id       gid_stable  gene_ID       annot_gene_id  \\\n",
       "0     TALONT001122495  ENSG00000143543     3272  ENSG00000143543.14   \n",
       "1     TALONT000930061  ENSG00000116783     1978  ENSG00000116783.14   \n",
       "2     TALONT000930071  ENSG00000116783     1978  ENSG00000116783.14   \n",
       "3     TALONT000930083  ENSG00000116783     1978  ENSG00000116783.14   \n",
       "4     TALONT000983329  ENSG00000156875     2365  ENSG00000156875.13   \n",
       "\n",
       "  annot_gene_name gene_novelty  \n",
       "0             JTB        Known  \n",
       "1          TNNI3K        Known  \n",
       "2          TNNI3K        Known  \n",
       "3          TNNI3K        Known  \n",
       "4         MFSD14A        Known  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp = temp.merge(gene_info, how='left', on='gid_stable')\n",
    "temp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "3ecc1f1f-53b1-4ebd-ba16-c6ce16a0b37a",
   "metadata": {},
   "outputs": [],
   "source": [
    "temp.drop('gid_stable', axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "48cfaa6a-70fa-4457-9bf7-62f3e61e41dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>k562_3_2</th>\n",
       "      <th>lower_lobe_of_right_lung_1_1</th>\n",
       "      <th>ovary_2_1</th>\n",
       "      <th>pgp1_endo_1_1</th>\n",
       "      <th>right_cardiac_atrium_3_1</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>gene_ID_fix</th>\n",
       "      <th>annot_gene_id_fix</th>\n",
       "      <th>annot_gene_name_fix</th>\n",
       "      <th>gene_novelty_fix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000469289.1</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2HG-201</td>\n",
       "      <td>2</td>\n",
       "      <td>535</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000417324.1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A-201</td>\n",
       "      <td>3</td>\n",
       "      <td>1187</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "      <td>ENSG00000237613.2</td>\n",
       "      <td>ENST00000461467.1</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>FAM138A-202</td>\n",
       "      <td>2</td>\n",
       "      <td>590</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "      <td>ENSG00000238009.6</td>\n",
       "      <td>ENST00000453576.2</td>\n",
       "      <td>AL627309.1</td>\n",
       "      <td>AL627309.1-204</td>\n",
       "      <td>2</td>\n",
       "      <td>336</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000238009</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 154 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_ID  transcript_ID      annot_gene_id annot_transcript_id  \\\n",
       "0        3              4  ENSG00000278267.1   ENST00000619216.1   \n",
       "1        4              6  ENSG00000243485.5   ENST00000469289.1   \n",
       "2        6              8  ENSG00000237613.2   ENST00000417324.1   \n",
       "3        6              9  ENSG00000237613.2   ENST00000461467.1   \n",
       "4       10             19  ENSG00000238009.6   ENST00000453576.2   \n",
       "\n",
       "  annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "0       MIR6859-1         MIR6859-1-201        1      68        Known   \n",
       "1     MIR1302-2HG       MIR1302-2HG-201        2     535        Known   \n",
       "2         FAM138A           FAM138A-201        3    1187        Known   \n",
       "3         FAM138A           FAM138A-202        2     590        Known   \n",
       "4      AL627309.1        AL627309.1-204        2     336        Known   \n",
       "\n",
       "  transcript_novelty  ... k562_3_2  lower_lobe_of_right_lung_1_1  ovary_2_1  \\\n",
       "0              Known  ...        0                             0          0   \n",
       "1              Known  ...        0                             0          0   \n",
       "2              Known  ...        0                             0          0   \n",
       "3              Known  ...        0                             0          0   \n",
       "4              Known  ...        0                             0          0   \n",
       "\n",
       "   pgp1_endo_1_1  right_cardiac_atrium_3_1       gid_stable  gene_ID_fix  \\\n",
       "0              0                         0  ENSG00000278267          NaN   \n",
       "1              0                         0  ENSG00000243485          NaN   \n",
       "2              0                         0  ENSG00000237613          NaN   \n",
       "3              0                         0  ENSG00000237613          NaN   \n",
       "4              0                         0  ENSG00000238009          NaN   \n",
       "\n",
       "   annot_gene_id_fix  annot_gene_name_fix  gene_novelty_fix  \n",
       "0                NaN                  NaN               NaN  \n",
       "1                NaN                  NaN               NaN  \n",
       "2                NaN                  NaN               NaN  \n",
       "3                NaN                  NaN               NaN  \n",
       "4                NaN                  NaN               NaN  \n",
       "\n",
       "[5 rows x 154 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.merge(temp, how='left', on='annot_transcript_id', suffixes=('', '_fix'))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "1e4f549b-a5f7-4ade-a487-90dd139994be",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>k562_3_2</th>\n",
       "      <th>lower_lobe_of_right_lung_1_1</th>\n",
       "      <th>ovary_2_1</th>\n",
       "      <th>pgp1_endo_1_1</th>\n",
       "      <th>right_cardiac_atrium_3_1</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>gene_ID_fix</th>\n",
       "      <th>annot_gene_id_fix</th>\n",
       "      <th>annot_gene_name_fix</th>\n",
       "      <th>gene_novelty_fix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132226</th>\n",
       "      <td>37107</td>\n",
       "      <td>242447</td>\n",
       "      <td>ENSG00000250709.1</td>\n",
       "      <td>TALONT000242447</td>\n",
       "      <td>CCDC169-SOHLH2</td>\n",
       "      <td>TALONT000242447</td>\n",
       "      <td>4</td>\n",
       "      <td>1562</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000250709</td>\n",
       "      <td>37108.0</td>\n",
       "      <td>ENSG00000242715.7</td>\n",
       "      <td>CCDC169</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133785</th>\n",
       "      <td>40722</td>\n",
       "      <td>244006</td>\n",
       "      <td>ENSG00000187951.11</td>\n",
       "      <td>TALONT000244006</td>\n",
       "      <td>AC091057.1</td>\n",
       "      <td>TALONT000244006</td>\n",
       "      <td>14</td>\n",
       "      <td>2783</td>\n",
       "      <td>Known</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000187951</td>\n",
       "      <td>40720.0</td>\n",
       "      <td>ENSG00000285035.1</td>\n",
       "      <td>AC091057.7</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158514</th>\n",
       "      <td>45142</td>\n",
       "      <td>268735</td>\n",
       "      <td>ENSG00000262304.2</td>\n",
       "      <td>TALONT000268735</td>\n",
       "      <td>AC027796.3</td>\n",
       "      <td>TALONT000268735</td>\n",
       "      <td>6</td>\n",
       "      <td>2939</td>\n",
       "      <td>Known</td>\n",
       "      <td>NIC</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000262304</td>\n",
       "      <td>45144.0</td>\n",
       "      <td>ENSG00000197417.7</td>\n",
       "      <td>SHPK</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161398</th>\n",
       "      <td>45149</td>\n",
       "      <td>271619</td>\n",
       "      <td>ENSG00000257950.3</td>\n",
       "      <td>TALONT000271619</td>\n",
       "      <td>P2RX5-TAX1BP3</td>\n",
       "      <td>TALONT000271619</td>\n",
       "      <td>12</td>\n",
       "      <td>2291</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000257950</td>\n",
       "      <td>45152.0</td>\n",
       "      <td>ENSG00000083454.21</td>\n",
       "      <td>P2RX5</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193346</th>\n",
       "      <td>40992</td>\n",
       "      <td>303567</td>\n",
       "      <td>ENSG00000168970.22</td>\n",
       "      <td>TALONT000303567</td>\n",
       "      <td>JMJD7-PLA2G4B</td>\n",
       "      <td>TALONT000303567</td>\n",
       "      <td>20</td>\n",
       "      <td>5541</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000168970</td>\n",
       "      <td>40993.0</td>\n",
       "      <td>ENSG00000243708.10</td>\n",
       "      <td>PLA2G4B</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 154 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_ID  transcript_ID       annot_gene_id annot_transcript_id  \\\n",
       "132226    37107         242447   ENSG00000250709.1     TALONT000242447   \n",
       "133785    40722         244006  ENSG00000187951.11     TALONT000244006   \n",
       "158514    45142         268735   ENSG00000262304.2     TALONT000268735   \n",
       "161398    45149         271619   ENSG00000257950.3     TALONT000271619   \n",
       "193346    40992         303567  ENSG00000168970.22     TALONT000303567   \n",
       "\n",
       "       annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "132226  CCDC169-SOHLH2       TALONT000242447        4    1562        Known   \n",
       "133785      AC091057.1       TALONT000244006       14    2783        Known   \n",
       "158514      AC027796.3       TALONT000268735        6    2939        Known   \n",
       "161398   P2RX5-TAX1BP3       TALONT000271619       12    2291        Known   \n",
       "193346   JMJD7-PLA2G4B       TALONT000303567       20    5541        Known   \n",
       "\n",
       "       transcript_novelty  ... k562_3_2  lower_lobe_of_right_lung_1_1  \\\n",
       "132226                ISM  ...        1                             0   \n",
       "133785                NIC  ...        3                             0   \n",
       "158514                NIC  ...        2                             0   \n",
       "161398                ISM  ...        1                             0   \n",
       "193346                ISM  ...        1                             4   \n",
       "\n",
       "        ovary_2_1  pgp1_endo_1_1  right_cardiac_atrium_3_1       gid_stable  \\\n",
       "132226          0              0                         0  ENSG00000250709   \n",
       "133785          0              0                         0  ENSG00000187951   \n",
       "158514          1              0                         0  ENSG00000262304   \n",
       "161398          0              0                         2  ENSG00000257950   \n",
       "193346          5              0                         0  ENSG00000168970   \n",
       "\n",
       "        gene_ID_fix   annot_gene_id_fix  annot_gene_name_fix  gene_novelty_fix  \n",
       "132226      37108.0   ENSG00000242715.7              CCDC169             Known  \n",
       "133785      40720.0   ENSG00000285035.1           AC091057.7             Known  \n",
       "158514      45144.0   ENSG00000197417.7                 SHPK             Known  \n",
       "161398      45152.0  ENSG00000083454.21                P2RX5             Known  \n",
       "193346      40993.0  ENSG00000243708.10              PLA2G4B             Known  \n",
       "\n",
       "[5 rows x 154 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.gene_ID_fix.notnull()].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "127b8da8-83e7-4615-ad85-5e26632c8950",
   "metadata": {},
   "outputs": [],
   "source": [
    "fix_inds = df.loc[df.gene_ID_fix.notnull()].index\n",
    "gene_cols = [g for g in gene_cols if g!='gid_stable']\n",
    "for g in gene_cols:\n",
    "    fix_col = f'{g}_fix'\n",
    "    df.loc[fix_inds, g] = df.loc[fix_inds, fix_col]\n",
    "    df.drop(fix_col, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "6d9c508c-b69e-436c-83ed-97bc0cd6ce8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>annot_gene_id</th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>annot_gene_name</th>\n",
       "      <th>annot_transcript_name</th>\n",
       "      <th>n_exons</th>\n",
       "      <th>length</th>\n",
       "      <th>gene_novelty</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>...</th>\n",
       "      <th>h9_panc_progen_1_2</th>\n",
       "      <th>hepg2_2_1</th>\n",
       "      <th>hl60_m1_12hr_1_2</th>\n",
       "      <th>hl60_m2_72hr_1_2</th>\n",
       "      <th>k562_3_2</th>\n",
       "      <th>lower_lobe_of_right_lung_1_1</th>\n",
       "      <th>ovary_2_1</th>\n",
       "      <th>pgp1_endo_1_1</th>\n",
       "      <th>right_cardiac_atrium_3_1</th>\n",
       "      <th>gid_stable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132226</th>\n",
       "      <td>37108.0</td>\n",
       "      <td>242447</td>\n",
       "      <td>ENSG00000242715.7</td>\n",
       "      <td>TALONT000242447</td>\n",
       "      <td>CCDC169</td>\n",
       "      <td>TALONT000242447</td>\n",
       "      <td>4</td>\n",
       "      <td>1562</td>\n",
       "      <td>Known</td>\n",
       "      <td>ISM</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ENSG00000250709</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 150 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_ID  transcript_ID      annot_gene_id annot_transcript_id  \\\n",
       "132226  37108.0         242447  ENSG00000242715.7     TALONT000242447   \n",
       "\n",
       "       annot_gene_name annot_transcript_name  n_exons  length gene_novelty  \\\n",
       "132226         CCDC169       TALONT000242447        4    1562        Known   \n",
       "\n",
       "       transcript_novelty  ... h9_panc_progen_1_2  hepg2_2_1  \\\n",
       "132226                ISM  ...                  0          0   \n",
       "\n",
       "        hl60_m1_12hr_1_2  hl60_m2_72hr_1_2  k562_3_2  \\\n",
       "132226                 0                 0         1   \n",
       "\n",
       "        lower_lobe_of_right_lung_1_1  ovary_2_1  pgp1_endo_1_1  \\\n",
       "132226                             0          0              0   \n",
       "\n",
       "        right_cardiac_atrium_3_1       gid_stable  \n",
       "132226                         0  ENSG00000250709  \n",
       "\n",
       "[1 rows x 150 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df.transcript_ID==242447]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "92392d97-25ed-46a9-9133-9f55707607aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5395598"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "463d609e-0c0b-4e46-8e7c-1cfbf118f9a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ok, think this is working, now need to change the whole thing into a function\n",
    "# to run systematically"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
