{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "95b343f5-ba40-42a9-b399-9c07bff5ca12",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *\n",
    "from scripts.mane_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7054de98-06a8-43aa-a8a4-37156b14cb8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../snakemake/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7f83a46f-9f8a-473e-b8a2-3a357c56ff25",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = '../'+expand(config['data']['ab'], species='human')[0]\n",
    "filt_ab = '../'+expand(config['data']['filt_ab'], species='human')[0]\n",
    "read_annot = '../'+expand(config['data']['read_annot'], species='human')[0]\n",
    "t_metadata = '../'+expand(config['ref']['cerberus']['t_info'], species='human')[0]\n",
    "lib_meta = '../'+expand(config['data']['meta'], species='human')[0]\n",
    "swan_file = '../'+expand(config['data']['sg'], species='human')[0]\n",
    "cerberus_h5 = '../'+expand(config['data']['cerb_annot'], species='human')[0]\n",
    "cerb_t_metadata = '../'+expand(config['data']['t_info'], species='human')[0]\n",
    "major_isos = '../'+expand(config['data']['major_isos'], species='human', obs_col='sample')[0]\n",
    "pi_tpm_table = '../'+expand(config['data']['pi_tpm']['triplet'], species='human', obs_col='sample')[0]\n",
    "pp_summary = '../'+expand(config['data']['p_pred']['summary'], species='human')[0]\n",
    "ref_t_metadata = '../'+expand(config['ref']['t_info'], species='human')[0]\n",
    "ref_g_metadata = '../'+expand(config['ref']['g_info'], species='human')[0]\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "min_tpm = 1\n",
    "gene_subset = 'polya'\n",
    "obs_col = 'sample'\n",
    "go_gene_subset = 'protein_coding'\n",
    "predom_iso_subset = 'protein_coding'\n",
    "\n",
    "m_lib_meta = '../'+expand(config['data']['meta'], species='mouse')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "be7455f8-9f93-4b49-9d61-bb15a4ce5946",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_browser_isos(ca, sg, gene,\n",
    "                      obs_col, obs_condition, \n",
    "                      filt_ab, pp_summary, \n",
    "                      major_set,\n",
    "                      h=0.1, w=56, fig_w=14, species='human',\n",
    "                      add_tss=False,\n",
    "                      add_ccre=False,\n",
    "                      major=False,\n",
    "                      order='expression',\n",
    "                      light_shade=None,\n",
    "                      dark_shade=None):\n",
    "    \"\"\"\n",
    "    Plot browser style isoform models for a given sample\n",
    "    \"\"\"\n",
    "\n",
    "    def plot_tss(ca, sg, tpm_df, x, y, h, ax):\n",
    "        tpm_df = add_feat(tpm_df, kind='tss', col='transcript_id')\n",
    "        tpm_df.head()\n",
    "        tss_df = ca.tss.loc[ca.tss.Name.isin(tpm_df.tss.tolist())]\n",
    "        tss_df.head()\n",
    "        regions = [(entry.Start, entry.End) for ind, entry in tss_df.iterrows()]\n",
    "        color = get_sector_colors()[0]['tss']\n",
    "        ax = sg.pg.plot_regions(regions, sg.pg.scale, sg.pg.strand, sg.pg.g_min, sg.pg.g_max, x, y, h, color, ax)\n",
    "        return ax\n",
    "    \n",
    "    def plot_cds(entry, sg, x, y, h, ax):\n",
    "        c_dict = get_cds_colors()\n",
    "        \n",
    "#         def get_inc_region(entry, how, scale):\n",
    "#             inc = 55000*scale\n",
    "#             if how == 'start' and entry.Strand == '+':\n",
    "#                 col = 'CDS_Start'\n",
    "#             elif how == 'stop' and entry.Strand == '+':\n",
    "#                 col = 'CDS_Stop'\n",
    "#             elif how == 'start' and entry.Strand == '-':\n",
    "#                 col = 'CDS_Stop'\n",
    "#             elif how == 'stop' and entry.Strand == '-':\n",
    "#                 col = 'CDS_Start'\n",
    "#             if entry.Strand == '+':\n",
    "#                 regions = (entry[col],entry[col]+inc)\n",
    "#             elif entry.Strand == '-':\n",
    "#                 regions = (entry[col],entry[col]-inc)\n",
    "#             regions = [regions]\n",
    "#             print(f'Scale: {scale}')\n",
    "#             print(f'CDS inc: {inc}')\n",
    "            \n",
    "#             return regions \n",
    "\n",
    "        # get start and stop \n",
    "        pdb.set_trace()\n",
    "    \n",
    "        if entry.Strand == '+':\n",
    "            orf_start = 'CDS_Start'\n",
    "            orf_stop = 'CDS_Stop'\n",
    "        elif entry.Strand == '-':\n",
    "            orf_stop = 'CDS_Start'\n",
    "            orf_start = 'CDS_Stop'\n",
    "\n",
    "        # iterate through exons\n",
    "        loc_path = sg.pg.loc_path\n",
    "        exons = [(sg.pg.loc_df.loc[v1, 'coord'],\n",
    "            sg.pg.loc_df.loc[v2, 'coord']) \\\n",
    "            for v1,v2 in zip(loc_path[:-1],loc_path[1:])][::2]\n",
    "        \n",
    "        \n",
    "        # start codon\n",
    "        color = c_dict['start']\n",
    "        regions = get_inc_region(entry, 'start', sg.pg.scale)\n",
    "        ax = sg.pg.plot_regions(regions, sg.pg.scale, sg.pg.strand, sg.pg.g_min, sg.pg.g_max, x, y, h, color, ax)\n",
    "        \n",
    "        # stop codon\n",
    "        color = c_dict['stop']\n",
    "        regions = get_inc_region(entry, 'stop', sg.pg.scale)\n",
    "        ax = sg.pg.plot_regions(regions, sg.pg.scale, sg.pg.strand, sg.pg.g_min, sg.pg.g_max, x, y, h, color, ax)\n",
    "       \n",
    "        return ax\n",
    "        \n",
    "\n",
    "    def plot_ccre(ca, sg, x, y, h, ax):\n",
    "\n",
    "        # ccre regions\n",
    "        sources = ['pls', 'pels', 'dels']\n",
    "        ccre = ca.tss_map.loc[ca.tss_map.source.isin(sources)].copy(deep=True)\n",
    "\n",
    "        # get ranges w/i this region\n",
    "        min_coord = sg.pg.g_min\n",
    "        max_coord = sg.pg.g_max\n",
    "        chrom = sg.pg.chrom\n",
    "\n",
    "        ccre['min_coord'] = ccre[['Start', 'End']].min(axis=1)\n",
    "        ccre['max_coord'] = ccre[['Start', 'End']].max(axis=1)\n",
    "\n",
    "        # subset on regions\n",
    "        ccre = pr.PyRanges(ccre)\n",
    "        region = pr.from_dict({'Chromosome': [chrom],\n",
    "                                        'Start': [min_coord],\n",
    "                                        'End': [max_coord]})\n",
    "        ccre = ccre.intersect(region, strandedness=None)\n",
    "        ccre = ccre.as_df()\n",
    "\n",
    "        # colors\n",
    "        c_dict, _ = get_ccre_colors()\n",
    "        colors = [c_dict[s] for s in ccre.source.tolist()]\n",
    "        regions = [(entry.Start, entry.End) for ind, entry in ccre.iterrows()]\n",
    "        ax = sg.pg.plot_regions(regions, sg.pg.scale, sg.pg.strand, sg.pg.g_min, sg.pg.g_max, x, y, h, colors, ax)\n",
    "\n",
    "        return ax\n",
    "\n",
    "    def get_major_isos(major_set, gene, sample=None):\n",
    "        \"\"\"\n",
    "        Get list of major isfoorms in a given sample\n",
    "        \"\"\"\n",
    "        df = pd.read_csv(major_set, sep='\\t')\n",
    "        df = df.loc[df.gname == gene]\n",
    "        if sample:\n",
    "            df = df.loc[df['sample'] == sample]\n",
    "        tids = df.tid.unique().tolist()\n",
    "        return tids\n",
    "\n",
    "    def get_isos(ca, filt_ab, gene, sample, species):\n",
    "        df = pd.read_csv(filt_ab, sep='\\t')\n",
    "        df = get_det_table(df,\n",
    "                       groupby='sample',\n",
    "                       how='iso',\n",
    "                       min_tpm=1,\n",
    "                       gene_subset='polya',\n",
    "                       species=species)\n",
    "        df = df.loc[sample]\n",
    "        df = df.to_frame()\n",
    "        df = df.loc[df[sample]==True]\n",
    "        gid = ca.triplets.loc[ca.triplets.gname==gene, 'gid'].values[0]\n",
    "        df.reset_index(inplace=True)\n",
    "        df['gid'] = df['index'].str.split('[', expand=True)[0]\n",
    "        df = df.loc[df.gid == gid]\n",
    "        tids = df['index'].tolist()\n",
    "        return tids\n",
    "\n",
    "\n",
    "    def get_tpm_df(sg, tids, obs_col, obs_condition):\n",
    "        tpm_df = swan.calc_tpm(sg.adata, obs_col=obs_col).sparse.to_dense()\n",
    "        tpm_df = tpm_df.transpose()\n",
    "        tpm_df = tpm_df.loc[tids, obs_condition].to_frame()\n",
    "        return tpm_df\n",
    "\n",
    "    if major:\n",
    "        tids = get_major_isos(major_set, gene, obs_condition)\n",
    "    else:\n",
    "        tids = get_isos(ca, filt_ab, gene, obs_condition, species)\n",
    "    tpm_df = get_tpm_df(sg, tids, obs_col, obs_condition)\n",
    "\n",
    "    # colormap definition\n",
    "    if not light_shade:\n",
    "        light_shade = get_sector_colors()[0]['mixed']\n",
    "    if not dark_shade:\n",
    "        dark_shade = get_sector_colors()[0]['simple']\n",
    "    cmap = mpl.colors.LinearSegmentedColormap.from_list('', [light_shade, dark_shade])\n",
    "\n",
    "\n",
    "    # # font sizes\n",
    "    # # small_text = 6/(6.11/16)\n",
    "    # small_text = 20.3\n",
    "    # big_text = 6.71/(6.11/16)\n",
    "    # # print('small text size: {}'.format(small_text))\n",
    "    # # print('big text size: {}'.format(big_text))\n",
    "\n",
    "\n",
    "#     # height spacing b/w models\n",
    "#     if not h_space:\n",
    "#         h_space = h*1.75\n",
    "#     # print('h_space : {}'.format(h_space))\n",
    "#     # print('h: {}'.format(h))\n",
    "\n",
    "#     # plotting settings\n",
    "#     fig_len = len(tids)\n",
    "#     fig_len += 1 # for scale\n",
    "#     if add_tss:\n",
    "#         fig_len += 1\n",
    "#     if add_ccre:\n",
    "#         fig_len += 1\n",
    "#     fig_h = h_space*(fig_len-1)+h\n",
    "#     fig_h += 0.3 # vertical spacing adjustment\n",
    "#     # print('fig h: {}'.format(fig_h))\n",
    "#     # print('fig w: {}'.format(fig_w))\n",
    "    \n",
    "    # trying to make this work w/ heatmap\n",
    "    # plt.figure(1, figsize=(fig_w, fig_h), frameon=False)\n",
    "    # ax = plt.gca()\n",
    "    \n",
    "    fig_h = len(tids)/4\n",
    "    \n",
    "    w1_rat = (fig_w/6)*2\n",
    "    w2_rat = (fig_w/6)*4\n",
    "    fig, (ax, ax2) = plt.subplots(1,2, figsize=(fig_w, fig_h),\n",
    "                                  gridspec_kw={'width_ratios':(w1_rat, w2_rat)},\n",
    "                                  frameon=False, sharey=True)\n",
    "    fig.subplots_adjust(wspace=0.00)\n",
    "    fig.subplots_adjust(hspace=0.00)\n",
    "    \n",
    "    mpl.rcParams['font.family'] = 'Arial'\n",
    "    mpl.rcParams['pdf.fonttype'] = 42\n",
    "\n",
    "    # plotting order\n",
    "    tpm_df = tpm_df.sort_values(by=obs_condition, ascending=False)\n",
    "    tpm_df = tpm_df[[obs_condition]]\n",
    "\n",
    "    if order == 'tss':\n",
    "        tpm_df = add_feat(tpm_df, kind='tss', col='index')\n",
    "        tpm_df.sort_values(by=['tss', obs_condition], ascending=[True, False], inplace=True)\n",
    "        tpm_df.drop('tss', axis=1, inplace=True)\n",
    "\n",
    "    # add reference ids\n",
    "    if species == 'human':\n",
    "        ref_source = 'v40'\n",
    "    elif species == 'mouse':\n",
    "        ref_source = 'vM25'\n",
    "    df = ca.t_map.loc[ca.t_map.source == ref_source]\n",
    "    df = df[['transcript_id','original_transcript_id', 'original_transcript_name']]\n",
    "    tpm_df = tpm_df.merge(df, how='left', left_index=True, right_on='transcript_id')\n",
    "\n",
    "    # clean up for transcripts that aren't known\n",
    "    df = ca.t_map.loc[ca.t_map.source == 'lapa']\n",
    "    df = df[['gene_name', 'transcript_id', 'transcript_name',\n",
    "             'tss_id', 'ic_id', 'tes_id']].drop_duplicates()\n",
    "    tpm_df = tpm_df.merge(df, how='left', on='transcript_id')\n",
    "    tpm_df.fillna('N/A', inplace=True)\n",
    "\n",
    "    if species == 'human':\n",
    "        refs = ['v40', 'v29']\n",
    "    elif species == 'mouse':\n",
    "        refs = ['vM21', 'vM25']\n",
    "    known_tids = ca.t_map.loc[ca.t_map.source.isin(refs)].transcript_id.unique().tolist()\n",
    "    tpm_df.loc[tpm_df.transcript_id.isin(known_tids), 'Known_str'] = '*'\n",
    "    tpm_df['Known'] = tpm_df.transcript_id.isin(known_tids)\n",
    "    \n",
    "\n",
    "    # triplets rather than entire transcript name\n",
    "    tpm_df['triplet'] = tpm_df.transcript_id.str.split('[', n=1, expand=True)[1]\n",
    "    tpm_df['triplet'] = tpm_df.triplet.str.split(']', n=1, expand=True)[0]\n",
    "    tpm_df['triplet'] = '['+tpm_df.triplet+']'\n",
    "    \n",
    "    # NMD and ORF \n",
    "    pp_df = pd.read_csv(pp_summary, sep='\\t')\n",
    "    pp_df.rename({'tid': 'transcript_id'}, axis=1, inplace=True)\n",
    "    tpm_df = tpm_df.merge(pp_df[['transcript_id', 'nmd', 'full_orf', \n",
    "                                 'seq', 'Strand', 'CDS_Start', 'CDS_Stop']], \n",
    "                          how='left', \n",
    "                          on='transcript_id')\n",
    "    \n",
    "    # are these the mane orf\n",
    "    if species == 'human':\n",
    "        tpm_df['gid'] = tpm_df['ic_id'].str.split('_', expand=True)[0]\n",
    "        gid = tpm_df.gid.values[0]\n",
    "        mane_orf = get_mane_orf(pp_summary,\n",
    "                                'v40_cerberus',\n",
    "                                gid=gid).seq.values[0]\n",
    "        tpm_df['is_mane_orf'] = tpm_df['seq'] == mane_orf        \n",
    "    \n",
    "    def add_bool_heatmap(tpm_df, ax):\n",
    "        keep_cols = ['triplet', 'gene_name', 'is_mane_orf',\n",
    "                     'full_orf', 'nmd', 'Known']\n",
    "        temp = tpm_df[keep_cols]\n",
    "        temp['iso_trip'] = temp.gene_name+' '+temp.triplet\n",
    "        temp.drop(['triplet', 'gene_name'], axis=1, inplace=True)\n",
    "        temp.set_index('iso_trip', inplace=True)\n",
    "        temp.index.name = ''\n",
    "        temp.rename({'is_mane_orf': 'MANE ORF',\n",
    "                     'full_orf': 'Full ORF',\n",
    "                     'nmd': 'NMD'}, axis=1, inplace=True)\n",
    "        # bool_cols = ['is_mane_orf', 'full_orf', 'nmd', 'Known']\n",
    "        bool_cols = ['MANE ORF', 'Full ORF', 'NMD', 'Known']\n",
    "        temp2 = pd.DataFrame()\n",
    "\n",
    "        m = {True: '*', False: ''}\n",
    "        for c in bool_cols:\n",
    "            new_col = c+'_str'\n",
    "            temp2[new_col] = temp[c].map(m)\n",
    "\n",
    "        ax = sns.heatmap(temp, cbar=False, cmap='Purples',\n",
    "                         linewidths=0.5, linecolor='k',\n",
    "                         annot=temp2, square=True, \n",
    "                         fmt='', ax=ax)\n",
    "        ax.tick_params(left=False,\n",
    "                       right=False, labelright=False,\n",
    "                       bottom=False, labelbottom=False,\n",
    "                       labeltop=True)\n",
    "        ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "        return ax\n",
    "    \n",
    "    # first add the labels\n",
    "    ax = add_bool_heatmap(tpm_df, ax)\n",
    "    \n",
    "    # get y coords from heatmap\n",
    "    y_locs = ax.get_yticks()\n",
    "    \n",
    "    # get x coords from heatmap\n",
    "    x = 0    \n",
    "    i = 0\n",
    "    # tpm_df = tpm_df.iloc[::-1]\n",
    "    for index, entry in tpm_df.iterrows():\n",
    "        \n",
    "        y_ytick = y_locs[i]\n",
    "\n",
    "        # y coords\n",
    "        # y = (len(tpm_df.index) - i)*(h_space)\n",
    "        y = y_ytick-(h/2)\n",
    "\n",
    "        # tid\n",
    "        tid = entry['transcript_id']\n",
    "        # color by TPM\n",
    "        if len(tpm_df.index) == 1:\n",
    "            norm_val = entry[obs_condition]\n",
    "        else:\n",
    "            norm_val = (entry[obs_condition]-tpm_df[obs_condition].min())/(tpm_df[obs_condition].max()-tpm_df[obs_condition].min())\n",
    "        color = cmap(norm_val)\n",
    "        \n",
    "        # plot models + cds\n",
    "        ax2 = sg.plot_browser(tid, y=y, x=x, h=h, w=w, color=color,\n",
    "                              ax=ax2, linewidth=1.5)\n",
    "        plot_cds(entry, sg, x, y, h, ax2)\n",
    "        \n",
    "        i+=1\n",
    "        \n",
    "    plt.tight_layout()\n",
    "    ax2.get_xaxis().set_visible(False)\n",
    "    ax2.get_yaxis().set_visible(False)\n",
    "    ax2.set_frame_on(False)     \n",
    "    \n",
    "    return ax, tpm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "912be149-5a8d-443b-b3eb-288b67eb09e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_browser_isos_2(h5,\n",
    "                        swan_file,\n",
    "                        filt_ab,\n",
    "                        pp_summary,\n",
    "                        major_isos,\n",
    "                        gene,\n",
    "                        obs_col,\n",
    "                        obs_condition,\n",
    "                        ofile,\n",
    "                        h=0.2,\n",
    "                        **kwargs):\n",
    "    \n",
    "    ca = cerberus.read(h5)\n",
    "    sg = swan.read(swan_file)\n",
    "\n",
    "    ax, tpm_df = plot_browser_isos(ca,\n",
    "                                   sg,\n",
    "                                   gene,\n",
    "                                   obs_col,\n",
    "                                   obs_condition,\n",
    "                                   filt_ab,\n",
    "                                   pp_summary,\n",
    "                                   major_isos,\n",
    "                                   h=h,\n",
    "                                   w=40,\n",
    "                                   **kwargs)\n",
    "    plt.savefig(ofile, dpi=500)\n",
    "\n",
    "    return ax, tpm_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "85107dae-2188-4cce-b293-9db77e921f02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ca = cerberus.read(cerberus_h5)\n",
    "# sg = swan.read(swan_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e1c0525-9c8b-4d93-b707-3453a161e8bc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for polya genes\n",
      "Enforcing minimum TPM\n",
      "Total # isos detected: 231136\n",
      "# isos >= 1 tpm: 210239\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 206806\n",
      "Found 50 total samples\n"
     ]
    }
   ],
   "source": [
    "ax, tpm_df = plot_browser_isos(ca,\n",
    "                               sg,\n",
    "                               'PKM',\n",
    "                               obs_col,\n",
    "                               'ovary',\n",
    "                               filt_ab,\n",
    "                               pp_summary,\n",
    "                               major_isos,\n",
    "                               h=0.6,\n",
    "                               w=10,\n",
    "                               fig_w=6)\n",
    "plt.savefig('test.pdf', dpi=300)\n",
    "# ax, tpm_df = plot_browser_isos_2(cerberus_h5,\n",
    "#                     swan_file, \n",
    "#                     filt_ab,\n",
    "#                     pp_summary,\n",
    "#                     major_isos,\n",
    "#                     'PKM',\n",
    "#                     obs_col,\n",
    "#                     'ovary',\n",
    "#                     'figures/pkm_ovary_browser_isos.pdf',\n",
    "#                     fig_w=8,\n",
    "#                     h=0.3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c59be875-87d2-4cf6-9c14-224f1d809de8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_bool_heatmap(tpm_df, ax):\n",
    "    keep_cols = ['triplet', 'gene_name', 'is_mane_orf', 'full_orf', 'nmd', 'Known']\n",
    "    temp = tpm_df[keep_cols]\n",
    "    temp['iso_trip'] = temp.gene_name+' '+temp.triplet\n",
    "    temp.drop(['triplet', 'gene_name'], axis=1, inplace=True)\n",
    "    temp.set_index('iso_trip', inplace=True)\n",
    "    temp.index.name = ''\n",
    "    temp.rename({'is_mane_orf': 'MANE ORF',\n",
    "                 'full_orf': 'Full ORF',\n",
    "                 'nmd': 'NMD'}, axis=1, inplace=True)\n",
    "    # bool_cols = ['is_mane_orf', 'full_orf', 'nmd', 'Known']\n",
    "    bool_cols = ['MANE ORF', 'Full ORF', 'NMD', 'Known']\n",
    "    temp2 = pd.DataFrame()\n",
    "\n",
    "    m = {True: '*', False: ''}\n",
    "    # annot_kws = dict()\n",
    "    for c in bool_cols:\n",
    "        new_col = c+'_str'\n",
    "        temp2[new_col] = temp[c].map(m)\n",
    "        # annot_kws[c] = new_col\n",
    "\n",
    "    temp.head()\n",
    "    temp2.head()\n",
    "\n",
    "    ax = sns.heatmap(temp, cbar=False, cmap='Purples',\n",
    "                     linewidths=0.5, linecolor='k',\n",
    "                     annot=temp2, square=True, \n",
    "                     fmt='', ax=ax)\n",
    "    ax.tick_params(left=False,\n",
    "                   right=False, labelright=False,\n",
    "                   bottom=False, labelbottom=False,\n",
    "                   labeltop=True)\n",
    "    _ = plt.xticks(rotation=90)\n",
    "    \n",
    "    return ax\n",
    "\n",
    "\n",
    "ax = plt.gca()\n",
    "ax = add_bool_heatmap(tpm_df, ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92934a71-c5ce-4714-a154-5b682c185176",
   "metadata": {},
   "outputs": [],
   "source": [
    "ytickslocs = ax.get_yticks()\n",
    "ytickslocs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e112b18-c0ee-4628-b37a-68f6b8482a2e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0.5, 1.5, 2.5, 3.5])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xtickslocs = ax.get_xticks()\n",
    "xtickslocs\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3410218-1055-45b3-8d4b-2e81eec951c1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d61320c1-1c8f-40a1-84af-e5f74867a7a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from ../data/human/swan_graph.p\n",
      "Calculating iso TPM values\n",
      "Subsetting for polya genes\n",
      "Enforcing minimum TPM\n",
      "Total # isos detected: 231136\n",
      "# isos >= 1 tpm: 210239\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 206806\n",
      "Found 50 total samples\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAygAAAA4CAYAAAAWytz3AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/NK7nSAAAACXBIWXMAAAsTAAALEwEAmpwYAAAM0ElEQVR4nO3dfbBcRZ3G8e+DEAQFYSGWa0WTCKhJQGJU8KUQoluiARRfsqyBtXAVFVylSvEVleBLARvFl6KQQkpKEIQsiu6uUCwWBGR9WRV5jQFFkyqUoEYSlQDh5ecf3eM9TM7M9Jk7N3fu3OdTdWvOnNP96z4z5947Pd19WhGBmZmZmZnZMNhusitgZmZmZmbW4gaKmZmZmZkNDTdQzMzMzMxsaLiBYmZmZmZmQ8MNFDMzMzMzGxpuoJiZmZmZ2dBo1ECRdKCkCyStlfSgpE2Sfizpg5Ke3CXfkySdJOl6Sb+XtCU/fk/S8ZJm1OSZIynyz6qmJyZpV0l/zPnPKUh/WaW8nzUoZ24+n5D04cI81XNr/dzUIe3RlTQvLq3XZMSQdKSka/J18ZCkOyR9StLO+fjymvMOSXNqYr2zcvxN/dTZzMzMzKaeogaKpO0kfQ74EfCvwGxgR2BX4ADgDODnkp5Vk3c+cBuwAjgImAnskB9fCZwN3CJpn3GfzViZAr4A7FGYfg/giMquRZIWFuSbAZxDOp+BkzQLOG0qxJB0CnA5sJh0XcwAng18DLhG0hMbFru0R3l3Svpow5hmZmZmNuS2L0x3KvC+vH0fcBawGpgLvAf4R2Bv4FuSXhARjwJIehpwNfD0nHcN8FVgHTALeCuwL/Ac0ofYRRHxh/GcUP62/qwcu9Qy0gfqqreRzq1TOXsAFwCvalrHNpcClwCb2uLvB3wLeEa/gbdVDEl7AR/PT/8EfIZ0nbwfWAAcyONfy/uA3bvEmwkcAvwF2KWyf3vgn4D/zeX8SdLuwP4RsarZmdWWW79q6cb8uBtEhMZbzmTodG5159NKWz0mKdqfd4sxKHV1aZKm17Hx1r1T/KaxS17zJvt61a9TutI61KUtKa/9Oqy7ppqUV5evWxnbQq/z6HXOpe9Zk/Kb/r4O6jVsEqfXdVzyN6C9nCbvRV2cXmm6XY+d8jaN102/vzPd4vVbh/ayOv4/rSh5/evSldSlJEZ7+nX5K/vZj5WUNv1s67+lPXtQJO0NtL6p3gC8KCI+ERGXRMRpwCLgt/n4/sBrK9lXMNY4uRJYGBErImJlRJyZ8/5XPj4LOH08JyNpKXAzzRonAMfmxwcY+zh6dKdv/SUdT+oVWtK8lltZExHfjohrc+wnS1oB/ITU6GtsEmK8HnhC3j45Is6MiPOBoyppDq9sX9Ij3htyvP9u238I6Tr6JfA04BjStbdSkudTmZmZmY2Akg9176ik+4+IuKt6MCLWA58ELgKOI32gRdIujH1A3QK8NSIeasv7MPB2oLX/GElP6eM8Wk4mfZh+CDi/JEPuIViUn/4PsDJv7076oFznNNIH5I30/rDd1N7ASaQhdKuB66dAjFuBz5Neu2q6dZXt6vt6HbC5S7x/Bm4iNUSqrgGeT3rNZ5OGkL0XeElE+DsPMzMzsxFQ0kA5pLJ9ZV2CiDg3Io6JiPMi4u68+2DG5mZcGxH3dsj7B+Da/HQG8LKCOnUSwA05xgWFeaq9LRcDF1aev61DnseA7wIvBK5qWMcSD5KGqR0A/GbYY0TEVRHxvog4KiJWVw4trmxX8z8K/KIuVh7edTBjDcVqOY+RGi1L8vHtgDntjWYzMzMzm7pK5qDsVdn+dYPYz6xsr+6YKlkDvDpvbzXRvoGjIuJOAEmH9Eqc5zQcnZ/eB1wBPEz6MD0XWCxpbkS0fzh/caWcg8ZR3zq/AWZHxO9z/CkZQ9I/AJ+t7Pomjx8qthp4QU3WN5KGd60k3ZCh3ZGk92YJadjYGZLOjojfNaqgmZmZmQ2lkh6UXSvbDzSIXZ0EfX+PtNXjfQ/xajUaGlgCPDVvfzMitkREkIarAQj4twGUUywiNrUaBVM1Rh6mdyVpCBbALWw9FG5th+xLgRs79YpExEXAvIi4JyK+Asx348TMzMxsdJT0oGxmrJHyJNKdlUpUGx27dEy19fE/F8YfhPbhXS0Xkm6PC3CspFM8x6GMpN1Id9l6Ud71Z2BZRDzS1gtTfT1nAWslPZU0vOvkbmVExD2V7fUDqHaxkjuTTCXdzqf9WKe02+I1KSmjybk0iVuiLk4/sUte837KalKX0ve927HS8pqey2S+x001qc+grp9+y+831qDj9Pve90o7ntdiPHUa5PXdtJx+Y47nvZ6osgZx/Y3a/+7poKQHpTrRea+6BJL2kfTKPGSqZW1le78eZexb2e53vkQjkvYEDqvsukZ5YUDgjsr+WYwNP7MuJO1Kuq10q3FyP3BERNzeI+uh+bF1966t5p+YmZmZ2fRQ0kBZVdk+tEOafwe+B6yXtCzvu4E0GRrg5ZJq19HI+1+enz4M/F9BnQbhaMoXWOw0Wd4ySTsA3yHdOADgr8CSiCi5g9hr8uNS4Kc1c37MzMzMbJooGeJ1LqkBIuAkSZdGxNrWQUkLSLcXhrRy+y8BIuKPklYCb87lfE3SYRHxQCXvTsDXKvW4OCLuG98pFTu2sv1Z6ufJfJh0m90jJM2McS4iOeI+x9gd3zYDr46I0sbmIknPIw3v+sgE1G1gtvVCRYPSZAGsTgu9jWfht36VLFrXLU2vY+Ote7eF0poukNYep9Oibk3KKnn9qulK61CXtqS89uuwZIHCXtdae75uZWwLvc6j1zmXvmdNym/6+zqo17BJnF7XccnfgPZymrwXdXF6pem1+F/p72+3eN30+zvTLV6/dWgvq2RIVcnrX5eupC4lMTzsa7j1bKBExG2SziStCr4n8FNJXwLuBOYBJwI75eSXRcRPKtk/QFr5eybplrO3SjqXNPxrNqlhs09Oew/woQ7VmCNpeZdqnh4RD/Y6lxZJC4GF+entEfGBDun2Jd1VagfgLaQP4UND0s6MrWR/b0T8cJJi7EtqxLZ8F5gp6cjKvm5zl0RqCHt4l5mZmdk0V9KDAqnhsDNwPKmX5NSaND9grCcFgIj4raRXkFaLn0uaw3JGTd5fkeYq1K6VQmrMnNKlfl8grdlR6tjK9qVd0l1IaqBAGuY1VA0U0h3ILs/bV9HfXJlBxHg3qZHRsjT/VN1B50Ut/wIcCPx/RKzrkMbMzMzMpoGSOShExKMRcQKpF+QyUm/Hw8Am4PvAu4CDI2JjTd7bgAWkFb9XARuAR/LjKuA9wH4RsWZ8p1Imz5VYVtnV7Rv7K0j1BJgn6aUTVrHxG8RdxvqNsbh3kq6uzo/uPTEzMzOb5pSW/bBtSdIcxu5WdmpELB9nvFuBWyNiWc/EExijYXnLGesVm1ud1zRZOo5H3Zgfd/MclLp4noPiOSil5XkOiuegdEvrOSieg1KSrqQuJTHa06/LX9nP9qIStbb139KiHhSbUM+VdKSkvnohJB1Fuk3zqn4rMIgYDcpanOemPHeiyzIzMzOzqcc9KJOgrQel5eaIWNgwznzgduB84LiIeLRHlgmJ0bC8m4D923YPRQ+KmZmZmU2+0knyNoQiYrWkvSPirsmMYWZmZmY2KO5BMTMzMzOzoeE5KGZmZmZmNjTcQDEzMzMzs6HhBoqZmZmZmQ0NN1DMzMzMzGxouIFiZmZmZmZDww0UMzMzMzMbGm6gmJmZmZnZ0PBCjWZmNjCS/r64VkRoMuti05evQ7OpzT0oZmYjRtKbJUXbz7c7pF0u6Yamx8zMzCaKe1DMzEbPAuBy4ITKvgcnqS5mZmaNuIFiZjZ65gO3RMT6ya6ImZlZUx7iZWY2euYDdzTNJGlHSddKul7STnn39pK+ImmzpF9L+pfBVtXMzOzx3EAxMxshkmYAewGHS/qVpLsknS5pxx75BFwAPAU4IiIeyIcOBAQsAs4Gvi7pORN3BmZmNt15iJeZ2WjZh/S3/X7gjaTGyheBXYB3d8n3eWAhcFBEbKrsXw+cEBFbgDWSDgOOA04afNXNzMzcQDEzGykRcbukPSNiQ951c+4d+YakEyPikZpsLwReCtwEbGg7dnNunLTcCMxrUqfqLV/NJpJvKWw2GjzEy8xsxFQaJy2/AHYAZnbIshk4GJgLHN927LG259sBWzAzM5sg7kExMxshkt4AfBl4RqXn4/nARtJwrTqrI+L7kk4BPi3pPyPi3nxsgSRFRKsX5ADguiZ18rfaZmbWhHtQzMxGy3WkSe3nSnp2njOyAlhRaWR0cjZwd07fMgs4S9I8SR9nbLK8mZnZhHAPipnZCImIDZIOBc4kzRfZBJwDnFaQ9xFJJwJXSzov776CdGevG4F1wOsi4u4JqbyZmRmg3l+omZmZlalOiPfQLpssvg7NpjYP8TIzMzMzs6HhBoqZmZmZmQ0ND/EyMzMzM7Oh4R4UMzMzMzMbGm6gmJmZmZnZ0HADxczMzMzMhoYbKGZmZmZmNjTcQDEzMzMzs6HhBoqZmZmZmQ2NvwF+DNuyy60CVQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1008x61.2 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_browser_isos_2(cerberus_h5,\n",
    "                    swan_file, \n",
    "                    filt_ab,\n",
    "                    pp_summary,\n",
    "                    major_isos,\n",
    "                    'COL1A1',\n",
    "                    obs_col,\n",
    "                    'ovary',\n",
    "                    'figures/pkm_ovary_browser_isos.pdf',\n",
    "                    h=0.2,\n",
    "                    ref_sources=['v29', 'v40'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30489b12-5251-4be3-a408-fbf8aa31da7e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
