{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3bea4128-5ca7-4afd-97f5-b833d23bd778",
   "metadata": {},
   "outputs": [],
   "source": [
    "from encoded_client.encoded import ENCODED\n",
    "import pdb\n",
    "import pandas as pd\n",
    "import glob\n",
    "import pysam\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "p = os.path.dirname(os.getcwd())\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0d3062c5-d7e5-4e67-ab79-2914aa1466e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = 'snakemake/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f5341805-392d-4b33-8ada-c45886494920",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = '../'+expand(config['data']['ab'], species='human')[0]\n",
    "filt_ab = expand(config['data']['filt_ab'], species='human')[0]\n",
    "gtf = expand(config['data']['cerb_gtf'], species='human')[0]\n",
    "read_annot = '../'+expand(config['data']['read_annot'], species='human')[0]\n",
    "t_metadata = '../'+expand(config['ref']['cerberus']['t_info'], species='human')[0]\n",
    "lib_meta = '../'+expand(config['data']['meta'], species='human')[0]\n",
    "swan_file = '../'+expand(config['data']['sg'], species='human')[0]\n",
    "cerberus_h5 = expand(config['data']['cerb_annot'], species='human')[0]\n",
    "cerb_t_metadata = '../'+expand(config['data']['t_info'], species='human')[0]\n",
    "major_isos = '../'+expand(config['data']['major_isos'], species='human', obs_col='sample')[0]\n",
    "pp_summary = '../'+expand(config['data']['p_pred']['summary'], species='human')[0]\n",
    "ref_t_metadata = '../'+expand(config['ref']['t_info'], species='human')[0]\n",
    "ref_g_metadata = '../'+expand(config['ref']['g_info'], species='human')[0]\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "min_tpm = 1\n",
    "gene_subset = 'protein_coding'\n",
    "obs_col = 'sample'\n",
    "go_gene_subset = 'protein_coding'\n",
    "\n",
    "m_lib_meta = '../'+expand(config['data']['meta'], species='mouse')[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be9a8420-3af2-40dc-8ff4-da335cbecfcd",
   "metadata": {},
   "source": [
    "## 231009 adding pertinent information to the filtered biggenepred or w/e"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "ba362379-fe05-43ab-b059-3095473ef909",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = 'huvec'\n",
    "bgp = expand(config['data']['tracks']['sample']['bgp_sort_filt'], species='human', sample='huvec')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "752538eb-3712-42e9-a17d-eb216f3efe8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get transcript name of each thing / novelty / novelty of each of three pieces?\n",
    "df = pd.read_csv(bgp, sep='\\t', header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "6fbcfd49-2dc0-424d-82a1-86df46bf5327",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.rename({3:'transcript_id'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "42b5d36e-32f7-4f37-912d-6a8353c579a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ca = cerberus.read(cerberus_h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "939ada67-784f-4366-aeee-edbfcbfe5b4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_map = ca.t_map.copy(deep=True)\n",
    "\n",
    "tss = ca.tss.copy(deep=True)\n",
    "tss.rename({'Name': 'tss_id', \n",
    "            'novelty': 'tss_novelty'}, \n",
    "             axis=1, inplace=True)\n",
    "tss = tss[['tss_id', 'tss_novelty']]\n",
    "\n",
    "tes = ca.tes.copy(deep=True)\n",
    "tes.rename({'Name': 'tes_id',\n",
    "            'novelty': 'tes_novelty'}, \n",
    "           axis=1, inplace=True)\n",
    "tes = tes[['tes_id', 'tes_novelty']]\n",
    "\n",
    "ic = ca.ic.copy(deep=True)\n",
    "ic.rename({'Name': 'ic_id', \n",
    "           'novelty': 'ic_novelty'}, \n",
    "          axis=1, inplace=True)\n",
    "ic = ic[['ic_id', 'ic_novelty']]\n",
    "\n",
    "t_map = t_map.merge(tss, how='left', on='tss_id')\n",
    "t_map = t_map.merge(tes, how='left', on='tes_id')\n",
    "t_map = t_map.merge(ic, how='left', on='ic_id')\n",
    "\n",
    "t_map = t_map[['transcript_id', 'gene_name', 'transcript_name',\n",
    "               'tss_id', 'ic_id', 'tes_id', \n",
    "               'tss_novelty', 'ic_novelty', 'tes_novelty']]\n",
    "t_map = t_map.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "280529fa-07c2-4dbe-904e-40e53762bc7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.merge(t_map,\n",
    "              how='left',\n",
    "              on='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "984aff58-6edb-4fff-89c2-bc8c9652fe89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>ic_id</th>\n",
       "      <th>tes_id</th>\n",
       "      <th>tss_novelty</th>\n",
       "      <th>ic_novelty</th>\n",
       "      <th>tes_novelty</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14353</td>\n",
       "      <td>29408</td>\n",
       "      <td>ENSG00000227232[2,5,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>29408</td>\n",
       "      <td>29408</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASH7P[2,5,1]</td>\n",
       "      <td>ENSG00000227232_2</td>\n",
       "      <td>ENSG00000227232_5</td>\n",
       "      <td>ENSG00000227232_1</td>\n",
       "      <td>Novel</td>\n",
       "      <td>NNC</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14353</td>\n",
       "      <td>29620</td>\n",
       "      <td>ENSG00000227232[1,2,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>29620</td>\n",
       "      <td>29620</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASH7P[1,2,1]</td>\n",
       "      <td>ENSG00000227232_1</td>\n",
       "      <td>ENSG00000227232_2</td>\n",
       "      <td>ENSG00000227232_1</td>\n",
       "      <td>Known</td>\n",
       "      <td>NNC</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>184823</td>\n",
       "      <td>199925</td>\n",
       "      <td>ENSG00000279457[2,6,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>199925</td>\n",
       "      <td>199925</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[2,6,2]</td>\n",
       "      <td>ENSG00000279457_2</td>\n",
       "      <td>ENSG00000279457_6</td>\n",
       "      <td>ENSG00000279457_2</td>\n",
       "      <td>Novel</td>\n",
       "      <td>NNC</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>184823</td>\n",
       "      <td>199925</td>\n",
       "      <td>ENSG00000279457[2,9,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>199925</td>\n",
       "      <td>199925</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[2,9,2]</td>\n",
       "      <td>ENSG00000279457_2</td>\n",
       "      <td>ENSG00000279457_9</td>\n",
       "      <td>ENSG00000279457_2</td>\n",
       "      <td>Novel</td>\n",
       "      <td>NNC</td>\n",
       "      <td>Novel</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>185166</td>\n",
       "      <td>195461</td>\n",
       "      <td>ENSG00000279457[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>195461</td>\n",
       "      <td>195461</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[1,1,1]</td>\n",
       "      <td>ENSG00000279457_1</td>\n",
       "      <td>ENSG00000279457_1</td>\n",
       "      <td>ENSG00000279457_1</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "      <td>Known</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1       2           transcript_id  4  5       6       7  8   9  \\\n",
       "0  chr1   14353   29408  ENSG00000227232[2,5,1]  0  -   29408   29408  0  11   \n",
       "1  chr1   14353   29620  ENSG00000227232[1,2,1]  0  -   29620   29620  0   9   \n",
       "2  chr1  184823  199925  ENSG00000279457[2,6,2]  0  -  199925  199925  0  11   \n",
       "3  chr1  184823  199925  ENSG00000279457[2,9,2]  0  -  199925  199925  0  11   \n",
       "4  chr1  185166  195461  ENSG00000279457[1,1,1]  0  -  195461  195461  0  10   \n",
       "\n",
       "   ...                 18  19 gene_name transcript_name             tss_id  \\\n",
       "0  ...  ENSG00000227232.5 NaN    WASH7P   WASH7P[2,5,1]  ENSG00000227232_2   \n",
       "1  ...  ENSG00000227232.5 NaN    WASH7P   WASH7P[1,2,1]  ENSG00000227232_1   \n",
       "2  ...  ENSG00000279457.4 NaN    WASH9P   WASH9P[2,6,2]  ENSG00000279457_2   \n",
       "3  ...  ENSG00000279457.4 NaN    WASH9P   WASH9P[2,9,2]  ENSG00000279457_2   \n",
       "4  ...  ENSG00000279457.4 NaN    WASH9P   WASH9P[1,1,1]  ENSG00000279457_1   \n",
       "\n",
       "               ic_id             tes_id tss_novelty ic_novelty  tes_novelty  \n",
       "0  ENSG00000227232_5  ENSG00000227232_1       Novel        NNC        Known  \n",
       "1  ENSG00000227232_2  ENSG00000227232_1       Known        NNC        Known  \n",
       "2  ENSG00000279457_6  ENSG00000279457_2       Novel        NNC        Novel  \n",
       "3  ENSG00000279457_9  ENSG00000279457_2       Novel        NNC        Novel  \n",
       "4  ENSG00000279457_1  ENSG00000279457_1       Known      Known        Known  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "92eba59b-f46c-4206-8451-610d6c0563b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add gene biotype from gencode\n",
    "t_df, _, _ = get_gtf_info(how='iso', ver=ver, add_stable_gid=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "5e0ae773-84a4-43a5-8ecc-ef61b21afe3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['gid_stable'] = cerberus.get_stable_gid(df, 18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "a04f3e65-a109-418e-bd11-3aea9b441c10",
   "metadata": {},
   "outputs": [],
   "source": [
    "t_df.rename({'tid':'transcript_id'}, axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "5db8dc77-ca76-482a-8000-2813ddb1e094",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.merge(t_df[['MANE_Select', 'MANE_Plus_Clinical', 't_len', 'transcript_id']],\n",
    "              how='left',\n",
    "              on='transcript_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "127beff1-a8dc-43c3-a0aa-e2ead3bcc82e",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_df, _, _ = get_gtf_info(how='gene', ver=ver, add_stable_gid=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "49fc1916-60a6-484d-ae30-b1c69758c315",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>gid_stable_x</th>\n",
       "      <th>MANE_Select</th>\n",
       "      <th>MANE_Plus_Clinical</th>\n",
       "      <th>gid_stable_y</th>\n",
       "      <th>t_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14353</td>\n",
       "      <td>29408</td>\n",
       "      <td>ENSG00000227232[2,5,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>29408</td>\n",
       "      <td>29408</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000227232[2,5,1]</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASH7P[2,5,1]</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14353</td>\n",
       "      <td>29620</td>\n",
       "      <td>ENSG00000227232[1,2,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>29620</td>\n",
       "      <td>29620</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000227232[1,2,1]</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASH7P[1,2,1]</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>184823</td>\n",
       "      <td>199925</td>\n",
       "      <td>ENSG00000279457[2,6,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>199925</td>\n",
       "      <td>199925</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000279457[2,6,2]</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[2,6,2]</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>184823</td>\n",
       "      <td>199925</td>\n",
       "      <td>ENSG00000279457[2,9,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>199925</td>\n",
       "      <td>199925</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000279457[2,9,2]</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[2,9,2]</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>185166</td>\n",
       "      <td>195461</td>\n",
       "      <td>ENSG00000279457[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>195461</td>\n",
       "      <td>195461</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000279457[1,1,1]</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[1,1,1]</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>1507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48747</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705366</td>\n",
       "      <td>19744811</td>\n",
       "      <td>ENSG00000012817[1,1,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>19744811</td>\n",
       "      <td>19744811</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000012817[1,1,2]</td>\n",
       "      <td>ENSG00000012817.15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>KDM5D[1,1,2]</td>\n",
       "      <td>ENSG00000012817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48748</th>\n",
       "      <td>chrY</td>\n",
       "      <td>19705366</td>\n",
       "      <td>19744989</td>\n",
       "      <td>ENSG00000012817[2,1,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>19744989</td>\n",
       "      <td>19744989</td>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000012817[2,1,2]</td>\n",
       "      <td>ENSG00000012817.15</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>KDM5D[2,1,2]</td>\n",
       "      <td>ENSG00000012817</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48749</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575725</td>\n",
       "      <td>20592680</td>\n",
       "      <td>ENSG00000198692[1,1,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>20592680</td>\n",
       "      <td>20592680</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000198692[1,1,2]</td>\n",
       "      <td>ENSG00000198692.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>EIF1AY[1,1,2]</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48750</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575725</td>\n",
       "      <td>20592680</td>\n",
       "      <td>ENSG00000198692[1,2,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>20592680</td>\n",
       "      <td>20592680</td>\n",
       "      <td>0</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000198692[1,2,2]</td>\n",
       "      <td>ENSG00000198692.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>EIF1AY[1,2,2]</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>870.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48751</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575725</td>\n",
       "      <td>20593204</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>20593204</td>\n",
       "      <td>20593204</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>ENSG00000198692.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>EIF1AY[1,1,1]</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>1446.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48752 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0         1         2           transcript_id  4  5         6  \\\n",
       "0      chr1     14353     29408  ENSG00000227232[2,5,1]  0  -     29408   \n",
       "1      chr1     14353     29620  ENSG00000227232[1,2,1]  0  -     29620   \n",
       "2      chr1    184823    199925  ENSG00000279457[2,6,2]  0  -    199925   \n",
       "3      chr1    184823    199925  ENSG00000279457[2,9,2]  0  -    199925   \n",
       "4      chr1    185166    195461  ENSG00000279457[1,1,1]  0  -    195461   \n",
       "...     ...       ...       ...                     ... .. ..       ...   \n",
       "48747  chrY  19705366  19744811  ENSG00000012817[1,1,2]  0  -  19744811   \n",
       "48748  chrY  19705366  19744989  ENSG00000012817[2,1,2]  0  -  19744989   \n",
       "48749  chrY  20575725  20592680  ENSG00000198692[1,1,2]  0  +  20592680   \n",
       "48750  chrY  20575725  20592680  ENSG00000198692[1,2,2]  0  +  20592680   \n",
       "48751  chrY  20575725  20593204  ENSG00000198692[1,1,1]  0  +  20593204   \n",
       "\n",
       "              7  8   9  ...                      17                  18  19  \\\n",
       "0         29408  0  11  ...  ENSG00000227232[2,5,1]   ENSG00000227232.5 NaN   \n",
       "1         29620  0   9  ...  ENSG00000227232[1,2,1]   ENSG00000227232.5 NaN   \n",
       "2        199925  0  11  ...  ENSG00000279457[2,6,2]   ENSG00000279457.4 NaN   \n",
       "3        199925  0  11  ...  ENSG00000279457[2,9,2]   ENSG00000279457.4 NaN   \n",
       "4        195461  0  10  ...  ENSG00000279457[1,1,1]   ENSG00000279457.4 NaN   \n",
       "...         ... ..  ..  ...                     ...                 ...  ..   \n",
       "48747  19744811  0  27  ...  ENSG00000012817[1,1,2]  ENSG00000012817.15 NaN   \n",
       "48748  19744989  0  27  ...  ENSG00000012817[2,1,2]  ENSG00000012817.15 NaN   \n",
       "48749  20592680  0   7  ...  ENSG00000198692[1,1,2]   ENSG00000198692.9 NaN   \n",
       "48750  20592680  0   6  ...  ENSG00000198692[1,2,2]   ENSG00000198692.9 NaN   \n",
       "48751  20593204  0   7  ...  ENSG00000198692[1,1,1]   ENSG00000198692.9 NaN   \n",
       "\n",
       "      gene_name transcript_name     gid_stable_x  MANE_Select  \\\n",
       "0        WASH7P   WASH7P[2,5,1]  ENSG00000227232          NaN   \n",
       "1        WASH7P   WASH7P[1,2,1]  ENSG00000227232          NaN   \n",
       "2        WASH9P   WASH9P[2,6,2]  ENSG00000279457          NaN   \n",
       "3        WASH9P   WASH9P[2,9,2]  ENSG00000279457          NaN   \n",
       "4        WASH9P   WASH9P[1,1,1]  ENSG00000279457        False   \n",
       "...         ...             ...              ...          ...   \n",
       "48747     KDM5D    KDM5D[1,1,2]  ENSG00000012817          NaN   \n",
       "48748     KDM5D    KDM5D[2,1,2]  ENSG00000012817          NaN   \n",
       "48749    EIF1AY   EIF1AY[1,1,2]  ENSG00000198692          NaN   \n",
       "48750    EIF1AY   EIF1AY[1,2,2]  ENSG00000198692        False   \n",
       "48751    EIF1AY   EIF1AY[1,1,1]  ENSG00000198692         True   \n",
       "\n",
       "      MANE_Plus_Clinical     gid_stable_y   t_len  \n",
       "0                    NaN              NaN     NaN  \n",
       "1                    NaN              NaN     NaN  \n",
       "2                    NaN              NaN     NaN  \n",
       "3                    NaN              NaN     NaN  \n",
       "4                  False  ENSG00000279457  1507.0  \n",
       "...                  ...              ...     ...  \n",
       "48747                NaN              NaN     NaN  \n",
       "48748                NaN              NaN     NaN  \n",
       "48749                NaN              NaN     NaN  \n",
       "48750              False  ENSG00000198692   870.0  \n",
       "48751              False  ENSG00000198692  1446.0  \n",
       "\n",
       "[48752 rows x 27 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "09a34ef4-29c0-45b3-8017-4d284f6ceec5",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'gid_stable'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-107-b4e1b85cd4c3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m df = df.merge(gene_df[['biotype_category', 'biotype', 'gid_stable']],\n\u001b[1;32m      2\u001b[0m               \u001b[0mhow\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'left'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m               on='gid_stable')\n\u001b[0m",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mmerge\u001b[0;34m(self, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m   9201\u001b[0m             \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   9202\u001b[0m             \u001b[0mindicator\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindicator\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 9203\u001b[0;31m             \u001b[0mvalidate\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalidate\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   9204\u001b[0m         )\n\u001b[1;32m   9205\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36mmerge\u001b[0;34m(left, right, how, on, left_on, right_on, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m    117\u001b[0m         \u001b[0mcopy\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcopy\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    118\u001b[0m         \u001b[0mindicator\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mindicator\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 119\u001b[0;31m         \u001b[0mvalidate\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalidate\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    120\u001b[0m     )\n\u001b[1;32m    121\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, left, right, how, on, left_on, right_on, axis, left_index, right_index, sort, suffixes, copy, indicator, validate)\u001b[0m\n\u001b[1;32m    697\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mright_join_keys\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    698\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin_names\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 699\u001b[0;31m         ) = self._get_merge_keys()\n\u001b[0m\u001b[1;32m    700\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    701\u001b[0m         \u001b[0;31m# validate the merge keys dtypes. We may need to coerce\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/reshape/merge.py\u001b[0m in \u001b[0;36m_get_merge_keys\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1107\u001b[0m                         \u001b[0mright_keys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1108\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0mlk\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1109\u001b[0;31m                         \u001b[0mleft_keys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mleft\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_label_or_level_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1110\u001b[0m                         \u001b[0mjoin_names\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1111\u001b[0m                     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/base_clone/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_get_label_or_level_values\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1777\u001b[0m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_level_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_values\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1778\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1779\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1780\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1781\u001b[0m         \u001b[0;31m# Check for duplicates\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'gid_stable'"
     ]
    }
   ],
   "source": [
    "df = df.merge(gene_df[['biotype_category', 'biotype', 'gid_stable']],\n",
    "              how='left',\n",
    "              on='gid_stable')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "97a9c25e-fd12-4022-a615-32de9527f80d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>transcript_name</th>\n",
       "      <th>gid_stable</th>\n",
       "      <th>MANE_Select</th>\n",
       "      <th>MANE_Plus_Clinical</th>\n",
       "      <th>t_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14353</td>\n",
       "      <td>29408</td>\n",
       "      <td>ENSG00000227232[2,5,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>29408</td>\n",
       "      <td>29408</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000227232[2,5,1]</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASH7P[2,5,1]</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1462.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>14353</td>\n",
       "      <td>29620</td>\n",
       "      <td>ENSG00000227232[1,2,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>29620</td>\n",
       "      <td>29620</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000227232[1,2,1]</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASH7P[1,2,1]</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1462.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>184823</td>\n",
       "      <td>199925</td>\n",
       "      <td>ENSG00000279457[2,6,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>199925</td>\n",
       "      <td>199925</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000279457[2,6,2]</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[2,6,2]</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>184823</td>\n",
       "      <td>199925</td>\n",
       "      <td>ENSG00000279457[2,9,2]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>199925</td>\n",
       "      <td>199925</td>\n",
       "      <td>0</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000279457[2,9,2]</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[2,9,2]</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>185166</td>\n",
       "      <td>195461</td>\n",
       "      <td>ENSG00000279457[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>-</td>\n",
       "      <td>195461</td>\n",
       "      <td>195461</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000279457[1,1,1]</td>\n",
       "      <td>ENSG00000279457.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>WASH9P</td>\n",
       "      <td>WASH9P[1,1,1]</td>\n",
       "      <td>ENSG00000279457</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>1507.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639443</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575725</td>\n",
       "      <td>20593204</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>20593204</td>\n",
       "      <td>20593204</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>ENSG00000198692.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>EIF1AY[1,1,1]</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>1446.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639444</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575725</td>\n",
       "      <td>20593204</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>20593204</td>\n",
       "      <td>20593204</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>ENSG00000198692.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>EIF1AY[1,1,1]</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>870.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639445</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575725</td>\n",
       "      <td>20593204</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>20593204</td>\n",
       "      <td>20593204</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>ENSG00000198692.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>EIF1AY[1,1,1]</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>962.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639446</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575725</td>\n",
       "      <td>20593204</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>20593204</td>\n",
       "      <td>20593204</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>ENSG00000198692.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>EIF1AY[1,1,1]</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3250.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639447</th>\n",
       "      <td>chrY</td>\n",
       "      <td>20575725</td>\n",
       "      <td>20593204</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>20593204</td>\n",
       "      <td>20593204</td>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000198692[1,1,1]</td>\n",
       "      <td>ENSG00000198692.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EIF1AY</td>\n",
       "      <td>EIF1AY[1,1,1]</td>\n",
       "      <td>ENSG00000198692</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>744.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>639448 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           0         1         2           transcript_id  4  5         6  \\\n",
       "0       chr1     14353     29408  ENSG00000227232[2,5,1]  0  -     29408   \n",
       "1       chr1     14353     29620  ENSG00000227232[1,2,1]  0  -     29620   \n",
       "2       chr1    184823    199925  ENSG00000279457[2,6,2]  0  -    199925   \n",
       "3       chr1    184823    199925  ENSG00000279457[2,9,2]  0  -    199925   \n",
       "4       chr1    185166    195461  ENSG00000279457[1,1,1]  0  -    195461   \n",
       "...      ...       ...       ...                     ... .. ..       ...   \n",
       "639443  chrY  20575725  20593204  ENSG00000198692[1,1,1]  0  +  20593204   \n",
       "639444  chrY  20575725  20593204  ENSG00000198692[1,1,1]  0  +  20593204   \n",
       "639445  chrY  20575725  20593204  ENSG00000198692[1,1,1]  0  +  20593204   \n",
       "639446  chrY  20575725  20593204  ENSG00000198692[1,1,1]  0  +  20593204   \n",
       "639447  chrY  20575725  20593204  ENSG00000198692[1,1,1]  0  +  20593204   \n",
       "\n",
       "               7  8   9  ...  16                      17                 18  \\\n",
       "0          29408  0  11  ... NaN  ENSG00000227232[2,5,1]  ENSG00000227232.5   \n",
       "1          29620  0   9  ... NaN  ENSG00000227232[1,2,1]  ENSG00000227232.5   \n",
       "2         199925  0  11  ... NaN  ENSG00000279457[2,6,2]  ENSG00000279457.4   \n",
       "3         199925  0  11  ... NaN  ENSG00000279457[2,9,2]  ENSG00000279457.4   \n",
       "4         195461  0  10  ... NaN  ENSG00000279457[1,1,1]  ENSG00000279457.4   \n",
       "...          ... ..  ..  ...  ..                     ...                ...   \n",
       "639443  20593204  0   7  ... NaN  ENSG00000198692[1,1,1]  ENSG00000198692.9   \n",
       "639444  20593204  0   7  ... NaN  ENSG00000198692[1,1,1]  ENSG00000198692.9   \n",
       "639445  20593204  0   7  ... NaN  ENSG00000198692[1,1,1]  ENSG00000198692.9   \n",
       "639446  20593204  0   7  ... NaN  ENSG00000198692[1,1,1]  ENSG00000198692.9   \n",
       "639447  20593204  0   7  ... NaN  ENSG00000198692[1,1,1]  ENSG00000198692.9   \n",
       "\n",
       "        19 gene_name transcript_name       gid_stable MANE_Select  \\\n",
       "0      NaN    WASH7P   WASH7P[2,5,1]  ENSG00000227232       False   \n",
       "1      NaN    WASH7P   WASH7P[1,2,1]  ENSG00000227232       False   \n",
       "2      NaN    WASH9P   WASH9P[2,6,2]  ENSG00000279457       False   \n",
       "3      NaN    WASH9P   WASH9P[2,9,2]  ENSG00000279457       False   \n",
       "4      NaN    WASH9P   WASH9P[1,1,1]  ENSG00000279457       False   \n",
       "...     ..       ...             ...              ...         ...   \n",
       "639443 NaN    EIF1AY   EIF1AY[1,1,1]  ENSG00000198692        True   \n",
       "639444 NaN    EIF1AY   EIF1AY[1,1,1]  ENSG00000198692       False   \n",
       "639445 NaN    EIF1AY   EIF1AY[1,1,1]  ENSG00000198692       False   \n",
       "639446 NaN    EIF1AY   EIF1AY[1,1,1]  ENSG00000198692       False   \n",
       "639447 NaN    EIF1AY   EIF1AY[1,1,1]  ENSG00000198692       False   \n",
       "\n",
       "       MANE_Plus_Clinical   t_len  \n",
       "0                   False  1462.0  \n",
       "1                   False  1462.0  \n",
       "2                   False  1507.0  \n",
       "3                   False  1507.0  \n",
       "4                   False  1507.0  \n",
       "...                   ...     ...  \n",
       "639443              False  1446.0  \n",
       "639444              False   870.0  \n",
       "639445              False   962.0  \n",
       "639446              False  3250.0  \n",
       "639447              False   744.0  \n",
       "\n",
       "[639448 rows x 26 columns]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "771ba3a1-7aa0-4d9f-bf58-5e28458e1166",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>length</th>\n",
       "      <th>biotype</th>\n",
       "      <th>biotype_category</th>\n",
       "      <th>tf</th>\n",
       "      <th>MANE_Select</th>\n",
       "      <th>MANE_Plus_Clinical</th>\n",
       "      <th>gid_stable</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000000460.17</td>\n",
       "      <td>C1orf112</td>\n",
       "      <td>192174</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000000971.17</td>\n",
       "      <td>CFH</td>\n",
       "      <td>100823</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000000971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000001461.17</td>\n",
       "      <td>NIPAL3</td>\n",
       "      <td>57275</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000001461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000004487.18</td>\n",
       "      <td>KDM1A</td>\n",
       "      <td>64347</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000004487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000007933.13</td>\n",
       "      <td>FMO3</td>\n",
       "      <td>27019</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>ENSG00000007933</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  gid     gname  length         biotype biotype_category  \\\n",
       "0  ENSG00000000460.17  C1orf112  192174  protein_coding   protein_coding   \n",
       "1  ENSG00000000971.17       CFH  100823  protein_coding   protein_coding   \n",
       "2  ENSG00000001461.17    NIPAL3   57275  protein_coding   protein_coding   \n",
       "3  ENSG00000004487.18     KDM1A   64347  protein_coding   protein_coding   \n",
       "4  ENSG00000007933.13      FMO3   27019  protein_coding   protein_coding   \n",
       "\n",
       "      tf  MANE_Select  MANE_Plus_Clinical       gid_stable  \n",
       "0  False         True               False  ENSG00000000460  \n",
       "1  False         True               False  ENSG00000000971  \n",
       "2  False         True               False  ENSG00000001461  \n",
       "3  False         True               False  ENSG00000004487  \n",
       "4  False         True               False  ENSG00000007933  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c323209-c394-40c5-94ce-21c950c160ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6dfe081-298f-4456-a42e-02ac498be486",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ea45d2f-f276-438d-8a11-007502e1d17b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# also add predicted CDS start / ends"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56a614cc-5a19-4891-9d94-e2f302f6f6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add tpm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "467f4bf2-e819-4177-9a67-bd42b72761bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40092ea1-6942-4a6d-bc9a-3c536cff4877",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8f07809e-7154-4c56-ad1f-d8b912881751",
   "metadata": {},
   "source": [
    "## 231004 making a GTF for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "19deb6c5-e482-4496-88f2-ff6109b5b1d3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/scripts/utils.py:2176: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[tpm_col] = (df[d]*1000000)/df[total_col]\n",
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/scripts/utils.py:2175: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[total_col] = df[d].sum()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enforcing minimum TPM\n",
      "Total # isos detected: 231136\n",
      "# isos >= 1 tpm: 210239\n",
      "Number of isos reported: 210239\n",
      "Found 50 total samples\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(filt_ab, sep='\\t')\n",
    "df = get_det_table(df,\n",
    "                   groupby=obs_col,\n",
    "                   how='iso',\n",
    "                   min_tpm=min_tpm)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7037026a-80a4-4189-a2f7-8b0f9442db28",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "782a7e86-c7c3-47b6-a2d8-c89b40ee9ccd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "210239\n",
      "56341\n",
      "['ENSG00000279457[1,1,1]', 'ENSG00000279457[2,4,2]', 'ENSG00000279457[2,6,2]', 'ENSG00000279457[2,9,2]', 'ENSG00000225880[1,1,1]']\n"
     ]
    }
   ],
   "source": [
    "print(len(df.index))\n",
    "for c in df.columns:\n",
    "    tids = df.loc[df[c]==True].index.tolist()\n",
    "    # print(len(tids))\n",
    "    # print(tids[:5])\n",
    "    # break\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d91d27c9-17b4-4603-aed7-f78eeed6d4c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n"
     ]
    }
   ],
   "source": [
    "gtf_df = pr.read_gtf(gtf, rename_attr=True).as_df()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ccd725cb-6e4c-4020-b21d-7aa6e64bec63",
   "metadata": {},
   "outputs": [],
   "source": [
    "gids = gtf_df.loc[gtf_df.transcript_id.isin(tids), 'gene_id'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f3d99d1f-c92e-4288-907d-4ca9c7b7546b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gids[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f02bea40-99be-434d-8d0b-c420256fad66",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = gtf_df.loc[gtf_df.transcript_id.isin(tids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8618e710-7aa5-413f-87f0-e4c7e7f8ba69",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df = pr.PyRanges(gtf_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c2db2bd9-222a-4f00-8268-e9a3ee41d9cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df.to_gtf('test_sample_gtf.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ab40c2c-8704-4e82-a3a4-b9cb43e21c38",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26906d89-ba78-4b3f-b8d3-fa73a21728fa",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "883ecef5-2ddb-4bbc-a98f-7881b5886a61",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b584464e-a8b8-4676-85ac-c39cb08ad20a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "6129eafe-71c5-4d9a-8243-b5961f5911a8",
   "metadata": {},
   "source": [
    "## snakemake debug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e1b2df4c-45db-44eb-b575-4e107ad713da",
   "metadata": {},
   "outputs": [],
   "source": [
    "species = ['human', 'mouse']\n",
    "meta_df = get_meta_df(config, species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f45209f0-dbcb-4ec9-8ded-2a38877c33d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENCODE_experiment_id</th>\n",
       "      <th>dataset</th>\n",
       "      <th>sample</th>\n",
       "      <th>sample_display</th>\n",
       "      <th>general_tissue_cell_type</th>\n",
       "      <th>fig1_tissue_label</th>\n",
       "      <th>health_status</th>\n",
       "      <th>tissue_or_cell_line</th>\n",
       "      <th>sample_color_hex_code</th>\n",
       "      <th>matching_mouse_samples</th>\n",
       "      <th>...</th>\n",
       "      <th>document_labels</th>\n",
       "      <th>platform</th>\n",
       "      <th>RIN</th>\n",
       "      <th>spikeins</th>\n",
       "      <th>species</th>\n",
       "      <th>age</th>\n",
       "      <th>adult</th>\n",
       "      <th>sex</th>\n",
       "      <th>genotype</th>\n",
       "      <th>matching_human_samples</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENCSR242UEB</td>\n",
       "      <td>adrenal_36d_m_2</td>\n",
       "      <td>adrenal_36d</td>\n",
       "      <td>Adrenal gland PND36</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tissue</td>\n",
       "      <td>#c87961</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>'PacBio libraries v3 (October, 2020) Protocol ...</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "      <td>6.4</td>\n",
       "      <td>True</td>\n",
       "      <td>mouse</td>\n",
       "      <td>36d</td>\n",
       "      <td>True</td>\n",
       "      <td>m</td>\n",
       "      <td>b6/cast</td>\n",
       "      <td>['adrenal gland']</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENCSR242UEB</td>\n",
       "      <td>adrenal_36d_m_1</td>\n",
       "      <td>adrenal_36d</td>\n",
       "      <td>Adrenal gland PND36</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tissue</td>\n",
       "      <td>#c87961</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>'PacBio libraries v3 (October, 2020) Protocol ...</td>\n",
       "      <td>Pacific Biosciences Sequel II</td>\n",
       "      <td>6.4</td>\n",
       "      <td>True</td>\n",
       "      <td>mouse</td>\n",
       "      <td>36d</td>\n",
       "      <td>True</td>\n",
       "      <td>m</td>\n",
       "      <td>b6/cast</td>\n",
       "      <td>['adrenal gland']</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   ENCODE_experiment_id          dataset       sample       sample_display  \\\n",
       "19          ENCSR242UEB  adrenal_36d_m_2  adrenal_36d  Adrenal gland PND36   \n",
       "20          ENCSR242UEB  adrenal_36d_m_1  adrenal_36d  Adrenal gland PND36   \n",
       "\n",
       "   general_tissue_cell_type fig1_tissue_label health_status  \\\n",
       "19            adrenal gland     adrenal gland           NaN   \n",
       "20            adrenal gland     adrenal gland           NaN   \n",
       "\n",
       "   tissue_or_cell_line sample_color_hex_code matching_mouse_samples  ...  \\\n",
       "19              tissue               #c87961                    NaN  ...   \n",
       "20              tissue               #c87961                    NaN  ...   \n",
       "\n",
       "                                      document_labels  \\\n",
       "19  'PacBio libraries v3 (October, 2020) Protocol ...   \n",
       "20  'PacBio libraries v3 (October, 2020) Protocol ...   \n",
       "\n",
       "                         platform  RIN spikeins species  age adult  sex  \\\n",
       "19  Pacific Biosciences Sequel II  6.4     True   mouse  36d  True    m   \n",
       "20  Pacific Biosciences Sequel II  6.4     True   mouse  36d  True    m   \n",
       "\n",
       "    genotype matching_human_samples  \n",
       "19   b6/cast      ['adrenal gland']  \n",
       "20   b6/cast      ['adrenal gland']  \n",
       "\n",
       "[2 rows x 25 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.loc[meta_df.dataset.str.contains('adrenal_36d_m')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4ef8dfb-9dfd-416a-a0da-c7591f630d9b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65c12e33-612f-4569-bbb5-4731bc42e03e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fef9d399-04e8-435e-beee-3d50d780d9d2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8515750-361b-4a90-8514-bd0f218d75ec",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcd9484a-f569-433c-903d-47647728d658",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd3e1a8f-15a9-4038-8e07-6c0777997e8b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a34c55ce-40c4-44e9-a68a-51ef21748340",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 230722 test get_lr_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cd92997-6f53-41f9-b049-fe8d77a19113",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eb51c13-d12d-41b2-8806-4ea51e65e33f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a728589-1ceb-4e2e-8733-4e8824d49a44",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65adb8a5-c2a2-4971-b2c2-185ed4ea1bee",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6940d5ce-1a98-4ae6-a818-5f22968fae7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_lr_exp_meta():\n",
    "    def get_genome_bam(experiment_id):\n",
    "        experiment = server.get_json(experiment_id)\n",
    "\n",
    "        metadata = []\n",
    "        default_analysis = experiment[\"default_analysis\"]\n",
    "        for f in experiment[\"files\"]:\n",
    "            analyses = [x[\"@id\"] for x in f[\"analyses\"]]\n",
    "            # this check only applies to processed data not reads (default_analysis in analyses and)\n",
    "            if f[\"output_type\"] in (\"reads\", \"unfiltered alignments\"):\n",
    "                if f[\"output_type\"] == \"unfiltered alignments\" and not default_analysis in analyses:\n",
    "                    # skip alignments for older analyses.\n",
    "                    continue\n",
    "                reps = f[\"technical_replicates\"]\n",
    "                assert len(reps) == 1\n",
    "                reps = reps[0]\n",
    "\n",
    "                extension = {\n",
    "                    \"fastq\": \".fastq.gz\",\n",
    "                    \"bam\": \".bam\",\n",
    "                }[f[\"file_format\"]]\n",
    "    \n",
    "                metadata.append({\n",
    "                    \"experiment\": experiment[\"accession\"],\n",
    "                    \"description\": experiment[\"description\"],\n",
    "                    \"simple_biosample_summary\": f[\"simple_biosample_summary\"],\n",
    "                    \"file\": f[\"accession\"],\n",
    "                    \"output_type\": f[\"output_type\"],\n",
    "                    \"file_size\": f[\"file_size\"],\n",
    "                    \"bio_rep\": f[\"biological_replicates\"][0],\n",
    "                    \"tech_rep\": f[\"technical_replicates\"][0]\n",
    "                })\n",
    "\n",
    "        return metadata\n",
    "\n",
    "    server = ENCODED(\"www.encodeproject.org\") \n",
    "    cart_url = \"https://www.encodeproject.org/carts/829d339c-913c-4773-8001-80130796a367/\"    \n",
    "    \n",
    "    cart = server.get_json(cart_url)\n",
    "\n",
    "    metadata = []\n",
    "    for experiment_id in cart[\"elements\"]:\n",
    "        metadata.extend(get_genome_bam(experiment_id))\n",
    "        \n",
    "    metadata = pd.DataFrame(metadata)\n",
    "    metadata['name'] = metadata['experiment']+'_'+metadata['tech_rep']\n",
    "    return metadata    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5a1d9ff8-cfd4-423e-9b13-f46d03bd42e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_lr_fastq_ids(file_format='bam'):\n",
    "    metadata = get_lr_exp_meta()\n",
    "    if file_format == 'bam':\n",
    "        metadata = metadata.loc[metadata.output_type=='unfiltered alignments']\n",
    "    elif file_format == 'fastq':\n",
    "        metadata = metadata.loc[metadata.output_type=='reads']\n",
    "    \n",
    "    return metadata.file.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d1e14952-4609-42ec-a4fe-139694753d44",
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = get_lr_fastq_ids(file_format='bam')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "55c8d7fb-323f-4cff-9a18-0b57b3a7a349",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "138"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "d227f142-803d-4268-9bbb-fb4cd43712c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "ids = get_lr_fastq_ids(file_format='fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6bd85c45-a301-4109-a2c0-097ad3a618bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "138"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "abfe10e7-6cf7-4ce0-87a0-1cb224a30e32",
   "metadata": {},
   "outputs": [],
   "source": [
    "def score_aligned_reads(filename):\n",
    "    with pysam.AlignmentFile(filename, \"rb\") as inbam:\n",
    "        aligned_reads = 0\n",
    "        aligned_query_len = []\n",
    "        aligned_positions = []\n",
    "\n",
    "        for read in inbam:\n",
    "            if not read.is_unmapped:\n",
    "                aligned_reads += 1\n",
    "                aligned_query_len.append(read.query_length)\n",
    "                aligned_positions.append(len(read.positions))\n",
    "\n",
    "        return {\n",
    "            \"aligned_reads\": len(aligned_query_len),\n",
    "            \"query_len_median\": np.median(aligned_query_len),\n",
    "            \"query_len\": aligned_query_len,\n",
    "            \"positions_median\": np.median(aligned_positions),\n",
    "            \"positions_len\": aligned_positions,\n",
    "        }\n",
    "def score_fastq_reads(filename):\n",
    "    read_len = []\n",
    "    with xopen(filename, \"rt\") as instream:\n",
    "        for record in SeqIO.parse(instream, \"fastq\"):\n",
    "            read_len.append(len(record))\n",
    "    \n",
    "    return {\n",
    "        \"raw_reads\": len(read_len),\n",
    "        \"read_len_median\": np.median(read_len),\n",
    "        \"read_len\": read_len,\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "9d5c6375-bf24-48dc-89db-02e5ea348c32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>experiment</th>\n",
       "      <th>description</th>\n",
       "      <th>simple_biosample_summary</th>\n",
       "      <th>file</th>\n",
       "      <th>output_type</th>\n",
       "      <th>file_size</th>\n",
       "      <th>bio_rep</th>\n",
       "      <th>tech_rep</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENCSR293MOX</td>\n",
       "      <td>W62 right lobe of liver</td>\n",
       "      <td>female child (16 years)</td>\n",
       "      <td>ENCFF318SKH</td>\n",
       "      <td>reads</td>\n",
       "      <td>255510907</td>\n",
       "      <td>1</td>\n",
       "      <td>1_1</td>\n",
       "      <td>ENCSR293MOX_1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENCSR293MOX</td>\n",
       "      <td>W62 right lobe of liver</td>\n",
       "      <td>female child (16 years)</td>\n",
       "      <td>ENCFF430RUG</td>\n",
       "      <td>unfiltered alignments</td>\n",
       "      <td>362638099</td>\n",
       "      <td>1</td>\n",
       "      <td>1_1</td>\n",
       "      <td>ENCSR293MOX_1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENCSR081NRO</td>\n",
       "      <td>Adrenal Gland 1-2 ENTEX50</td>\n",
       "      <td>male adult (37 years)</td>\n",
       "      <td>ENCFF211SQY</td>\n",
       "      <td>reads</td>\n",
       "      <td>387217252</td>\n",
       "      <td>1</td>\n",
       "      <td>1_1</td>\n",
       "      <td>ENCSR081NRO_1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENCSR081NRO</td>\n",
       "      <td>Adrenal Gland 1-2 ENTEX50</td>\n",
       "      <td>male adult (37 years)</td>\n",
       "      <td>ENCFF967OHL</td>\n",
       "      <td>unfiltered alignments</td>\n",
       "      <td>521377683</td>\n",
       "      <td>1</td>\n",
       "      <td>1_1</td>\n",
       "      <td>ENCSR081NRO_1_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENCSR257JBF</td>\n",
       "      <td>excitatory cortical neurons</td>\n",
       "      <td>male adult (53 years)</td>\n",
       "      <td>ENCFF919JFJ</td>\n",
       "      <td>reads</td>\n",
       "      <td>345607718</td>\n",
       "      <td>1</td>\n",
       "      <td>1_1</td>\n",
       "      <td>ENCSR257JBF_1_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    experiment                  description simple_biosample_summary  \\\n",
       "0  ENCSR293MOX      W62 right lobe of liver  female child (16 years)   \n",
       "1  ENCSR293MOX      W62 right lobe of liver  female child (16 years)   \n",
       "2  ENCSR081NRO    Adrenal Gland 1-2 ENTEX50    male adult (37 years)   \n",
       "3  ENCSR081NRO    Adrenal Gland 1-2 ENTEX50    male adult (37 years)   \n",
       "4  ENCSR257JBF  excitatory cortical neurons    male adult (53 years)   \n",
       "\n",
       "          file            output_type  file_size  bio_rep tech_rep  \\\n",
       "0  ENCFF318SKH                  reads  255510907        1      1_1   \n",
       "1  ENCFF430RUG  unfiltered alignments  362638099        1      1_1   \n",
       "2  ENCFF211SQY                  reads  387217252        1      1_1   \n",
       "3  ENCFF967OHL  unfiltered alignments  521377683        1      1_1   \n",
       "4  ENCFF919JFJ                  reads  345607718        1      1_1   \n",
       "\n",
       "              name  \n",
       "0  ENCSR293MOX_1_1  \n",
       "1  ENCSR293MOX_1_1  \n",
       "2  ENCSR081NRO_1_1  \n",
       "3  ENCSR081NRO_1_1  \n",
       "4  ENCSR257JBF_1_1  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = get_lr_exp_meta()\n",
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "941e82e6-bb39-4fc1-8c10-564133c39173",
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "no BGZF EOF marker; file may be truncated",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-35-d3a4e377bb78>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutput_type\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'unfiltered alignments'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscore_aligned_reads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      9\u001b[0m     \u001b[0;32melif\u001b[0m \u001b[0mtemp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0moutput_type\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'reads'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscore_fastq_reads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-33-b4d4b5920659>\u001b[0m in \u001b[0;36mscore_aligned_reads\u001b[0;34m(filename)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mscore_aligned_reads\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mpysam\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mAlignmentFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"rb\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0minbam\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m         \u001b[0maligned_reads\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m         \u001b[0maligned_query_len\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0maligned_positions\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpysam/libcalignmentfile.pyx\u001b[0m in \u001b[0;36mpysam.libcalignmentfile.AlignmentFile.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpysam/libcalignmentfile.pyx\u001b[0m in \u001b[0;36mpysam.libcalignmentfile.AlignmentFile._open\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpysam/libchtslib.pyx\u001b[0m in \u001b[0;36mpysam.libchtslib.HTSFile.check_truncation\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: no BGZF EOF marker; file may be truncated"
     ]
    }
   ],
   "source": [
    "counts = {}\n",
    "d = 'data/human/bam/*bam'\n",
    "for f in glob.glob(d):\n",
    "    encid = f.rsplit('/', maxsplit=1)[1].split('.')[0]\n",
    "    temp = metadata.loc[metadata.file == encid].iloc[0]\n",
    "    name = temp.name\n",
    "    if temp.output_type == 'unfiltered alignments':\n",
    "        result = score_aligned_reads(f)\n",
    "    elif temp.output_type == 'reads':\n",
    "        result = score_fastq_reads(f)\n",
    "    counts.setdefault(name, {}).update(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feb575af-b8d7-400a-a3d7-dbbc70941536",
   "metadata": {},
   "outputs": [],
   "source": [
    "array_attributes = ['read_len', 'query_len', 'positions_len']\n",
    "lengths = {}\n",
    "\n",
    "medians = {}\n",
    "for name in counts:\n",
    "    for k in counts[name]:\n",
    "        if k in array_attributes:\n",
    "            lengths[(k, name)] = counts[name][k]\n",
    "        else:\n",
    "            medians.setdefault(k, {})[name] = counts[name][k]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "44279e69-8597-459c-9ba1-079d79d53434",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_lr_read_lens(bams, fastqs, out):\n",
    "    \n",
    "    def score_aligned_reads(filename):\n",
    "        with pysam.AlignmentFile(filename, \"rb\") as inbam:\n",
    "            aligned_reads = 0\n",
    "            aligned_query_len = []\n",
    "            aligned_positions = []\n",
    "\n",
    "            for read in inbam:\n",
    "                if not read.is_unmapped:\n",
    "                    aligned_reads += 1\n",
    "                    aligned_query_len.append(read.query_length)\n",
    "                    aligned_positions.append(len(read.positions))\n",
    "\n",
    "            return {\n",
    "                \"aligned_reads\": len(aligned_query_len),\n",
    "                \"query_len_median\": np.median(aligned_query_len),\n",
    "                \"query_len\": aligned_query_len,\n",
    "                \"positions_median\": np.median(aligned_positions),\n",
    "                \"positions_len\": aligned_positions,\n",
    "            }\n",
    "    def score_fastq_reads(filename):\n",
    "        read_len = []\n",
    "        with xopen(filename, \"rt\") as instream:\n",
    "            for record in SeqIO.parse(instream, \"fastq\"):\n",
    "                read_len.append(len(record))\n",
    "\n",
    "        return {\n",
    "            \"raw_reads\": len(read_len),\n",
    "            \"read_len_median\": np.median(read_len),\n",
    "            \"read_len\": read_len,\n",
    "        }\n",
    "    \n",
    "    counts = {}\n",
    "    for f in bams+fastqs:\n",
    "        encid = f.rsplit('/', maxsplit=1)[1].split('.')[0]\n",
    "        temp = metadata.loc[metadata.file == encid].iloc[0]\n",
    "        name = temp.name\n",
    "        if temp.output_type == 'unfiltered alignments':\n",
    "            result = score_aligned_reads(f)\n",
    "        elif temp.output_type == 'reads':\n",
    "            result = score_fastq_reads(f)\n",
    "        counts.setdefault(name, {}).update(result)\n",
    "        \n",
    "    array_attributes = ['read_len', 'query_len', 'positions_len']\n",
    "    lengths = {}\n",
    "\n",
    "    medians = {}\n",
    "    for name in counts:\n",
    "        for k in counts[name]:\n",
    "            if k in array_attributes:\n",
    "                lengths[(k, name)] = counts[name][k]\n",
    "            else:\n",
    "                medians.setdefault(k, {})[name] = counts[name][k]\n",
    "                \n",
    "    various_counts = pd.DataFrame(medians)\n",
    "    various_counts.to_csv(out, sep='\\t')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
