{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "58ea5c04-a35f-4643-b407-5beb687830f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as st\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import os\n",
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt\n",
    "import swan_vis as swan\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.mane_utils import *\n",
    "from scripts.plotting import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b4c45c9-c569-48f2-9df5-24da5deee406",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../snakemake/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cf91884a-351c-4d15-b452-59b8cbccb963",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = '../'+expand(config['data']['ab'], species='human')[0]\n",
    "filt_ab = '../'+expand(config['data']['filt_ab'], species='human')[0]\n",
    "read_annot = '../'+expand(config['data']['read_annot'], species='human')[0]\n",
    "t_metadata = '../'+expand(config['ref']['cerberus']['t_info'], species='human')[0]\n",
    "lib_meta = '../'+expand(config['data']['meta'], species='human')[0]\n",
    "swan_file = '../'+expand(config['data']['sg'], species='human')[0]\n",
    "cerberus_h5 = '../'+expand(config['data']['cerb_annot'], species='human')[0]\n",
    "major_isos = '../'+expand(config['data']['major_isos'], species='human', obs_col='sample')[0]\n",
    "pp_summary = '../'+expand(config['data']['p_pred']['summary'], species='human')[0]\n",
    "ref_t_metadata = '../'+expand(config['ref']['t_info'], species='human')[0]\n",
    "ref_g_metadata = '../'+expand(config['ref']['g_info'], species='human')[0]\n",
    "\n",
    "\n",
    "sr_ab = '../'+config['sr']['ab']\n",
    "\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "min_tpm = 1\n",
    "gene_subset = 'protein_coding'\n",
    "obs_col = 'sample'\n",
    "go_gene_subset = 'protein_coding'\n",
    "\n",
    "m_lib_meta = '../'+expand(config['data']['meta'], species='mouse')[0]\n",
    "m_ab = '../'+expand(config['data']['ab'], species='mouse')[0]\n",
    "m_filt_ab = '../'+expand(config['data']['filt_ab'], species='mouse')[0]\n",
    "m_read_annot = '../'+expand(config['data']['read_annot'], species='mouse')[0]\n",
    "m_t_metadata = '../'+expand(config['ref']['cerberus']['t_info'], species='mouse')[0]\n",
    "m_swan_file = '../'+expand(config['data']['sg'], species='mouse')[0]\n",
    "m_cerberus_h5 = '../'+expand(config['data']['cerb_annot'], species='mouse')[0]\n",
    "m_pi_tpm_table = '../'+expand(config['data']['pi_tpm']['triplet'], species='mouse', obs_col='sample')[0]\n",
    "m_major_isos = '../'+expand(config['data']['major_isos'], species='mouse', obs_col='sample')[0]\n",
    "m_pp_summary = '../'+expand(config['data']['p_pred']['summary'], species='mouse')[0]\n",
    "m_ref_t_metadata = '../'+expand(config['ref']['t_info'], species='mouse')[0]\n",
    "m_ref_g_metadata = '../'+expand(config['ref']['g_info'], species='mouse')[0]\n",
    "m_ref_gtf = '../'+expand(config['ref']['cerberus']['gtf'], species='mouse')[0]\n",
    "m_gtf = '../'+expand(config['data']['cerb_gtf'], species='mouse')[0]\n",
    "\n",
    "mouse_ver = 'vM25_cerberus'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3449f376-4092-427a-b840-fabccfb96b14",
   "metadata": {},
   "source": [
    "I want a GTF that's the union of the reference cerberus-ized GTF and the mouse GTF. First I'll try to do that, then I'll try to limit the novel transcripts in the GTF to those that've passed our filtering."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "e40c8eaf-988b-40b1-98db-c01ad94885ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_union_cerb_gtf(gtf1,\n",
    "                       gtf2, \n",
    "                       h5):\n",
    "    \"\"\"\n",
    "    Merge 2 GTFs. Preferentially keep entries from gtf1 vs. gtf2.\n",
    "    \"\"\"\n",
    "    \n",
    "    def preproc_gtf(gtf):\n",
    "        df = pr.read_gtf(gtf,\n",
    "                         duplicate_attr=True,\n",
    "                         rename_attr=True).df\n",
    "        df['gene_id'] = cerberus.get_stable_gid(df, 'gene_id')\n",
    "        df = df.loc[df.gene_id.str.contains('ENSMUS')]\n",
    "        tids = df.transcript_id.unique().tolist()    \n",
    "        return df, tids\n",
    "    \n",
    "    gtf1_df, gtf1_tids = preproc_gtf(gtf1)\n",
    "    gtf2_df, gtf2_tids = preproc_gtf(gtf2)    \n",
    "    \n",
    "    # filter gtf2 based on unique transcript ids\n",
    "    # already in gtf1\n",
    "    new_ids = list(set(gtf1_tids)-set(gtf2_tids))\n",
    "    gtf2_df = gtf2_df.loc[gtf2_df.transcript_id.isin(new_ids)]\n",
    "    \n",
    "    # filter gtf2 based on genes that are in gtf1\n",
    "    gtf2_df = gtf2_df.loc[gtf2_df.gene_id.isin(gtf1_df.gene_id.tolist())]\n",
    "    \n",
    "    # use cerberus functions to fix GTF entries\n",
    "    ca = cerberus.read(h5)\n",
    "    tss = pr.PyRanges(ca.tss.rename({'Name': 'tss_id'}, axis=1))\n",
    "    tes = pr.PyRanges(ca.tes.rename({'Name': 'tes_id'}, axis=1))\n",
    "    df = pd.concat([gtf1_df, gtf2_df], axis=0) \n",
    "    df = cerberus.sort_gtf(df)\n",
    "    df = cerberus.update_gtf_ends(df, tss, tes)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "fc2236d3-7156-4b78-98e8-750b65095361",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n"
     ]
    }
   ],
   "source": [
    "df = get_union_cerb_gtf(m_ref_gtf,\n",
    "                        m_gtf,\n",
    "                        m_cerberus_h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "72d60830-9c6a-4e1e-a992-72a07bc820ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr.PyRanges(df).to_gtf('vM25_cerberus_2.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "3ef2ff0d-0f2b-4792-a72d-c9478b950dc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = pr.read_gtf(m_ref_gtf, duplicate_attr=True).df\n",
    "ref_df['gene_id'] = cerberus.get_stable_gid(ref_df, 'gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "23700b88-4447-473d-96cf-5a6edb66d2bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_ids = ref_df.transcript_id.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "ee90a59d-d31a-494c-8f7f-4fb0cfb95a2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found attributes with reserved names: ['source'].\n",
      "Renaming attributes with suffix '_attr'\n"
     ]
    }
   ],
   "source": [
    "df = pr.read_gtf(m_gtf, duplicate_attr=True, rename_attr=True).df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "8af78d97-1413-4a2e-8017-74a9db83640c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.loc[df.gene_id.str.contains('ENSMUS')]\n",
    "df['gene_id'] = cerberus.get_stable_gid(df, 'gene_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "a781092c-123b-4cb2-a8a5-e2235debb0d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "149102\n",
      "85674\n"
     ]
    }
   ],
   "source": [
    "# filter the new gtf based on unique transcript ids\n",
    "new_ids = df.transcript_id.unique().tolist()\n",
    "print(len(new_ids))\n",
    "new_ids = list(set(new_ids)-set(ref_ids))\n",
    "print(len(new_ids))\n",
    "df = df.loc[df.transcript_id.isin(new_ids)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "ec834ef0-8384-484a-a34f-0b5a8860e774",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate how many transcripts we have and how many new ones we should get"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "2a79008c-6ddf-45ec-bee1-c72e6e8f2dc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca = cerberus.read(m_cerberus_h5)\n",
    "tss = pr.PyRanges(ca.tss.rename({'Name': 'tss_id'}, axis=1))\n",
    "tes = pr.PyRanges(ca.tes.rename({'Name': 'tes_id'}, axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "98fce7eb-4d5c-4a5c-97f0-7da9e36277ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = pd.concat([ref_df, df], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "d17eb13c-700c-45b9-8b52-fecbcb0246e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = cerberus.sort_gtf(new_df)\n",
    "new_df = cerberus.update_gtf_ends(new_df, tss, tes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "05a543c5-8e61-4f7d-89fb-698613838633",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "b982046e-ff0f-4518-a595-dae4b0cc713d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>ISM-suffix_to_IDs</th>\n",
       "      <th>ISM_to_IDs</th>\n",
       "      <th>ISM_transcript</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>ISM-suffix_transcript</th>\n",
       "      <th>ISM-prefix_transcript</th>\n",
       "      <th>ISM-prefix_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>gene</td>\n",
       "      <td>161082474</td>\n",
       "      <td>161258263</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>161162699</td>\n",
       "      <td>161258263</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>161162699</td>\n",
       "      <td>161163248</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>161171489</td>\n",
       "      <td>161171535</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrX</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>161177246</td>\n",
       "      <td>161177317</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000000037</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Source     Feature      Start        End Score Strand Frame  \\\n",
       "0       chrX     NaN        gene  161082474  161258263     .      +     .   \n",
       "1       chrX  HAVANA  transcript  161162699  161258263     .      +     .   \n",
       "2       chrX  HAVANA        exon  161162699  161163248     .      +     .   \n",
       "3       chrX  HAVANA        exon  161171489  161171535     .      +     .   \n",
       "4       chrX  HAVANA        exon  161177246  161177317     .      +     .   \n",
       "\n",
       "              gene_id       gene_type  ... transcript_status source_attr  \\\n",
       "0  ENSMUSG00000000037  protein_coding  ...               NaN         NaN   \n",
       "1  ENSMUSG00000000037  protein_coding  ...               NaN         NaN   \n",
       "2  ENSMUSG00000000037  protein_coding  ...               NaN         NaN   \n",
       "3  ENSMUSG00000000037  protein_coding  ...               NaN         NaN   \n",
       "4  ENSMUSG00000000037  protein_coding  ...               NaN         NaN   \n",
       "\n",
       "  NNC_transcript ISM-suffix_to_IDs ISM_to_IDs ISM_transcript  \\\n",
       "0            NaN               NaN        NaN            NaN   \n",
       "1            NaN               NaN        NaN            NaN   \n",
       "2            NaN               NaN        NaN            NaN   \n",
       "3            NaN               NaN        NaN            NaN   \n",
       "4            NaN               NaN        NaN            NaN   \n",
       "\n",
       "  transcript_novelty ISM-suffix_transcript ISM-prefix_transcript  \\\n",
       "0                NaN                   NaN                   NaN   \n",
       "1                NaN                   NaN                   NaN   \n",
       "2                NaN                   NaN                   NaN   \n",
       "3                NaN                   NaN                   NaN   \n",
       "4                NaN                   NaN                   NaN   \n",
       "\n",
       "  ISM-prefix_to_IDs  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2               NaN  \n",
       "3               NaN  \n",
       "4               NaN  \n",
       "\n",
       "[5 rows x 48 columns]"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df.head()\n",
    "# new_df.gene_id.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "f6f7a950-3782-434b-a2d8-24332dd379eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>mgi_id</th>\n",
       "      <th>transcript_type</th>\n",
       "      <th>level</th>\n",
       "      <th>original_transcript_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>tag</th>\n",
       "      <th>ont</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Source, Feature, Start, End, Score, Strand, Frame, gene_id, gene_type, gene_name, ic, ic_id, tss_id, tss, tes_id, tes, transcript_id, transcript_name, tes_last_sa_issue, tss_first_sd_issue, original_transcript_name, transcript_support_level, havana_gene, havana_transcript, mgi_id, transcript_type, level, original_transcript_id, protein_id, ccdsid, tag, ont]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 33 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_df.loc[ref_df.gene_id=='ENCODEMG000055991']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "5c8f8451-b013-48d2-a192-cb124cc346ea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gene', 'transcript', 'exon']\n",
       "Categories (3, object): ['exon', 'gene', 'transcript']"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df.Feature.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "9839f8e2-09b1-4c4a-972b-1f4559c752e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove things that don't have a gene in v40\n",
    "t_df = new_df[['transcript_id', 'gene_id']].drop_duplicates()\n",
    "t_gids = t_df.gene_id.unique().tolist()\n",
    "g_df = new_df.loc[new_df.Feature=='gene']\n",
    "g_gids = g_df.gene_id.unique().tolist()\n",
    "t_df.loc[~t_df.gene_id.isin(g_gids)]\n",
    "\n",
    "len(new_df.loc[new_df.gene_id.isin(g_gids)])\n",
    "new_df = new_df.loc[new_df.gene_id.isin(g_gids)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "20aeaae2-34a4-406d-9c88-a5942a667ebe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>ISM-suffix_to_IDs</th>\n",
       "      <th>ISM_to_IDs</th>\n",
       "      <th>ISM_transcript</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>ISM-suffix_transcript</th>\n",
       "      <th>ISM-prefix_transcript</th>\n",
       "      <th>ISM-prefix_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Chromosome, Source, Feature, Start, End, Score, Strand, Frame, gene_id, gene_type, gene_name, ic, ic_id, tss_id, tss, tes_id, tes, transcript_id, transcript_name, tes_last_sa_issue, tss_first_sd_issue, original_transcript_name, transcript_support_level, havana_gene, havana_transcript, mgi_id, transcript_type, level, original_transcript_id, protein_id, ccdsid, tag, ont, gene_status, talon_gene, talon_transcript, source_cerberus, NIC_transcript, transcript_status, source_attr, NNC_transcript, ISM-suffix_to_IDs, ISM_to_IDs, ISM_transcript, transcript_novelty, ISM-suffix_transcript, ISM-prefix_transcript, ISM-prefix_to_IDs]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 48 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_df.loc[new_df.gene_id =='ENSMUSG00000097131']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "id": "ab35d9e7-629e-440c-94a3-d0d2556861db",
   "metadata": {},
   "outputs": [],
   "source": [
    "pr.PyRanges(new_df).to_gtf('vM25_cerberus.gtf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "08d99ba0-caa9-4d58-889b-9da923a55496",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Source</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Score</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Frame</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>...</th>\n",
       "      <th>transcript_status</th>\n",
       "      <th>source_attr</th>\n",
       "      <th>NNC_transcript</th>\n",
       "      <th>ISM-suffix_to_IDs</th>\n",
       "      <th>ISM_to_IDs</th>\n",
       "      <th>ISM_transcript</th>\n",
       "      <th>transcript_novelty</th>\n",
       "      <th>ISM-suffix_transcript</th>\n",
       "      <th>ISM-prefix_transcript</th>\n",
       "      <th>ISM-prefix_to_IDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>973796</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72236414</td>\n",
       "      <td>72266186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973797</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>72236414</td>\n",
       "      <td>72236528</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973798</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>72237865</td>\n",
       "      <td>72237923</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973799</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>72241370</td>\n",
       "      <td>72241503</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973800</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>72264742</td>\n",
       "      <td>72266186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973801</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72243970</td>\n",
       "      <td>72274919</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973802</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>72243970</td>\n",
       "      <td>72244163</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973803</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>72264742</td>\n",
       "      <td>72264892</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973804</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>72269467</td>\n",
       "      <td>72269548</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973805</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>72272147</td>\n",
       "      <td>72274919</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973806</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>transcript</td>\n",
       "      <td>72254764</td>\n",
       "      <td>72266780</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973807</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>72254764</td>\n",
       "      <td>72254994</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973808</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>72261106</td>\n",
       "      <td>72261186</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973809</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>72262172</td>\n",
       "      <td>72262330</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>973810</th>\n",
       "      <td>chr1</td>\n",
       "      <td>TALON</td>\n",
       "      <td>exon</td>\n",
       "      <td>72264742</td>\n",
       "      <td>72266780</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ENSMUSG00000097131</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>...</td>\n",
       "      <td>NOVEL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15 rows × 48 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome   Source     Feature     Start       End Score Strand Frame  \\\n",
       "973796       chr1  ENSEMBL  transcript  72236414  72266186     .      +     .   \n",
       "973797       chr1  ENSEMBL        exon  72236414  72236528     .      +     .   \n",
       "973798       chr1  ENSEMBL        exon  72237865  72237923     .      +     .   \n",
       "973799       chr1  ENSEMBL        exon  72241370  72241503     .      +     .   \n",
       "973800       chr1  ENSEMBL        exon  72264742  72266186     .      +     .   \n",
       "973801       chr1  ENSEMBL  transcript  72243970  72274919     .      +     .   \n",
       "973802       chr1  ENSEMBL        exon  72243970  72244163     .      +     .   \n",
       "973803       chr1  ENSEMBL        exon  72264742  72264892     .      +     .   \n",
       "973804       chr1  ENSEMBL        exon  72269467  72269548     .      +     .   \n",
       "973805       chr1  ENSEMBL        exon  72272147  72274919     .      +     .   \n",
       "973806       chr1    TALON  transcript  72254764  72266780     .      +     .   \n",
       "973807       chr1    TALON        exon  72254764  72254994     .      +     .   \n",
       "973808       chr1    TALON        exon  72261106  72261186     .      +     .   \n",
       "973809       chr1    TALON        exon  72262172  72262330     .      +     .   \n",
       "973810       chr1    TALON        exon  72264742  72266780     .      +     .   \n",
       "\n",
       "                   gene_id gene_type  ... transcript_status source_attr  \\\n",
       "973796  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973797  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973798  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973799  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973800  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973801  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973802  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973803  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973804  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973805  ENSMUSG00000097131   lincRNA  ...             KNOWN     ENSEMBL   \n",
       "973806  ENSMUSG00000097131   lincRNA  ...             NOVEL         NaN   \n",
       "973807  ENSMUSG00000097131   lincRNA  ...             NOVEL         NaN   \n",
       "973808  ENSMUSG00000097131   lincRNA  ...             NOVEL         NaN   \n",
       "973809  ENSMUSG00000097131   lincRNA  ...             NOVEL         NaN   \n",
       "973810  ENSMUSG00000097131   lincRNA  ...             NOVEL         NaN   \n",
       "\n",
       "       NNC_transcript ISM-suffix_to_IDs ISM_to_IDs ISM_transcript  \\\n",
       "973796            NaN               NaN        NaN            NaN   \n",
       "973797            NaN               NaN        NaN            NaN   \n",
       "973798            NaN               NaN        NaN            NaN   \n",
       "973799            NaN               NaN        NaN            NaN   \n",
       "973800            NaN               NaN        NaN            NaN   \n",
       "973801            NaN               NaN        NaN            NaN   \n",
       "973802            NaN               NaN        NaN            NaN   \n",
       "973803            NaN               NaN        NaN            NaN   \n",
       "973804            NaN               NaN        NaN            NaN   \n",
       "973805            NaN               NaN        NaN            NaN   \n",
       "973806            NaN               NaN        NaN            NaN   \n",
       "973807            NaN               NaN        NaN            NaN   \n",
       "973808            NaN               NaN        NaN            NaN   \n",
       "973809            NaN               NaN        NaN            NaN   \n",
       "973810            NaN               NaN        NaN            NaN   \n",
       "\n",
       "       transcript_novelty ISM-suffix_transcript ISM-prefix_transcript  \\\n",
       "973796                NaN                   NaN                   NaN   \n",
       "973797                NaN                   NaN                   NaN   \n",
       "973798                NaN                   NaN                   NaN   \n",
       "973799                NaN                   NaN                   NaN   \n",
       "973800                NaN                   NaN                   NaN   \n",
       "973801                NaN                   NaN                   NaN   \n",
       "973802                NaN                   NaN                   NaN   \n",
       "973803                NaN                   NaN                   NaN   \n",
       "973804                NaN                   NaN                   NaN   \n",
       "973805                NaN                   NaN                   NaN   \n",
       "973806                NaN                   NaN                   NaN   \n",
       "973807                NaN                   NaN                   NaN   \n",
       "973808                NaN                   NaN                   NaN   \n",
       "973809                NaN                   NaN                   NaN   \n",
       "973810                NaN                   NaN                   NaN   \n",
       "\n",
       "       ISM-prefix_to_IDs  \n",
       "973796               NaN  \n",
       "973797               NaN  \n",
       "973798               NaN  \n",
       "973799               NaN  \n",
       "973800               NaN  \n",
       "973801               NaN  \n",
       "973802               NaN  \n",
       "973803               NaN  \n",
       "973804               NaN  \n",
       "973805               NaN  \n",
       "973806               NaN  \n",
       "973807               NaN  \n",
       "973808               NaN  \n",
       "973809               NaN  \n",
       "973810               NaN  \n",
       "\n",
       "[15 rows x 48 columns]"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# new_df.gene_id.head()\n",
    "# new_df.gene_id.tail()\n",
    "# new_df.loc[new_df.gene_id.str.contains('.')]\n",
    "new_df.loc[new_df.gene_id=='ENSMUSG00000097131']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "1fb0cf7c-fd7f-478f-85ae-2158783c2160",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try to make this into a talon db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae0886d-d2c0-4773-9ffb-df83ba306780",
   "metadata": {},
   "outputs": [],
   "source": [
    "# talon_initialize_database \\\n",
    "#     --f vM25_cerberus.gtf \\\n",
    "#     --g mm10 \\\n",
    "#     --a vM25+cerberus \\\n",
    "#     --o vM25_cerberus"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
