{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c711a6cc-e326-4061-af4c-186ad4cbb6d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import sys\n",
    "import os\n",
    "from Bio import SearchIO\n",
    "import yaml\n",
    "from snakemake.io import expand\n",
    "import pdb\n",
    "import swan_vis as swan\n",
    "\n",
    "p = os.path.dirname(os.path.dirname(os.getcwd()))\n",
    "sys.path.append(p)\n",
    "\n",
    "from scripts.utils import *\n",
    "from scripts.plotting import *\n",
    "from scripts.mane_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "90a49224-0249-4215-8e56-bd31deae72c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "config_file = '../snakemake/config.yml'\n",
    "with open(config_file) as f:\n",
    "    config = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "12e77f56-fb55-4b5d-b717-c25b29e7c375",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab = '../'+expand(config['data']['ab'], species='human')[0]\n",
    "filt_ab = '../'+expand(config['data']['filt_ab'], species='human')[0]\n",
    "read_annot = '../'+expand(config['data']['read_annot'], species='human')[0]\n",
    "ref_t_metadata = '../'+expand(config['ref']['t_info'], species='human')[0]\n",
    "ref_g_metadata = '../'+expand(config['ref']['g_info'], species='human')[0]\n",
    "t_metadata = '../'+expand(config['ref']['cerberus']['t_info'], species='human')[0]\n",
    "lib_meta = '../'+expand(config['data']['meta'], species='human')[0]\n",
    "swan_file = '../'+expand(config['data']['sg'], species='human')[0]\n",
    "cerberus_h5 = '../'+expand(config['data']['cerb_annot'], species='human')[0]\n",
    "cerb_t_metadata = '../'+expand(config['data']['t_info'], species='human')[0]\n",
    "major_isos = '../'+expand(config['data']['major_isos'], species='human', obs_col='sample')[0]\n",
    "pp_summary = '../'+expand(config['data']['p_pred']['summary'], species='human')[0]\n",
    "\n",
    "ver = 'v40_cerberus'\n",
    "min_tpm = 1\n",
    "gene_subset = 'polya'\n",
    "obs_col = 'sample'\n",
    "\n",
    "m_lib_meta = '../'+expand(config['data']['meta'], species='mouse')[0]\n",
    "m_ab = '../'+expand(config['data']['ab'], species='mouse')[0]\n",
    "m_filt_ab = '../'+expand(config['data']['filt_ab'], species='mouse')[0]\n",
    "m_read_annot = '../'+expand(config['data']['read_annot'], species='mouse')[0]\n",
    "m_ref_t_metadata = '../'+expand(config['ref']['t_info'], species='mouse')[0]\n",
    "m_ref_g_metadata = '../'+expand(config['ref']['g_info'], species='mouse')[0]\n",
    "m_t_metadata = '../'+expand(config['ref']['cerberus']['t_info'], species='mouse')[0]\n",
    "m_lib_meta = '../'+expand(config['data']['meta'], species='mouse')[0]\n",
    "m_swan_file = '../'+expand(config['data']['sg'], species='mouse')[0]\n",
    "m_cerberus_h5 = '../'+expand(config['data']['cerb_annot'], species='mouse')[0]\n",
    "m_cerb_t_metadata = '../'+expand(config['data']['t_info'], species='mouse')[0]\n",
    "m_major_isos = '../'+expand(config['data']['major_isos'], species='mouse', obs_col='sample')[0]\n",
    "m_pp_summary = '../'+expand(config['data']['p_pred']['summary'], species='mouse')[0]\n",
    "\n",
    "\n",
    "mouse_ver = 'vM25_cerberus'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4ea0cf8-7b80-42a4-847b-8c21ef0a1b67",
   "metadata": {},
   "source": [
    "## Save TPMs of transcripts expressed in 16 cell lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "67aa2dc2-dd6f-448e-8c0d-0b7b995e9bf0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for polya genes\n",
      "Enforcing minimum TPM\n",
      "Total # isos detected: 231136\n",
      "# isos >= 1 tpm: 210239\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 206806\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>huvec_1_1</th>\n",
       "      <th>huvec_1_2</th>\n",
       "      <th>hmec_1_1</th>\n",
       "      <th>hmec_1_2</th>\n",
       "      <th>panc1_1_1</th>\n",
       "      <th>pc3_1_1</th>\n",
       "      <th>k562_1_1</th>\n",
       "      <th>hepg2_1_1</th>\n",
       "      <th>hct116_1_1</th>\n",
       "      <th>imr90_1_1</th>\n",
       "      <th>...</th>\n",
       "      <th>ocily7_1_1</th>\n",
       "      <th>ocily7_1_2</th>\n",
       "      <th>mcf10a_1_1</th>\n",
       "      <th>mcf10a_1_2</th>\n",
       "      <th>pc9_1_1</th>\n",
       "      <th>pc9_1_2</th>\n",
       "      <th>a673_1_1</th>\n",
       "      <th>a673_1_2</th>\n",
       "      <th>calu3_1_1</th>\n",
       "      <th>calu3_1_2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232[1,2,1]</th>\n",
       "      <td>0.57069</td>\n",
       "      <td>1.138878</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.875113</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.226145</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.834033</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.856421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232[1,3,1]</th>\n",
       "      <td>0.57069</td>\n",
       "      <td>0.569439</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584375</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.875113</td>\n",
       "      <td>0.815517</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.919458</td>\n",
       "      <td>...</td>\n",
       "      <td>1.064405</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.821249</td>\n",
       "      <td>2.578556</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.712842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232[1,4,1]</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.815517</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.183814</td>\n",
       "      <td>1.306486</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232[1,5,1]</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.569439</td>\n",
       "      <td>1.086655</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.061271</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.856421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232[2,3,1]</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.856421</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        huvec_1_1  huvec_1_2  hmec_1_1  hmec_1_2  panc1_1_1  \\\n",
       "annot_transcript_id                                                           \n",
       "ENSG00000227232[1,2,1]    0.57069   1.138878  0.000000  0.000000        0.0   \n",
       "ENSG00000227232[1,3,1]    0.57069   0.569439  0.000000  0.584375        0.0   \n",
       "ENSG00000227232[1,4,1]    0.00000   0.000000  0.000000  0.000000        0.0   \n",
       "ENSG00000227232[1,5,1]    0.00000   0.569439  1.086655  0.000000        0.0   \n",
       "ENSG00000227232[2,3,1]    0.00000   0.000000  0.000000  0.000000        0.0   \n",
       "\n",
       "                         pc3_1_1  k562_1_1  hepg2_1_1  hct116_1_1  imr90_1_1  \\\n",
       "annot_transcript_id                                                            \n",
       "ENSG00000227232[1,2,1]  0.875113  0.000000   1.226145    0.000000   0.000000   \n",
       "ENSG00000227232[1,3,1]  0.875113  0.815517   0.000000    0.000000   3.919458   \n",
       "ENSG00000227232[1,4,1]  0.000000  0.815517   0.000000    3.183814   1.306486   \n",
       "ENSG00000227232[1,5,1]  0.000000  0.000000   0.000000    1.061271   0.000000   \n",
       "ENSG00000227232[2,3,1]  0.000000  0.000000   0.000000    0.000000   0.000000   \n",
       "\n",
       "                        ...  ocily7_1_1  ocily7_1_2  mcf10a_1_1  mcf10a_1_2  \\\n",
       "annot_transcript_id     ...                                                   \n",
       "ENSG00000227232[1,2,1]  ...    0.000000         0.0         0.0         0.0   \n",
       "ENSG00000227232[1,3,1]  ...    1.064405         0.0         0.0         0.0   \n",
       "ENSG00000227232[1,4,1]  ...    0.000000         0.0         0.0         0.0   \n",
       "ENSG00000227232[1,5,1]  ...    0.000000         0.0         0.0         0.0   \n",
       "ENSG00000227232[2,3,1]  ...    0.000000         0.0         0.0         0.0   \n",
       "\n",
       "                         pc9_1_1   pc9_1_2  a673_1_1  a673_1_2  calu3_1_1  \\\n",
       "annot_transcript_id                                                         \n",
       "ENSG00000227232[1,2,1]  0.000000  0.000000       0.0  0.834033        0.0   \n",
       "ENSG00000227232[1,3,1]  0.821249  2.578556       0.0  0.000000        0.0   \n",
       "ENSG00000227232[1,4,1]  0.000000  0.000000       0.0  0.000000        0.0   \n",
       "ENSG00000227232[1,5,1]  0.000000  0.000000       0.0  0.000000        0.0   \n",
       "ENSG00000227232[2,3,1]  0.000000  0.000000       0.0  0.000000        0.0   \n",
       "\n",
       "                        calu3_1_2  \n",
       "annot_transcript_id                \n",
       "ENSG00000227232[1,2,1]   0.856421  \n",
       "ENSG00000227232[1,3,1]   1.712842  \n",
       "ENSG00000227232[1,4,1]   0.000000  \n",
       "ENSG00000227232[1,5,1]   0.856421  \n",
       "ENSG00000227232[2,3,1]   0.856421  \n",
       "\n",
       "[5 rows x 24 columns]"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets = get_ljungman_datasets()\n",
    "\n",
    "ab_df = pd.read_csv(filt_ab, sep='\\t')\n",
    "ab_df, _ = get_tpm_table(ab_df,\n",
    "               how='iso',\n",
    "               min_tpm=min_tpm,\n",
    "               gene_subset=gene_subset)\n",
    "ab_df = ab_df[datasets]\n",
    "ab_df.to_csv('transcript_abundance.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1a159a6-e930-4bab-8574-97ca4f45df72",
   "metadata": {},
   "source": [
    "## Get transcripts expressed in 16 cell lines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9acc9873-fa6e-45e9-b2a3-7941c22fc82c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for polya genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/scripts/utils.py:2096: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[tpm_col] = (df[d]*1000000)/df[total_col]\n",
      "/Users/fairliereese/Documents/programming/mortazavi_lab/data/paper_rnawg/scripts/utils.py:2095: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  df[total_col] = df[d].sum()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enforcing minimum TPM\n",
      "Total # isos detected: 231136\n",
      "# isos >= 1 tpm: 210239\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 206806\n",
      "Found 138 total libraries\n"
     ]
    }
   ],
   "source": [
    "# get tids that are expressed in ljungman datasets\n",
    "datasets = get_ljungman_datasets()\n",
    "df = pd.read_csv(filt_ab, sep='\\t')\n",
    "df = get_det_table(df,\n",
    "               how='iso',\n",
    "               min_tpm=min_tpm,\n",
    "               gene_subset=gene_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c4d93bf6-d4b4-4cda-98ed-4ae1a23fd1fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# limit to 16 cell lines\n",
    "df = df.loc[datasets]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5297dbc0-3253-420f-9215-c924643dc5ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.transpose()\n",
    "df = df.loc[df.sum(axis=1)>=1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b4d6f9c8-fd5a-452e-82a4-a69ebc2173a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "tids = df.index.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2ec11f2-c575-4b5a-b6e0-a7d0b59286ef",
   "metadata": {},
   "source": [
    "## Get v29 transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "bc510b60-b80e-40bd-990c-60ec66c9c4bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "ca = cerberus.read(cerberus_h5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "38878795-9313-4a7d-87b6-b49f1703a1e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_df = ca.t_map.loc[ca.t_map.source=='v29']\n",
    "ref_tids = ref_df.transcript_id.unique().tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9303e1b9-995d-48fc-9860-e50bb8776f40",
   "metadata": {},
   "source": [
    "## Find transcripts w/ evidence of fusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9f87b1a9-f2f0-403d-b2a7-91e62c80ad73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read in graph from ../data/human/swan_graph.p\n"
     ]
    }
   ],
   "source": [
    "sg = swan.read(swan_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "b9a49270-a7c5-4669-a845-4167d533dc51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the edges for each transcript\n",
    "# get the genes associated with each edge\n",
    "# get the transcripts that have edges assc. w/ more than 1 gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "60dcfc02-3c5e-447e-b979-be8b106da081",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a list of exons / introns\n",
    "df = swan.pivot_path_list(sg.t_df, 'path')\n",
    "df.reset_index(inplace=True)\n",
    "df = df.merge(sg.t_df[['gid', 'gname']].reset_index(), on='tid', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "6e3f9915-7e4c-4d08-bd12-398f40c27b82",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a table of edge:gid\n",
    "eg_df = df[['edge_id', 'gid']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6cc6fc46-9473-430e-9eaf-27b3696d0ce1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge in the edge:gid to the tid:edge table\n",
    "df.drop(['gid', 'gname'], axis=1, inplace=True)\n",
    "df = df.merge(eg_df[['edge_id', 'gid']], how='left', on='edge_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "dcf78f71-a83f-43c1-be80-088bc2183f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# how many genes are represented by the edges of each transcript?\n",
    "gt_df = df.drop('edge_id', axis=1).copy(deep=True)\n",
    "gt_df = gt_df.groupby('tid').nunique().rename({'gid':'n_genes'}, axis=1).reset_index()\n",
    "\n",
    "# get gene ids and gene names assc. with this as well and merge in\n",
    "temp = df.drop('edge_id', axis=1).copy(deep=True)\n",
    "temp = temp.merge(sg.t_df[['gname', 'gid']].drop_duplicates(), how='left', on='gid').drop_duplicates()\n",
    "temp = temp.groupby('tid').agg(','.join).reset_index()\n",
    "gt_df = gt_df.merge(temp, on='tid', how='left')\n",
    "\n",
    "# limit to transcripts w/ >1 gene\n",
    "gt_df = gt_df.loc[gt_df.n_genes>1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "2000c5b8-390a-45af-9c90-b499c6bbe78a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# limit entries in to those expressed in the 16 cell lines\n",
    "gt_df = gt_df.loc[gt_df.tid.isin(tids)]\n",
    "\n",
    "# label v29 annotated transcripts\n",
    "gt_df['Known'] = gt_df.tid.isin(ref_tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "dd4991de-88be-497a-9878-8a97902ccb9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gt_df.to_csv('fusion_transcripts.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff996449-f404-4271-b358-8cf18fbbb369",
   "metadata": {},
   "source": [
    "## Get all unspliced transcripts w/ gene name, gene id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a13c0e86-4432-44ea-ac81-0f2198cbad29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# label monoexonic transcripts\n",
    "t_df = sg.t_df.copy(deep=True)\n",
    "t_df = t_df[['tid', 'gid', 'gname', 'path']].reset_index(drop=True)\n",
    "t_df['unspliced'] = t_df.apply(lambda x: len(x.path)==1, axis=1)\n",
    "t_df = t_df.loc[t_df.unspliced==True]\n",
    "t_df.set_index('tid', inplace=True)\n",
    "\n",
    "# get edges used in each monoexonic transcript\n",
    "ue_t_df = swan.pivot_path_list(t_df, path_col='path')\n",
    "ue_df = sg.edge_df.loc[ue_t_df.edge_id.tolist()]\n",
    "\n",
    "# add coords for these edges\n",
    "ue_df = ue_df.merge(sg.loc_df[['chrom', 'coord']],\n",
    "            how='left', left_on='v1', right_on='vertex_id')\n",
    "ue_df.rename({'coord': 'start'}, axis=1, inplace=True)\n",
    "ue_df = ue_df.merge(sg.loc_df[['coord']],\n",
    "            how='left', left_on='v2', right_on='vertex_id')\n",
    "ue_df.rename({'coord': 'stop'}, axis=1, inplace=True)\n",
    "ue_df.drop(['v1', 'v2'], axis=1, inplace=True)\n",
    "\n",
    "# merge info in with transcript info\n",
    "ue_t_df.reset_index(inplace=True)\n",
    "ue_t_df = ue_t_df.merge(ue_df, on='edge_id', how='left')\n",
    "ue_t_df = ue_t_df.merge(sg.t_df.reset_index(drop=True)[['tid', 'gid', 'gname']], on='tid', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "30267bd9-e116-46a2-8dee-6731a0a9da2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# limit entries in to those expressed in the 16 cell lines\n",
    "ue_t_df = ue_t_df.loc[ue_t_df.tid.isin(tids)]\n",
    "\n",
    "# label v29 annotated transcripts\n",
    "ue_t_df['Known'] = ue_t_df.tid.isin(ref_tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9223e303-31ee-43bc-a9f5-2418ae8f9965",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>edge_id</th>\n",
       "      <th>strand</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>annotation</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>Known</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000001460[12,13,1]</td>\n",
       "      <td>947924</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1</td>\n",
       "      <td>24357356</td>\n",
       "      <td>24356949</td>\n",
       "      <td>ENSG00000001460</td>\n",
       "      <td>STPG1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000002834[9,9,1]</td>\n",
       "      <td>974058</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>False</td>\n",
       "      <td>chr17</td>\n",
       "      <td>38919325</td>\n",
       "      <td>38921820</td>\n",
       "      <td>ENSG00000002834</td>\n",
       "      <td>LASP1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000003249[5,6,1]</td>\n",
       "      <td>339329</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>True</td>\n",
       "      <td>chr16</td>\n",
       "      <td>90007811</td>\n",
       "      <td>90004821</td>\n",
       "      <td>ENSG00000003249</td>\n",
       "      <td>DBNDD1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000004399[17,17,1]</td>\n",
       "      <td>997925</td>\n",
       "      <td>-</td>\n",
       "      <td>exon</td>\n",
       "      <td>False</td>\n",
       "      <td>chr3</td>\n",
       "      <td>129555524</td>\n",
       "      <td>129555164</td>\n",
       "      <td>ENSG00000004399</td>\n",
       "      <td>PLXND1</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000004487[11,18,1]</td>\n",
       "      <td>287</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>23080823</td>\n",
       "      <td>23083739</td>\n",
       "      <td>ENSG00000004487</td>\n",
       "      <td>KDM1A</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        tid edge_id strand edge_type  annotation  chrom  \\\n",
       "0  ENSG00000001460[12,13,1]  947924      -      exon       False   chr1   \n",
       "3    ENSG00000002834[9,9,1]  974058      +      exon       False  chr17   \n",
       "4    ENSG00000003249[5,6,1]  339329      -      exon        True  chr16   \n",
       "5  ENSG00000004399[17,17,1]  997925      -      exon       False   chr3   \n",
       "6  ENSG00000004487[11,18,1]     287      +      exon        True   chr1   \n",
       "\n",
       "       start       stop              gid   gname  Known  \n",
       "0   24357356   24356949  ENSG00000001460   STPG1  False  \n",
       "3   38919325   38921820  ENSG00000002834   LASP1  False  \n",
       "4   90007811   90004821  ENSG00000003249  DBNDD1   True  \n",
       "5  129555524  129555164  ENSG00000004399  PLXND1  False  \n",
       "6   23080823   23083739  ENSG00000004487   KDM1A   True  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ue_t_df.to_csv('unspliced_transcripts.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48d291ed-9f13-4405-92f6-26dc7280f2de",
   "metadata": {},
   "source": [
    "## 230516 Find transcripts with evidence of gene fusion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57cf2d84-9e12-46ac-b298-9de1ef173248",
   "metadata": {},
   "outputs": [],
   "source": [
    "sg = swan.read(swan_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95de6f52-48a6-4b5e-92b6-3bda742d879d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a list of splice junctions\n",
    "df = swan.pivot_path_list(sg.t_df, 'path')\n",
    "df.reset_index(inplace=True)\n",
    "df = df.merge(sg.t_df[['gid', 'gname']].reset_index(), on='tid', how='left')\n",
    "\n",
    "# limit to those from transcripts in 16 cell lines\n",
    "print(len(df.index))\n",
    "df = df.loc[df.tid.isin(tids)]\n",
    "print(len(df.index))\n",
    "\n",
    "# remove transcripts already annotated\n",
    "df = df.loc[~df.tid.isin(ref_tids)]\n",
    "print(len(df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "edfa71d3-014f-4a3e-853a-0c7349c8f34e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# edges that occur in more than one gene\n",
    "df2 = df[['edge_id', 'gid']].groupby(['edge_id']).nunique().reset_index().rename({'gid':'n_genes'}, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cb0fc76-9601-4e3a-aaaa-5a5f06456df4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df3 = df2.loc[df2.n_genes>1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee6bd900-8da7-4865-8306-95dbc3b01e85",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.displot(df3, x='n_genes', kind='hist', discrete=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc7b6c38-ccd5-45c9-a935-600a590675ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get names of transcripts / genes that have more than 1 \n",
    "# gene that share the same splice junction\n",
    "print(len(df3.index))\n",
    "df3 = df3.merge(df, on='edge_id', how='left')\n",
    "print(len(df3.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29e2404f-a716-42d3-93c6-f4e2a5a4edb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05447cac-8c0b-455d-ba5a-b0f169c0aa47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add coords for these edges\n",
    "temp = sg.edge_df.merge(sg.loc_df[['chrom', 'coord']],\n",
    "            how='left', left_on='v1', right_on='vertex_id')\n",
    "temp.rename({'coord': 'start'}, axis=1, inplace=True)\n",
    "temp = temp.merge(sg.loc_df[['coord']],\n",
    "            how='left', left_on='v2', right_on='vertex_id')\n",
    "temp.rename({'coord': 'stop'}, axis=1, inplace=True)\n",
    "temp.drop(['v1', 'v2'], axis=1, inplace=True)\n",
    "df3 = df3.merge(temp, how='left', on='edge_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "abdea77c-2bde-4fe9-a9d9-bee7331c553b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edge_id</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>tid</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>strand</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>annotation</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>843</th>\n",
       "      <td>17886</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000143369[1,1,2]</td>\n",
       "      <td>ENSG00000143369</td>\n",
       "      <td>ECM1</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>150508279</td>\n",
       "      <td>150509531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>845</th>\n",
       "      <td>17887</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000143369[1,1,2]</td>\n",
       "      <td>ENSG00000143369</td>\n",
       "      <td>ECM1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>150509531</td>\n",
       "      <td>150509581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>847</th>\n",
       "      <td>17888</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000143369[1,1,2]</td>\n",
       "      <td>ENSG00000143369</td>\n",
       "      <td>ECM1</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>150509581</td>\n",
       "      <td>150509661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>849</th>\n",
       "      <td>17889</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000143369[1,1,2]</td>\n",
       "      <td>ENSG00000143369</td>\n",
       "      <td>ECM1</td>\n",
       "      <td>+</td>\n",
       "      <td>exon</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>150509661</td>\n",
       "      <td>150509762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>851</th>\n",
       "      <td>17890</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000143369[1,1,2]</td>\n",
       "      <td>ENSG00000143369</td>\n",
       "      <td>ECM1</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>150509762</td>\n",
       "      <td>150509922</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    edge_id  n_genes                     tid              gid gname strand  \\\n",
       "843   17886        2  ENSG00000143369[1,1,2]  ENSG00000143369  ECM1      +   \n",
       "845   17887        2  ENSG00000143369[1,1,2]  ENSG00000143369  ECM1      +   \n",
       "847   17888        2  ENSG00000143369[1,1,2]  ENSG00000143369  ECM1      +   \n",
       "849   17889        2  ENSG00000143369[1,1,2]  ENSG00000143369  ECM1      +   \n",
       "851   17890        2  ENSG00000143369[1,1,2]  ENSG00000143369  ECM1      +   \n",
       "\n",
       "    edge_type  annotation chrom      start       stop  \n",
       "843    intron        True  chr1  150508279  150509531  \n",
       "845      exon        True  chr1  150509531  150509581  \n",
       "847    intron        True  chr1  150509581  150509661  \n",
       "849      exon        True  chr1  150509661  150509762  \n",
       "851    intron        True  chr1  150509762  150509922  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gnames = df3.gname.unique().tolist()\n",
    "df3.loc[df3.gname==gnames[8]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "506d5964-99a5-4f84-a0fc-927dd00bfd7d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edge_id</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>tid</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "      <th>strand</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>annotation</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2130</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000058673[1,17,1]</td>\n",
       "      <td>ENSG00000058673</td>\n",
       "      <td>ZC3H11A</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>203803016</td>\n",
       "      <td>203816926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2130</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000058673[1,18,1]</td>\n",
       "      <td>ENSG00000058673</td>\n",
       "      <td>ZC3H11A</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>203803016</td>\n",
       "      <td>203816926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2130</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000058673[1,19,1]</td>\n",
       "      <td>ENSG00000058673</td>\n",
       "      <td>ZC3H11A</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>203803016</td>\n",
       "      <td>203816926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2130</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000058673[1,2,1]</td>\n",
       "      <td>ENSG00000058673</td>\n",
       "      <td>ZC3H11A</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>203803016</td>\n",
       "      <td>203816926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2130</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000058673[1,20,1]</td>\n",
       "      <td>ENSG00000058673</td>\n",
       "      <td>ZC3H11A</td>\n",
       "      <td>+</td>\n",
       "      <td>intron</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1</td>\n",
       "      <td>203803016</td>\n",
       "      <td>203816926</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  edge_id  n_genes                      tid              gid    gname strand  \\\n",
       "0    2130        2  ENSG00000058673[1,17,1]  ENSG00000058673  ZC3H11A      +   \n",
       "1    2130        2  ENSG00000058673[1,18,1]  ENSG00000058673  ZC3H11A      +   \n",
       "2    2130        2  ENSG00000058673[1,19,1]  ENSG00000058673  ZC3H11A      +   \n",
       "3    2130        2   ENSG00000058673[1,2,1]  ENSG00000058673  ZC3H11A      +   \n",
       "4    2130        2  ENSG00000058673[1,20,1]  ENSG00000058673  ZC3H11A      +   \n",
       "\n",
       "  edge_type  annotation chrom      start       stop  \n",
       "0    intron        True  chr1  203803016  203816926  \n",
       "1    intron        True  chr1  203803016  203816926  \n",
       "2    intron        True  chr1  203803016  203816926  \n",
       "3    intron        True  chr1  203803016  203816926  \n",
       "4    intron        True  chr1  203803016  203816926  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "65de98f3-1122-444f-a0a1-4dac808d9f3f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating iso TPM values\n",
      "Subsetting for polya genes\n",
      "Enforcing minimum TPM\n",
      "Total # isos detected: 231136\n",
      "# isos >= 1 tpm: 210239\n",
      "Applying gene type and novelty subset\n",
      "Number of isos reported: 206806\n"
     ]
    }
   ],
   "source": [
    "datasets = get_ljungman_datasets()\n",
    "\n",
    "ab_df = pd.read_csv(filt_ab, sep='\\t')\n",
    "ab_df, _ = get_tpm_table(ab_df,\n",
    "               how='iso',\n",
    "               min_tpm=min_tpm,\n",
    "               gene_subset=gene_subset)\n",
    "print(ab_df.shape)\n",
    "ab_df = ab_df.loc[tids, datasets]\n",
    "print(ab_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "70c351a4-4a12-490c-bc0b-3b0db0da1868",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(206806, 138)\n",
      "(2989, 24)\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "f8695a94-e743-4a39-8c42-922039aa84b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "ab_df.to_csv('novel_fusions_transcript_tpm.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ceab4c27-4745-46cd-80c4-f2a4e0eb0658",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>annot_transcript_id</th>\n",
       "      <th>huvec_1_1</th>\n",
       "      <th>huvec_1_2</th>\n",
       "      <th>hmec_1_1</th>\n",
       "      <th>hmec_1_2</th>\n",
       "      <th>panc1_1_1</th>\n",
       "      <th>pc3_1_1</th>\n",
       "      <th>k562_1_1</th>\n",
       "      <th>hepg2_1_1</th>\n",
       "      <th>hct116_1_1</th>\n",
       "      <th>...</th>\n",
       "      <th>ocily7_1_1</th>\n",
       "      <th>ocily7_1_2</th>\n",
       "      <th>mcf10a_1_1</th>\n",
       "      <th>mcf10a_1_2</th>\n",
       "      <th>pc9_1_1</th>\n",
       "      <th>pc9_1_2</th>\n",
       "      <th>a673_1_1</th>\n",
       "      <th>a673_1_2</th>\n",
       "      <th>calu3_1_1</th>\n",
       "      <th>calu3_1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000058673[1,17,1]</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.584375</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.815517</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.122543</td>\n",
       "      <td>...</td>\n",
       "      <td>4.257620</td>\n",
       "      <td>3.604625</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.409037</td>\n",
       "      <td>0.821249</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.064829</td>\n",
       "      <td>3.336130</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.856421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000058673[1,18,1]</td>\n",
       "      <td>1.71207</td>\n",
       "      <td>2.277755</td>\n",
       "      <td>1.629982</td>\n",
       "      <td>2.337500</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.631034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.061271</td>\n",
       "      <td>...</td>\n",
       "      <td>11.708455</td>\n",
       "      <td>8.410793</td>\n",
       "      <td>4.300363</td>\n",
       "      <td>4.931630</td>\n",
       "      <td>0.821249</td>\n",
       "      <td>4.297593</td>\n",
       "      <td>6.388973</td>\n",
       "      <td>5.004195</td>\n",
       "      <td>3.677955</td>\n",
       "      <td>8.564210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000058673[1,19,1]</td>\n",
       "      <td>0.57069</td>\n",
       "      <td>0.569439</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.201542</td>\n",
       "      <td>1.433454</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000058673[1,2,1]</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.543327</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.01183</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.061271</td>\n",
       "      <td>...</td>\n",
       "      <td>2.128810</td>\n",
       "      <td>2.403084</td>\n",
       "      <td>3.583636</td>\n",
       "      <td>0.704519</td>\n",
       "      <td>4.106243</td>\n",
       "      <td>2.578556</td>\n",
       "      <td>1.064829</td>\n",
       "      <td>5.004195</td>\n",
       "      <td>0.919489</td>\n",
       "      <td>0.856421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000058673[1,20,1]</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>4.257620</td>\n",
       "      <td>1.201542</td>\n",
       "      <td>1.433454</td>\n",
       "      <td>3.522593</td>\n",
       "      <td>0.821249</td>\n",
       "      <td>0.859519</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.834033</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       annot_transcript_id  huvec_1_1  huvec_1_2  hmec_1_1  hmec_1_2  \\\n",
       "0  ENSG00000058673[1,17,1]    0.00000   0.000000  0.000000  0.584375   \n",
       "1  ENSG00000058673[1,18,1]    1.71207   2.277755  1.629982  2.337500   \n",
       "2  ENSG00000058673[1,19,1]    0.57069   0.569439  0.000000  0.000000   \n",
       "3   ENSG00000058673[1,2,1]    0.00000   0.000000  0.543327  0.000000   \n",
       "4  ENSG00000058673[1,20,1]    0.00000   0.000000  0.000000  0.000000   \n",
       "\n",
       "   panc1_1_1  pc3_1_1  k562_1_1  hepg2_1_1  hct116_1_1  ...  ocily7_1_1  \\\n",
       "0    0.00000      0.0  0.815517        0.0    2.122543  ...    4.257620   \n",
       "1    0.00000      0.0  1.631034        0.0    1.061271  ...   11.708455   \n",
       "2    0.00000      0.0  0.000000        0.0    0.000000  ...    0.000000   \n",
       "3    1.01183      0.0  0.000000        0.0    1.061271  ...    2.128810   \n",
       "4    0.00000      0.0  0.000000        0.0    0.000000  ...    4.257620   \n",
       "\n",
       "   ocily7_1_2  mcf10a_1_1  mcf10a_1_2   pc9_1_1   pc9_1_2  a673_1_1  a673_1_2  \\\n",
       "0    3.604625    0.000000    1.409037  0.821249  0.000000  1.064829  3.336130   \n",
       "1    8.410793    4.300363    4.931630  0.821249  4.297593  6.388973  5.004195   \n",
       "2    1.201542    1.433454    0.000000  0.000000  0.000000  0.000000  0.000000   \n",
       "3    2.403084    3.583636    0.704519  4.106243  2.578556  1.064829  5.004195   \n",
       "4    1.201542    1.433454    3.522593  0.821249  0.859519  0.000000  0.834033   \n",
       "\n",
       "   calu3_1_1  calu3_1_2  \n",
       "0   0.000000   0.856421  \n",
       "1   3.677955   8.564210  \n",
       "2   0.000000   0.000000  \n",
       "3   0.919489   0.856421  \n",
       "4   0.000000   0.000000  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ab_df = pd.read_csv('novel_fusions_transcript_tpm.tsv', sep='\\t')\n",
    "ab_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "c04602df-6e01-4339-825b-c14abdb223f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# table for Ariel - edge that's shared along with tids, gids, and gnames\n",
    "df3 = df3[['edge_id', 'edge_type', 'chrom', 'strand', 'start', 'stop', 'tid', 'gid', 'gname']]\n",
    "df3 = df3.drop_duplicates()\n",
    "\n",
    "# one for tids\n",
    "tid_df = df3.drop(['gname', 'gid'], axis=1)\n",
    "tid_df = tid_df.groupby(['edge_id', 'edge_type', 'chrom', 'strand', 'start', 'stop']).agg(','.join).reset_index()\n",
    "\n",
    "# one for gids\n",
    "gid_df = df3.drop('tid', axis=1)\n",
    "gid_df.drop_duplicates(inplace=True)\n",
    "gid_df = gid_df.groupby(['edge_id', 'edge_type', 'chrom', 'strand', 'start', 'stop']).agg(','.join).reset_index()\n",
    "\n",
    "# merge in \n",
    "tid_df = tid_df.merge(gid_df, how='left', on=['edge_id', 'edge_type', 'chrom', 'strand', 'start', 'stop'])\n",
    "\n",
    "tid_df.to_csv('novel_fusion_edges.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "7993b39f-bc27-4f98-8e5d-de1ae08f5787",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>edge_id</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>tid</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2130</td>\n",
       "      <td>intron</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203803016</td>\n",
       "      <td>203816926</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2131</td>\n",
       "      <td>exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203816926</td>\n",
       "      <td>203817124</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2132</td>\n",
       "      <td>intron</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203817124</td>\n",
       "      <td>203818570</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2133</td>\n",
       "      <td>exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203818570</td>\n",
       "      <td>203818689</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2134</td>\n",
       "      <td>intron</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203818689</td>\n",
       "      <td>203828299</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   edge_id edge_type chrom strand      start       stop  \\\n",
       "0     2130    intron  chr1      +  203803016  203816926   \n",
       "1     2131      exon  chr1      +  203816926  203817124   \n",
       "2     2132    intron  chr1      +  203817124  203818570   \n",
       "3     2133      exon  chr1      +  203818570  203818689   \n",
       "4     2134    intron  chr1      +  203818689  203828299   \n",
       "\n",
       "                                                 tid  \\\n",
       "0  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "1  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "2  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "3  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "4  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "\n",
       "                               gid          gname  \n",
       "0  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  \n",
       "1  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  \n",
       "2  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  \n",
       "3  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  \n",
       "4  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  "
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tid_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3d18844-96d1-4b64-94cb-b2c3c8377051",
   "metadata": {},
   "source": [
    "## 230517"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a1629e4d-4dc1-42f3-9982-ab6865391bd5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>edge_id</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>chrom</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>tid</th>\n",
       "      <th>gid</th>\n",
       "      <th>gname</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>2130</td>\n",
       "      <td>intron</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203803016</td>\n",
       "      <td>203816926</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>2131</td>\n",
       "      <td>exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203816926</td>\n",
       "      <td>203817124</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2132</td>\n",
       "      <td>intron</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203817124</td>\n",
       "      <td>203818570</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>2133</td>\n",
       "      <td>exon</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203818570</td>\n",
       "      <td>203818689</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>2134</td>\n",
       "      <td>intron</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>203818689</td>\n",
       "      <td>203828299</td>\n",
       "      <td>ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...</td>\n",
       "      <td>ENSG00000058673,ENSG00000257315</td>\n",
       "      <td>ZC3H11A,ZBED6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  edge_id edge_type chrom strand      start       stop  \\\n",
       "0           0     2130    intron  chr1      +  203803016  203816926   \n",
       "1           1     2131      exon  chr1      +  203816926  203817124   \n",
       "2           2     2132    intron  chr1      +  203817124  203818570   \n",
       "3           3     2133      exon  chr1      +  203818570  203818689   \n",
       "4           4     2134    intron  chr1      +  203818689  203828299   \n",
       "\n",
       "                                                 tid  \\\n",
       "0  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "1  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "2  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "3  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "4  ENSG00000058673[1,17,1],ENSG00000058673[1,18,1...   \n",
       "\n",
       "                               gid          gname  \n",
       "0  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  \n",
       "1  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  \n",
       "2  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  \n",
       "3  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  \n",
       "4  ENSG00000058673,ENSG00000257315  ZC3H11A,ZBED6  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edge_df = pd.read_csv('novel_fusion_edges.tsv', sep='\\t')\n",
    "edge_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5819dc70-5dbb-4be9-83e8-c9debc2076b8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
